---
title: "SPCG figure RNA seq summary " 
author: "ASC Lab"
output: 
  html_document:
    toc: TRUE
    toc_float: FALSE
editor_options: 
  chunk_output_type: console
---


<style type="text/css">
.main-container {
  max-width: 2500px;
  margin-left: 5px;
  margin-right: auto;
}
.toc-content {
  max-width: 2500px;
  margin-left: 50px;
  margin-right: auto;
}

div {

  margin-left: 5px;
}


hr.new1 {
  border-top: 1px solid #84a8e0;
}


</style>



version: 1.0 <br />
originally created 06/15/2020/ <br />
Run at "`r format(Sys.time())`"

 
# Overview/Method:
  * only patient tissue are plotted ( no cell or pdx )

# version update
  * last version was 7 this is a complete migration so think of this as a new version.  
  * Mistake last versions was not having fusion and outlier in two seperate chunks. This version will be a bit more contigous. 
  
```{r setup, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE, include = FALSE , cache.lazy = FALSE )
sample.source = c("patient")
library ( ggpattern )
library(dplyr)
library(Matrix)
library (RColorBrewer )
library(cowplot)
library(biomaRt)
library(knitr)
library(kableExtra)
library(tidyverse)
library(ggpubr)
library(limma)
library ( fgsea )
library(openxlsx)
library ( ComplexHeatmap)
library ( feather )
library ( ggplot2)
library("RColorBrewer")
library ( patchwork)
library(survival)
library ( survminer )
library ( ggbeeswarm )
library(ggridges)
library(gridExtra)
library(grid)
library ( stringr)
library ( data.table)
library ( gridExtra)
library ( forestplot )
library ( reshape2)
library  ( edgeR)
library ( limma )
library ( treemapify)
library ( patchwork)
library ( ggforce)
library ( pkr)
library(wesanderson)
library(HGNChelper)
library ( ggpattern)
getPalette = colorRampPalette(brewer.pal(9, "Set1")) # expand color pallete
options(scipen=999)
library(circlize)
library ( metap )
library ( patchwork)
library(ggforce)
library(ggh4x)
library(DiagrammeR)
library(magrittr)
library(DiagrammeRsvg)

source ( "draw.p.R")

# library ( conflicted)
print.it = 1


### get necessary variables 
### 
config = data.frame ( t ( read.table ( "config.new.txt", header = T  ) ) , stringsAsFactors = F) 
colnames ( config ) = config[1:1, ]
config = config [-1,]

#################

main.data = config$main.data
resource =  config$resource
source (  config$source )
source ( "/ehome/scripts/config.ext/ext.plot.R")

out.dir = "./summary1/"
dir.create( out.dir )
# get ready to slurp up data! 
  wb <- createWorkbook()
  
  
  

  subcolor <- c(
   
"ALL" ="#bf812d"
,"AML"= "#dfc27d"
,"CNS" ="#33a02c"
,"EWS"= "#ff7f00"
,"HEME" ="#ef3b2c"
,"HBL" ="#a6cee3"
,"HCC"= "#b2df8a"
,"NBL"= "#ffed6f"
,"NRSTS" ="#7a0177"
,"OS" ="#cab2d6"
,"OTHER_HEME" ="#ef3b2c"
,"O/H"= "#ef3b2c"
,"Other Heme" ="#ef3b2c"
,"RARE/OTHER"= "#4DA7D2"
,"R/O"= "#4DA7D2"
,"FPRMS"= "#d94801"
,"FNRMS" = "#fdbf6f"
,"WILMS"= "#fb9a99"
)
  
  
  

```


```{r}
  
# get all the necessary dataset loaded up 

genelist = readRDS ( config$genelist)
hugo = genelist$hugo
genelist = genelist$gene
# from Final.fusion_2.3_SPCGc_newSummary.R, this is the latest with new filters to be more consistent. 
fusion = readRDS(config$FUSION_2021 )     # FUSION_2021_FINAL.rds
itd = readRDS("/data/SUMMARY/FUSION/ITD.rds")
drug =  readRDS ( config$DRUG)
drug = drug [ drug$clinical_sig == "positive", ]
drug$uid = seq ( 1,  nrow ( drug ), 1)
# remove anything that reads chemotherapy 


out = readRDS( config$OUT )
cna = readRDS( config$CNA.hugo)
snv = readRDS( config$SNV.hugo)
tpm = readRDS(config$TPM)
# config STAR raw original is fine and should contain all the genes, use SNHG1 linc rna as test. 
star = read.table( config$star , header=TRUE,sep="\t",stringsAsFactors = FALSE,na.strings=".",  quote = "", fill = TRUE)
gc()
```

```{r}
# loading more up and modifying databses 

cosmic2 = genelist [genelist$Cosmic_v.92 == "Y" , c("gene","cosmic.annot")]
cosmic2$cosmic.annot = gsub ( "^N$","", cosmic2$cosmic.annot)

cosmic2 = cosmic2 %>% 
  mutate(role = strsplit(as.character(cosmic.annot), ",")) %>% 
  unnest(role) %>% data.frame()

cosmic2$cosmic.annot = NULL 



#### don't merge the drugs... yet may be not necessary
drug.up = drug[ drug$alteration!="UNDEREXPRESSION" & drug$type == "expression" & drug$clinical_sig == "positive",  ]
drug.down = drug[ drug$alteration=="UNDEREXPRESSION" & drug$type == "expression" & drug$clinical_sig == "positive" ,  ]

# sanity 
unique ( drug.up$alteration)
unique ( drug.down$alteration)
unique ( drug.up$clinical_sig)

drug.fusion = drug[ drug$type == "fusion", ]






tsg = genelist[ grepl("TSG", genelist$cosmic.annot , ignore.case = T) & ! grepl("ONCOGENE", genelist$cosmic.annot , ignore.case = T) , ]
oncogene =  genelist[ !grepl("TSG", genelist$cosmic.annot , ignore.case = T) &  grepl("ONCOGENE", genelist$cosmic.annot , ignore.case = T) , ]
tsg = as.character ( tsg$gene)
oncogene = as.character ( oncogene$gene)
tsg.onco =  genelist[ grepl("TSG", genelist$cosmic.annot , ignore.case = T) &  grepl("ONCOGENE", genelist$cosmic.annot , ignore.case = T) , ]






genecard = config$GENECARD
# for the genes in question lets get some information. 
genecard  <- readRDS(genecard)
genecard = genecard[ , c("InputTerm","EntrezGene")]
genecard = genecard[!is.na(genecard$EntrezGene), ]
colnames(genecard) = c("GENE","description")


# get family genecard 
familyclass = read.xlsx("https://www.dropbox.com/s/bmlz5xd0q9gnmox/genefamily.xlsx?dl=1")
familyclass = familyclass [ familyclass$Source == "UniProtKB/Swiss-Prot", ]
familyclass = familyclass[ grepl ( "^Belongs", familyclass$Name), ]
familyclass$family = stringr::str_match(familyclass$Name, "(Belongs.*?\\.)")[, 2]
familyclass$family = gsub ( "Belongs to the |\\.$", "", familyclass$family )     
familyclass$family = gsub ( "\\.| family$", "", familyclass$family )   
familyclass$family = gsub ( " superfamily$", "", familyclass$family )  
 




```

```{r}
# getting the keys and clearning up 

manaul_rm = c ( "BD6", "AJ6", "AD12" ) 

key.file <-  config$spcg.key

### read in data
new.key = 0

rna.key  <- read.xlsx(key.file , sheet="masterkey.stable.txt")
raw_key = rna.key 
rna.key[ is.na(rna.key) ] <- ""

nrow ( rna.key)

table ( rna.key$sample.source )
#rna.key$sample.source = tolower( rna.key$sample.source )
table ( rna.key$cancer.subtype )
#rna.key$cancer.subtype = toupper( rna.key$cancer.subtype )
#rna.key$cancer.subtype = gsub ( " ", "_", rna.key$cancer.subtype)
table ( rna.key$sex  )
#rna.key$sex = gsub ( " ", "", rna.key$sex)
table ( rna.key$patient.institution )
#rna.key = rna.key[rna.key$patient.institution %in% c("Stanford","UCSF", "CHO"), ]
table ( rna.key$treatment)
#rna.key$treatment = gsub ( "Naive", "Naïve",  rna.key$treatment  )



nrow(rna.key)
nrow(raw_key)

rna.key = rna.key[rna.key$patient.institution %in% c("Stanford","UCSF", "CHO") , ]
nrow(rna.key)


rna.key = rna.key[rna.key$sample.source == "patient" , ]
nrow(rna.key)

table ( rna.key$exclude.sample)

rna.key = rna.key[rna.key$exclude.sample == 0 , ]
nrow(rna.key)

table ( rna.key$exclude.paper )
rna.key = rna.key[rna.key$exclude.paper == 0 , ]
nrow(rna.key)


wgs.key  = rna.key
sort ( wgs.key$WGS.id )
wgs.key = wgs.key[wgs.key$WGS.id != "None", ]
nrow ( wgs.key)
sort ( wgs.key$WGS.id )
nrow ( wgs.key)
unique ( wgs.key$WGS.exclude)
#wgs.key = wgs.key[wgs.key$WGS.exclude == "PASS", ]

# give me all that are dups 
dupw = wgs.key [ duplicated ( wgs.key$WGS.id ) , ]$WGS.id
# View (wgs.key[ wgs.key$WGS.id %in% dupw, ] )
wgs.key = wgs.key [!duplicated (  wgs.key$WGS.id), ]

nrow ( rna.key )
rna.key=rna.key[rna.key$RNAseq.id != "None", ]
nrow ( rna.key )

rna.key_failed = rna.key[rna.key$RNAseq.exclude == "FAIL", ]
rna.key = rna.key[rna.key$RNAseq.exclude == "PASS", ] # this one is only for RNAseq so make a copy of WGS after this. 
nrow ( rna.key )
sort  ( rna.key$RNAseq.id)
nrow ( rna.key )
unique ( rna.key$RNAseq.exclude)
# rna.key = rna.key [ ! rna.key$RNAseq.id %in% manaul_rm , ]






rna.key$cancer.subtype = gsub ( "RARE/OTHER", "R/O", rna.key$cancer.subtype  )
rna.key$cancer.subtype = gsub ( "OTHER_HEME", "O/H", rna.key$cancer.subtype  )
wgs.key$cancer.subtype = gsub ( "RARE/OTHER", "R/O", wgs.key$cancer.subtype  )
wgs.key$cancer.subtype = gsub ( "OTHER_HEME", "O/H", wgs.key$cancer.subtype  )


sort ( unique ( rna.key$RNAseq.id) ) 
sort ( unique ( rna.key$patient.id) )




```


```{r}

# finding missing values if ther are any. 

missing_fusion = setdiff ( rna.key$RNAseq.id, unique ( fusion$raw$RNAseq.id))
missing_outlier = setdiff ( rna.key$RNAseq.id, unique ( out$RNAseq.id))
missing_tpm = setdiff ( rna.key$RNAseq.id, unique ( names ( tpm )  ))


setdiff( rna.key$RNAseq.id, names ( star ))




nofusion = setdiff ( rna.key$RNAseq.id, unique ( fusion$tier3$RNAseq.id))



```



```{r}
# QC is important because it will determine which samples to eliminate 


# here be dragons, temporary fix only 

# rna.key [rna.key$RNAseq.id == "BO2", ]$cancer.subtype = "FNRMS"

qc= read.table ( 
    paste0( config$QC, "/QC_wide.tsv" ) 
    ,header=TRUE,sep="\t",stringsAsFactors = FALSE,
    na.strings=".", quote = "", fill = TRUE)

qc$Mitochondrialratio = round ( qc$Mitochondrial/qc$Mapped.reads, 1)

qc.txn= read.table ( 
    paste0( config$QC, "/QC_Txn2.tsv" ) 
    ,header=TRUE,sep="\t",stringsAsFactors = FALSE,
    na.strings=".", quote = "", fill = TRUE)


qc = merge ( qc, qc.txn [ , !colnames ( qc.txn) %in% colnames ( qc ) [ colnames ( qc ) != "tube"]], by="tube")


setdiff  ( rna.key$RNAseq.id, qc$tube )


# going to combine the key with QC 
rna.key = merge ( rna.key, qc , by.x="RNAseq.id", "tube" , all.x=T, all.y=F  )

# this is how many samples failed and how many pass our QC 
qc_thres = 20000000

# ok lets loop through the QC and define each sample as PASS or FAIL 
# QC chunk 
rna.key$qc = ifelse ( 
    ( rna.key$Uniquely.mapped.reads > qc_thres | rna.key$UMEND > 10000000 ) & rna.key$Mitochondrialratio <= .3  & rna.key$UniqueMultiple_ratio > 1, "PASS", "FAIL"
    )

nrow ( rna.key[rna.key$qc == "FAIL", ] ) 
nrow ( rna.key[rna.key$qc == "FAIL", ] )  / nrow ( rna.key)
table (rna.key$qc )


rna.key$group = rna.key$cancer.subtype

# there should be no more duplications so these codes are obsolete now. 

dups = rna.key[duplicated(rna.key$parent.sample), ]$parent.sample

dup.lose = c()

for ( d in dups){
  d = rna.key[ rna.key$parent.sample == d, ]
  lose = min ( d$Uniquely.mapped.reads)
  dup.lose = c ( dup.lose, d[d$Uniquely.mapped.reads == lose, ]$RNAseq.id )
}


# warn if missing colors 
swarn = setdiff ( rna.key$cancer.subtype, names ( subcolor))
if ( length ( swarn) != 0 ){
  warning("some colors are missing ")
}


# Removal sample chunk 
#rna.key [ rna.key$RNAseq.id %in%  manaul_rm,  ]

# only have to run once, saving qc directly into key 
if ( 1>2 ){
fix_table <- createWorkbook()
addWorksheet(fix_table, 'QC')

writeData(fix_table
          ,'QC'
          ,rna.key [ 
                   , c("RNAseq.id", 'Uniquely.mapped.reads', 'UMEND'
                       , 'Mitochondrialratio', 'UniqueMultiple_ratio'
                       , "RNAlibrary.chemistry" , "parent.preservation"
                       , "RNAseq.platform", "qc")]
           
          , rowNames=F  )

saveWorkbook(fix_table, file = paste0(out.dir,"fix_include_qc.xlsx"), overwrite = TRUE)
}
```


```{r}


paper_table <- createWorkbook()

addWorksheet(paper_table, 'QC')
#writeData(wb, 'QC' , df  , rowNames=T  )
# writeData(paper_table, 'QC', )
writeData(paper_table
          ,'QC'
          ,rna.key [ 
                   , c("RNAseq.id", 'Uniquely.mapped.reads', 'UMEND'
                       , 'Mitochondrialratio', 'UniqueMultiple_ratio'
                       , "RNAlibrary.chemistry" , "parent.preservation"
                       , "RNAseq.platform", "qc")]
           
          , rowNames=F  )
#saveWorkbook(wb, file = paste0(out.dir,"file.xlsx"), overwrite = TRUE)




qc.plot <- qcbar(rna.key,key=rna.key,group.col="group",tube="RNAseq.id", colqc = subcolor, uthresh = qc_thres )

          ( qc.plot$Uniquely.mapped.reads + theme ( legend.position = "none") ) /
          (qc.plot$UMEND + theme ( legend.position = "none") )/
         ( qc.plot$mitoratio + theme ( legend.position = "none") ) / 
         (  qc.plot$unique2multiple   + theme ( legend.position = "bottom")  )




# this part will be skipped because these are excluded above already 

if ( 1 > 2) {
transplant = rna.key[!is.na(rna.key$exclude.sample) & rna.key$exclude.sample == "Transplant", ] # keep transplant 
}

# failed failedChem are kept for fusions 

failedqc = rna.key[rna.key$qc == "FAIL" , ]
failedChem = rna.key[rna.key$RNAlibrary.chemistry == "RNA Access", ]$RNAseq.id 

stable2_key = rna.key
rna.key = stable2_key[! stable2_key$RNAseq.id %in% failedqc$RNAseq.id , ]

nrow ( rna.key )

# sanity check 
rna.key[duplicated(rna.key$parent.sample), ]$parent.sample
intersect ( rna.key$RNAseq.id , failedqc$RNAseq.id )


rna_access = rna.key[rna.key$RNAlibrary.chemistry == "RNA Access", ]
nrow ( rna_access )
rna_ffpe = rna_access[rna_access$parent.preservation == "FFPE",  ]
nrow ( rna_ffpe )

rna_fft = rna_access[rna_access$parent.preservation == "FFT",  ]
nrow ( rna_fft )

nrow (  rna.key[rna.key$RNAlibrary.chemistry != "RNA Access", ] )


```

```{r}
# miscallenous making it easier to read the key 
kname = "RNAseq.id patient.id sex sample.type sample.classification treatment cancer.type diagnosis cancer.subtype oncotree group top6 top6_rho UMEND qc"
unlist ( strsplit(kname, " ") ) 
#View ( rna.key[ , unlist ( strsplit(kname, " ") )  ])

```


# SPCG {.tabset}

## Samples {.tabset}


### QC Plots 


```{r fig=TRUE,fig.width=22, fig.height=9, echo=FALSE, include=TRUE }
( qc.plot$Uniquely.mapped.reads + theme ( legend.position = "none") ) /
          (qc.plot$UMEND + theme ( legend.position = "none") )/
         ( qc.plot$mitoratio + theme ( legend.position = "none") ) / 
         (  qc.plot$unique2multiple   + theme ( legend.position = "bottom")  )
```

```{r}
# get estimate scores 
# run maketest you can create esimate scorres
makeest = function (tpm){
# make estimate 
library(estimate)


gtest = function (input.f, output.f, id = c("GeneSymbol", "EntrezID")) 
{
  stopifnot((is.character(output.f) && length(output.f) == 
               1 && nzchar(output.f)))
  id <- match.arg(id)
  merged.df <- merge(common_genes, input.f, by.x = id, by.y = "row.names")
  rownames(merged.df) <- merged.df$GeneSymbol
  merged.df <- merged.df[, -1:-ncol(common_genes)]
  print(sprintf("Merged dataset includes %d genes (%d mismatched).", 
                nrow(merged.df), nrow(common_genes) - nrow(merged.df)))
  outputGCT(merged.df, output.f)
}

est.dir = "./ESTIMATE/"
dir.create( est.dir )

counts = tpm[ , unique ( rna.key$RNAseq.id  )   ] # it does not matter what the other inputs are so just chose the first x columns

gtest(input.f= counts  , output.f=paste0(est.dir, "ematrix.gct"), id="GeneSymbol")
estimateScore(paste0(est.dir, "ematrix.gct"), paste0(est.dir, "score.gct"), platform="illumina")
est = t(data.frame(fread(paste0(est.dir, "score.gct"), skip = 2)[,-1], row.names = T))

# estiamte purity 
estimate_purity = cos(0.6049872018 + 0.0001467884*est[,3])
est = data.frame ( est, stringsAsFactors = F)
est$purity = as.numeric ( estimate_purity  )

saveRDS( est, "ESTIMATE.rds")

return ( est )
}

est = readRDS( "ESTIMATE.rds" )

if ( length ( setdiff ( rna.key$RNAseq.id, row.names ( est )) ) > 1 ) {
  makeest (tpm)
  est = readRDS( "ESTIMATE.rds" )
}

# add estimate to qc

dim ( rna.key )


rna.key = merge ( rna.key, est, by.x="RNAseq.id", by.y="row.names")

```


### Tabulations {.tabset}

#### Diagnosis Match 

* Note so far: "AJ6", "AD12", "BD6" were removed. 
AD12, pid 533-5 ( CIC-DUX ) --> matches to sarcomas
AJ6, quote from ASC " Ah yes I remember this now.  So we should remove then. Please indicate on masterkey this samples should not be used.  It is not “leukemia” just normal bone marrow "
--> also no bone marrow
BD6 as per conversaion in dignostic fusion for CRTC1-MAML2 and Marcus


```{r fig=TRUE,fig.width=4, echo=FALSE, include=TRUE }


kable( rna.key[ , unlist ( strsplit(kname, " ") )  ]  , format = "html" , row.names = F, caption = "Reasons for Exclusion" ) %>% kable_classic(full_width = F, position = "center")

# rna.key =rna.key[!rna.key$RNAseq.id %in% rm , ]

```



#### Total 

```{r fig=TRUE,fig.width=4, echo=FALSE, include=TRUE }
# sanity check! 
#unique ( rna.key$qc)
#unique ( rna.key$WGS.id)



tally = data.frame ( 
  "Total" = nrow ( rna.key )
  ,"Total.WGS" = nrow ( rna.key[rna.key$WGS.id != "None", ])
  , "Total.unique" = nrow ( rna.key[!duplicated ( rna.key$patient.id), ] )
  , "Total.unique.WGS" = nrow ( rna.key[!duplicated ( rna.key$patient.id) & rna.key$WGS.id != "None", ] )
  )

kable( tally  , format = "html" , row.names = F ) %>% kable_classic(full_width = F, position = "center")


```

#### Overall without Exome Seq. 

```{r fig=TRUE,fig.width=4, echo=FALSE, include=TRUE }

tally = data.frame ( 
  "Total" = nrow ( rna.key[rna.key$RNAlibrary.chemistry != "RNA Access", ] )
  ,"Total.Matched_WGS" = nrow ( rna.key[rna.key$WGS.id != "None", ])
  , "Total.unique" = nrow ( rna.key[!duplicated ( rna.key$patient.id), ] )
  , "Total.unique.WGS" = nrow ( rna.key[!duplicated ( rna.key$patient.id) & rna.key$WGS.id != "None", ] )
  )

kable( tally  , format = "html" , row.names = F ) %>% kable_classic(full_width = F, position = "center")

failed.df = data.frame (QC = length ( failedqc) , 
                         chemistry= nrow ( rna.key[rna.key$RNAlibrary.chemistry == "RNA Access", ] )  )


```

#### Samples Removed 

```{r fig=TRUE,fig.width=4, echo=FALSE, include=TRUE }

ccc= c("")

failed.df = data.frame (QC = nrow ( failedqc) ,  
                         chemistry= nrow ( rna.key[rna.key$RNAlibrary.chemistry == "RNA Access", ] )  )


kable( failed.df  , format = "html" , row.names = F, caption = "Reasons for Exclusion" ) %>% kable_classic(full_width = F, position = "center")




```

#### Total RNA.seq 
```{r fig=TRUE,fig.width=4, echo=FALSE, include=TRUE }

tally = data.frame ( 
  "Total" = nrow ( rna.key_failed ) + nrow ( rna.key)
  ,"PASSED QC" = nrow ( rna.key)
  ,"Failed QC" = nrow ( rna.key_failed )
  , "Matched_WGS" = nrow ( rna.key[rna.key$WGS.id != "None", ])
  ,"Unique Patient" = nrow ( rna.key[!duplicated ( rna.key$patient.id), ] )
  , "Fresh Frozen" = nrow ( rna.key[rna.key$parent.preservation == "FFT", ] )
  , "FFPE" = nrow ( rna.key[rna.key$parent.preservation == "FFPE", ] )

  )

kable( tally  , format = "html" , row.names = F ) %>% kable_classic(full_width = F, position = "center")



```


#### Alkylator

```{r fig=TRUE,fig.width=4, echo=FALSE, include=TRUE }
# sanity check! 




tally = as.data.frame ( 
      table ( rna.key$alkylator), stringAsFactor=F
  )
colnames(tally) = c("alkylator", "total")
tally = tally[ order ( tally$total, decreasing = T), ]
kable( tally  , format = "html" , row.names = F, caption="alkylator" ) %>% kable_classic(full_width = F, position = "center")


```


#### Treatment

```{r fig=TRUE,fig.width=4, echo=FALSE, include=TRUE }


tally = as.data.frame ( 
      table ( rna.key$treatment), stringAsFactor=F
  )
colnames(tally) = c("treatment", "total")
tally = tally[ order ( tally$total, decreasing = T), ]
kable( tally  , format = "html" , row.names = F, caption="treatment" ) %>% kable_classic(full_width = F, position = "center")


```

#### Radiation

```{r fig=TRUE,fig.width=4, echo=FALSE, include=TRUE }



tally = as.data.frame ( 
      table ( rna.key$radiation), stringAsFactor=F
  )
colnames(tally) = c("radiation", "total")
tally = tally[ order ( tally$total, decreasing = T), ]
kable( tally  , format = "html" , row.names = F, caption="radiation" ) %>% kable_classic(full_width = F, position = "center")


```


#### Sample Type

```{r fig=TRUE,fig.width=4, echo=FALSE, include=TRUE }



temp = rna.key$sample.type
temp = gsub ( ".*_", "", temp )

tally = as.data.frame ( 
      table ( temp), stringAsFactor=F
  )
colnames(tally) = c("sample.type", "total")
tally = tally[ order ( tally$total, decreasing = T), ]
kable( tally  , format = "html" , row.names = F, caption="sample.type" ) %>% kable_classic(full_width = F, position = "center")


```

#### Major groups

```{r fig=TRUE,fig.width=4, echo=FALSE, include=TRUE }



temp = rna.key$cancer.type


tally = as.data.frame ( 
      table ( temp), stringAsFactor=F
  )
colnames(tally) = c("Type", "total")
tally = tally[ order ( tally$total, decreasing = T), ]
kable( tally  , format = "html" , row.names = F, caption="sample.type" ) %>% kable_classic(full_width = F, position = "center")


```



```{r}

# summarize. 

failed = failedqc[  order ( failedqc$cancer.subtype), ]




tally = as.data.frame ( 
      table ( rna.key$patient.institution), stringAsFactor=F
  )



tab.inst = make.pie( tally
                , name.first="Institution"
                #, cc = c("#83bf43","#cf4051", "#6ec0db", "grey")
                , leg.pos = "right"
                , title = "Institution"
                )

tally = as.data.frame ( 
      table ( rna.key$patient.institution), stringAsFactor=F
  )


tally = as.data.frame ( 
      table ( rna.key$sex), stringAsFactor=F
  )
tab.sex = make.pie( tally
                , name.first="Sex"
                , cc = c("#e38ae3","#38a9ff", "grey")
                , leg.pos = "right"
                , title = "Sex"
                )


tally = as.data.frame ( 
      table ( rna.key$cancer.type), stringAsFactor=F
  )
tab.ct = make.pie( tally
                , name.first="Major Type"
                , cc = c("#8be386" ,"#eb4034", "#79c4d1" )
                , leg.pos = "right"
                , title = "TYPE"
                )


```

### Plot Tabulation

```{r fig=TRUE,fig.width=12, fig.height=10, echo=FALSE, include=TRUE , results='asis'}
# tabulate 
cancer.subtype.order = c ( "ALL", "AML", "O/H", "OS","EWS","HCC","HBL","WILMS","NRSTS","FPRMS","FNRMS","NBL","CNS","R/O")



key.plot = data.frame ( table ( rna.key [ , c( "treatment", "cancer.subtype")] ) )
key.plot=key.plot[key.plot$Freq != 0 , ]

total.by.cancer.type = data.frame ( table ( rna.key$cancer.subtype), stringsAsFactors = F )
colnames ( total.by.cancer.type ) = c("cancer.subtype", "Total")

total.by.cancer.type = merge ( total.by.cancer.type, key.plot, by="cancer.subtype" )
total.by.cancer.type$sub = paste0( total.by.cancer.type$cancer.subtype, "(", total.by.cancer.type$Total, ")")

key.plot$label = paste0 ( " ", key.plot$treatment, "(", key.plot$Freq, ")")

subcolor_freq =  subcolor

names ( subcolor_freq ) = as.character ( sapply ( names ( subcolor) , function ( x)  
{
  unique ( total.by.cancer.type[total.by.cancer.type$cancer.subtype ==x, ]$sub )
}
  )
)

total.by.cancer.type$label = paste0 ( " ", total.by.cancer.type$treatment, "(", total.by.cancer.type$Freq, ")")

plot_tab1 =  ggplot(total.by.cancer.type, aes(area = Freq, fill = sub, label = label, subgroup = sub, subgroup2 = treatment)) +
    geom_treemap(position = "identity" ) + 
    geom_treemap_text(colour = "black", place = "top", min.size=0, grow=F, reflow = T, size= 12, angle=0) +
    geom_treemap_subgroup_text(place = "bottomleft", reflow = F, alpha = 1, colour =
                                   "white", fontface = "italic", min.size = 0,  angle=0, size= 20 ) +
    geom_treemap_subgroup2_border(colour = "grey", size = 3) +
    geom_treemap_subgroup_border(colour = "white", size = 5) + scale_fill_manual(values = subcolor_freq )  +
    theme(legend.position="bottom") 






#nrow ( rna.key )
#nrow ( rna.key[!duplicated( rna.key$patient.id), ] )

plot_tab1 / (tab.inst + tab.sex + tab.ct)  + plot_layout(height =c ( .7,.3 ) )


```


## Multi-Samples 

```{r}

# multi chunk

key.long = rna.key[ rna.key$sample.source == "patient", ]
dups = key.long$patient.id[ duplicated(key.long$patient.id) ]
key.long= key.long[key.long$patient.id %in% c( dups), c("patient.id","temporal.order","cancer.subtype", "sample.classification", "treatment",  "WGS.id")]

# make unique names so we can "stack" them one at at time 

key.long$label =  "R"
# get the color for each sample;its difficult to do it by factor here so each row will have a color base off cancers type
key.long$col = subcolor[key.long$cancer.subtype ]

key.long = key.long[ order ( key.long$cancer.subtype, key.long$patient.id), ]
key.long$patient.id = factor ( key.long$patient.id, levels= unique ( key.long$patient.id )) 


# we are going to include WGS only as well. 



# going to plot stack if its the same region and x axis as the temporal 
# then use facet to plot each patient id seperately 

 temp = key.long 
 temp$temporal.order = as.character ( temp$temporal.order )

# in order to make things "solid" I needed change pattern_density to 0  
 
multi_g1 = ggplot(temp, aes(x = temporal.order, y=1 )) + 
   geom_col_pattern( aes ( pattern   =sample.classification, fill = cancer.subtype, pattern_density =sample.classification  ) , 
                     stat = 'identity', position = 'stack'
                     ,  colour = "white" # color between the plots 
                     , pattern_colour  = '#666262' # color of pattern
                   ) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  # note I can change fill color as usual, however this really messes with the legend
  scale_fill_manual(values =  unique ( key.long$col )  ) +
  scale_pattern_manual(values= c( "NonPrimary" = "stripe","Primary"="circle") ) +
  # change density to white will make it solid 
  scale_pattern_density_manual(values= c( "NonPrimary" = .01,"Primary"= 0  ) ) + 
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        , axis.ticks.y=element_blank()
        , axis.text.y=element_blank()
  ) + facet_wrap(~ patient.id, scales = "free" ,  ncol = 10 ) +
  theme(legend.position="none") + theme(legend.key.size = unit(2, 'cm')    ) +
  geom_text(data=temp,aes(x=temporal.order, label=label), position = position_stack(vjust = 0.1) , colour="black") 




# key is weird so make a new dummy one 
temp2 = temp 
temp2$sample.classification = factor ( temp2$sample.classification, levels = c("Primary","NonPrimary"))
d_legend=  ggplot(head ( temp2 ) , aes( sample.classification, temporal.order)) +
   geom_col_pattern(
     aes(pattern = sample.classification, fill = sample.classification, pattern_fill = sample.classification , pattern_density = sample.classification ),
     colour                   = 'black') +
   theme_bw() +
   labs(
     title    = "Longitudinal Samples"
   ) +
   scale_pattern_fill_manual(values = c(a='blue', b='red', c='grey')) + # the color of the dots
   scale_fill_manual( values=c ("white", "white" ) ) +
     scale_pattern_manual(values= c( "NonPrimary" = "stripe","Primary"="circle") ) +
  scale_pattern_density_manual(values= c( "NonPrimary" = .01,"Primary"= 0  ) ) + 
   theme(legend.position = 'bottom') +
   coord_fixed(ratio = 1) + theme(legend.key.size = unit(2, 'cm'))
 





temp[temp$patient.id == "490", ]

temp[temp$patient.id == "572", ]
temp[temp$patient.id == "104", ]
temp[temp$patient.id == "579", ]


# we need to add samples where there is WGS but no RNAseq 


key.wgs_long = wgs.key[ wgs.key$sample.source == "patient", ]
dups = key.wgs_long$patient.id[ duplicated(key.wgs_long$patient.id) ]
key.wgs_long= key.wgs_long[key.wgs_long$patient.id %in% c( dups), c("patient.id","temporal.order","cancer.subtype", "sample.classification", "treatment", "WGS.id")]

# make unique names so we can "stack" them one at at time 

key.wgs_long$label =  "W"
# get the color for each sample;its difficult to do it by factor here so each row will have a color base off cancers type
key.wgs_long$col = subcolor[key.wgs_long$cancer.subtype ]

key.wgs_long = key.wgs_long[ order ( key.wgs_long$cancer.subtype, key.wgs_long$patient.id), ]
key.wgs_long$patient.id = factor ( key.wgs_long$patient.id, levels= unique ( key.wgs_long$patient.id )) 


not_in_rna = setdiff (key.wgs_long$WGS.id, temp$WGS.id )
temp2 = key.wgs_long[key.wgs_long$WGS.id %in% not_in_rna, ]

temp3 = rbind ( temp, temp2)



temp3 = temp3[ order ( temp3$cancer.subtype, temp3$patient.id), ]
temp3$patient.id = factor ( temp3$patient.id, levels= unique ( temp3$patient.id )) 

temp3$temporal.order = as.character ( temp3$temporal.order )


temp3$label  = ifelse ( temp3$label == "R" & temp3$WGS.id != "None", "R/W",  temp3$label   )



multi_g1 <- ggplot(temp3, aes(x = temporal.order, y=1 )) + 
   geom_col_pattern( aes ( pattern   =sample.classification, fill = cancer.subtype, pattern_density =sample.classification  ) , 
                     stat = 'identity', position = 'stack'
                     ,  colour = "white" # color between the plots 
                     , pattern_colour  = '#666262' # color of pattern
                   ) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  # note I can change fill color as usual, however this really messes with the legend
  scale_fill_manual(values =  unique ( temp3$col )  ) +
  scale_pattern_manual(values= c( "NonPrimary" = "stripe","Primary"="circle") ) +
  # change density to white will make it solid 
  scale_pattern_density_manual(values= c( "NonPrimary" = .01,"Primary"= 0  ) ) + 
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        , axis.ticks.y=element_blank()
        , axis.text.y=element_blank()
  ) + facet_wrap(~ patient.id, scales = "free" ,  ncol = 10 ) +
  theme(legend.position="none") + theme(legend.key.size = unit(2, 'cm')    ) +
  geom_text(data=temp3,aes(x=temporal.order, label=label), position = position_stack(vjust = 0.1) , colour="black") 


# r & w will not be included
#rna.key$label = "R"
#wgs.key$label = "W"
#temp3 = temp3[ !duplicated ( temp3$WGS.id), ] 
temp4 = temp3

cname = c( "patient.id" ,"temporal.order"   ,"cancer.subtype" , "sample.classification" ,"treatment"  , "WGS.id", "RNAseq.id", "RNAseq.exclude" )

wgs.key2 = wgs.key [ !wgs.key$RNAseq.id %in% rna.key$RNAseq.id,  cname ]
wgs.key2  = rbind ( rna.key[ , cname ], wgs.key2[ , cname ])
wgs.key2$RNAseq.id = ifelse ( wgs.key2$RNAseq.exclude != "PASS", "None", wgs.key2$RNAseq.id )
dups =wgs.key2$WGS.id[ duplicated ( wgs.key2$WGS.id)]
dups =dups[dups!="None"]
dups = wgs.key2[wgs.key2$WGS.id %in% dups, ]
dups$RNAseq.id2 = ifelse( dups$RNAseq.id == "None", "123", dups$RNAseq.id  )
dups = dups [ order ( dups$RNAseq.id2 , decreasing = T), ]
dups = dups[ !duplicated ( dups$WGS.id), ]
dups$RNAseq.id2 = NULL 
# remove dups 
wgs.key2=rbind ( wgs.key2[!wgs.key2$WGS.id %in% dups$WGS.id, ], dups) 
#wgs.key2[ duplicated ( wgs.key2$WGS.id), ]


temp5 =wgs.key2
temp5$label = ''
for ( row in 1:nrow ( wgs.key2) ){
r = wgs.key2[ row, ]  

#rna_1 = rna.key[ rna.key$patient.id == r$patient.id & rna.key$temporal.order == r$temporal.order, ]
#wgs_1 = wgs.key2[ wgs.key2$patient.id == r$patient.id & wgs.key2$temporal.order == r$temporal.order, ]

if ( r$RNAseq.id != "None" & r$WGS.id != "None" ){
  temp5[row, ]$label = "R/W"
} else if (  r$RNAseq.id != "None" & r$WGS.id == "None"   ){
  temp5[row, ]$label = "R"
} else if ( r$RNAseq.id == "None" & r$WGS.id != "None" ){
  temp5[row, ]$label = "W"
}

}

temp5$label_treat = ifelse ( temp5$treatment == "Treated", "*", 
                             ifelse ( temp5$treatment == "Naive", "", "^")
                             )
temp4 = temp5 
# we are going to partition these into W only R only and RW 
rw = data.frame()
w_only = data.frame()
r_only = data.frame()
# 
temp4$patient.id = as.character(temp4$patient.id )
for ( p in rev ( unique ( as.character ( temp4$patient.id ) ) ) ){
  p2 = temp4[ temp4$patient.id == p, ]
  
  rwt  = p2 [ grepl ( "R/W", p2$label), ]
  w =  p2 [ grepl ( "W$", p2$label), ]
  r = p2 [ grepl ( "^R", p2$label), ]
  
  if ( nrow ( rwt) >=2 ){
    rw = rbind ( rw, p2)
 
    next; 
  } else if ( nrow ( r) >=2 ){
    r_only = rbind ( r_only, p2)
    next; 
  }else if ( nrow ( w) >=2 ){
    w_only = rbind ( w_only, p2)
    next;
  }

}


rw = rw[ order ( rw$cancer.subtype, rw$patient.id), ]
rw$patient.id = factor ( rw$patient.id, levels= unique ( rw$patient.id )) 
rw_only = rw 


w_only = w_only[ order ( w_only$cancer.subtype, w_only$patient.id), ]
w_only$patient.id = factor ( w_only$patient.id, levels= unique ( w_only$patient.id )) 

r_only = r_only[ order ( r_only$cancer.subtype, r_only$patient.id), ]
r_only$patient.id = factor ( r_only$patient.id, levels= unique ( r_only$patient.id )) 


rw_only_plot =  ggplot(rw, aes(x = temporal.order, y=1 )) + 
   geom_col_pattern( aes ( pattern   =sample.classification, fill = cancer.subtype, pattern_density =sample.classification  ) , 
                     stat = 'identity', position = 'stack'
                     ,  colour = "white" # color between the plots 
                     , pattern_colour  = '#666262' # color of pattern
                   ) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  # note I can change fill color as usual, however this really messes with the legend
  scale_fill_manual(values =  subcolor ) +
  scale_pattern_manual(values= c( "NonPrimary" = "stripe","Primary"="circle") ) +
  # change density to white will make it solid 
  scale_pattern_density_manual(values= c( "NonPrimary" = .01,"Primary"= 0  ) ) + 
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        , axis.ticks.y=element_blank()
        , axis.text.y=element_blank()
        , plot.title = element_text(hjust = 0.5, size=30)
  ) + facet_wrap(~ patient.id, scales = "free" ,  ncol = 5 ) +
  theme(legend.position="none") + theme(legend.key.size = unit(2, 'cm')    ) +
  theme(strip.text.x = element_text(size = 20)) +
  geom_text(data=rw,aes(x=temporal.order, label=label), position = position_stack(vjust = 0.5) , colour="black", size=6) +
  geom_text(data=rw,aes(x=temporal.order, label=label_treat), position = position_stack(vjust = 0.7) , colour="black", size=13) +
  labs(title="RNA + WGS")




w_only_plot =  ggplot(w_only, aes(x = temporal.order, y=1 )) + 
   geom_col_pattern( aes ( pattern   =sample.classification, fill = cancer.subtype, pattern_density =sample.classification  ) , 
                     stat = 'identity', position = 'stack'
                     ,  colour = "white" # color between the plots 
                     , pattern_colour  = '#666262' # color of pattern
                   ) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  # note I can change fill color as usual, however this really messes with the legend
  scale_fill_manual(values =  subcolor ) +
  scale_pattern_manual(values= c( "NonPrimary" = "stripe","Primary"="circle") ) +
  # change density to white will make it solid 
  scale_pattern_density_manual(values= c( "NonPrimary" = .01,"Primary"= 0  ) ) + 
    theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        , axis.ticks.y=element_blank()
        , axis.text.y=element_blank()
        , plot.title = element_text(hjust = 0.5, size=30)
  )  + facet_wrap(~ patient.id, scales = "free" ,  ncol = 7 ) +
  theme(legend.position="none") + theme(legend.key.size = unit(2, 'cm')    ) +
  theme(strip.text.x = element_text(size = 20)) +
  geom_text(data=w_only,aes(x=temporal.order, label=label), position = position_stack(vjust = 0.5) , colour="black", size=7)  +
  geom_text(data=w_only,aes(x=temporal.order, label=label_treat), position = position_stack(vjust = 0.7) , colour="black", size=13) +
  labs(title="WGS ONLY")


r_only_plot =  ggplot(r_only, aes(x = temporal.order, y=1 )) + 
   geom_col_pattern( aes ( pattern   =sample.classification, fill = cancer.subtype, pattern_density =sample.classification  ) , 
                     stat = 'identity', position = 'stack'
                     ,  colour = "white" # color between the plots 
                     , pattern_colour  = '#666262' # color of pattern
                   ) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  # note I can change fill color as usual, however this really messes with the legend
  scale_fill_manual(values =  subcolor  ) +
  scale_pattern_manual(values= c( "NonPrimary" = "stripe","Primary"="circle") ) +
  # change density to white will make it solid 
  scale_pattern_density_manual(values= c( "NonPrimary" = .01,"Primary"= 0  ) ) + 
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        , axis.ticks.y=element_blank()
        , axis.text.y=element_blank()
        , plot.title = element_text(hjust = 0.5, size=30)
  )+ facet_wrap(~ patient.id, scales = "free" ,  ncol = 5 ) +
  theme(legend.position="none") + theme(legend.key.size = unit(2, 'cm')    ) +
  theme(strip.text.x = element_text(size = 20)) +
  geom_text(data=r_only,aes(x=temporal.order, label=label), position = position_stack(vjust = 0.5) , colour="black", size=10)  +
  geom_text(data=r_only,aes(x=temporal.order, label=label_treat), position = position_stack(vjust = 0.7) , colour="black", size=13) +
  labs(title="RNA ONLY")







temp_r = r_only[!duplicated ( r_only$patient.id), ]
temp_r = as.data.frame ( 
      table ( c( temp_r$cancer.subtype )  ), stringAsFactor=F
  )
colnames(temp_r) = c("cancer.subtype", "RNA only")

temp_rw = rw_only[!duplicated ( rw_only$patient.id), ]
temp_rw = as.data.frame ( 
      table ( c( temp_rw$cancer.subtype )  ), stringAsFactor=F
  )
colnames(temp_rw) = c("cancer.subtype", "RNA+WGS")

temp_w = w_only[!duplicated ( w_only$patient.id), ]
temp_w = as.data.frame ( 
      table ( c( temp_w$cancer.subtype )  ), stringAsFactor=F
  )
colnames(temp_w) = c("cancer.subtype", "WGS only")

### make pies 
colnames(temp_rw) = c("cancer.subtype", "Freq")
pie.rw= make.pie( temp_rw
                , name.first= 'cancer.subtype'
                , cc = subcolor
                , leg.pos = "right"
                , title = "RNA+WGS"
                )

colnames(temp_rw) = c("cancer.subtype", "RNA+WGS")

####


temp_rw = merge ( temp_rw, temp_r, all=T)
temp_rw = merge ( temp_rw, temp_w, all=T)
temp_rw[is.na(temp_rw)] <-0
temp_rw = temp_rw[ order ( temp_rw$`RNA+WGS`, decreasing = T), ]
temp_rw$cancer.subtype = as.character ( temp_rw$cancer.subtype)

tally = rbind ( temp_rw, c ( "TOTAL", sum ( temp_rw$`RNA+WGS`), sum ( temp_rw$`RNA only`), sum ( temp_rw$`WGS only`) ) )

tall_multi = tally 

kable( tall_multi  , format = "html" , row.names = F, caption="multi.sample" ) %>% kable_classic(full_width = F, position = "center", ) %>% row_spec(row= nrow ( tally), bold = T, color = "black", background = "#c6c9cf")

multi_t1 <- ggtexttable(tall_multi, rows = NULL, theme = ttheme("blank")) %>%
  tab_add_hline(at.row = c( 1:2, nrow ( tall_multi) + 1) , row.side = "top", linewidth = 2) %>% table_cell_bg( row = nrow ( tall_multi)+1, column = 1:ncol(tall_multi), linewidth = 5, fill="#d7d7d9", color = "#d7d7d9") %>%
  table_cell_font(row= c(1:nrow(tall_multi)+1 ), column=1:ncol(tall_multi),
                size = 16.5 )


hold_table  = paste("\n   place holder for a table insertion.\n",
         "       There will be a  table will be here \n",
         "       ")

hold_table = ggplot() + 
  annotate("text", x = 4, y = 25, size=8, label = hold_table) + 
  theme_void()



# tally 





fusion.main = fusion 


# make fake lengend 

temp2 = temp
temp2$sample.classification = ifelse ( temp2$sample.classification  == "NonPrimary", "Metastasis",temp2$sample.classification )

temp2$sample.classification = factor ( temp2$sample.classification, levels = c("Primary","Metastasis"))
d_legend=  ggplot(head ( temp2 ) , aes( sample.classification, temporal.order)) +
   geom_col_pattern(
     aes(pattern = sample.classification, fill = sample.classification, pattern_fill = sample.classification , pattern_density = sample.classification ),
     colour                   = 'black') +
   theme_bw() +
   labs(
     title    = "Longitudinal Samples"
   ) +
   scale_pattern_fill_manual(values = c(a='blue', b='red', c='grey')) + # the color of the dots
   scale_fill_manual( values=c ("white", "white" ) ) +
     scale_pattern_manual(values= c( "Metastasis" = "stripe","Primary"="circle") ) +
  scale_pattern_density_manual(values= c( "Metastasis" = .01,"Primary"= 0  ) ) + 
   theme(legend.position = 'bottom') +
   coord_fixed(ratio = 1) + theme(legend.key.size = unit(2, 'cm'))
 



```

```{r fig=TRUE,fig.width=16, fig.height=18, echo=FALSE, include=TRUE}


leg <- get_legend(d_legend)
# Convert to a ggplot and print
leg = as_ggplot(leg)


laym = "
AAAAAAAABBBBBB
AAAAAAAACCCCCC
AAAAAAAADDD###
"

rw_only_plot + w_only_plot+r_only_plot+ multi_t1+ plot_layout(design = laym)




addWorksheet(paper_table, 'mult_sample')
#writeData(wb, 'QC' , df  , rowNames=T  )
# writeData(paper_table, 'QC', )
writeData(paper_table
          ,'mult_sample'
          ,tall_multi
          , rowNames=F  )





```



```{r fig=TRUE,fig.width=5, fig.height=3, echo=FALSE, include=TRUE}
leg
multi_legend = leg 
```


```{r}
# fromal tabualtion - don't really know what the bottom of this is for yet. 
# how many samples have multi-regional
# how many have multi-longitudnal. 
# of those above how many have just R, W or both ( at least 2 pr ,pre )
# this is not finished and will just leave it to manaul coutning, too complicated! 

t=multi_g1$data

for ( i in t$patient.id ){
  t2 = t[t$patient.id == i, ]
  # is there longitudinal ? 
  tunique = unique ( t2$temporal.order)
  long = "no"
    if ( length ( tunique ) >= 2 ){
      long = "yes"
      longdf = data.frame ( table ( t2[t2$temporal.order == tunique, ]$label))
    }
  ml = "no"
  for ( m in tunique){
    if ( nrow ( t[t$patient.id == i & t$temporal.order == m , ] ) >=2 ){
      ml = "yes"
    }
  }

  
  }


```




```{r}


# tablulations. 

t=multi_g1$data
length ( unique ( t$patient.id ))
length ( unique ( t$cancer.subtype))
# how many samples are longitudnal 
t$match1 = paste ( t$patient.id, t$temporal.order)
t2 = t[!duplicated ( t$match1), ]
dup = t2$patient.id [ duplicated ( t2$patient.id) ]
length ( unique ( dup ) )


howmanylong = 0 
howmanyspatial = 0

howmanylong_rw = 0 
howmanylong_ro = 0
howmanylong_wo = 0 

howmanyspatial_rw = 0 
howmanyspatial_ro = 0
howmanyspatial_wo = 0 



for ( n in unique ( t$patient.id)){
  
  df = t[t$patient.id == n, ]
  rsample = unique ( df$temporal.order)
  if ( length ( rsample ) > 1 ){
    howmanylong = howmanylong + 1 
    if ( length ( df$label [ df$label == "R/W"] ) > 1 ){
      howmanylong_rw = howmanylong_rw + 1 
    } else if ( length ( df$label [  grepl ( "R", df$label ) ] ) > 1){
      howmanylong_ro = howmanylong_ro + 1 
    }  else if ( length ( df$label [  grepl ( "W", df$label ) ] ) > 1){
      howmanylong_wo = howmanylong_wo + 1 
    }
    
  }else if ( length ( rsample) == 1 ){
    howmanyspatial = howmanyspatial +1 
    
    if ( length ( df$label [ df$label == "R/W"] ) > 1 ){
      howmanyspatial_rw = howmanyspatial_rw + 1 
    } else if ( length ( df$label [  grepl ( "R", df$label ) ] ) > 1){
      howmanyspatial_ro = howmanyspatial_ro + 1 
    }  else if ( length ( df$label [  grepl ( "W", df$label ) ] ) > 1){
      howmanyspatial_wo = howmanyspatial_wo + 1 
    }
    
    
  }
}

# sanity checks. 

howmanylong +  howmanyspatial == length ( unique ( t$patient.id) )


howmanylong_rw +
howmanylong_ro +
howmanylong_wo + 
howmanyspatial_rw + 
howmanyspatial_ro +
howmanyspatial_wo  == length ( unique ( t$patient.id) )


ls_df = rbind ( data.frame ( 
  total = c ( howmanylong_rw , 
howmanylong_ro ,
howmanylong_wo ),
modality = c("RNAseq + WGS", "RNAseq", "WGS"),
type=rep("longitudinal", 3) 
  )


, data.frame ( 
  total = c ( howmanyspatial_rw , 
howmanyspatial_ro ,
howmanyspatial_wo ), 
modality = c("RNAseq + WGS", "RNAseq", "WGS"),
type=rep("spatial", 3) 
  
  ) ) 


set.seed(2)
sample ( getPalette(30), 5 )
tree.long.tab = ggplot(ls_df, aes(area = total, fill = modality, label = paste (  total ) , subgroup = type, subgroup2 = modality)) +
    geom_treemap(position = "identity" ) + 
    geom_treemap_text(colour = "black", place = "middle", min.size=15, grow=F, reflow = T, size= 40, angle=0
                      
                      ) +
    geom_treemap_subgroup_text(place = "bottomleft", reflow = T, alpha = 1, colour =
                                   "white", fontface = "italic", min.size = 20,  angle=0
                               , size= 35
                               , padding.y = grid::unit(10, "mm"), padding.x = grid::unit(5, "mm")
                               ) +
    geom_treemap_subgroup2_border(colour = "grey", size = 3) +
    geom_treemap_subgroup_border(colour = "white", size = 15 ) + scale_fill_manual(values = sample ( getPalette(30), 5 ) ) +
    theme(legend.position="bottom") 





```


## Sample Clustering {.tabset}


```{r}

### this is preparing for the hclust
# hclust chunk 1
# series of necessary functions

library(jsonlite)
get.onco = 0 
# there are a few API but the one below will download the latest version of the entire set and we can query locally

if ( get.onco == 1){
    onco<- fromJSON(file("http://oncotree.mskcc.org/api/tumorTypes?version=oncotree_latest_stable"))
    onco2<- flatten(onco)
    saveRDS(onco2, "oncotree.rds")
}else{
    onco2 = readRDS("oncotree.rds")
}



# write function to get lineage 
# this function will take an input "code" and find the the entry, then it would recursively go up each node until until it hits the the main node denoted ast "parent" == TISSUE 

getoncotree = function ( code, onco2, df=NA ){
    if (is.na(df)){ df = data.frame()}
    x = onco2[onco2$code == code, ]
    p = x$parent
    if ( p == "TISSUE"){
        df = rbind ( x, df)
        return ( df )
    }else{
        code = x$parent
        df = rbind ( x, df)
        getoncotree(code, onco2,df)
    }
}
oc = getoncotree(code="AML", onco2)
oc[ , c(1:3,5:6, 8)]
```



```{r}

# hclust chunk 2 

hr.key = rna.key [ rna.key$RNAlibrary.chemistry != "RNA Access" | rna.key$qc != "FAIL",  ]
rm = c("AJ6", "AD12")
hr.key = hr.key[!hr.key$RNAseq.id %in% rm,   ]

level = 1

# get parent sample for oncotree:  this is tissue type
hr.key$headlevel = as.character ( sapply(hr.key$oncotree, 
                                      function(x){
                                          #print (x)
                                          x = getoncotree(code=x, onco2)
                                          return ( as.character ( unique ( x[x$level == level, ]$code ) ) )
                                      }
)
)

level =2 

hr.key$level2 = as.character ( sapply(hr.key$oncotree, 
                                      function(x){
                                          #print (x)
                                          x = getoncotree(code=x, onco2)
                                          return ( as.character ( unique ( x[x$level == 2, ]$code ) ) )
                                      }
)
)

# set colors 
set.seed(123)

# these are the major colors 

colrs <-  unique ( c ( brewer.pal(8, "Set2"), brewer.pal(8, "Dark2") ,  
                       brewer.pal(9, "Pastel1"), brewer.pal(8, "Pastel2")
                       , brewer.pal(12, "Set3")
) )

# set parent color 
ncolrs <- length(unique( hr.key$headlevel )) 

cc <- setNames(as.list(   colrs[1:ncolrs]    ),  unique( hr.key$headlevel  ) )
hr.key$head.color <- as.character ( cc[ hr.key$headlevel  ] )
# set level two colors
ncolrs <- length(unique( hr.key$level2 )) 
cc <- setNames(as.list(  colrs[1:ncolrs]     ),  unique( hr.key$level2  ) )
hr.key$level2.color <- as.character ( cc[ hr.key$level2  ] )

# set oncotree color: the level in which it was assigned in the masterkey 

hr.key$cancer.subtype.color <- as.character ( 
  sapply ( hr.key$cancer.subtype, function(x)
    
    subcolor[x]
    
    )
  )


```

```{r}

# hclust chunk 3
# do clustering of diseases 


star = star[ , hr.key$RNAseq.id ]


y.match <- DGEList( star )
dim ( y.match )
## low counts have not been removed yet. 
threshold <- ncol(star)/3
keep <- rowSums( cpm(star)  > 0) >= threshold
y.match  <- y.match [keep, ]
cpm <- as.data.frame(cpm(y.match, prior.count=2, log=T))

all.equal( hr.key$RNAseq.id, colnames ( cpm ))

## end getting raw star 


```



```{r}

#library ( pals )
# get tpm 
cm = cpm 
sd <- apply (cm, 1,  function(x) cov(x)  )
cm$sd <- as.numeric ( sd )
# remove rows with all zeros which will result in NaN 
cm = cm [ !is.na(cm$sd), ]

### do a single HR you might need to run scan on the bottom to find the optimal HR 
p = c(.1,.2,.3,.4,.5,.6, .7, .8, .9,.95)
q <- quantile (cm$sd, probs = p )


cm <- cm[cm$sd > as.numeric ( q[9] ),]
nrow (cm)
cm$sd <- NULL

dim ( cm )


### create color annotations 

diag.type = data.frame ( table ( hr.key$cancer.subtype) )

head.color =  setNames( hr.key$head.color , hr.key$headlevel )
head.subcolor = setNames( hr.key$cancer.subtype.color, hr.key$cancer.subtype   )


annt.color = list()

for ( n in as.character ( diag.type$Var1) ){
  
  annt.color[[n]]=
    setNames( 
             ifelse ( names (  head.subcolor ) == n, head.subcolor, "#f0f1f5")
            , hr.key$cancer.subtype 
)

  }

annt.color[['class']] = setNames(  ifelse (
  
  hr.key$sample.classification == "NonPrimary", "black", ifelse(
    hr.key$sample.classification == "Primary", "grey", "#f0f2aa"
  )
  
), hr.key$RNAseq.id )



annt.color[['treatment']] = setNames(  ifelse (
  
  hr.key$treatment == "Treated", "#ebab34", ifelse(
    hr.key$treatment == "Naive", "#dfedb9", "#f0f2aa"
  )
  
), hr.key$RNAseq.id )





# type 
dotype = 0 
if ( dotype == 1){
unique ( rna.key$sample.type)

stype = hr.key$sample.type
stype = gsub ( ".*_", "", stype)



annt.color[['type']] = setNames(  ifelse (
  
  stype== "Biopsy", "green", ifelse(
    stype== "Metastasis", "red", "#f0f2aa"
  )
  
), hr.key$RNAseq.id )

}



#immuno_annt = rowAnnotation (purity = est_annt$purity) 

set.seed(1)

topannt = HeatmapAnnotation( 
                              ALL = hr.key$cancer.subtype
                              ,AML =hr.key$cancer.subtype
                              ,"Other Heme" =hr.key$cancer.subtype
                             ,OS = hr.key$cancer.subtype
                             , EWS = hr.key$cancer.subtype
                            
                                 
                             
                              ,"HBL"  = hr.key$cancer.subtype      
                              ,"HCC" = hr.key$cancer.subtype
                             ,"WILMS"= hr.key$cancer.subtype
                             
                              , NRSTS = hr.key$cancer.subtype
                              ,"FPRMS"  = hr.key$cancer.subtype 
                             ,"FNRMS"  = hr.key$cancer.subtype 
                              ,"NBL"= hr.key$cancer.subtype
                             , "CNS"  = hr.key$cancer.subtype
                             , "R/O"   = hr.key$cancer.subtype 
                            # , "class" = hr.key$RNAseq.id
                             , "class" = hr.key$RNAseq.id
                             , "treatment"= hr.key$RNAseq.id
                            , "purity" = hr.key$purity # since this is contibous it is not necessary
                             ,col =  annt.color 
                            , show_legend =  F
) 


km2 = 10 
dend2 = makehr ( t( cm ), km=km2
                 , dist.this = "geodesic", aggreg= "ward.D2", meta1 = hr.key$cancer.subtype, colorbranch = 0 )    
dend2$dend = dend2$dend %>% set("labels_cex", 0) %>% set("labels_col", "white")


dend_gene = makehr ( as.matrix ( cm) , km=km2
                 , dist.this = "geodesic", aggreg= "ward.D2", meta1 = NA, colorbranch = 0 ) 



par(mar=c(0,0,0,15))

cm2 = scale ( data.matrix(cm) ) 
cm2 = cm 
cm2[cm2 > 10] = 10 
#cm2[cm2 < -5]= -5 



hmmain = ComplexHeatmap::Heatmap(  data.matrix(cm2)  
           ,  name = "Sample Exp"
           , show_row_dend = T
           , show_column_names = F
           , show_row_names = F
           #, name= heat.title
           ,column_split = km2
           ,top_annotation = topannt
           #, col = pals::brewer.rdylbu(25) 
           , col =  colorRampPalette(rev(brewer.pal(n = 7, name ="RdYlBu")))(25)
           , cluster_columns = sort ( dend2$dend , decreasing = T)
           #, column_split = 2
            , cluster_rows  = dend_gene$dend 
           , row_split = 7
           ,row_title = NULL
           #,column_title = NULL
         ,show_heatmap_legend = T
         , row_names_gp = grid::gpar(fontsize = 8) 
)



type.color = annt.color[['class']] 
names ( type.color) = as.character ( sapply ( names ( type.color) , function (x) hr.key[hr.key$RNAseq.id == x, ]$sample.classification) )
type.color2 =  unique ( type.color ) 
names ( type.color2 ) = unique ( names (type.color ))

treat.color = annt.color[['treatment']] 
names ( treat.color) = as.character ( sapply ( names ( treat.color) , function (x) hr.key[hr.key$RNAseq.id == x, ]$treatment) )

treat.color2 =  unique ( treat.color ) 
names ( treat.color2 ) = unique ( names (treat.color ))

# purity

l.type = Legend(labels = names ( type.color2 ), title = "CLASS", legend_gp = gpar(fill = type.color2,fontsize = 12 ),
                title_position = "leftcenter-rot"
                , labels_gp = gpar(fontsize = 12)
                , title_gp = gpar(col = "black", fontsize = 12)
                ,grid_height = unit(8, "mm"), grid_width = unit(8, "mm"))


l.treat = Legend(labels = names ( treat.color2 ), title = "TREATMENT", legend_gp = gpar(fill = treat.color2,fontsize = 12 ),
                title_position = "leftcenter-rot"
                , labels_gp = gpar(fontsize = 12)
                , title_gp = gpar(col = "black", fontsize = 12)
                ,grid_height = unit(8, "mm"), grid_width = unit(8, "mm"))



pd = packLegend(l.type, l.treat, row_gap = unit(1, "cm") , direction="vertical")

#draw(hmmain,annotation_legend_list = pd)

```

### hclust

```{r fig=TRUE,fig.width=16, fig.height=10, echo=FALSE, include=TRUE}

figs_out = list()
draw(hmmain,annotation_legend_list = pd, merge_legend = TRUE   )

figs_out[['heatmap1']] =hmmain
figs_out[['heatmap1_leg']]=pd


```

```{r}

# DEG hclust chunk 
cluster.info = column_order(hmmain) 
total.cluster = length ( cluster.info )
cluster.df = data.frame()

for (n in  1:total.cluster){
  samples = names ( cm2) [ column_order(hmmain)[[n]] ] 
  cluster.name = rep ( paste0("C",n) , length ( samples ))
  cluster.df = rbind ( cluster.df
          , data.frame(cluster=cluster.name 
          , RNAseq.id= samples  ) 
          )
}



# this will now set up the analysis for DEG either with most of the genes are just cm2 
# reorder the clusters to match the y.match 
cluster.df = cluster.df[match( colnames ( y.match$counts ) , cluster.df$RNAseq.id ),]

sanity_c = merge ( cluster.df, rna.key[ , c("RNAseq.id", "cancer.subtype")], by.x = "RNAseq.id", by.y="RNAseq.id")
sanity_c = sanity_c[ order ( sanity_c$cluster), ]


# save to excel 

wb <- createWorkbook()

addWorksheet(wb, 'clusterInfo')
writeData(wb, 'clusterInfo' , df  , rowNames=T  )

saveWorkbook(wb, file = paste0(out.dir,"cluster.xlsx"), overwrite = TRUE)

##################


# 


all.equal( colnames ( y.match$counts), cluster.df$RNAseq.id)

design.plate <- model.matrix(~0  +  cluster.df$cluster )
colnames(design.plate) = gsub ( "cluster.df\\$cluster", "", colnames ( design.plate) )

# sanity check to make sure the cluster 10 corresonds to the actual samples! 
# this assumes that cluster 10 is where most of the liver cancers are
rna.key [ rna.key$RNAseq.id  %in% names ( cm2 ) [ which ( design.plate [ , 2] == 1 )] , ]$cancer.subtype


voom <- voom(y.match , plot = T, design = design.plate)  


fit  <- lmFit( voom, design.plate  )

# contrast.matrix is difficult because you would need to compare 1 to many so you would need to create one first 
cname = colnames(design.plate)
formt = c()
for ( c in cname ){
  all = cname[cname!=c]
  allt = length ( all )
  form = paste ( c, "=", c , "-(", paste ( all , collapse="+"), ")/", allt )
  formt = c( formt, form )
  
}

  cat ( formt,  sep="\n,")

contrast.matrix = makeContrasts( 
 C1 = C1 -( C10+C2+C3+C4+C5+C6+C7+C8+C9 )/ 9
,C10 = C10 -( C1+C2+C3+C4+C5+C6+C7+C8+C9 )/ 9
,C2 = C2 -( C1+C10+C3+C4+C5+C6+C7+C8+C9 )/ 9
,C3 = C3 -( C1+C10+C2+C4+C5+C6+C7+C8+C9 )/ 9
,C4 = C4 -( C1+C10+C2+C3+C5+C6+C7+C8+C9 )/ 9
,C5 = C5 -( C1+C10+C2+C3+C4+C6+C7+C8+C9 )/ 9
,C6 = C6 -( C1+C10+C2+C3+C4+C5+C7+C8+C9 )/ 9
,C7 = C7 -( C1+C10+C2+C3+C4+C5+C6+C8+C9 )/ 9
,C8 = C8 -( C1+C10+C2+C3+C4+C5+C6+C7+C9 )/ 9
,C9 = C9 -( C1+C10+C2+C3+C4+C5+C6+C7+C8 )/ 9
#,C7_C4 = C7-C4
#,C7_C8 = C7-C8
  ,    levels=design.plate
)


#cname = unique ( c( cname, "C7_C4", "C7_C8"))

fit  <- lmFit(voom, design.plate )

fit <- contrasts.fit(fit, contrast.matrix)
fit <- eBayes(fit)

dg.voom = list ()

for ( g in cname ){
ALL <- topTable(fit, coef=g, n=Inf )
ALL = ALL [ ALL$adj.P.Val < .05, ]
ALL = ALL [ order ( abs(ALL$logFC), decreasing = T), ]
dg.voom[[g]]= ALL
}
```

```{r}




sarcoma = rna.key[ grepl ( "sarcoma", rna.key$diagnosis), ]


cm_sarcoma = cm [ , sarcoma$RNAseq.id]

pca.sarcoma <- PCA( t( cm_sarcoma ), graph = FALSE)

pca.sarcoma <- fviz_pca_ind(pca.sarcoma, label="none", addEllipses=TRUE, ellipse.level=0.95,  axes = c(1, 2),  alpha=0  )
pca.sarcoma <- pca.sarcoma +geom_point( aes(colour=  sarcoma$cancer.subtype, shape=factor ( sarcoma$treatment)  ), size=9.0 , alpha=.4 ) + scale_colour_manual(values=subcolor)

 pca.sarcoma  + 
scale_colour_discrete(name  ="cancer subtype") + scale_shape_discrete ( name="Treatment") + ggtitle("Unsupervised") 

 
 

library("Rtsne")
tsne_out <- Rtsne(t( cm_sarcoma ), check_duplicates=FALSE) # Run TSNE
# Show the objects in the 2D tsne representation

df <- data.frame(x = tsne_out$Y[,1],
                 y = tsne_out$Y[,2],
                    cancer.subtype = sarcoma$cancer.subtype, 
                 treatment = sarcoma$treatment
                 
                 )

ggplot(df, aes(x, y, colour = cancer.subtype, shape=treatment )) +
  geom_point(size=7, alpha=.8) +
     theme_minimal()   + xlab("tSNE 1") + ylab ( "tSNE 2 ") +
  theme(legend.position="right",  legend.key = element_blank(),
          
          axis.text.y = element_text(size= 15 ),
          #axis.text.x = element_blank(),
          axis.title.x = element_text(size=20),
          
          axis.title.y     = element_text(size=20), 
          legend.text      =element_text(size=12)
    ) + scale_colour_manual(values=subcolor)




library(umap)

cpmlog2_2 = as.data.frame(cpm(y.match, prior.count=2, log=F))
cpmlog2_2 = log2 ( cpmlog2_2 + 1 )
cm_sarcoma = cpmlog2_2[ , sarcoma$RNAseq.id]

stype = sarcoma$sample.type
stype = gsub ( ".*_", "", stype)


sarcoma.umap = umap(t( cm_sarcoma   ))

#plot.umap(sarcoma.umap, sarcoma$cancer.subtype, colors = subcolor, cex=3)


df <- data.frame(x = sarcoma.umap$layout[,1],
                 y = sarcoma.umap$layout[,2],
                 cancer.subtype = sarcoma$cancer.subtype, 
                 type = stype
                 
                 )

ggplot(df, aes(x, y, colour = cancer.subtype, shape=type )) +
  geom_point(size=7, alpha=.8) +
     theme_minimal()   + xlab("UMAP 1") + ylab ( "UMAP 2 ") +
  theme(legend.position="right",  legend.key = element_blank(),
          
          axis.text.y = element_text(size= 15 ),
          #axis.text.x = element_blank(),
          axis.title.x = element_text(size=20),
          
          axis.title.y     = element_text(size=20), 
          legend.text      =element_text(size=12)
    ) + scale_colour_manual(values=subcolor)



## plot all 

rkey2 = rna.key[ rna.key$cancer.type == "Solid",    ]

stype = rkey2$sample.type
stype = gsub ( ".*_", "", stype)

rna.umap = umap(t( cpmlog2_2[ , rkey2$RNAseq.id]   ))

#plot.umap(rna.umap, rna$cancer.subtype, colors = subcolor, cex=3)


df <- data.frame(x = rna.umap$layout[,1],
                 y = rna.umap$layout[,2],
                 cancer.subtype = rkey2$cancer.subtype, 
                 type = stype
                 
)

ggplot(df, aes(x, y, colour = cancer.subtype, shape=type )) +
  geom_point(size=7, alpha=.8) +
  theme_minimal()   + xlab("UMAP 1") + ylab ( "UMAP 2 ") +
  theme(legend.position="right",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 15 ),
        #axis.text.x = element_blank(),
        axis.title.x = element_text(size=20),
        
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12)
  ) + scale_colour_manual(values=subcolor)






```


```{r}
# get function for GSEA 
# gsea chunk 

pathgsea =  readRDS( paste0( resource, "/gsea/limma/go_path_fgsea_genename.rds"))
ens = readRDS( paste0( resource, "annotation/gene.detail.rds"))


dsig_d1 = qusage::read.gmt( paste0 (  resource,  "drug/Dsigdb/D1_FDA.gmt")  ) 
names ( dsig_d1 ) = paste0 ( "FDA_" , names ( dsig_d1 ))

dsig_d2 = qusage::read.gmt( paste0 (  resource,  "drug/Dsigdb/D2.gmt")  ) 
names ( dsig_d2 ) = paste0 ( "KINASE_" , names ( dsig_d2 ))

dsig_d3 = qusage::read.gmt( paste0 (  resource,  "drug/Dsigdb/D3_CMAP.gmt")  ) 
names ( dsig_d3 ) = paste0 ( "CMAP_" , names ( dsig_d3 ))
dsig_d3 = dsig_d3[ grepl ( "_DOWN", names ( dsig_d3))]

pathgsea = c ( pathgsea, dsig_d1 )
pathgsea = c ( pathgsea, dsig_d2 )
pathgsea = c ( pathgsea, dsig_d3 )


        

cluster.df = merge ( cluster.df, rna.key[ , c("cancer.subtype", "RNAseq.id")],  by="RNAseq.id")

  
library(fgsea)



do.gsea4 = function ( gfold, gname, pathways.in, db=NA, entrez="entrezgene_id" , genename = "external_gene_name", 
                      ss=123, fdr= .05 ,  custom = ''){
  
  
  set.seed (ss)
  g.this2 = gfold # grab logfc
  names ( g.this2 ) = gname # associate with name of gene 
  g.this2 = sort ( g.this2, decreasing = T ) # probably not necessary but I do it anyway
  
  
  
  gsea.result <- fgseaMultilevel(pathways.in , g.this2, minSize=15, maxSize=50000 )
  #gsea.result <- fgsea (stats=g.this2, pathways=pathways.in , minSize=2, maxSize=Inf)
  
  gsea.result=gsea.result[!is.na(gsea.result$padj), ]
  gsea.result = gsea.result[ order ( gsea.result$pval), ]
  
  prob.gsea = as.numeric ( quantile ( gsea.result$padj, probs=(.05)))  
  
  pstandard = .05 
  if ( prob.gsea > fdr & prob.gsea < 1 ){
    fdr = prob.gsea
    pstandard = .01
  }
  
  raw.result = gsea.result
  raw.result = raw.result[ order ( raw.result$pval), ]
  
  
  gsea.result = raw.result[ raw.result$padj <= fdr & raw.result$pval < pstandard , ] 
  gsea.result = gsea.result[ order ( gsea.result$pval), ]
  
  
  gsea.result$leadingEdge  <- 
    vapply(gsea.result$leadingEdge, function(.x) {
      y <- unlist(.x, use.names = TRUE)
      paste(names(y), y, sep = '', collapse = ', ')
    }, FUN.VALUE = character(1))
  
  
  
  
  
  gsea.result$Up_regulated =unlist ( gsea.result$leadingEdge )
  gsea.result$Down_regulated =""
  
  
  
  
  
  #### 
  #### 
  
  
  do.this = unique ( c ( custom, c("KEGG", "HALLMARK", "REACT", "BIOCA", "BP", "MF", "CC", "CELL")) )
  
  gsea_plots = list ()
  top_up = data.frame ()
  top_down = data.frame()
  
  for ( dt in do.this){
  
  custom.gsea = gsea.result[ grepl(paste0("^", dt), gsea.result$pathway), ]
  gsea_plots [[dt]] = plot.gsea( dfg= custom.gsea , thres = 10, title='' )
  top_up = rbind (top_up,  gsea_plots [[dt]]$plot$data[ gsea_plots [[dt]]$plot$data$ES > 0 , c(1,6,3) ] ) 
  top_down = rbind (top_down,  gsea_plots [[dt]]$plot$data[ gsea_plots [[dt]]$plot$data$ES > 0 , c(1,6,3) ] )                 
  gsea_plots [[dt]] = gsea_plots [[dt]] $plot 
  }
  
  
  names ( gsea_plots )= paste0(names ( gsea_plots ), ".plot")
  
  # combine up and down 
  
  
  top_up$type = str_match(top_up$pathway, "(.*?)_")[, 2]
  top_down$type = str_match(top_down$pathway, "(.*?)_")[, 2]   
  
  
  
  # convert to gene.id 
  
  # instead of the grids I'm going to go with the other plots intead. 
  
  
  
  
  return ( c ( 
    
    gsea_plots, list ( 
    all_df = gsea.result,
    top_down = top_down,
    top_up = top_up, 
    fdr=fdr
    )
  ))
}

plot.gsea = function ( dfg, thres = 8 , title = "" ){
  
  dfg = dfg[ order ( dfg$padj, -abs ( dfg$NES) ), ]
  dfg.up  = dfg [ dfg$NES > 0, ]
  dfg.up = dfg.up[ order ( dfg.up$padj), ]
  dfg.down  = dfg [ dfg$NES < 0, ]
  dfg.down = dfg.down[ order ( dfg.down$padj), ]
  
  
  dfg = rbind ( head ( dfg.up, thres ) , head ( dfg.down, thres) )
  dfg = dfg[ order ( dfg$NES , decreasing = F), ]
  
  dfg$group = ifelse ( dfg$NES > 0, "up", "down")
  dfg$pathway = str_trunc ( as.character ( dfg$pathway) , 60 )
  
  dfg$pathway = factor ( dfg$pathway, levels =  unique ( dfg$pathway) )
  dfg$group = factor ( dfg$group , levels = c("up","down"))  
  
  g1 = ggplot(data=dfg, aes(x=NES, y=pathway, fill=group)) +
    geom_bar(stat="identity") + scale_fill_manual(values=c( "#8bb4c9", "#f2c830" ), name="NES direction") +
    theme(legend.position="right",  legend.key = element_blank(),
          
          axis.text.y = element_text(size=12),
          axis.text.x = element_text(angle = 90, size=12),
          axis.title.x = element_text(size=12),
          
          axis.title.y     = element_text(size=12), 
          legend.text      =element_text(size=12)
    ) + xlab("Enrichment Score") + ylab("") +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black")) + ggtitle ( paste( "GSEA: TOP positive/negative: ", title ) )
  
  
  dfgFULL = rbind ( dfg.up , dfg.down )
  dfgFULL = dfgFULL[ order ( dfgFULL$NES , decreasing = T), ]
  
  dfgFULL$group = ifelse ( dfgFULL$NES > 0, "up", "down")
  dfgFULL$pathway = str_trunc ( as.character ( dfgFULL$pathway) , 60 )
  
  
  return ( list ( plot=g1, df=dfgFULL ) )
  
}


# test run 
temp =   do.gsea4 (
      gfold = dg.voom[["C1"]] $logFC
      , gname =  row.names ( dg.voom[["C1"]]  ) 
        , pathways.in=pathgsea
       ,  db=ens
      )



```



```{r}

# GSEA chunck
redog = 1
set.seed(123)
if ( redog == 1 ){
dg.gsea= list ()
bubble_up = data.frame()

for ( d in cname ){
gsea = do.gsea4 (
      gfold = dg.voom[[d]] $logFC
      , gname =  row.names ( dg.voom[[d]]  ) 
        , pathways.in=pathgsea
       ,  db=ens, 
      custom = c ( "FDA", "KINASE", "CMAP") 
      )
dg.gsea[[d]]= gsea
gsea$top_up$group = d
bubble_up = rbind ( bubble_up,gsea$top_up )
}



saveRDS( list ( dg.gsea = dg.gsea, bubble_up= bubble_up   ), "hclustGSEA")

}else {
  dg.gsea = readRDS( "hclustGSEA")
  bubble_up = dg.gsea$bubble_up
  dg.gsea = dg.gsea$dg.gsea
}



reorder_within <- function(x, by, within, fun = mean, sep = "___", ...) {
  new_x <- paste(x, within, sep = sep)
  stats::reorder(new_x, by, FUN = fun)
}

scale_x_reordered <- function(..., sep = "___") {
  reg <- paste0(sep, ".+$")
  ggplot2::scale_x_discrete(labels = function(x) gsub(reg, "", x), ...)
}


```

### GSEA, plots {.tabset}


```{r fig=TRUE,fig.width=10, fig.height=6, echo=FALSE, include=TRUE, results='asis'}


ethis =   toupper( c( 'bp' , "mf", "hallmark" , "kegg", "REACTOME" , "BIOCARTA", "KINASE", "FDA", "CMAP"
) )


scale_this <- function(x){
  (x - mean(x, na.rm=TRUE)) / sd(x, na.rm=TRUE)
}


bar_path = list()
mx = 6 

for ( g in ethis ){

cat("#### ", g , "\n")
  
  bubble_pretty = bubble_up [ bubble_up$type == g, ]
  bubble_pretty = bubble_pretty [ ! grepl ("_", bubble_pretty$group ), ]
  
  bubble_pretty$pathway = as.character ( bubble_pretty$pathway )
  bubble_pretty$pathway = sub ( ".*?_", "", bubble_pretty$pathway )
  bubble_pretty$pathway = str_trunc( bubble_pretty$pathway , 35)
  bubble_pretty$logp = -log(bubble_pretty$padj) 

 
bubble_pretty2 <- bubble_pretty %>% group_by(group) %>%
  mutate(NES_s = scale_this(NES)   )  %>%  data.frame(stringsAsFactors = F)

bubble_pretty2 = bubble_pretty2[ order ( bubble_pretty2$NES_s), ]
bubble_pretty2$group = factor ( bubble_pretty2$group, levels = c( "C1", "C2", "C3", "C4", "C5", "C6", "C7", "C8", "C9", "C10" ))
bubble_pretty2$pathway = factor ( bubble_pretty2$pathway, levels = unique( bubble_pretty2$pathway )  )

g2 = ggplot( bubble_pretty2 , aes(x = group , y = pathway , colour = NES_s)) +
  geom_point(size=4) + 
  scale_color_gradientn(colours = c("red", "yellow", "blue" )) +
  #scale_color_gradientn(colours = colorRampPalette(rev(brewer.pal(n = 7, name ="RdYlBu")))(7) ) +   
   theme_minimal() +
  theme(panel.grid = element_line(size = 1, color="grey") 
        
        ,panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank()
         
        )  + facet_grid(group ~., scales = "free" , space = "free" )  + xlab("") + ylab ( "") +
  theme(legend.position="right",  legend.key = element_blank(),
          
          axis.text.y = element_text(size= 9 ),
          axis.text.x = element_blank(),
          axis.title.x = element_text(size=20),
          
          axis.title.y     = element_text(size=10), 
          legend.text      =element_text(size=10)
    ) + scale_size_continuous(range = c(12,12)) +
  # change size of facet 
  theme(strip.text.y = element_text(size = 20, colour = "black", angle = 0)) + ggtitle ( g)

#print ( g2 )


# stack plots instead. 

bp2 = bubble_pretty2 %>%  group_by(group) %>% top_n(mx,NES)


bar_path[[ g  ]] = ggplot( bp2  , aes(x=reorder_within ( pathway,  NES  , group) , y=NES, fill=group   )) +
   geom_bar(stat = 'identity', position = 'stack') +
    theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black")) +
    ylab("") + 
    xlab("") + 
    theme(legend.position="none",  legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=12),
          axis.title.y     = element_text(size=20), 
          legend.text      =element_text(size=12)
          , plot.title = element_text(size = 30, face = "bold")
    ) + 
    scale_x_reordered() +
  scale_fill_manual( values= rep ("#67A9CF", 10) ) + 
  #scale_fill_brewer(palette="Set3") +
  coord_flip() + ggtitle (  g ) +
 # scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) + # plot in intergers
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  facet_grid(group ~., scales = "free" , space = "free" ) 

print (bar_path[[ g  ]])
cat ( "\n\n")  
  
}  
  
```  

```{r}
# write to excel 

for ( g in c("BP", "MF", "HALLMARK", "KEGG", "REACTOME", "BIOCARTA" ) ){


  bbb=  bubble_up [ bubble_up$type == g, ]
  addWorksheet(paper_table, paste0("hclust_", g  ) )
writeData(paper_table
          , paste0("hclust_", g  ) 
          ,bbb
          , rowNames=F  )
}

```

```{r}

bar_path[["BP"]] + scale_fill_manual( values= rep ("#67A9CF", 10) ) +ylab("NES")
bar_path[["MF"]] + scale_fill_manual( values= rep ("#CCC7C7", 10) )+ylab("NES")
bar_path[["REACTOME"]] + scale_fill_manual( values= rep ("#EF8A62", 10) )+ylab("NES")

```

### GSEA, top 10 {.tabset}


```{r fig=TRUE,fig.width=8, fig.height=5, echo=FALSE, include=TRUE, results='asis'}
#kable(  dfpath[["C1"]] , format = "html" , row.names = F ) %>% kable_classic(full_width = F, position = "center") 

for ( g in names ( bar_path ))
{
  cat("#### ", g , "\n")
  print ( bar_path[[g]] )
cat ( "\n\n")  
}


```




```{r}





# Fusion chunk 1 
tfusion = fusion 

fusion = fusion$tier3 [ fusion$tier3$RNAseq.id %in% rna.key$RNAseq.id, ]
nofusion = setdiff( rna.key$RNAseq.id, unique ( fusion$RNAseq.id))
cname = names ( fusion)
is.stuff = cname[grepl ("^is.", cname)]
is.stuff = is.stuff[is.stuff!="is.germ"]
is.stuff = is.stuff[! grepl ( "is.cancer", is.stuff )]

fuse1 = fusion[ , c("RNAseq.id","fusion","frameness","SV_RESULT","cancer.subtype", is.stuff, "right", "uid"  )]

# this is druggable but its a better progonosis marker
fuse1 [ fuse1$fusion == "KIAA1549-BRAF", ]

#  tabulate type, in the is.column 

fusion.type =  reshape2::melt ( fuse1, id.vars = c("RNAseq.id","fusion","frameness","SV_RESULT","cancer.subtype", "uid"))
fusion.type =fusion.type[fusion.type$value != 0 , ]
fusion.type$type = as.character ( fusion.type$variable)
fusion.type$variable = NULL 



# druggable supercedes > canonical > is.known 
# calculate total type -  here we are going to subset every category, some rename some merge and rbind it all back. 
# the end result should equal amount of rows but prioritize with the above order and relabeld. 


temp.drug = fusion.type[fusion.type$type == "is.drug", ]
temp.can = fusion.type[fusion.type$type == "is.can" & !fusion.type$uid %in% temp.drug$uid, ]
# convert is.can to is.diag
temp.can$type = "is.diag"
# sanity check temp.can should no longer have drug! 
temp.can[temp.can$type == "is.drug", ]
temp.all = rbind ( temp.drug, temp.can)
temp.obs = fusion.type[fusion.type$type %in% c( "is.known", "is.seq" )  & !fusion.type$uid %in% temp.all$uid, ]
temp.obs[temp.obs$type %in% c( "is.drug", "is.drug" ) , ]
# change name to is.obsv 
temp.obs$type = "is.obsv" # there may be dups in here becaues is.known and is.seq is duplicated when melted so you need to remove this.
temp.obs = temp.obs[!duplicated ( temp.obs$uid), ]
temp.all = rbind ( temp.all, temp.obs)

temp.novel = fusion.type[fusion.type$type %in% c( "is.novel" )  & !fusion.type$uid %in% temp.all$uid, ]
temp.all = rbind ( temp.all, temp.novel)

# there should equal amount of rows for the new and older 
nrow ( fuse1) - nrow ( temp.all )
unique ( temp.all$type)

fusion.reclass = temp.all [ temp.all$frameness == "in.frame" , ]
is.can_noframe = temp.all[temp.all$type == "is.diag" & temp.all$frameness != "in.frame",   ]
 
fusion.reclass_outofframe =  temp.all [ temp.all$frameness == "out.of.frame", ]
######################################################################################################################

# I need a reclassifiction in wide form because fusion.reclass removes all the samples with zero. 


```


```{r}
# tabulate total cancer subtype
total_subcancer = data.frame ( table ( rna.key$cancer.subtype), stringsAsFactors = F )
colnames ( total_subcancer) = c("cancer.subtype","Freq")


```

```{r}

# change Other Heme to RO 
rna.key_og = rna.key 
rna.key$cancer.subtype = ifelse (rna.key$cancer.subtype == "Other Heme", "O/H",  rna.key$cancer.subtype  )

```

```{r}

# function for some fusion plots 

getfusion <- function ( type = "in.frame", fuse1, gfont=5){

fusion.frame = data.frame ( table ( fuse1[fuse1$frameness == type, ]$RNAseq.id ) , stringsAsFactors = F)
colnames ( fusion.frame ) = c("RNAseq.id","fusion.frame")
fusion.frame =  merge ( fusion.frame, rna.key[ , c ( "RNAseq.id","cancer.subtype") ], by="RNAseq.id")
fusion.frame = fusion.frame[ order ( fusion.frame$cancer.subtype, fusion.frame$fusion.frame), ]
fusion.frame$RNAseq.id = factor (fusion.frame$RNAseq.id, levels = fusion.frame$RNAseq.id )
fusion.frame$cancer.subtype = factor ( fusion.frame$cancer.subtype, levels = cancer.subtype.order)

# table median per disease 

fusion.frame.median  = fusion.frame %>% dplyr::group_by(cancer.subtype) %>%
  dplyr::summarise(
    med = median ( fusion.frame  ) # count occurence but leave out pizzly
  ) %>% data.frame()

# paper 
# create lollipop to summarize numbers 
# we might add more data here not sure yet. 
lolli_med = fusion.frame.median
lolli_med = lolli_med[ order ( lolli_med$med, decreasing = T), ]
lolli_med$cancer.subtype  = factor ( as.character ( lolli_med$cancer.subtype),  levels = rev ( as.character ( lolli_med$cancer.subtype)   ) )

lolliorder = levels ( lolli_med$cancer.subtype)

lolli_med = ggplot(lolli_med, aes(x=cancer.subtype, y=med)) +
  geom_segment( aes(x=cancer.subtype, xend=cancer.subtype, y=0, yend=med), color="black") +
  geom_point( color=subcolor [ rev ( levels ( lolli_med$cancer.subtype)) ], size=10, alpha=0.75) +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  ) + coord_flip() + xlab("") + ylab("median") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) 


fusion.frame.percent  = fusion.frame %>% dplyr::group_by(cancer.subtype) %>%
  dplyr::summarise(
    med = sum ( fusion.frame  ) / n() # count occurence but leave out pizzly
  ) %>% data.frame()



# fusion boxpolot by disease 

fusion.frame$cancer.subtype = factor ( fusion.frame$cancer.subtype,    levels=lolliorder)

fusion.by.sample.box=  ggplot( fusion.frame, aes(x=cancer.subtype, y=fusion.frame, fill=cancer.subtype)) +
  geom_boxplot(outlier.shape = NA) +
  scale_fill_viridis(discrete = TRUE, alpha=0.6) +
  geom_jitter(color="black", size=0.4, alpha=0.9) + coord_flip() + scale_fill_manual(values= subcolor ) +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  )  + xlab("") + ylab("Total Fusion") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) 




fusion.frame = fusion.frame[ order ( match(fusion.frame$cancer.subtype,lolliorder ) , fusion.frame$fusion.frame), ]
fusion.frame$cancer.subtype = factor ( fusion.frame$cancer.subtype, levels=as.character ( unique(fusion.frame$cancer.subtype)))
fusion.frame$RNAseq.id = factor ( fusion.frame$RNAseq.id, levels= as.character(fusion.frame$RNAseq.id))


sm = fusion.frame %>% group_by( cancer.subtype )  %>%
  dplyr::summarise  (
    median = median (  fusion.frame  ),
    RNAseq.id = RNAseq.id[ceiling(n()/2)]
  )



# first you want to add blank rows between groups. 
# I'm thinking maybe 5 is good 

buffer = 2 

add_blank <- function(x, n=5) {
  tibble::add_row(x, fusion.frame=rep(-5, n), RNAseq.id=rep('N', n))
}

fusion.frame2 = fusion.frame %>% 
  group_by(cancer.subtype) %>% 
  group_modify(~add_blank(., buffer)) %>% data.frame()


fusion.frame2 = fusion.frame2 %>% 
  group_split(cancer.subtype) %>% 
  map_dfr(~ .x %>% 
            add_row(cancer.subtype = dplyr::first(.x$cancer.subtype), fusion.frame= -5, .after = 0) %>%
            uncount(rep(c(buffer, 1), c(1, n()-1)))) %>% data.frame()


fusion.frame2 = fusion.frame2[- ( 1:buffer), ]

fusion.frame2[is.na(fusion.frame2)] <- "N"



# create a pseudocolumn to plot 
fusion.frame2$RNAseq.id = make.unique( fusion.frame2$RNAseq.id )

first.last = fusion.frame2  %>% 
  group_by(cancer.subtype) %>%
  dplyr::summarise(
    first = dplyr::first(RNAseq.id),
    last = dplyr::last(RNAseq.id)
  )


fusion.frame2$RNAseq.id = factor ( fusion.frame2$RNAseq.id, levels=fusion.frame2$RNAseq.id)

bgcolor = rep("grey", length(unique ( fusion.frame2$cancer.subtype ))  )
bgcolor[base::which ( seq_along(bgcolor) %% 2 > 0 )] = "white"
names ( bgcolor) = unique ( fusion.frame2$cancer.subtype )


maxp = max (fusion.frame2$fusion.frame ) + 1



fusion.frame2 = merge ( fusion.frame2, rna.key [ , c("RNAseq.id", "sample.classification")], by="RNAseq.id", all.x=T, all.y=F  )
fusion.frame2[is.na(fusion.frame2)] <- "Unknown"

set.seed(2)

g1 = ggplot( fusion.frame2, aes(x=RNAseq.id, y=fusion.frame  )) +
  #geom_jitter(aes ( color=cancer.subtype, shape= sample.classification   ), size=5, alpha=.8, width = 0)  + 
  geom_point(aes ( color=cancer.subtype
                   #, shape= sample.classification   
                   ) , position=position_jitter(h=0.1, w=0.1), alpha = 0.5, size = 5) + 
  scale_color_manual(values= subcolor ) +
  theme(
    
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  )  + xlab("sample") + ylab("Total Fusion") +
  
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("Total Fusions") + 
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
        , legend.position = "none"
  )  + geom_crossbar(data=sm, aes(x=RNAseq.id, y=median, ymin = median, ymax = median ),
                    size=1,col="black", width = 5.5)  +
  geom_rect(data=first.last, aes(NULL,NULL,xmin=first,xmax=last,fill=cancer.subtype),
            ymin=-.5,ymax=maxp, colour="white", size=.2, alpha=.2) + 
  scale_fill_manual(values=bgcolor)  +
  geom_text(
    aes(x = first, y = maxp, label = cancer.subtype),
    data = first.last,
    size = gfont, vjust = 0, hjust = 0, check_overlap = F
  )  + ylim(0, maxp ) 


df = merge (  fusion.frame, rna.key[ , 1:10], by="RNAseq.id")
return ( list ( g1=g1, lolli_med = lolli_med, df = df  ))

}

getfusion_pid <- function ( type = "in.frame", fuse1, gfont=5){

fusion.frame = fuse1[fuse1$frameness == type, ]  
fusion.frame = merge ( fusion.frame, rna.key[ , c ("patient.id",  "RNAseq.id") ], by="RNAseq.id") 


  
fusion.frame = data.frame ( table ( fusion.frame$patient.id ) , stringsAsFactors = F)
colnames ( fusion.frame ) = c("patient.id","fusion.frame")


fusion.frame$cancer.subtype =  as.character ( sapply ( fusion.frame$patient.id, function(x)   { unique ( rna.key[rna.key$patient.id == x, ]$cancer.subtype)  }) )
fusion.frame$cancer.subtype  = ifelse(grepl ( "OS.*NRST",  fusion.frame$cancer.subtype  ), "OS", fusion.frame$cancer.subtype  )

  
zero = setdiff( rna.key$patient.id, fusion.frame$patient.id)
zero4 = rna.key[rna.key$patient.id %in% zero,   c("patient.id","cancer.subtype") ] 
zero4$fusion.frame = 0 
  
fusion.frame = rbind ( fusion.frame, zero4[, names ( fusion.frame)])

fusion.frame$dup = paste ( fusion.frame$fusion, fusion.frame$patient.id) # remove duplicates fusion and patient.id 
fusion.frame = fusion.frame[!duplicated ( fusion.frame$dup), ]

#fusion.frame$patient.id = NULL 
fusion.frame$dup = NULL 




fusion.frame = fusion.frame[ order ( fusion.frame$cancer.subtype, fusion.frame$fusion.frame), ]
fusion.frame$patient.id = factor (fusion.frame$patient.id, levels = fusion.frame$patient.id )


fusion.frame$cancer.subtype = factor ( fusion.frame$cancer.subtype, levels = cancer.subtype.order)

# table median per disease 

fusion.frame.median  = fusion.frame %>% dplyr::group_by(cancer.subtype) %>%
  dplyr::summarise(
    med = median ( fusion.frame  ) # count occurence but leave out pizzly
  ) %>% data.frame()

# paper 
# create lollipop to summarize numbers 
# we might add more data here not sure yet. 
lolli_med = fusion.frame.median
lolli_med = lolli_med[ order ( lolli_med$med, decreasing = T), ]
lolli_med$cancer.subtype  = factor ( as.character ( lolli_med$cancer.subtype),  levels = rev ( as.character ( lolli_med$cancer.subtype)   ) )

lolliorder = levels ( lolli_med$cancer.subtype)

lolli_med = ggplot(lolli_med, aes(x=cancer.subtype, y=med)) +
  geom_segment( aes(x=cancer.subtype, xend=cancer.subtype, y=0, yend=med), color="black") +
  geom_point( color=subcolor [ rev ( levels ( lolli_med$cancer.subtype)) ], size=10, alpha=0.75) +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  ) + coord_flip() + xlab("") + ylab("median") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) 


fusion.frame.percent  = fusion.frame %>% dplyr::group_by(cancer.subtype) %>%
  dplyr::summarise(
    med = sum ( fusion.frame  ) / n() # count occurence but leave out pizzly
  ) %>% data.frame()



# fusion boxpolot by disease 

fusion.frame$cancer.subtype = factor ( fusion.frame$cancer.subtype,    levels=lolliorder)

fusion.by.sample.box=  ggplot( fusion.frame, aes(x=cancer.subtype, y=fusion.frame, fill=cancer.subtype)) +
  geom_boxplot(outlier.shape = NA) +
  scale_fill_viridis(discrete = TRUE, alpha=0.6) +
  geom_jitter(color="black", size=0.4, alpha=0.9) + coord_flip() + scale_fill_manual(values= subcolor ) +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  )  + xlab("") + ylab("Total Fusion") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) 




fusion.frame = fusion.frame[ order ( match(fusion.frame$cancer.subtype,lolliorder ) , fusion.frame$fusion.frame), ]
fusion.frame$cancer.subtype = factor ( fusion.frame$cancer.subtype, levels=as.character ( unique(fusion.frame$cancer.subtype)))
fusion.frame$patient.id = factor ( fusion.frame$patient.id, levels= as.character(fusion.frame$patient.id))


sm = fusion.frame %>% group_by( cancer.subtype )  %>%
  dplyr::summarise  (
    median = median (  fusion.frame  ),
    patient.id = patient.id[ceiling(n()/2)]
  )



# first you want to add blank rows between groups. 
# I'm thinking maybe 5 is good 

buffer = 2 

add_blank <- function(x, n=5) {
  tibble::add_row(x, fusion.frame=rep(-5, n), patient.id=rep('N', n))
}

fusion.frame2 = fusion.frame %>% 
  group_by(cancer.subtype) %>% 
  group_modify(~add_blank(., buffer)) %>% data.frame()


fusion.frame2 = fusion.frame2 %>% 
  group_split(cancer.subtype) %>% 
  map_dfr(~ .x %>% 
            add_row(cancer.subtype = dplyr::first(.x$cancer.subtype), fusion.frame= -5, .after = 0) %>%
            uncount(rep(c(buffer, 1), c(1, n()-1)))) %>% data.frame()


fusion.frame2 = fusion.frame2[- ( 1:buffer), ]

fusion.frame2[is.na(fusion.frame2)] <- "N"



# create a pseudocolumn to plot 
fusion.frame2$patient.id = make.unique( fusion.frame2$patient.id )

first.last = fusion.frame2  %>% 
  group_by(cancer.subtype) %>%
  dplyr::summarise(
    first = dplyr::first(patient.id),
    last = dplyr::last(patient.id)
  )


fusion.frame2$patient.id = factor ( fusion.frame2$patient.id, levels=fusion.frame2$patient.id)

bgcolor = rep("grey", length(unique ( fusion.frame2$cancer.subtype ))  )
bgcolor[base::which ( seq_along(bgcolor) %% 2 > 0 )] = "white"
names ( bgcolor) = unique ( fusion.frame2$cancer.subtype )


maxp = max (fusion.frame2$fusion.frame ) + 1



fusion.frame2 = merge ( fusion.frame2, rna.key [ , c("patient.id", "sample.classification")], by="patient.id", all.x=T, all.y=F  )
fusion.frame2[is.na(fusion.frame2)] <- "Unknown"

set.seed(2)

g1 = ggplot( fusion.frame2, aes(x=patient.id, y=fusion.frame  )) +
  #geom_jitter(aes ( color=cancer.subtype, shape= sample.classification   ), size=5, alpha=.8, width = 0)  + 
  geom_point(aes ( color=cancer.subtype
                   #, shape= sample.classification   
                   ) , position=position_jitter(h=0.1, w=0.1), alpha = 0.5, size = 5) + 
  scale_color_manual(values= subcolor ) +
  theme(
    
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  )  + xlab("sample") + ylab("Total Fusion") +
  
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("Total Fusions") + 
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
        , legend.position = "none"
  )  + geom_crossbar(data=sm, aes(x=patient.id, y=median, ymin = median, ymax = median ),
                    size=1,col="black", width = 5.5)  +
  geom_rect(data=first.last, aes(NULL,NULL,xmin=first,xmax=last,fill=cancer.subtype),
            ymin=-.5,ymax=maxp, colour="white", size=.2, alpha=.2) + 
  scale_fill_manual(values=bgcolor)  +
  geom_text(
    aes(x = first, y = maxp, label = cancer.subtype),
    data = first.last,
    size = gfont, vjust = 0, hjust = 0, check_overlap = F
  )  + ylim(0, maxp ) 


df = merge (  fusion.frame, rna.key[ , 1:10], by="patient.id")
return ( list ( g1=g1, lolli_med = lolli_med, df = df  ))

}

getfusion_patient <- function ( type = "in.frame", f2){




  
fusion.frame = f2  %>% group_by(  patient.id )  %>%
  dplyr::summarise  (
    fusion.frame = n (),
    cancer.subtype = unique ( paste0(unique ( cancer.subtype),collapse = ';')) 
  ) %>% data.frame()


# make sure to remove samples that have two diagnosis 

fusion.frame$cancer.subtype = gsub ( ";.*", "", fusion.frame$cancer.subtype )

zero = setdiff( rna.key$patient.id, fusion.frame$patient.id)
zero = rna.key[rna.key$patient.id %in% zero, c("patient.id","cancer.subtype")]
zero$fusion.frame = 0 

fusion.frame = rbind ( fusion.frame, zero[ , colnames(fusion.frame)])

# table median per disease 

fusion.frame.median  = fusion.frame %>% dplyr::group_by(cancer.subtype) %>%
  dplyr::summarise(
    med = median ( fusion.frame  ) # count occurence but leave out pizzly
  ) %>% data.frame()

# paper 
# create lollipop to summarize numbers 
# we might add more data here not sure yet. 
lolli_med = fusion.frame.median
lolli_med = lolli_med[ order ( lolli_med$med, decreasing = T), ]
lolli_med$cancer.subtype  = factor ( as.character ( lolli_med$cancer.subtype),  levels = rev ( as.character ( lolli_med$cancer.subtype)   ) )

lolliorder = levels ( lolli_med$cancer.subtype)

lolli_med = ggplot(lolli_med, aes(x=cancer.subtype, y=med)) +
  geom_segment( aes(x=cancer.subtype, xend=cancer.subtype, y=0, yend=med), color="black") +
  geom_point( color=subcolor [ rev ( levels ( lolli_med$cancer.subtype)) ], size=10, alpha=0.75) +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  ) + coord_flip() + xlab("") + ylab("median") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) 


fusion.frame.percent  = fusion.frame %>% dplyr::group_by(cancer.subtype) %>%
  dplyr::summarise(
    med = sum ( fusion.frame  ) / n() # count occurence but leave out pizzly
  ) %>% data.frame()



# fusion boxpolot by disease 

fusion.frame$cancer.subtype = factor ( fusion.frame$cancer.subtype,    levels=lolliorder)

fusion.by.sample.box=  ggplot( fusion.frame, aes(x=cancer.subtype, y=fusion.frame, fill=cancer.subtype)) +
  geom_boxplot(outlier.shape = NA) +
  scale_fill_viridis(discrete = TRUE, alpha=0.6) +
  geom_jitter(color="black", size=0.4, alpha=0.9) + coord_flip() + scale_fill_manual(values= subcolor ) +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  )  + xlab("") + ylab("Total Fusion") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) 




fusion.frame = fusion.frame[ order ( match(fusion.frame$cancer.subtype,lolliorder ) , fusion.frame$fusion.frame), ]
fusion.frame$cancer.subtype = factor ( fusion.frame$cancer.subtype, levels=as.character ( unique(fusion.frame$cancer.subtype)))
fusion.frame$RNAseq.id = factor ( fusion.frame$patient.id, levels= unique (fusion.frame$patient.id))


sm = fusion.frame %>% group_by( cancer.subtype )  %>%
  dplyr::summarise  (
    median = median (  fusion.frame  ),
    RNAseq.id = RNAseq.id[ceiling(n()/2)]
  )



# first you want to add blank rows between groups. 
# I'm thinking maybe 5 is good 

buffer = 2 

add_blank <- function(x, n=5) {
  tibble::add_row(x, fusion.frame=rep(-5, n), RNAseq.id=rep('N', n))
}

fusion.frame2 = fusion.frame %>% 
  group_by(cancer.subtype) %>% 
  group_modify(~add_blank(., buffer)) %>% data.frame()


fusion.frame2 = fusion.frame2 %>% 
  group_split(cancer.subtype) %>% 
  map_dfr(~ .x %>% 
            add_row(cancer.subtype = first(.x$cancer.subtype), fusion.frame= -5, .after = 0) %>%
            uncount(rep(c(buffer, 1), c(1, n()-1)))) %>% data.frame()


fusion.frame2 = fusion.frame2[- ( 1:buffer), ]

fusion.frame2[is.na(fusion.frame2)] <- "N"



# create a pseudocolumn to plot 
fusion.frame2$RNAseq.id = make.unique( fusion.frame2$RNAseq.id )

first.last = fusion.frame2  %>% 
  group_by(cancer.subtype) %>%
  dplyr::summarise(
    first = dplyr::first(RNAseq.id),
    last = dplyr::last(RNAseq.id)
  )


fusion.frame2$RNAseq.id = factor ( fusion.frame2$RNAseq.id, levels=fusion.frame2$RNAseq.id)

bgcolor = rep("grey", length(unique ( fusion.frame2$cancer.subtype ))  )
bgcolor[which ( seq_along(bgcolor) %% 2 > 0 )] = "white"
names ( bgcolor) = unique ( fusion.frame2$cancer.subtype )


maxp = max (fusion.frame2$fusion.frame ) + 1



fusion.frame2 = merge ( fusion.frame2, rna.key [ , c("patient.di", "sample.classification")], by.x="RNAseq.id", by.y="patient.id" , all.x=T, all.y=F  )
fusion.frame2[is.na(fusion.frame2)] <- "Unknown"

set.seed(2)

g1 = ggplot( fusion.frame2, aes(x=RNAseq.id, y=fusion.frame  )) +
  #geom_jitter(aes ( color=cancer.subtype, shape= sample.classification   ), size=5, alpha=.8, width = 0)  + 
  geom_point(aes ( color=cancer.subtype, shape= sample.classification   ) , position=position_jitter(h=0.1, w=0.1), alpha = 0.5, size = 5) + 
  scale_color_manual(values= subcolor ) +
  theme(
    
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  )  + xlab("sample") + ylab("Total Fusion") +
  
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("Total Fusions") + 
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
        , legend.position = "bottom"
  )  + geom_crossbar(data=sm, aes(x=RNAseq.id, y=median, ymin = median, ymax = median ),
                    size=1,col="black", width = 5.5)  +
  geom_rect(data=first.last, aes(NULL,NULL,xmin=first,xmax=last,fill=cancer.subtype),
            ymin=-.5,ymax=maxp, colour="white", size=.2, alpha=.2) + 
  scale_fill_manual(values=bgcolor)  +
  geom_text(
    aes(x = first, y = maxp, label = cancer.subtype),
    data = first.last,
    size = 3, vjust = 0, hjust = 0, check_overlap = F
  )  + ylim(0, maxp ) 


df = merge (  fusion.frame, rna.key[ , 1:10], by="RNAseq.id")
return ( list ( g1=g1, lolli_med = lolli_med, df = df  ))

}




summarise_fusion <- function ( class="in.frame", fusion.reclass=fusion.reclass ){
  
  fusion.in =  fusion.reclass[fusion.reclass$frameness == class, ]
  fusion.in$cancer.subtype = NULL 
  fusion.in$patient.id = NULL 
  
  fusion.in= merge ( fusion.in, rna.key[ , c ( "patient.id", "RNAseq.id", "cancer.subtype") ], by="RNAseq.id")
  
  fusion.in$dup = paste ( fusion.in$fusion, fusion.in$patient.id) # remove duplicates fusion and patient.id 
  fusion.in = fusion.in[!duplicated ( fusion.in$dup), ]
  fusion.in$RNAseq.id = factor ( fusion.in$RNAseq.id, levels= rna.key$RNAseq.id )
  fusion.in$cancer.subtype =  gsub ( "OS;NRSTS", "OS",  fusion.in$cancer.subtype  )
  # we need to add zero in to get an accurate median estimate 
  
  zero = setdiff( rna.key$patient.id, fusion.in$patient.id)
  fusion.in2=fusion.in
  fusion.in2$total = 1 
  temp4 = rna.key[rna.key$patient.id %in% zero,   c("RNAseq.id","patient.id","cancer.subtype") ] 
  temp4$total = 0 
  temp4$fusion = "NONE" 
  temp4$type = "no.fusion" 
  fusion.in2 = rbind ( fusion.in2[ , c("RNAseq.id","patient.id","cancer.subtype","fusion","total","type")], temp4  )
  
  fusion.in2 = fusion.in2[ , c("RNAseq.id","patient.id","cancer.subtype","fusion","total","type")]
  # fusion chunk 4  tally 
  
  
  # tally patient 
  
  # first calculate total for each patient 
  
  tallyfusion.pid =  fusion.in2 %>% group_by(  patient.id )  %>%
    dplyr::summarise  (
      total_patient = sum (total),
      cancer.subtype = unique ( paste0(unique ( cancer.subtype),collapse = ';')) 
    ) %>% data.frame()
  
  
  tallyfusion.pid$cancer.subtype = gsub ( "OS;NRSTS", "OS", tallyfusion.pid$cancer.subtype  )
  
  # now sum by cancer type 
  
  tallyfusion.pid_tb = tallyfusion.pid %>% group_by(  cancer.subtype )  %>%
    dplyr::summarise  (
      "Unique By Patient" = sum ( total_patient ) ,
      "Median" = median ( total_patient ), 
      "Ave" = round ( mean ( total_patient), 1),
      "Min" = min ( total_patient ), 
      "Max" = max ( total_patient), 
      # "CV" = cov ( total_patient )
    ) %>% data.frame()
  
  
  ### calculate by unique, that is we want to know how many total unique fusion in that group 
  
  tallyfusion.cs =  fusion.in2[ !duplicated (  paste(fusion.in2$fusion, fusion.in2$cancer.subtype)  )  , ]  %>% 
    group_by(  patient.id )  %>%
    dplyr::summarise  (
      total_patient = sum (total),
      cancer.subtype = unique ( paste0(unique ( cancer.subtype),collapse = ';')) 
    ) %>% data.frame()
  
  
  tallyfusion.cs$cancer.subtype = gsub ( "OS;NRSTS", "OS", tallyfusion.cs$cancer.subtype  )
  
  # now sum by cancer type 
  
  tallyfusion.cs = tallyfusion.cs %>% group_by(  cancer.subtype )  %>%
    dplyr::summarise  (
      "Unique by Cancer subtype" = sum ( total_patient ) 
    ) %>% data.frame()
  
  
  tally_f = merge ( tallyfusion.pid_tb, tallyfusion.cs, by="cancer.subtype")
  
  
  ############################# 
  
  # calculating unique left and right by cancer type 
  
  left_right = stringr::str_match(fusion.in2$fusion, "(.*)-(.*)")
  fusion.in2$left = left_right[ , 2]
  fusion.in2$right = left_right[ , 3]
  
  
  left_unique_cs =  fusion.in2[ !duplicated (  paste(fusion.in2$left, fusion.in2$cancer.subtype)  )  , ]  %>% 
    group_by(  cancer.subtype )  %>%
    dplyr::summarise  (
      Unique.5 = sum (total),
      #cancer.subtype = unique ( paste0(unique ( cancer.subtype),collapse = ';')) 
    ) %>% data.frame()
  
  
  
  
  right_unique_cs =  fusion.in2[ !duplicated (  paste(fusion.in2$right, fusion.in2$cancer.subtype)  )  , ]  %>% 
    group_by(  cancer.subtype )  %>%
    dplyr::summarise  (
      "Unique.3" = sum (total),
      #cancer.subtype = unique ( paste0(unique ( cancer.subtype),collapse = ';')) 
    ) %>% data.frame()
  
  
  ################################
  tally_f = merge ( tally_f, left_unique_cs, by="cancer.subtype")
  tally_f = merge ( tally_f, right_unique_cs, by="cancer.subtype")
  
  # get each type 
  
  temp2 =  fusion.in2 %>% group_by(  patient.id )  %>%
    dplyr::mutate  (
      total_patient = sum (total),
      cancer.subtype = unique ( paste0(unique ( cancer.subtype),collapse = ';')) 
    ) %>% data.frame()
  
  
  temp2$cancer.subtype = gsub ( "OS;NRSTS", "OS", temp2$cancer.subtype  )
  
  
  
  
  temp2 = data.frame ( table ( 
    temp2[!duplicated (paste ( temp2$cancer.subtype, temp2$fusion, temp2$patient.id ) ) , c("cancer.subtype", "type") ]
  ))
  
  temp2 <- dcast( temp2 ,cancer.subtype   ~ type, value.var= "Freq"  )
  
  
  
  tallyfusion.pid_tb = merge ( tally_f, temp2, by="cancer.subtype")
  
  
  # calculate how many unique pid per cancersubtyp 
  temps = rna.key[!duplicated ( rna.key$patient.id), ] %>% group_by(   cancer.subtype  )  %>%
    dplyr::summarise  (
      total.patient = n () ) %>%
    data.frame ( stringsAsFactors = F )
  
  # calculate percent with no fusion. 
  #cs_total = data.frame ( table (   rna.key[ rna.key$patient.id %in%  fusion.in2$patient.id, ]$cancer.subtype      ))
  #colnames ( cs_total ) = c("cancer.subtype","total.patient")
  #tallyfusion.pid_tb = merge ( tallyfusion.pid_tb, cs_total, by="cancer.subtype")
  #tallyfusion.pid_tb = tallyfusion.pid_tb[ order ( tallyfusion.pid_tb$Unique.By.Patient , decreasing = T ), ]
  
  ## add in how many samples did not have fusions. 
 # nofusionsample = temp4[!duplicated ( temp4$patient.id), ]
  
  #nofusionsample = nofusionsample %>% group_by(   cancer.subtype  )  %>%
  #  dplyr::summarise  (
   #   no.fusion = n (),
    #  cancer.subtype = unique ( paste0(unique ( cancer.subtype),collapse = ';')) 
    #) %>% data.frame()
  
  #tallyfusion.pid_tb = merge ( tallyfusion.pid_tb, nofusionsample, by="cancer.subtype", all.x=T, all.y=F )
  #tallyfusion.pid_tb[is.na(tallyfusion.pid_tb)] <- 0 
  
  tallyfusion.pid_tb = merge ( tallyfusion.pid_tb, temps, by="cancer.subtype" , all=T)
  
  
  #tallyfusion.pid_tb$no.fusion = round ( tallyfusion.pid_tb$no.fusion / tallyfusion.pid_tb$total.patient, 2) 
  
  #tallyfusion.pid_tb$total.patient = NULL 
  
  tallyfusion.pid_tb$fusion_positive = (tallyfusion.pid_tb$total.patient-tallyfusion.pid_tb$no.fusion) / tallyfusion.pid_tb$total.patient
  tallyfusion.pid_tb = tallyfusion.pid_tb[ order ( tallyfusion.pid_tb$Unique.By.Patient , decreasing = T ), ]
  
  return ( tallyfusion.pid_tb )
  
}



```


```{r}
# redefining what is actionable here.
# UPDATE: should not need to do this anymore but you can sanity check it! 
# there is a TK216 PMID 19584866 for out database is level L3 that is in dish 

is.drug = sapply ( fuse1$fusion , function(x){
  lr = unlist ( stringr::str_split(x,"-"))
  p2 = lr[2]
  ac = drug.fusion[ (drug.fusion$drugmatch == x ) | 
                    (drug.fusion$drugmatch == lr[2]  ) 
                    , ]
 
   if ( grepl ( "FLI", x)){
    ac = data.frame()
  }

  
  if ( nrow ( ac ) == 0 ){ return ( "0" )}else { return (  paste(ac$uid, collapse = ",") )}
}
)
is.drug[is.drug != "0" ]

rm ( is.drug )
# skip redefing druggable fusion. Every druggable fusion would need to be manually look through anyways. 
#fuse1$is.drug = as.character ( is.drug )


# fusion chunk 3  main plot



fuse1$RNAseq.id = factor ( fuse1$RNAseq.id, levels= rna.key$RNAseq.id )


# create plots, can use fuse1 because we don't care about clasification here. 

fusion.plot = getfusion (  type = "in.frame", fuse1, gfont= 5  )
tallyfusion.pid_tb = summarise_fusion (class="in.frame", fusion.reclass = fusion.reclass )
in.frame.plot_patient = getfusion_pid  (type = "in.frame", fuse1, gfont= 5 )

summary_non_inframe = summarise_fusion (class="out.of.frame", fusion.reclass = fusion.reclass_outofframe )
# to summarise total by patient its important to remove duplicates  
# the reason is that a single pid can have multiple samples --> biopsy/met which can end up with the same fusion! 



kable( tallyfusion.pid_tb[ , c(
  "cancer.subtype" 
  ,"Unique.By.Patient" 
  ,"Unique.by.Cancer.subtype" 
  , "Unique.5", "Unique.3"
  
  ,"Median" , "Ave", "Min" ,"Max"  
  ,"is.diag" ,"is.drug" ,"is.novel" ,"is.obsv" ,"fusion_positive"
)
 ] , format = "html" , row.names = F, caption="", table.attr = "style='width:40%;'" ) %>% kable_classic(full_width = T, position = "center", ) 


fusion_tally = tallyfusion.pid_tb[ , c(
  "cancer.subtype" 
  ,"Unique.By.Patient" 
  ,"Unique.by.Cancer.subtype" 
  ,"Median" , "Min" ,"Max"  
  ,"is.diag" ,"is.drug" ,"is.novel" ,"is.obsv" ,"fusion_positive"
)
 ] 

colnames (fusion_tally) = c("CS", "UnqP", "UnqCS", "Median","Min","Max","is.diag" ,"is.drug" ,"is.novel" ,"is.obsv" ,"pct.pos")
fusion_tally$pct.pos = round ( fusion_tally$pct.pos, 2)


kable( fusion_tally , format = "html" , row.names = F, caption="", table.attr = "style='width:40%;'" ) %>% kable_classic(full_width = T, position = "center"
                                                                                                          ,bootstrap_options = c("striped")               
                                                                                                                         )%>%
  footnote(
           symbol  = c("CS: Cancer Subtype", "UnqP: Unique by patient", "UnqCS: Unique by cancer subtype", "pct.pos: Percent of Patient samples with at least 1 in.frame fusion"),
         
           )

### fusion_tally 
fuse_tally_tbl <- ggtexttable(fusion_tally[, 1:10], rows = NULL, theme = ttheme("light") ) %>%
  tab_add_hline(at.row = c( 1:2) , row.side = "top", linewidth = 2) %>%
  tab_add_vline(at.column = 2, column.side = "left", from.row = 2, linetype =1) %>%
  tab_add_footnote(text = "CS: Cancer Subtype
  UnqP: Unique by patient
  UnqCS: Unique by cancer subtype
  pct.pos: Percent of Patient samples with at least 1 in.frame fusion
                   
                   ", size = 12, face = "italic")
                 


addWorksheet(paper_table, 'fusion_tab')
#writeData(wb, 'QC' , df  , rowNames=T  )
# writeData(paper_table, 'QC', )
writeData(paper_table
          ,'fusion_tab'
          ,fusion_tally
          , rowNames=F  )






#top6_t1 <- table_cell_bg(top6_t1, row = 0, column = 1:ncol(df6), linewidth = 5,
 #                   fill="#d7d7d9", color = "#d7d7d9")








# get total 
sum ( tallyfusion.pid_tb$Unique.By.Patient)
median ( tallyfusion.pid_tb$fusion_positive)
sum ( summary_non_inframe$Unique.By.Patient)
sum ( tallyfusion.pid_tb$is.obsv)

sum ( summary_non_inframe$Unique.By.Patient)

sum ( tallyfusion.pid_tb$is.diag)

####### create a small plot for no fusion 

temp = tallyfusion.pid_tb
temp$fuse = "plot"

q = quantile ( temp$fusion_positive )

fusion_percent = ggplot( temp, aes( x=fuse,y=fusion_positive)) +
  #geom_violin()+ 
  geom_jitter(shape=19, position=position_jitter(0.1), aes( colour=cancer.subtype), size=10, alpha = 0.9  ) + 
  scale_color_manual(values= subcolor ) +
  theme(
    
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  )  + ylab("Percent Fusion") +
  
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
        , legend.position = "bottom"
        , axis.text.y = element_text(size= 22 )
        , axis.title.y     = element_text(size=22)
        , legend.title = element_blank()
  ) + 

  geom_hline(yintercept=q[2], linetype="dashed", 
             color = "grey", size=2) +
  geom_hline(yintercept=q[4], linetype="dashed", 
             color = "grey", size=2) + ggtitle ( "Cancer Subtype")


## add percent of cancer_type
# # sanity check this should equal above by percent 
total_type =   rna.key[!duplicated ( rna.key$patient.id), ] %>% group_by(   cancer.subtype  )  %>%
    dplyr::summarise  (
      total.patient = n () ) %>%
    data.frame ( stringsAsFactors = F )

fusion.reclass_type = merge ( fusion.reclass[, c("RNAseq.id", "fusion", "frameness")], rna.key [ , c("RNAseq.id", "cancer.subtype", "patient.id")], by="RNAseq.id")
fusion.reclass_type = fusion.reclass_type[!duplicated ( fusion.reclass_type$patient.id), ]
fusion.reclass_type = merge ( fusion.reclass_type, total_type, by="cancer.subtype")

fusion.reclass_type %>% dplyr:: group_by (cancer.subtype,total.patient) %>% dplyr::summarise ( pct.pos = n() ) %>%  mutate( round ( pct.pos /total.patient,2 ) )

# redo again with cancer.typ 

cancer.type_color = c( "Leukemia"='#b54536', Solid = "#333232", CNS='#bf9c43' )

 total_type =   rna.key[!duplicated ( rna.key$patient.id), ] %>% group_by(   cancer.type  )  %>%
    dplyr::summarise  (
      total.patient = n () ) %>%
    data.frame ( stringsAsFactors = F )

fusion.reclass_type = merge ( fusion.reclass[, c("RNAseq.id", "fusion", "frameness")], rna.key [ , c("RNAseq.id", "cancer.type", "patient.id")], by="RNAseq.id")
fusion.reclass_type = fusion.reclass_type[!duplicated ( fusion.reclass_type$patient.id), ]
fusion.reclass_type = merge ( fusion.reclass_type, total_type, by="cancer.type")

fusion_pct_cancerType  = fusion.reclass_type %>% dplyr:: group_by (cancer.type,total.patient) %>% dplyr::summarise ( pct.pos = n() ) %>%  mutate( fusion_positive = round ( pct.pos /total.patient,2 ) ) %>%
   ggplot( ., aes( x=1,y=fusion_positive)) +
  #geom_violin()+ 
  geom_jitter(shape=19, position=position_jitter(0.0), aes( colour=cancer.type), size=10, alpha = 0.9  ) + 
 # scale_color_manual(values= subcolor ) +
  theme(
    
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  )  + ylab("Percent Fusion") +
  
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  xlab("") + 
  scale_colour_manual(values= cancer.type_color ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
        , legend.position = "bottom"
        , axis.text.y = element_text(size= 22 )
        , axis.title.y     = element_text(size=22)
        , legend.title = element_blank()
  ) + ggtitle ( "Cancer Type")


lay2 = "AAAAABBBBBBBBB"

fusion_pct_cancerType + ( fusion_percent + ylab('') ) + plot_layout(design = lay2)

  

```



```{r}

temp3 = fusion.reclass[ fusion.reclass$type == "is.diag", ]
temp3$cancer.subtype = NULL 

nrow ( temp3 )
temp3 = merge ( temp3, rna.key[ , c("patient.id", "RNAseq.id", "diagnosis", "cancer.subtype")], by="RNAseq.id")

temp3=  temp3[ !duplicated ( paste (temp3$patient.id, temp3$fusion )), ]


temp3_ranke = data.frame ( table ( temp3$fusion ) )
temp3_ranke = temp3_ranke[ order ( temp3_ranke$Freq,temp3_ranke$Var1,  decreasing = T), ]

temp3$fusion = factor ( temp3$fusion  , levels= rev ( as.character(temp3_ranke$Var1) )  )

diag_table2 = diag_table
rthis = which ( as.character( diag_table$RNAseq.id ) %in% c(	"BD6", "Z9", "BE5" )  )
diag_table2$label = ifelse(  diag_table$RNAseq.id %in% c(	"BD6", "Z9", "BE5" ), "*", "")
#diag_table2$cancer.subtype = factor ( diag_table2$cancer.subtype )

well.known.diag = ggplot(data=diag_table2, aes(x=fusion, y=1,  fill=cancer.subtype  )) +
  geom_bar(stat="identity",  width = 1, position = "stack", color="grey") +  # color=""
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"),
          axis.text.y = element_text(size= 20 ),
      axis.text.x =  element_text(size= 20 ),
      axis.title.x = element_text(size=20),
      axis.title.y     = element_text(size=25), 
      legend.text      =element_text(size=25),
      legend.title = element_text(size=25),
      plot.title = element_text(size = 40, face = "bold")
        
        ) +
  ylab("") + 
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  coord_flip() + scale_fill_manual(values=subcolor) + ggtitle ( "Diagnostic") +
  geom_text(aes(label=label)
            #, position = position_stack(vjust = 0.5, hjust = 0.5) 
            , vjust = 0.75 , hjust = 2.0 #position = position
            , colour="black", size=10) + ggtitle ( "Diagnostic Fusions")



temp3[ order ( temp3$cancer.subtype, temp3$fusion), ]



diag_table = temp3[ order ( temp3$cancer.subtype, temp3$fusion), ] 

rthis = which ( as.character( diag_table$RNAseq.id ) %in% c(	"BD6", "Z9", "BE5" )  )


kable( diag_table[  , c(
  "RNAseq.id", "patient.id",  "fusion" , "diagnosis" , "cancer.subtype"
)
 ] , format = "html" , row.names = F, caption="", table.attr = "style='width:40%;'" ) %>% kable_classic(full_width = T, position = "center", )  %>%
    column_spec(1, bold = T, border_right = T) %>%
    row_spec(rthis, background = "#edebeb")


addWorksheet(paper_table, 'Diagnostic_fusion')
writeData(paper_table
          ,'Diagnostic_fusion'
          ,diag_table[  , c(
  "RNAseq.id", "patient.id",  "fusion" , "diagnosis" , "cancer.subtype"
)
 ] 
, rowNames=F  )


######### plotting drugs 

is.drug = temp.all[ temp.all$type == "is.drug" & temp.all$frameness == "in.frame", ]
is.drug$right =  gsub ( ".*-", '',is.drug$fusion)
unique ( is.drug$right)

is.drug = merge ( is.drug[ , c("RNAseq.id","fusion")], rna.key[ c("patient.id", "RNAseq.id","diagnosis", "cancer.subtype")], by="RNAseq.id" )
is.drug = is.drug[order ( is.drug$fusion), ]


### for paper 
kable( is.drug , format = "html" , row.names = F, caption="", table.attr = "style='width:40%;'" ) %>% kable_classic(full_width = T, position = "center", )  %>%
  column_spec(1, bold = T, border_right = T)

addWorksheet(paper_table, 'Druggable_fusion')
writeData(paper_table
          ,'Druggable_fusion'
          ,is.drug
          , rowNames=F  )



is_drug_tbl =  ggtexttable(is.drug, rows = NULL, theme = ttheme("blank") ) %>%
  tab_add_hline(at.row = c( 1:2) , row.side = "top", linewidth = 2) %>%
  tab_add_vline(at.column = 2, column.side = "left", from.row = 2, linetype =1)
 
 


is.diag2 = is.drug  %>% 
  group_by(fusion,cancer.subtype )  %>%
  dplyr::summarise  (
    Freq = n()
  )  %>% arrange ( desc ( Freq) ) %>% data.frame ( stringsAsFactors = F)


dup = is.diag2  %>% 
  group_by(fusion )  %>%
  dplyr::summarise  (
    Freq = sum(Freq)
  ) %>% arrange ( desc ( Freq) ) %>% data.frame ( stringsAsFactors = F)


is.diag2$fusion = factor ( is.diag2$fusion, rev( as.character ( dup$fusion )  )  )



druggable.plot = ggplot(data=is.diag2, aes(x=fusion, y=Freq,  fill=cancer.subtype  )) +
  geom_bar(stat="identity",  width = 1, position = "stack") +  # color=""
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
  ) + coord_flip() + scale_fill_manual(values=subcolor) + ggtitle ( "Druggable")







```


```{r}
sum ( tallyfusion.pid_tb$is.obsv )
sum ( tallyfusion.pid_tb$is.novel )
```

```{r}
# fusion chunk 5 fusion pie 


fusion.in =  fusion.reclass[fusion.reclass$frameness == "in.frame", ] # fusion.reclass is all inframe already
fusion.in$cancer.subtype = NULL 
fusion.in$patient.id = NULL 

fusion.in= merge ( fusion.in, rna.key[ , c ( "patient.id", "RNAseq.id", "cancer.subtype") ], by="RNAseq.id")

fusion.in$dup = paste ( fusion.in$fusion, fusion.in$patient.id) # remove duplicates fusion and patient.id 
fusion.in = fusion.in[!duplicated ( fusion.in$dup), ]
fusion.in$RNAseq.id = factor ( fusion.in$RNAseq.id, levels= rna.key$RNAseq.id )

# we need to add zero in to get an accurate median estimate 

zero = setdiff( rna.key$patient.id, fusion.in$patient.id)
fusion.in2=fusion.in
fusion.in2$total = 1 
temp4 = rna.key[rna.key$patient.id %in% zero,   c("RNAseq.id","patient.id","cancer.subtype") ] 
temp4$total = 0 
temp4$fusion = "NONE" 
temp4$type = "no.fusion" 
## lets not include no.fusion here
#fusion.in2 = rbind ( fusion.in2[ , c("RNAseq.id","patient.id","cancer.subtype","fusion","total","type")], temp4  )





cc5 = rev ( brewer.pal(5, "Set3") )
names ( cc5) =  rev ( c(
  "is.diag"  , "is.novel"  ,     "no.fusion",    "is.drug", "is.obsv" 
) )
cc5["no.fusion"]="#f5f5f5"
cc5["is.novel"]="#e8d558"


piethis = fusion.in2[fusion.in2$type != "no.fusion", ] %>% group_by(  patient.id )  %>%
  dplyr::mutate  (
    total_patient = sum (total),
    cancer.subtype = unique ( paste0(unique ( cancer.subtype),collapse = ';')) 
  ) %>% data.frame()




mp= piethis   %>%
  group_by(cancer.subtype, type ) %>%
  summarise( 
    countT= sum(total_patient),
    median.type = median ( total_patient )
  ) %>%  group_by(cancer.subtype)  %>% 
  mutate ( 
    per = sum ( countT)  
  ) %>% group_by(cancer.subtype, type ) %>%  mutate (
    per = round(100*countT/per,2)
  ) %>% data.frame( stringsAsFactors = F ) 

mp$cancer.subtype = factor ( mp$cancer.subtype )


## calculate median and range

fusion.pie = ggplot(mp, aes(x="", y=per, fill=type )) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0)     +
  xlab("") + ylab("") +
  scale_fill_manual(values= cc5 ) + theme_void() +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), 
        legend.position="bottom"
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))  +  facet_wrap  (cancer.subtype~., ncol = 4  )








```

```{r}
## tabulate wgs evidence 

plot_wgs_fusion <- function ( df , type, legend_pos="bottom", fs=25  ){

#this.sv  = temp.all [temp.all$type %in% c ( "is.diag"  )  , c( "RNAseq.id","fusion" , "SV_RESULT", "type","frameness")]  
this.sv  = df [df$type %in% type   , c( "RNAseq.id","fusion" , "SV_RESULT", "type","frameness")]  
this.sv[this.sv == "-" ]= "RNA_ONLY"

# as long as it does not say RNA_only or no_wgs then there is evidence. 
this.sv$SV_RESULT = ifelse( !this.sv$SV_RESULT  %in% c( "RNA_ONLY", "NO_WGS" ), "WGS.evidence", this.sv$SV_RESULT  )
# relabel no wgs that is samples without rna 
this.sv$SV_RESULT = gsub ( "NO_WGS", "no-wgs", this.sv$SV_RESULT ) # these sample do not have WGS pair 

# get frequency 
this.sv = plyr::count(this.sv, c("RNAseq.id","SV_RESULT"))


sv.freq <- this.sv %>% dplyr::group_by( SV_RESULT ) %>%
  dplyr::summarise(
    Freq = sum(freq)  
    
  ) %>% data.frame ( stringsAsFactors = F)

sv.freq = sv.freq[sv.freq$SV_RESULT != "no-wgs", ]



msv= sv.freq %>%
  #group_by(type) %>%
  mutate(countT= sum(Freq)) %>%
  mutate(per=round(100*Freq/countT,2))%>% data.frame()


# this part is necessary to place the label rotationally correctly 

msv <- msv %>% 
  # group_by(type )  %>%
  mutate(
    cs = rev(cumsum(rev(per))),
    text_yp = per/2 + lead(cs, 1),
    text_yp = if_else(is.na(text_yp), per/2, text_yp)
  )  %>%  data.frame()

colnames ( msv )[1] = "Gene.fusion"


pie.sv = ggplot(msv, aes(x="", y=per, fill=Gene.fusion )) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0)   + 
  theme_void()   +
  geom_label_repel(
    aes(
      label = Freq, y = text_yp
       
    ), show_guide  = F, size = fs
    
  ) + xlab("") + ylab("") +
  theme_void() + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(), 
        
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
  theme(
    strip.background = element_blank(),
    strip.text = element_blank(), 
    legend.position = legend_pos ,
    legend.text      =element_text(size=25),
    legend.title = element_text(size=25), 
    plot.title = element_text(size = 40, face = "bold", hjust = 0.5 ), 
 
                                
  )  +
  scale_fill_manual(values = c( 'RNA_ONLY'="#db7944", 'WGS.evidence' = '#8a7f79') ) + ggtitle ( paste (type  ))


return ( pie.sv )

}


table ( temp.all$type)

wgs_is.obsv = plot_wgs_fusion( df = temp.all , type = "is.obsv", legend_pos = "bottom", fs=18)
wgs_is.diag = plot_wgs_fusion( df = temp.all , type = "is.diag", legend_pos = "none" , fs=18)
wgs_is.novel = plot_wgs_fusion( df = temp.all , type = "is.novel", legend_pos = "none" , fs=18 )

wgs_is.diag  +   wgs_is.obsv + wgs_is.novel 

```


## Fusion {.tabset}

* Total unique fusion for each sample ( not aggregated by patient )

### Simple Median breakdown 

```{r fig=TRUE,fig.width=5, fig.height=5, echo=FALSE, include=TRUE}
fusion.plot$lolli_med + ggtitle ( "Median Fusion by Disease ") +  
  
       theme(
          
          axis.text.y = element_text(size= 20 ),
          #axis.text.x = element_blank(),
          axis.title.x = element_text(size=20),
          
          axis.title.y     = element_text(size=20 ) 
       )

```


### fusion by sample

```{r fig=TRUE,fig.width=12, fig.height=7, echo=FALSE, include=TRUE}
fusion.plot$g1 + ggtitle ( "Fusion by Sample ") + 
  
       theme(
          
          axis.text.y = element_text(size= 25 ),
          #axis.text.x = element_blank(),
          axis.title.x = element_text(size=25),
          
          axis.title.y     = element_text(size=25 ) 
       )

```

### fusion tabulations

```{r fig=TRUE,fig.width=10, fig.height=5, echo=FALSE, include=TRUE}

tallyfusion.pid_tb$fusion_positive = round ( tallyfusion.pid_tb$fusion_positive, 2)
kable( tallyfusion.pid_tb[ , c(
  "cancer.subtype" 
  ,"Unique.By.Patient" 
  ,"Unique.by.Cancer.subtype" 
  , "Unique.5", "Unique.3"
  
  ,"Median" , "Ave", "Min" ,"Max"  
  ,"is.diag" ,"is.drug" ,"is.novel" ,"is.obsv" ,"fusion_positive"
)
 ] , format = "html" , row.names = F, caption="", table.attr = "style='width:40%;'" ) %>% kable_classic(full_width = T, position = "center", ) 


#kable( tallyfusion.pid_tb[ , 1:13] , format = "html" , row.names = F, caption="", table.attr = "style='width:40%;'" ) %>% kable_classic(full_width = T, position = "center", )  %>%  save_kable("fusion1B.pdf")

```

### Percent with fusions 

```{r fig=TRUE,fig.width=3, fig.height=5, echo=FALSE, include=TRUE}
fusion_percent
```

### fusion is.x pie

```{r fig=TRUE,fig.width=7, fig.height=7, echo=FALSE, include=TRUE}
fusion.pie

```

```{r}
os_tab = fusion.in2[ fusion.in2$cancer.subtype == "OS" & fusion.in2$type != "no.fusion", ]
os_tab2 = tidyr::extract(os_tab, fusion, c("left", "right"), "(.*)-(.*)")


os_tab = data.frame ( table ( os_tab$fusion ) )
os_tab = os_tab[ order ( -os_tab$Freq), ]

os_right = data.frame ( table ( os_tab2$right ) )
os_right = os_right[ order ( -os_right$Freq), ]



```

```{r}
# functions for ranking facet
reorder_within <- function(x, by, within, fun = mean, sep = "___", ...) {
  new_x <- paste(x, within, sep = sep)
  stats::reorder(new_x, by, FUN = fun)
}

scale_x_reordered <- function(..., sep = "___") {
  reg <- paste0(sep, ".+$")
  ggplot2::scale_x_discrete(labels = function(x) gsub(reg, "", x), ...)
}






```



```{r}

total_subcancer_inframe = data.frame(table ( fusion.in2[!duplicated ( fusion.in2$patient.id ) ,   ]$cancer.subtype)) 
colnames ( total_subcancer_inframe ) = c( "cancer.subtype", "Freq")

# study left right frequency 



left_right = tidyr::extract(fusion.in2, fusion, c("left", "right"), "(.*)-(.*)")
fusion.in2$left = left_right$left
fusion.in2$right = left_right$right

left_fusion =    fusion.in2[!duplicated ( paste ( fusion.in2$patient.id, fusion.in2$left ) ), ] # important to make the patients unique


this.left_fusion_tab2 =  left_fusion  %>% dplyr::group_by (cancer.subtype, left) %>%
  #this.left_fusion_tab2 =  this.left_fusion_tab %>% dplyr::group_by (cancer.subtype, gene) %>%
  dplyr::summarise(
    total = n ( ) # 
  ) %>% data.frame()

this.left_fusion_tab2 = this.left_fusion_tab2[ order ( this.left_fusion_tab2$total, decreasing = T), ]

# none fusions will have left that is NA so need to remove 
this.left_fusion_tab2 = this.left_fusion_tab2[ !is.na(this.left_fusion_tab2$left), ]

fusion_left.out = data.frame()

for ( ut in unique ( this.left_fusion_tab2$cancer.subtype)){
  ut = this.left_fusion_tab2[this.left_fusion_tab2$cancer.subtype == ut, ]
  ut = ut [ ut$total >= 2, ]
  fusion_left.out = rbind ( head ( ut, 10),fusion_left.out )
}



#total_subcancer

fusion_left.out$total =  as.numeric( fusion_left.out$total) 
fusion_left.out = merge ( fusion_left.out, total_subcancer, by="cancer.subtype")
fusion_left.out$percent = round ( fusion_left.out$total / fusion_left.out$Freq, 2)

fusion_left.plot =  ggplot( fusion_left.out  , aes(x=reorder_within ( left, percent , cancer.subtype) , y=percent, fill=cancer.subtype   )) +
  geom_bar(stat = 'identity', position = 'stack') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 7.5 ),
        axis.text.x = element_text(angle = 0, size= 15 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12)
        , plot.title = element_text(size = 12, face = "bold")
  ) + 
  scale_x_reordered() +
  scale_fill_manual( values= subcolor ) + coord_flip() + ggtitle ( "Recurrent 5' Gene Fusion") +
  #scale_y_continuous(breaks = c(.1,.2,.3,.4,.5,.6,.7,.8,.9,1) ) + # plot in intergers
  scale_y_continuous(breaks = c(.2,.4,.6,.8,1) , expand = expansion(mult = c(0, .1))) +
  facet_grid(cancer.subtype~., scales = "free" , space = "free" ) 



############### Right version
right_fusion =   fusion.in2[!duplicated ( paste ( fusion.in2$patient.id, fusion.in2$right ) ), ]  # important to make the patients unique


this.right_fusion_tab2 =  right_fusion  %>% dplyr::group_by (cancer.subtype, right) %>%
  #this.right_fusion_tab2 =  this.right_fusion_tab %>% dplyr::group_by (cancer.subtype, gene) %>%
  dplyr::summarise(
    total = n ( ) # 
  ) %>% data.frame()

this.right_fusion_tab2 = this.right_fusion_tab2[ order ( this.right_fusion_tab2$total, decreasing = T), ]


this.right_fusion_tab2 = this.right_fusion_tab2[ !is.na(this.right_fusion_tab2$right), ]

fusion_right.out = data.frame()

for ( ut in unique ( this.right_fusion_tab2$cancer.subtype)){
  ut = this.right_fusion_tab2[this.right_fusion_tab2$cancer.subtype == ut, ]
  ut = ut [ ut$total >= 2, ]
  fusion_right.out = rbind ( head ( ut, 10),fusion_right.out )
}




fusion_right.out$total =  as.numeric( fusion_right.out$total) 
fusion_right.out = merge ( fusion_right.out, total_subcancer, by="cancer.subtype")
fusion_right.out$percent = round ( fusion_right.out$total / fusion_right.out$Freq, 2)

fusion_right.plot =  ggplot( fusion_right.out  , aes(x=reorder_within ( right, percent , cancer.subtype) , y=percent, fill=cancer.subtype   )) +
  geom_bar(stat = 'identity', position = 'stack') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 7.5 ),
        axis.text.x = element_text(angle = 0, size= 15 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12)
        , plot.title = element_text(size = 12, face = "bold")
  ) + 
  scale_x_reordered() +
  scale_fill_manual( values= subcolor ) + coord_flip() + ggtitle ( "Recurrent 3' Gene Fusion") +
  # scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) + # plot in intergers
  #scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  scale_y_continuous(breaks = c(.2,.4,.6,.8,1) , expand = expansion(mult = c(0, .1))) + 
  facet_grid(cancer.subtype~., scales = "free" , space = "free" ) 



################## 

```

```{r}

library  ( ggsci )
# 5.12.22 add cosmic annotation and tabulate for fusion 3' gene 
# ratio of TSG to oncogene
dim ( fuse2[fuse2$frameness == "in.frame", ])

fuse2$is.cancer = ifelse ( fuse2$right %in% genelist$gene, "yes", "no")

fusion_cosmic_right =  fuse2[fuse2$frameness == "in.frame" | fuse2$frameness == "out.of.frame" , ]  %>% dplyr::group_by(frameness,onco.tsg.2) %>%
  dplyr::summarise(
    Total = n()  
  ) %>% mutate (
    onco.tsg.2 =ifelse(onco.tsg.2 == "", "no_annt", onco.tsg.2), 
    onco.tsg.2 = gsub ( "0", "not_cancer", onco.tsg.2), 
    onco.tsg.2 = ifelse(onco.tsg.2 == "TSG,ONCOGENE", "ONCOGENE,TSG", onco.tsg.2),
    onco.tsg.2 = factor ( onco.tsg.2, levels = c("not_cancer" , "no_annt",  "ONCOGENE,TSG",  "ONCOGENE" , "TSG") )
  )%>%
  ggplot( .  , aes(x=frameness, y=Total, fill=onco.tsg.2   )) +
  geom_bar(position="fill", stat="identity") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="right",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=15), 
        legend.text      =element_text(size=15),
        legend.title =  element_blank()
        , plot.title = element_text(size = 12, face = "bold")
  ) + coord_flip() + scale_fill_manual(values =c(  ggsci::pal_jco()(10) )  )


### getting some examples. 


## this needs to get fix 


fuse2.1 = merge ( fuse2, rna.key[ , c("RNAseq.id", "patient.id", "patient.institution" )], by="RNAseq.id")
fuse2.1$frameness =  ifelse ( fuse2.1$frameness == "." & fuse2.1$annt1 == "exons" & fuse2.1$annt2 == "5UTRs", "CDS-UTR", fuse2.1$frameness   )
fuse2.1$frameness =  ifelse ( fuse2.1$frameness == "." & fuse2.1$annt1 == "5UTRs" & fuse2.1$annt2 == "exons", "UTR-CDS", fuse2.1$frameness   )

fuse2.1[fuse2.1$right == "TP53" & (fuse2.1$frameness == "in.frame" | fuse2.1$frameness == "out.of.frame")  , ] 

fuse_tp53 = fuse2.1[fuse2.1$right == "TP53"   , ]
fuse_tp53
fuse_tp53$reading_frame = fuse_tp53$frameness  

fuse_tp53 = fuse_tp53[ order ( fuse_tp53$frameness), ]

kable ( fuse_tp53[ , c ("patient.id" ,"RNAseq.id"  ,"cancer.subtype", "fusion"  , "reading_frame" , "family.1", "family.2" , "onco.tsg.1" , "onco.tsg.2" , "is.cancer.1" , "is.cancer.2" ) ]
        , format="html", row.names=F, table.attr = "style='width:10%;'" ) %>%
  kable_classic(full_width = F, position = "center",)  %>%
  column_spec(1, bold = T, border_right = T)  


kable ( fuse_tp53[ , c ("patient.id" ,"RNAseq.id"  ,"cancer.subtype", "fusion"  , "reading_frame" , "SV_RESULT"  ) ]
        , format="html", row.names=F, table.attr = "style='width:10%;'" ) %>%
  kable_classic(full_width = F, position = "center",)  %>%
  column_spec(1, bold = T, border_right = T)  %>% 
  footnote(
    symbol  = c("SV - means no WGS match")) 




fuse_tp53[ which ( fuse_tp53$RNAseq.id == "AI1"), ]$frameness 

out [ out$RNAseq.id %in% fuse2.1[fuse2.1$right == "TP53"   , ]$RNAseq.id & out$gene == "TP53", ] 


## more tp53 for left gene now 

fuse_tp53_left = fuse2.1[fuse2.1$left == "TP53"   , ]
fuse_tp53_left
fuse_tp53_left$reading_frame = fuse_tp53_left$frameness  

fuse_tp53_left = fuse_tp53_left[ order ( fuse_tp53_left$frameness), ]


kable ( fuse_tp53_left[, c ("patient.id" ,"RNAseq.id"  ,"cancer.subtype", "fusion"  , "reading_frame" , "SV_RESULT"  ) ]
        , format="html", row.names=F, table.attr = "style='width:10%;'" ) %>%
  kable_classic(full_width = F, position = "center",)  %>%
  column_spec(1, bold = T, border_right = T)  %>% 
  footnote(
    symbol  = c("SV - means no WGS pair")) 


out [ out$gene == "TP53" & out$RNAseq.id %in% fuse2.1[fuse2.1$left == "TP53" , ]$RNAseq.id , ]

# what other tsg might be influence by out.of.frame or utr swaps? 
temp = fuse2.1[fuse2.1$onco.tsg.2 == "TSG" & fuse2.1$frameness != "in.frame" , ] 

out [ out$ggene == "TP53" & out$RNAseq.id %in% fuse2.1[fuse2.1$left == "TP53" , ]$RNAseq.id , ]

# need to look up TP53 and see how this works. 

fuse2.1[fuse2.1$onco.tsg.2 == "TSG" & ( fuse2.1$frameness == "out.of.frame")  , ] 
fuse2.1[fuse2.1$onco.tsg.2 == "ONCOGENE" & ( fuse2.1$frameness == "in.frame")  , ] 

# more info 
fusion[fusion$RNAseq.id == "BD4" & grepl ( "-PTMA", fusion$fusion), c("RNAseq.id", "frameness", "fusion", "FRPM", "is.cancer.1", "annt1", "annt2", "domain1", "domain2") ]
fusion[grepl ( "FUS-FEV", fusion$fusion), c("RNAseq.id", "frameness", "fusion", "FRPM", "is.cancer.1", "annt1", "annt2", "domain1", "domain2") ]

```


### fusion Left Right tabulations 

```{r}

fusion_left.plot | fusion_right.plot

```

```{r}
# remove fuse1's cancer subtyp bc its not accurately reflecting the key 
fuse2  = fusion[ , c("RNAseq.id","fusion","frameness","SV_RESULT","cancer.subtype", is.stuff,"left", "right", "uid","annt1","annt2", "domain1", "domain2",  "family.1", "family.2", "onco.tsg.1", "onco.tsg.2", "is.cancer.1", "is.cancer.2"  )]
fuse2$cancer.subtype = NULL 

fuse2 = merge ( fuse2, rna.key[ , c("RNAseq.id", "cancer.subtype")], by="RNAseq.id" )

family_right = data.frame ( table ( fuse2 [ , c( "cancer.subtype" , "frameness" , "family.2")] ) )
family_right = family_right[ order ( -family_right$Freq), ]

family_right = family_right[family_right$family.2 != "0", ]

head ( family_right[family_right$cancer.subtype == "OS" & family_right$frameness == "in.frame",] , 20 )

```



```{r}

plot_domain <- function ( framethis , fusion_domain, type=NA ){
  



if ( !is.na ( type )) {
  ttt = temp.all[ temp.all$type %in% type, ] 
  ttt$uid = paste ( ttt$RNAseq.id, ttt$fusion)
  
  fusion_domain$final_left = merge (fusion_domain$final_left, ttt [ , c("type", "uid" )], by = "uid"  )
  fusion_domain$final_right = merge (fusion_domain$final_right, ttt [ , c("type", "uid" )], by = "uid"  )
    
  }

left_domain = fusion_domain$final_left
left_domain = left_domain[!duplicated ( left_domain$dup), ]
left_domain = left_domain[left_domain$Rnaseq.id %in% unique ( rna.key$RNAseq.id), ]
#left_domain$fusion= stringr::str_match( left_domain$uid, " (.+)")[,2]
left_domain$Rnaseq.id = NULL 
left_domain = merge (left_domain, fusetemp, by="uid", )
left_domain$cancer.subtype = NULL 
left_domain = merge ( left_domain, rna.key[ , c ( "RNAseq.id", "cancer.subtype") ], by="RNAseq.id")


left_tab = data.frame ( table ( left_domain [ left_domain$frameness %in% framethis, ]$proteinDomainName  ))
#left_tab = data.frame ( table ( left_domain $proteinDomainName  ))
left_tab = left_tab[ order ( left_tab$Freq, decreasing = T ), ]
head ( left_tab )




left_plot = left_domain [ left_domain$proteinDomainName %in% 
                        as.character ( left_tab$Var1[1:10] ) & left_domain$frameness %in% framethis, ] 

left_df = left_plot 

left_plot =left_plot %>% dplyr::group_by( cancer.subtype, proteinDomainName ) %>%
  dplyr::summarise(
    freq = n()  
  )  %>% data.frame ( stringsAsFactors = F )

left_plot$proteinDomainName = factor ( left_plot$proteinDomainName, levels =  rev ( as.character ( left_tab$Var1[1:10] ) )  )



left_plot = ggplot( left_plot  , aes(x=proteinDomainName , y=freq, fill=cancer.subtype   )) +
   geom_bar(stat = 'identity', position = 'stack') +
    theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black")) +
    ylab("") + 
    xlab("") + 
    theme(legend.position="none",  legend.key = element_blank(),
          
          axis.text.y = element_text(size= 18 ),
          axis.text.x = element_text(angle = 0, size= 18 ),
          axis.title.x = element_text(size=12),
          axis.title.y     = element_text(size=20), 
          legend.text      =element_text(size=12)
          , plot.title = element_text(size = 30, face = "bold")
    ) + 
  scale_fill_manual( values= subcolor ) + coord_flip() + ggtitle ( "5' Domain") +
  scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) +
   scale_x_discrete(labels = wrap_format(30))



### plot right 


right_domain = fusion_domain$final_right
right_domain = right_domain[!duplicated ( right_domain$dup), ]
right_domain = right_domain[right_domain$Rnaseq.id %in% unique ( rna.key$RNAseq.id), ]
#right_domain$fusion= stringr::str_match( right_domain$uid, " (.+)")[,2]
right_domain$Rnaseq.id = NULL 
right_domain = merge (right_domain, fusetemp, by="uid", )
right_domain$cancer.subtype = NULL 
right_domain = merge ( right_domain, rna.key[ , c ( "RNAseq.id", "cancer.subtype") ], by="RNAseq.id")


right_tab = data.frame ( table ( right_domain [ right_domain$frameness %in% framethis, ]$proteinDomainName  ))
right_tab = right_tab[ order ( right_tab$Freq, decreasing = T ), ]
head ( right_tab )




right_plot = right_domain [ right_domain$proteinDomainName %in% 
                            as.character ( right_tab$Var1[1:10] ) & right_domain$frameness %in% framethis, ] 

right_df = right_plot

right_plot =right_plot %>% dplyr::group_by( cancer.subtype, proteinDomainName ) %>%
  dplyr::summarise(
    freq = n()  
  )  %>% data.frame ( stringsAsFactors = F )

right_plot$proteinDomainName = factor ( right_plot$proteinDomainName, levels =  rev ( as.character ( right_tab$Var1[1:10] ) )  )



right_plot = ggplot( right_plot  , aes(x=proteinDomainName , y=freq, fill=cancer.subtype   )) +
  geom_bar(stat = 'identity', position = 'stack') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 18 ),
        axis.text.x = element_text(angle = 0, size= 18 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12)
        , plot.title = element_text(size = 30, face = "bold")
  ) + 
  scale_fill_manual( values= subcolor ) + coord_flip() + ggtitle ( "3' Domain") +
  scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) +
  scale_x_discrete(labels = wrap_format(30))

return ( 
  list ( 
    left_tab = left_tab 
    ,right_tab = right_tab
    , left_plot = left_plot
    , right_plot = right_plot 
    , left_df = left_df
    , right_df = right_df
    )
  )

}


```

```{r}

library ( scales)


#### finding domains 

fusetemp = fuse1 
fusetemp$uid = paste ( fusetemp$RNAseq.id, fusetemp$fusion)

# study fusions based on domain output from arriba ( modified version see desktop, took output of fusions and ran arriba to get the domain Dropbox\Oncology Lab\RNA-SEQ\2021\fusion.mod\arriba_v2.1.0\fusion_domain )
fusion_domain = readRDS( "../domainsresults_2022.rds"  )
# manaul relabeling 
fusion_domain$final_left$proteinDomainName = 
  gsub ( "(a.k.a. RRM, RBD, or RNP domain)"
         , "RRM, RBD, or RNP"
         , fusion_domain$final_left$proteinDomainName) 


fusion_domain$final_right$proteinDomainName = 
  gsub ( "(a.k.a. RRM, RBD, or RNP domain)"
         , "RRM, RBD, or RNP"
         , fusion_domain$final_right$proteinDomainName)

fusion_domain$final_left = fusion_domain$final_left[fusion_domain$final_left$Rnaseq.id %in% rna.key$RNAseq.id, ]
fusion_domain$final_right = fusion_domain$final_right[fusion_domain$final_right$Rnaseq.id %in% rna.key$RNAseq.id, ]

# here is not necessary to remove pid dupliates because we want to know if what the consequences are of the fusion. 
# this is why the the function will remove uid, which is id_fusion

in_frame_domain = plot_domain( framethis = 'in.frame', fusion_domain=fusion_domain )
out_frame_domain = plot_domain( framethis = 'out.of.frame', fusion_domain=fusion_domain  )

in_frame_domain_novel = plot_domain( framethis = 'in.frame', fusion_domain=fusion_domain, type="is.novel" )
out_frame_domain_novel = plot_domain( framethis = 'out.of.frame', fusion_domain=fusion_domain, type="is.novel"  )




in_frame_domain$right_plot


# maybe just with 3' domains. 

in_frame_domain_right_plot = in_frame_domain$right_plot + ggtitle ( "3' Domain, in-frame")
out_frame_domain_right_plot =  out_frame_domain$right_plot  + ggtitle ( "3' Domain, out-of-frame")



 ( in_frame_domain$right_plot + ggtitle ("3' Domain in.frame") )   + ( out_frame_domain$right_plot + ggtitle ("3' Domain out.of.frame") ) 

( in_frame_domain_novel$right_plot + ggtitle ( "3' Domain Novel") )  +  ( in_frame_domain$right_plot + ggtitle ( "3' Domain Known") )


```

### domain of fusions 
```{r}




# study the fusions. 
in_domain = in_frame_domain$right_df

in_domain = data.frame ( table ( in_domain[ , c("proteinDomainName", "fusion")]))
in_domain = in_domain[in_domain$Freq != 0, ]



out_domain = out_frame_domain$right_df
out_domain = data.frame ( table ( out_domain[ , c("proteinDomainName", "fusion")]))
out_domain = out_domain[out_domain$Freq != 0, ]


# draw network below. 


```


```{r}

draw_network1 <- function (df, draw="0", layout = "fr" ){
  
df$proteinDomainName = as.character ( df$proteinDomainName )
df$fusion = as.character( df$fusion)

nodes<- create_node_df(n= length(unique ( c ( df$proteinDomainName, df$fusion) ) ),
                       style= "filled",
                       label= unique ( c ( df$proteinDomainName, df$fusion) ),
                       fillcolor= "#E6BB58",
                       color = "grey",
                       shape= "circle", fontsize= 10, fontcolor="black", 
                       height  = .5 )


# data frame of edges
edges<-create_edge_df(from = match (df$proteinDomainName , ( nodes$label) ) ,
                      to=match (df$fusion , ( nodes$label) )  ,
                      label = df$Freq,
                      rel= "related",
                      color= "black", 
                      fontsize= 10 
                      #len = c(1200)
                      
                      )
edges




nodes$fillcolor = ifelse ( 
  nodes$label %in% unique ( df$proteinDomainName ), "#A7CD52", nodes$fillcolor
  )

nodes$fontcolor = ifelse ( 
  nodes$label %in% unique ( df$proteinDomainName ), "#1D9EE9", nodes$fontcolor
)


nodes$height = ifelse ( 
  nodes$label %in% unique ( df$proteinDomainName ), 1, nodes$height
)





graph<-create_graph(nodes, edges)
render_graph(graph, layout = layout )

#graph %>%
# export_graph(
 #  file_name = "graph.pdf",
  # width = 2500,
  # height = 2000
   
  #)



return ( graph )





}

```

```{r}

## modifying 3' networks to make it look better. 
in_frame_domain$right_plot + ggtitle ("3' Domain in.frame") 
in_domain_net = draw_network1(in_domain, layout="kk")
render_graph(in_domain_net, layout = "fr", height=1500, width=2000  )

out_frame_domain$right_plot + ggtitle ("3' Domain out.of.frame")
out_domain_net = draw_network1(out_domain, layout="kk")
render_graph(out_domain_net, layout = "fr", height=1500, width=2000 )

```




```{r}

## plotting family and domains of fusions 
# find top domain disruption &  family of genes 


## Now lets see which family of genes are involved 
frameness = "in.frame"

draw_family <- function( frameness=frameness ){

f1 = fusion[fusion$family.1 != "0" & fusion$frameness == frameness & !duplicated ( paste ( fusion$RNAseq.id, fusion$fusion)), ]   %>% dplyr::group_by( family.1 ) %>%
  dplyr::summarise(
    freq = n()  
  ) %>% arrange ( desc ( freq) ) %>% data.frame ( stringsAsFactors = F)

head ( f1, 10 )


f2 = fusion[fusion$family.2 != "0" & fusion$frameness == frameness & !duplicated ( paste ( fusion$RNAseq.id, fusion$fusion)) , ]   %>% dplyr::group_by( family.2 ) %>%
  dplyr::summarise(
    freq = n()  
  ) %>% arrange ( desc ( freq) ) %>% data.frame ( stringsAsFactors = F)





# plot stack plot disruption 
f1_order = as.character ( f1$family.1[1:10] )

f1 = fusion[fusion$family.1 != "0" & fusion$frameness == frameness , c("RNAseq.id","fusion", "family.1")] 
f1 = merge ( f1, rna.key [ , c("RNAseq.id","cancer.subtype")], by="RNAseq.id")

f1 = f1[f1$family.1 %in% f1_order, ]

f1$family.1 = factor ( f1$family.1, rev (f1_order) )

family_1plot = f1 %>% dplyr::group_by( cancer.subtype, family.1 ) %>%
  dplyr::summarise(
    freq = n()  
  ) %>% ggplot(., aes(x = family.1, y=freq,  fill = cancer.subtype)) + 
    scale_fill_manual(values =  subcolor ) +
    geom_bar(stat = 'identity', position = 'stack') +
    theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black")) +
    ylab("") + 
    xlab("") + coord_flip() +   theme(legend.position = "none" )  +  ggtitle ( "5' Gene Family") +
    scale_x_discrete(labels = wrap_format(30))




#########################################

f2_order = as.character ( f2$family.2[1:10] )

f2 = fusion[fusion$family.2 != "0" & fusion$frameness == frameness , c("RNAseq.id","fusion", "family.2")] 
f2 = merge ( f2, rna.key [ , c("RNAseq.id","cancer.subtype")], by="RNAseq.id")

f2 = f2[f2$family.2 %in% f2_order, ]
f2$family.2 = factor ( f2$family.2, rev (f2_order) )


family_2plot = f2 %>% dplyr::group_by( cancer.subtype, family.2 ) %>%
  dplyr::summarise(
    freq = n()  
  ) %>% ggplot(., aes(x = family.2, y=freq,  fill = cancer.subtype)) + 
    scale_fill_manual(values =  subcolor ) +
    geom_bar(stat = 'identity', position = 'stack',  colour = 'white') +
    theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black")) +
    ylab("") + 
    xlab("") + coord_flip() +   theme(legend.position = "right" )  +  ggtitle ( "3' Gene Family") +
  scale_x_discrete(labels = wrap_format(30))


return ( list  ( p1=family_1plot , 
                  p2=family_2plot ) 
         
         )

}

familyplots_inframe = draw_family(("in.frame"))
familyplots_outofframe = draw_family(("out.of.frame"))
```

### Gene family 
* The top plots the family that the gene is part of 
* The lower plots are domains that are disrupted as a result of the fusion breakage 

```{r fig=TRUE,fig.width=12, fig.height=10, echo=FALSE, include=TRUE}
(familyplots_inframe$p1  + familyplots_inframe$p2 ) /
(familyplots_outofframe$p1 + familyplots_outofframe$p2 )  
```




```{r}
piecc = c( higher_Sig = "#cf4051", higher_noSig = "#e0c126", lower_Sig = "#83bf43", lower_noSig= "#6ec0db" )
trueseq.key = rna.key[ !rna.key$RNAseq.id %in% failedChem, ]
fusion_truseq = fusion[ !fusion$RNAseq.id %in% failedChem,  ]
### here we want to test the the expression of the right gene with and without fusion 
out.frame = fusion_truseq[ , c("RNAseq.id","right", "frameness")]
out.frame = merge ( out.frame, trueseq.key[, c("RNAseq.id","cancer.subtype")], by="RNAseq.id")
out.frame = out.frame[ out.frame$right %in% unique ( c(tsg, tsg.onco)) & out.frame$frameness == "out.of.frame", ]


in.frame = fusion_truseq[ , c("RNAseq.id","right", "frameness")]
in.frame = merge ( in.frame, trueseq.key[, c("RNAseq.id","cancer.subtype")], by="RNAseq.id")
in.frame = in.frame[ in.frame$right %in% unique ( c(oncogene, tsg.onco)) & in.frame$frameness == "in.frame", ]

in.frame  %>% dplyr::group_by( right, cancer.subtype ) %>%
  dplyr::summarise(
    freq = n()  
  ) %>% arrange ( desc ( freq) ) %>% data.frame ( stringsAsFactors = F)


fusion_truseq$frameness2 = as.character ( 
  sapply(fusion_truseq$frameness, function(x)
  {
  
    if ( !x %in% c ( "out.of.frame", "in.frame")  ){
      x= "unknown"
    }
    return ( x )
  }
    )
  )
# find the most common right 
common.right  = fusion_truseq %>% dplyr::group_by( right, frameness2 ) %>%
  dplyr::summarise(
    freq = n()  
  ) %>% arrange ( desc ( freq) ) %>% data.frame ( stringsAsFactors = F)

# find the most common left 

common.left  = fusion_truseq %>% dplyr::group_by( left, frameness2 ) %>%
  dplyr::summarise(
    freq = n()  
  ) %>% arrange ( desc ( freq) ) %>% data.frame ( stringsAsFactors = F)



###  TEST single plot samples with and without fusion_truseq. 
# not this does not consider fusion_truseq type 

test.gene = "TP53"
grepl_this = paste0('^', test.gene, '-|-', test.gene, '$')
#test.sample = fusion_truseq[fusion_truseq$right == test.gene, c("RNAseq.id", "fusion")]
# change below criteria to meet your needs 
test.sample = fuse2.1 [ fuse2.1$right == test.gene & grepl ( "out|UTR", fuse2.1$frameness), c("RNAseq.id", "fusion") ]
test.sample = fuse2.1 [ grepl ( "^TP53-|-TP53$", fuse2.1$fusion) & grepl ( "out|UTR", fuse2.1$frameness), c("RNAseq.id", "fusion") ]
test.sample = fuse2.1 [ grepl ( grepl_this , fuse2.1$fusion) , c("RNAseq.id", "fusion") ]

test.tpm = melt ( tpm[tpm$gene==test.gene, test.sample$RNAseq.id] )
test.tpm$group = "fusion"


test.tpm = merge ( test.tpm, trueseq.key[trueseq.key$cancer.type %in% c("Solid", "CNS") , c("RNAseq.id", "cancer.subtype") ], by.x="variable",  by.y="RNAseq.id" )



nofusion.tpm = melt ( tpm[tpm$gene==test.gene, setdiff(trueseq.key$RNAseq.id, test.sample$RNAseq.id) ] ) 
nofusion.tpm$group = "no.fusion"
nofusion.tpm$cancer.subtype = "none"

testexpfusion = rbind ( nofusion.tpm, test.tpm )
testexpfusion$value = log2 ( testexpfusion$value + 1 )
testexpfusion[1] = "sample"
testexpfusion$fusion = test.gene




subcolor2 = subcolor 
subcolor2["none"] = "#d2d4d6"
## plot singles 

single_cmp_fusion = ggplot(testexpfusion, aes(y=value, x=group))+ 
  geom_violin(alpha = 0.5) +
  geom_point(aes(fill = cancer.subtype), size = 4, shape = 21, position = position_jitterdodge(dodge.width = 0.25), alpha = 0.75 )+
  theme_bw() +
  ylab("log2 ( tpm +1 ) ") +
  xlab("") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 12 ),
        axis.text.x = element_text(angle = 0, size= 12 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") + xlab ( "") +  ylab ( "") +
  scale_fill_manual(values = subcolor2 )

  my_add_pval(single_cmp_fusion, pairs = list(c(1, 2)), test='wilcox.test', textsize = 10) + ggtitle ( paste ( "Fusion vs No Fusion", test.gene ))
  
  
  
# how many fusions are TSG?   
 tsg_fusions = data.frame(); 
 # don't use tsg because this has fusion as well so could muddy the picture? 
 
for ( test.gene in  tsg ) { 

  grepl_this = paste0('^', test.gene, '-|-', test.gene, '$')
  test.sample = fuse2.1 [ grepl ( grepl_this , fuse2.1$fusion) , c("RNAseq.id", "fusion") ]
  tsg_fusions = rbind ( tsg_fusions, data.frame ( gene=test.gene, total=nrow ( test.sample)) )
}
 
 tsg_fusions = tsg_fusions[ order ( -tsg_fusions$total), ]
  
```




```{r}
# ITD chunk 
itd2 = fusion [ fusion$RNAseq.id %in% rna.key$RNAseq.id &  fusion$type == "itd", ]
# View ( itd2 [ itd2$gene1 == "EWSR1", ] )
itd2 = merge ( itd2[ , c("RNAseq.id", "gene1", "type", "frameness", "algo")], rna.key[, 1:10], by="RNAseq.id" )

itd3 = itd2 



itd_this = unique ( itd2$gene1) # ogiginall only wanted "FLT3", "BCOR", "EGFR"  but decided to plot everything. 
itd_this= itd_this[itd_this %in% genelist$gene ]

itd_rank = itd2[itd2$gene1 %in%  itd_this, ] %>% dplyr::group_by( gene1 ) %>%
  dplyr::summarise(
    freq = n()  
    
  )   %>% arrange(desc(freq)) 


itd2 = itd2[itd2$gene1 %in% itd_this, ] %>% dplyr::group_by( gene1,  cancer.subtype ) %>%
  dplyr::summarise(
    freq = n()  
  ) 

itd2$gene1 = factor (itd2$gene1  ,  levels=rev ( itd_rank$gene1  ) )


itd2_plot= ggplot(itd2 , aes(x=gene1, y=freq ,  fill=cancer.subtype   )) +
  geom_bar(stat="identity",  width = 1, position = "stack") +  
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
    theme(axis.title.x=element_blank(),
        #axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
        , legend.position = "none"
        , axis.text = element_text(size= 22 )
        , axis.title.y     = element_text(size=22)
        #, legend.position = "none"
  ) +
  xlab("ITD") + 
  scale_fill_manual( values= subcolor ) + coord_flip()




```

```{r}

top6 = read.xlsx("key_QC.xlsx")
top6 = top6[ top6$RNAseq.id %in%rna.key[rna.key$qc == "PASS", ]$RNAseq.id , ] 
match = top6$match
match = ifelse ( is.na(match), "NA", match )
match = ifelse ( match == "1", "match", match )
match = ifelse ( match == "0", "not_match", match )

match = data.frame ( table ( match) ) 
colnames ( match ) = c("Top6", "Total")

match =  match[match(match$Top6 , c("match", "not_match", "NA")),]
match_total =  sum ( match[match$Top6 %in%  c( "match", "not_match" ) , ]$Total )
match$percent = round ( match$Total / match_total, 2 ) 
match$percent = ifelse ( match$Top6 == "NA", "", match$percent )

kable( match , format = "html" , row.names = F, caption="", table.attr = "style='width:40%;'" ) %>% kable_classic(full_width = T, position = "center", )

```



### start







```{r}
# do sv + fusion 

# redo fusion plots with stack instead 
# 
fusion.sv =  reshape2::melt ( fuse1[ , c( "RNAseq.id","fusion" , "SV_RESULT")], id.vars = c("RNAseq.id","fusion" ))

this.sv  = fuse1[ , c( "RNAseq.id","fusion" , "SV_RESULT")]  
this.sv[this.sv == "-" ]= "RNA_ONLY"

# ver d: collapse all sv 
this.sv$SV_RESULT = ifelse( !this.sv$SV_RESULT  %in% c( "RNA_ONLY", "NO_WGS" ), "WGS.evidence", this.sv$SV_RESULT  )
 unique ( this.sv$SV_RESULT)
this.sv$SV_RESULT = gsub ( "NO_WGS", "no-wgs", this.sv$SV_RESULT ) # these sample do not have WGS pair 


this.sv = plyr::count(this.sv, c("RNAseq.id","SV_RESULT"))







zero = setdiff ( rna.key$RNAseq.id, unique ( this.sv$RNAseq.id)) 

for ( z in zero ){
  this.sv = rbind ( this.sv, data.frame ( RNAseq.id=z,SV_RESULT="No.Fusion",freq=0) )     
}

table ( this.sv$SV_RESULT)
colnames ( this.sv) = c("RNAseq.id", "SV_RESULT","freq")
this.sv = merge ( this.sv, rna.key[ , c("RNAseq.id", "cancer.subtype")], by = "RNAseq.id")


this.sv$cancer.subtype = factor ( this.sv$cancer.subtype, levels = cancer.subtype.order)






out.sv.media = this.sv %>% dplyr::group_by (RNAseq.id) %>%
  dplyr::summarise(
    total = sum ( freq  ) # count occurence but leave out pizzly
    ,cancer.subtype = unique ( cancer.subtype)
  ) %>% data.frame()



out.sv.media = out.sv.media %>% dplyr::group_by (cancer.subtype) %>%
  dplyr::summarise(
    med = median ( total  ) # count occurence but leave out pizzly
  ) %>% data.frame()





#unique ( this.sv$SV_RESULT )

this.sv$SV_RESULT = factor ( this.sv$SV_RESULT, levels = c(
  "RNA_ONLY", "WGS.evidence",  "No.Fusion" , "no-wgs"
  ) )

cc3 = brewer.pal(3, "Set3")
cc3 = c( cc3[1:2], "white", "grey")

# paper plot fusion

out.total.bar.sv = ggplot(data=this.sv, aes(x=RNAseq.id, y=freq ,  fill=SV_RESULT  )) +
  geom_bar(stat="identity",  width = 1, position = "stack") +  
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("Total Fusions") + 
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))  +
  facet_grid(~cancer.subtype, scales = "free" , space = "free" ) +
  scale_fill_manual(values = c( cc3  ) ) +
  geom_hline(data = out.sv.media, aes(yintercept =med), color="grey", size=1)


# pie chart of this.sv 
#sv.freq = data.frame ( table ( this.sv$SV_RESULT), stringsAsFactors = F)

sv.freq <- this.sv %>% dplyr::group_by( SV_RESULT ) %>%
  dplyr::summarise(
    Freq = sum(freq)  
    
  ) %>% data.frame ( stringsAsFactors = F)

sv.freq = sv.freq[!sv.freq$SV_RESULT %in% c ( "No.Fusion", "no-wgs"), ]

sv.freq$Var1 = factor ( sv.freq$SV_RESULT, levels = levels( this.sv$SV_RESULT ))



msv= sv.freq %>%
  #group_by(type) %>%
  mutate(countT= sum(Freq)) %>%
  mutate(per=round(100*Freq/countT,2))%>% data.frame()


# this part is necessary to place the label rotationally correctly 

msv <- msv %>% 
  # group_by(type )  %>%
  mutate(
    cs = rev(cumsum(rev(per))),
    text_yp = per/2 + lead(cs, 1),
    text_yp = if_else(is.na(text_yp), per/2, text_yp)
  )  %>%  data.frame()

colnames ( msv )[1] = "Gene.fusion"


pie.sv = ggplot(msv, aes(x="", y=per, fill=Gene.fusion )) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0)   + 
  theme_void()   +
  geom_label_repel(
    aes(
      label = Freq, y = text_yp
    ), show_guide  = F
    
  ) + xlab("") + ylab("") +
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_rect(fill="lightblue")
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
  theme(
    strip.background = element_blank(),
    strip.text = element_blank()
  ) + theme_void()  +
   scale_fill_manual(values = cc3)


fusion.total.disease = merge ( fusion [ , c("RNAseq.id", "fusion", "left","right")], rna.key[ , c ( "RNAseq.id","cancer.subtype") ], by="RNAseq.id")


# how many fusion can be explained by at least one wgs event? 
fusion.wgsE = fusion.sv[fusion.sv$value != "NO_WGS", ]
fusion.wgsE$wgs.evidence = ifelse( fusion.wgsE$value != "-", "wgs", "no.wgs")



```


```{r}
# creat a druggable pie 

# druggable pie

drug.pi.fusion = data.frame ( table ( fuse1 [ fuse1$is.drug != "0", c("cancer.subtype", "fusion") ]))
drug.pi.fusion = drug.pi.fusion[drug.pi.fusion$Freq != "0", ]


drug.pi.fusion$cancer.subtype = as.character( drug.pi.fusion$cancer.subtype )
drug.pi.fusion$fusion = as.character( drug.pi.fusion$fusion )

drug.pi.fusion$cs = drug.pi.fusion$cancer.subtype
druggable.fusion = ggplot(drug.pi.fusion, aes(x = fusion, y=Freq,  fill = cancer.subtype)) + 
  scale_fill_manual(values =  subcolor[as.character ( drug.pi.fusion$cancer.subtype) ]  ) +
  geom_bar(stat = 'identity', position = 'stack',  colour = 'black') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") +
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(), 
        axis.text.x = element_text(angle = 0, size= 15 ), legend.position = "right"
    ) +  scale_y_continuous(breaks= unique(drug.pi.fusion$Freq)) + coord_flip() + 
  geom_text(aes(y = Freq , hjust = 3, label = fusion) ) 
```











```{r}
top6a = 0 
# save trueseq for further QC'ing matching samples 
wb <- createWorkbook()

addWorksheet(wb, 'top6')
writeData(wb, 'top6' , trueseq.key[ , c("RNAseq.id", "cancer.subtype", "diagnosis", "sample.type",   "top6" )]  , rowNames=T  )
if ( top6a == 1 ){
  saveWorkbook(wb, file = paste0("top6_review.xlsx"), overwrite = TRUE)
}
```

```{r}



est = readRDS( "ESTIMATE.rds" )
est = data.frame ( t ( est ) )
disease = "OS"
gene_check = c ( "ENPP1", "TMEM173", "CD274", "CTLA4" ) 

count = tpm [ gene_check, trueseq.key$RNAseq.id ]
count = rbind ( count, est[ , trueseq.key$RNAseq.id])
count = data.frame ( t ( count ))

disease_id = trueseq.key[ trueseq.key$cancer.subtype == "OS", ]$RNAseq.id

cor_imm = cor(count[disease_id, ], method = c( "spearman"))

est_stable = count 

count = log2 ( count  )

## plot correlations 

library("ggpubr")
gene_c = "ENPP1"
all_c = ggscatter(count, x = gene_c, y = "ImmuneScore", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "spearman",
          xlab = gene_c, ylab = "ImmunoScore", title = "ALL Sample")
os_c = ggscatter(count[disease_id, ] , x = gene_c, y = "ImmuneScore", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "spearman",
          xlab = gene_c, ylab = "ImmunoScore", title = "OS only" )

all_c / os_c 





```

```{r}
# heat map of top 6

# get updates 

## had to be manaully curated

top6 = read.xlsx("https://www.dropbox.com/s/vranmbn6oabodl7/top6_review.xlsx?dl=1")
top6 = top6 [ !is.na ( top6$percent_matched), ]
top6 = top6 [ order ( top6$diagnosis, top6$percent_matched, decreasing = T ), ]
top6$diagnosis = tolower(top6$diagnosis)

top6$absmatch = ifelse ( top6$percent_matched >=1, 1, round ( top6$percent_matched, 2))
# right annotation start only 
# first find first occurance 

first_occ = c()
for ( f in unique(top6$diagnosis)){
  first_occ = c( first_occ, match( f , top6[,"diagnosis"] ) )

}

ha = rowAnnotation(foo = anno_mark(at = first_occ , labels = top6$diagnosis[ first_occ]    ))


## add left annotations 
top6$type = gsub ( ".*_", "", top6$sample.type)
type = top6$type
type_color = setNames( c("#2ac9f5","#a6790f","grey","#dc4ef2"), unique ( top6$type ) ) 


match_color = setNames( c("#257d2d","black"), c ( "yes", "no" ) ) 
match_binary = ifelse ( top6$match > 0 , "yes", "no"  )


type_annt = rowAnnotation(
  type = type
  ,match = match_binary
  ,col = list ( 
    type= type_color
    ,match = match_color
  ) , annotation_name_gp= gpar(fontsize = 0, col="white" )
)



est_annt = est_stable
est_annt$RNAseq.id = row.names ( est_annt)
est_annt = est_annt[match( top6$RNAseq.id , est_annt$RNAseq.id ),]



set.seed(1)
immuno_annt = rowAnnotation (purity = est_annt$purity) 


# add grey to the color scale 
cgradient = rev ( brewer.pal(5, "Set1") )
cgradient[1] = "black"
cgradient[2] = "#b75ae8"
cgradient[3] = "#64b6d9"
cgradient[4] = "#6cd156"
cgradient[5] = "#c7225e"


 ComplexHeatmap::Heatmap( top6[ , "percent_matched", drop=F] 
                                    , name= paste ( " Diagnosis matching" )
                                    #, row_order = type_row_order
                                    
                                    #, column_order = colnames ( pthis$df)
                                    , cluster_columns = F
                                    
                                  , show_row_names = F
                                    #, row_labels = ens_gene
                                    , show_row_dend = F
                                    , show_column_names = F
                                 #   , col= c("#eff5cb", "#f5946e")
                                    , col = cgradient                             
                                    ,cluster_column_slices = F
                                    , cluster_rows =F
                                    ,row_title = NULL
                                    #, right_annotation = r_annt
                                    ,row_split =  top6$diagnosis
                                    , column_title = NULL  
                                    , left_annotation = immuno_annt
                                    ,  right_annotation = ha 
                                    
                                    
  )

 
 
 top6L = top6[ , c( "cancer.subtype", "absmatch", "percent_matched","RNAseq.id"), drop=F] 

 
 ## summarize 
 
top6_results =  top6L %>% 
  group_by( cancer.subtype ) %>% 
  mutate(n = n()) %>%           #< get freq of cancer.subtype
  mutate(freq = n/sum(n)) %>%      #< find perc
  filter(absmatch > 0 ) %>%           #< filter  
  summarise(freq = sum(freq)) %>%   data.frame()
 
 
 top6_results = top6_results[ order ( top6_results$cancer.subtype), ]
 top6_results$freq = round ( top6_results$freq, 2)
  
  
 
 # build blocks up if 1 then 6 blocks if .87 then
 
howmany = data.frame ( table ( top6L$absmatch ) )
howmany$f = c( 0:6)
  
dft = data.frame ()
for ( u in unique ( top6L$RNAseq.id)){
  hw = top6L [ top6L$RNAseq.id == u, ]
  
  if (howmany[howmany$Var1== hw$absmatch, ]$f == 0  ){
    dft = rbind (dft, data.frame ( RNAseq.id=hw$RNAseq.id, Freq=0,  cancer.subtype=hw$cancer.subtype, match=hw$absmatch ) )
    next
  }
  
  for ( i in 1:howmany[howmany$Var1== hw$absmatch, ]$f ){
    dft = rbind (dft, data.frame ( RNAseq.id=hw$RNAseq.id, Freq=1, cancer.subtype=hw$cancer.subtype , match=hw$absmatch ) )
  }
}


dft2 = merge ( dft, est_annt , by="RNAseq.id")

plot_top6 = ggplot(dft2, aes( reorder_within ( RNAseq.id, match , cancer.subtype)  , Freq) ) +
  geom_bar(position="stack", stat="identity", color = "black", aes ( fill=cancer.subtype), width = 1   ) +
  scale_fill_manual( values= subcolor ) +
    geom_bar(aes(y=purity*-1 ), stat="identity", alpha=.5, fill='grey' ) + 
  #  scale_fill_gradient(low = "blue", high = "yellow")  +
   theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black")) +
    ylab("") + 
    xlab("") + 
    theme(legend.position="none",  legend.key = element_blank(),
          
          axis.text.y =  element_blank(),
          axis.text.x = element_text(angle = 0, size= 20 ),
          axis.title.x = element_text(size=12),
          axis.title.y     =  element_blank() , 
          legend.text      =element_text(size=12)
          , plot.title = element_text(size = 30, face = "bold")
    ) + 
    scale_x_reordered() + coord_flip() + ggtitle ( "Diagnosis Matches") +
 # scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) + # plot in intergers
  #scale_y_continuous(expand = expansion(mult = c(0, .1)) ) +
  scale_y_continuous(limit = c(-1,6),
                         breaks =  c( 0, 1,  2, 3, 4, 5, 6 ),
                         labels = c("0","1/6","2/6","3/6","4/6", "5/6", "6/6"), 
                     expand = expansion(mult = c(0, 0))
                     ) +
  facet_grid(cancer.subtype~., scales = "free" , space = "free" )



 dftemp = dft2[!duplicated ( dft2$RNAseq.id), ]
 df6 = data.frame ( table ( dftemp$cancer.subtype, dftemp$match))
  df6 = dcast( df6
, Var1   ~ Var2, value.var= "Freq"  )
 
colnames ( df6) = c( "CS","0/6", "1/6_", "2/6_", "3/6_", "4/6_", "5/6_", "6/6_")
df6 = df6 %>% mutate( match = rowSums(.[grepl('_',colnames(.))] )   / rowSums(.[grepl('\\/',colnames(.))] ) )
colnames ( df6) = c( "CS","0/6", "1/6", "2/6", "3/6", "4/6", "5/6", "6/6", "pct_match")
df6$pct_match = round ( df6$pct_match,2)

dft2[!duplicated ( dft2$RNAseq.id), ] %>% group_by(cancer.subtype) %>%
  summarise(min_purity = min(purity),
         mean_purity = mean ( purity), 
         max_purity = max ( purity)
         )  %>%  data.frame(stringsAsFactors = F)



df6 = df6[order ( -df6$pct_match, df6$CS), ]


kable(  df6 , format = "html" , row.names = F, caption="", table.attr = "style='width:30%;'" ) %>% kable_classic(full_width = T, position = "center"
                                                                                                                         ,bootstrap_options = c("striped")               
)%>%
  footnote(
    symbol  = c("CS: Cancer Subtype", "x/6: how many of the top 6 matches with the diagnosis", "pct_match: percent that at leat 1/6 matched with diagnosis")) %>% row_spec(0, background = "#e6e6e6" )  %>%
  column_spec( 1, border_right = T ) 


# insert purity scores .

purity_agg = dft2 %>% dplyr::group_by(cancer.subtype) %>%
  dplyr::summarise( 
    purity_median = median ( purity)
    ,purity_mean = mean ( purity )
    , purity_min = min ( purity )
    , purity_max = max ( purity )
    )

top6_t1 <- ggtexttable(df6, rows = NULL, theme = ttheme("blank") ) %>%
  tab_add_hline(at.row = c( 1:2) , row.side = "top", linewidth = 2) %>%
  tab_add_vline(at.column = 2, column.side = "left", from.row = 2, linetype =1)
                 

#top6_t1 <- table_cell_bg(top6_t1, row = 0, column = 1:ncol(df6), linewidth = 5,
 #                   fill="#d7d7d9", color = "#d7d7d9")


           
           
           

  
wbt <- createWorkbook()

addWorksheet(wbt, 'top6' )
writeData(wbt, "top6" , top6  , rowNames=T  )




topneg = top6 [ top6$absmatch == 0,  c( "RNAseq.id" , "cancer.subtype"  , "diagnosis" , "sample.type"  , "matched" , "top6" ) ]
topneg = topneg[topneg$cancer.subtype != "WILMS", ]

topneg = merge ( topneg, rna.key[ , c("RNAseq.id", "patient.id")], by="RNAseq.id")

addWorksheet(wbt, 'nomatch' )
writeData(wbt, "nomatch" , topneg  , rowNames=F  )


saveWorkbook(wbt, file = paste0("top6.xlsx"), overwrite = TRUE) 


```

```{r}
## immuno scores 

est = readRDS(   "/data/SUMMARY//ESTIMATE/EST.rds")  
setdiff( trueseq.key$RNAseq.id, c ( est$est.solid$pid, est$ALL$pid) )


```

```{r}
# going to use  Mann-Whitney U Test 

perc.rank <- function(x) trunc(rank(x))/length(x) # this function will return percentile 

unique ( common.right$frameness2)
frametest = c("in.frame") # in.frame 

genepair = "right"
fusion.test.set = unique ( common.right$right) 


genepair = "left"
fusion.test.set = unique ( common.left$left) 

#test = studyFusion(genepair = "right", fusion.test.set = unique ( common.right$right), frametest = c ( "in.frame" )  )

studyFusion = function(  genepair, fusion.test.set , frametest, minf=2, limy = NA, repel_this=1, size_repel= 3.5, total_column = 4  , xpos=1, rank="p", canprint=0 ){

statswu2 = data.frame ( stringsAsFactors = F  )
exp.fusion2 =  data.frame ( stringsAsFactors = F  )



for ( test.gene in unique ( fusion.test.set ) ){
  
  #IGH needs to be skipped 
  if ( test.gene == "IGH" ){
    next
  }  
  
  if ( ! any ( tpm$gene %in% test.gene )){
    next 
  }
  
  test.sample = fusion_truseq[fusion_truseq[[genepair]] == test.gene
                              & fusion_truseq$frameness2 %in% frametest
                              , c("RNAseq.id", "fusion")]
  
  # remove duplication that is sometimes the fusion gene can have mulitple parnter and should therefor not be counted >1 
  test.sample = test.sample[!duplicated ( test.sample$RNAseq.id), ]
  
  # add to this to fix too many. 
  
  if ( frametest == "out.of.frame"){
    test.sample_frame = fusion_truseq[fusion_truseq[[genepair]] == test.gene
                              & fusion_truseq$frameness2 %in% "in.frame"
                              , c("RNAseq.id", "fusion")]
    test.frame.sample = unique (test.sample_frame )
    test.sample = test.sample[! unique ( test.sample$RNAseq.id) %in%  unique ( test.sample_frame$RNAseq.id) , ]
    
  }
  
  
  
  if ( nrow ( test.sample) < minf ){
    next
  }
  
  test.tpm = melt ( tpm[tpm$gene==test.gene, test.sample$RNAseq.id, drop=F] )
  test.tpm$group = "fusion"
  type = trueseq.key[ trueseq.key$RNAseq.id %in% unique ( as.character ( test.tpm$variable) ),  ]$cancer.type
  
  # this is important to make sure we are comparing solid to solid and liquid to liquid 
  
  if ( any(type %in% c("CNS","Solid"))){
    type =   c("CNS","Solid")
  }
  
  
  test.tpm = merge ( test.tpm, trueseq.key[ , c("RNAseq.id", "cancer.subtype") ], by.x="variable",  by.y="RNAseq.id" )
  
  nofusesample = setdiff(trueseq.key$RNAseq.id, test.sample$RNAseq.id) 
  # the first line removes any nofusesample that does not have the same type 
  nofusesample = trueseq.key[trueseq.key$cancer.type %in% unique ( type) & trueseq.key$RNAseq.id %in% nofusesample , ]$RNAseq.id
  
  nofusion.tpm = melt ( tpm[tpm$gene==test.gene, nofusesample  ] )
  nofusion.tpm$group = "no.fusion"
  nofusion.tpm$cancer.subtype = "none"
  
  testexpfusion = rbind ( nofusion.tpm, test.tpm )
  testexpfusion$value = log2 ( testexpfusion$value + 1 )
  testexpfusion[1] = "sample"
  testexpfusion$fusion = test.gene
  testexpfusion$sample.size = nrow ( test.tpm )
  # Mann-Whitney U
  
  x1 = testexpfusion[testexpfusion$group == "no.fusion", ]$value
  x2 = testexpfusion[testexpfusion$group == "fusion", ]$value
  wu = wilcox.test(x2,x1)
  diff = median ( x2 ) - median ( x1 )
  
  
  statswu2 = rbind ( statswu2, data.frame ( p=wu$p.value, plog2 = -log2(wu$p.value), logFC = diff, gene = test.gene
                                            , freq =  nrow ( test.tpm )  )
                     
  )
  testexpfusion$zscale = as.numeric ( scale ( testexpfusion$value ) )
  testexpfusion$percentile = as.numeric ( perc.rank(testexpfusion$value) )
  exp.fusion2 = rbind ( exp.fusion2, testexpfusion )
}



statswu2$fdr = p.adjust(statswu2$p, n=nrow ( statswu2) )
# plotting the top 10 
exp.fusion2 = merge ( exp.fusion2 , statswu2, by.x="fusion", by.y="gene")

if ( rank == "p"){
  exp.fusion2 = exp.fusion2[ order ( exp.fusion2$p), ]
}else {
  exp.fusion2 = exp.fusion2[order ( exp.fusion2$freq, decreasing = T), ]
}

# add annotation 
exp.fusion2$annt = paste ( "p:" , round(exp.fusion2$p, 2),    "fdr:", round(exp.fusion2$fdr, 2))
exp.fusion2$fusion = factor ( exp.fusion2$fusion , levels = unique ( exp.fusion2$fusion  ))


topdo = unique ( exp.fusion2$fusion) [1:12]

todo2 = exp.fusion2[ exp.fusion2$freq >=minf & exp.fusion2$p < .05, ]   # this will be used to label the scatter plot
#todo2 = unique ( as.character ( todo2[ order ( todo2$p), ]$fusion ) )
todo2 = unique ( as.character ( todo2[ order ( -todo2$logFC), ]$fusion ) )
exp.fusion.plot = NA
if ( length ( todo2) >=12 ){
  ## plot singles, top genes only 
  edf_temp = exp.fusion2[exp.fusion2$fusion %in% todo2[1:15] , ] 
  
  exp.fusion.plot = ggplot(exp.fusion2[exp.fusion2$fusion %in% todo2[1:12] , ], aes(y=value, x=group))+ #geom_boxplot(outlier.shape = NA)+ 
    geom_violin(alpha = 0.5) +
    geom_point(aes(fill = cancer.subtype), size = 4, shape = 21, position = position_jitterdodge(dodge.width = 0.25), alpha = 0.75 )+
    theme_bw() +
    ylab("log2 ( tpm +1 ) ") +
    xlab("") +
    theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=22),
          
          axis.title.y     = element_text(size=22), 
          legend.text      =element_text(size=12)
    ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") + xlab ( "") +  ylab ( "") +
    scale_fill_manual(values = subcolor2 ) + 
    geom_text( aes(x = xpos, y = 0, label = annt), vjust = .20)  +
    facet_wrap(~ fusion, scales = "free" ,  ncol = total_column ) + ggtitle ( "TOP 12 Fusion Most Recurrent 3' Gene exp vs None")
  
    # special print below to seperate out canonica from non 
    if ( canprint == 1){
    # seperate canonical from none 
    is.can_sep = unique ( fusion [ fusion$right %in% unique ( edf_temp$fusion ) & fusion$is.can == 1, ]$right  )
    expthis = exp.fusion2[exp.fusion2$fusion %in% todo2[1:11] , ]
    expthis = expthis[!expthis$fusion %in% unique ( as.character (is.can_sep ) ) , ]
    
    
    noncanP = ggplot(expthis, aes(y=value, x=group))+ #geom_boxplot(outlier.shape = NA)+ 
    geom_violin(alpha = 0.5) +
    geom_point(aes(fill = cancer.subtype), size = 4, shape = 21, position = position_jitterdodge(dodge.width = 0.25), alpha = 0.75 )+
    theme_bw() +
    ylab("log2 ( tpm +1 ) ") +
    xlab("") +
    theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=22),
          
          axis.title.y     = element_text(size=22), 
          legend.text      =element_text(size=12)
    ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") + xlab ( "") +  ylab ( "") +
    scale_fill_manual(values = subcolor2 ) + 
    geom_text( aes(x = xpos, y = 0, label = annt), vjust = .20)  +
    facet_wrap(~ fusion, scales = "free" ,  ncol = total_column ) + ggtitle ( "TOP 12 Fusion Most Recurrent 3' Gene exp vs None")
    
    
    
    yescanP = ggplot(exp.fusion2[exp.fusion2$fusion %in% is.can_sep, ] , aes(y=value, x=group))+ #geom_boxplot(outlier.shape = NA)+ 
    geom_violin(alpha = 0.5) +
    geom_point(aes(fill = cancer.subtype), size = 4, shape = 21, position = position_jitterdodge(dodge.width = 0.25), alpha = 0.75 )+
    theme_bw() +
    ylab("log2 ( tpm +1 ) ") +
    xlab("") +
    theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=22),
          
          axis.title.y     = element_text(size=22), 
          legend.text      =element_text(size=12)
    ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") + xlab ( "") +  ylab ( "") +
    scale_fill_manual(values = subcolor2 ) + 
    geom_text( aes(x = xpos, y = 0, label = annt), vjust = .20)  +
    facet_wrap(~ fusion, scales = "free" ,  ncol = total_column ) + ggtitle ( "")
    
        
    exp_all = rbind ( exp.fusion2[exp.fusion2$fusion %in% is.can_sep, ], expthis  ) 
        
    exp_all$fusion = factor ( exp_all$fusion, levels= unique ( as.character ( c ( as.character ( expthis$fusion), is.can_sep ))) )
    
    exp_all=  ggplot(exp_all, aes(y=value, x=group))+ #geom_boxplot(outlier.shape = NA)+ 
    geom_violin(alpha = 0.5) +
    geom_point(aes(fill = cancer.subtype), size = 4, shape = 21, position = position_jitterdodge(dodge.width = 0.25), alpha = 0.75 )+
    theme_bw() +
    ylab("log2 ( tpm +1 ) ") +
    xlab("") +
    theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=22),
          
          axis.title.y     = element_text(size=22), 
          legend.text      =element_text(size=12)
    ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") + xlab ( "") +  ylab ( "") +
    scale_fill_manual(values = subcolor2 ) + 
    geom_text( aes(x = xpos, y = 0, label = annt), vjust = .20)  +
    facet_wrap(~ fusion, scales = "free" ,  ncol = total_column ) + ggtitle ( "")
    
    print ( exp_all )
    
    
    }
  
}

# scatter 

exp.fusion2$direction = ifelse ( exp.fusion2$logFC > 0, "higher", "lower")

exp.fusion2$piegroup = ifelse ( exp.fusion2$p < .05, paste0(exp.fusion2$direction,"_Sig"), paste0(exp.fusion2$direction,"_noSig") )
piecc = c( higher_Sig = "#cf4051", higher_noSig = "#d18a82", lower_Sig = "#83bf43", lower_noSig= "#c9e38d" )
piecc = c( higher_Sig = "#cf4051", higher_noSig = "#e0c126", lower_Sig = "#83bf43", lower_noSig= "#6ec0db" )

todo3 = unique ( c(todo2[1:12]) )

exp.fusion2$pgroup = ifelse(exp.fusion2$sample.size >= minf, paste0 ( ">=", minf) , "undetermine")
exp.fusion2$group = factor ( exp.fusion2$group)
exp.fusion2 <- exp.fusion2 %>%
  mutate(group = relevel(group, ref = "no.fusion"))
contrasts(exp.fusion2$group)

nondup =  exp.fusion2 %>% dplyr::group_by( fusion, group) %>%
  dplyr::summarise(
    #median = median ( percentile ) 
    median = median ( value ) 
  )  %>% data.frame( stringsAsFactors = F  )



pall = wilcox.test(nondup[nondup$group=="fusion", ]$median, nondup[nondup$group=="no.fusion", ]$median)
pall = round ( pall$p.value, 4 )

#wilcox.test(exp.fusion2[exp.fusion2$group=="fusion", ]$percentile, exp.fusion2[exp.fusion2$group=="no.fusion", ]$percentile)
exp.plot = exp.fusion2[!duplicated ( exp.fusion2$fusion) & exp.fusion2$sample.size >= minf, ] 
meta = sumz ( exp.plot$p, weights = exp.plot$sample.size)
xmax = max ( exp.fusion2[!duplicated ( exp.fusion2$fusion) & exp.fusion2$sample.size >= minf, ]$logFC )
xlabel = xmax - ( .10 * xmax )

if ( repel_this == 0 ){
  todo3=""
}

expfusion.scat.plot = ggplot(exp.plot, aes(y=plog2, x=logFC))+ #geom_boxplot(outlier.shape = NA)+ 
  geom_point(aes ( fill = piegroup  ), shape = 21, size = 5 ,position = position_jitterdodge(dodge.width = 0.25, jitter.height =.2), alpha = 0.75 )+
  theme_bw() +
  ylab(" -log2 ( pv)") +
  xlab(" log FC ") +
  theme(legend.position="bottom", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 12 ),
        axis.text.x = element_text(angle = 0, size= 12 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  )  + geom_hline(yintercept= 4.3, color = "grey", linetype = "dashed", size = 3.0 ) +  scale_fill_manual(values = piecc ) +
  annotate(geom="text", x=xlabel, y=4.8,  label=paste ( "pv=", ".05" ), size=5 ) +
  #annotate(geom="text", x=5, y=30,  label="p < .0001 ( Fusion vs Non.Fusion) ", size=5 )  + 
  geom_text_repel(
    data = exp.fusion2[!duplicated ( exp.fusion2$fusion) & exp.fusion2$fusion %in% todo3, ],
    aes(
      label = fusion
    )
    , 
    fontface="bold", 
    color="black",
    size = size_repel , 
    point.padding = unit(.25, "lines"),
    #  box.padding = unit(.25, "lines"),
    nudge_y = 1.3,
    show.legend = F,
    #segment.color = cc5
    #segment.alpha = 0
    #segment.linetype = 2, 
    #segment.curvature = 40, 
    #arrow = arrow(length = unit(0.015, "npc"))
  ) + ggtitle ( " 3' Fusion Gene expression ")




maincmp = 
ggplot()+ 
  geom_violin(data = exp.fusion2, alpha = 0.5 , aes(y=value, x=group,  fill = group ) ) +
  #geom_point(aes(fill = group), size = 4, shape = 21, position = position_jitterdodge(dodge.width = 0.25), alpha = 0.75 )+
  theme_bw() +
  ylab("log2 TPM ") +
  xlab("") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=32),
        
        axis.title.y     = element_text(size=32), 
        legend.text      =element_text(size=12)
  ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") + scale_fill_manual(values=c("#b3b1ad","#77a7bf")) 
  #annotate(geom="text", x=0, y=1.2,  label=paste ( "pv=", ".05" ), size=5 )

exp.plot$piegroup = ifelse ( exp.plot$p < .05, paste0(exp.plot$direction,"_Sig"), paste0(exp.plot$direction,"_noSig") )
tally = data.frame ( table ( exp.plot$piegroup))
#piecc = c( higher_Sig = "#cf4051", higher_noSig = "#e0c126", lower_Sig = "#83bf43", lower_noSig= "#6ec0db" )


addthis = setdiff ( names ( piecc), tally$Var1 )
tally2 = tally 
for ( a in addthis){
 tally2 = rbind ( tally2, data.frame(Var1=a, Freq=0))
}
 tally2$state = gsub ( ".*_", "", tally2$Var1)
  tally2$group = gsub ( "_.*", "", tally2$Var1)
  tally2$Var1 = NULL 
tally2 =  pivot_wider(data = tally2, names_from = group,values_from=Freq) %>% data.frame(stringsAsFactors = F)
row.names ( tally2) = tally2$state
tally2$state = NULL 
chi = chisq.test(tally2, simulate.p.value=T)
if ( is.na ( chi )){
chi = chisq.test(tally2[,1], simulate.p.value=T)
}

if ( is.na ( chi )){
chi = chisq.test(tally2[,2], simulate.p.value=T)
}




pie.ratio= make.pie( tally
                , name.first= paste ( "chisq=", round ( chi$p.value, 4 ))
                #,name.first= paste ( "")
                , cc = piecc
                , leg.pos = "right"
                , title = ""
                )



return ( list (
  maincmp=maincmp, 
  pvalue = pall, 
  expfusion.scat.plot = expfusion.scat.plot, 
  exp.fusion.plot = exp.fusion.plot, 
  meta = meta, 
  pie.ratio = pie.ratio, 
  chi = chi
  
))

}
```


```{r}

right_out = studyFusion(genepair = "right", fusion.test.set = unique ( common.right$right), frametest = c ( "out.of.frame" ) , minf = 2 , total_column=3 )
# I created canprint which is a fudge thing to seprate out known fusions from novel, however mainly this if used for in.frame only default is not to do this! 

right_in = studyFusion(genepair = "right", fusion.test.set = unique ( common.right$right), frametest = c ( "in.frame" )  , minf = 2 , total_column=3 , xpos= 1, canprint=1 )
left.in = studyFusion(genepair = "left", fusion.test.set = unique ( common.left$left), frametest = c ( "in.frame" )  , minf = 2 , total_column=3  )

maxx = max ( c( right_in$expfusion.scat.plot$data$logFC, right_out$expfusion.scat.plot$data$logFC, left.in$expfusion.scat.plot$data$logFC ) ) + .5  
minx = min ( c( right_in$expfusion.scat.plot$data$logFC, right_out$expfusion.scat.plot$data$logFC, left.in$expfusion.scat.plot$data$logFC ) ) - .5 

maxy = max ( c( right_in$expfusion.scat.plot$data$plog2, right_out$expfusion.scat.plot$data$plog2, left.in$expfusion.scat.plot$data$plog2 ) ) + .5  
miny = min ( c( right_in$expfusion.scat.plot$data$plog2, right_out$expfusion.scat.plot$data$plog2, left.in$expfusion.scat.plot$data$plog2 ) ) - .5 

(right_in$expfusion.scat.plot + ggtitle ( "3' Fusion Gene in-frame") + ylim(miny,maxy) + xlim(minx, maxx) + theme( legend.position = "none")  )/
(right_out$expfusion.scat.plot + ggtitle ( "3' Fusion Gene out-of-frame") + ylim(miny,maxy)  + xlim(minx, maxx) + theme( legend.position = "none") ) /
(left.in$expfusion.scat.plot +  ylim(miny,maxy) +  xlim(minx, maxx) + ggtitle ( "5' Fusion Gene in-frame") )

(right_in$maincmp + ggtitle ( paste ( "pv=", right_in$pvalue ) ) + theme( legend.position = "none")  )/
(right_out$maincmp + ggtitle ( paste ( "pv=", right_out$pvalue ) ) + theme( legend.position = "none") ) /
(left.in$maincmp + ggtitle ( paste ( "pv=", left.in$pvalue ) ) )

# may be just make a pie ratio here? 
right_in$pie.ratio
left.in$pie.ratio

(right_in$pie.ratio + ggtitle ( "3' Fusion Gene in-frame") + theme( legend.position = "right")  )/
(right_out$pie.ratio + ggtitle ( "3' Fusion Gene out-of-frame") + theme( legend.position = "right") ) /
(left.in$pie.ratio + ggtitle ( "5' Fusion Gene in-frame") )




```

### Expression of 3' Gene
* Unpaired Mann-Whitney test, those with fusion vs those without 
* only fusions with >= 2 samples were calculated
* Non Fusion group were defined as any sample without a gene fusion with the 3' gene AND same cancer type 
  + (CNS, Solid) OR leukemia

```{r fig=TRUE,fig.width=14, fig.height=10, echo=FALSE, include=TRUE}
(right_in$expfusion.scat.plot + ggtitle ( "3' Gene in-frame") + ylim(miny,maxy) + xlim(minx, maxx) + theme( legend.position = "none")  )/
(right_out$expfusion.scat.plot + ggtitle ( "3' Gene out-of-frame") + ylim(miny,maxy)  + xlim(minx, maxx) + theme( legend.position = "none") ) |
  
  (
    (right_in$pie.ratio + ggtitle ( "") + theme( legend.position = "right")  )/
(right_out$pie.ratio + ggtitle ( "") + theme( legend.position = "right") )
  )
#/(left.in$expfusion.scat.plot +  ylim(miny,maxy) +  xlim(minx, maxx) + ggtitle ( "5' Gene in-frame") )
```

### Pie Chart and chisquare

```{r fig=TRUE,fig.width=14, fig.height=10, echo=FALSE, include=TRUE}
(right_in$pie.ratio + ggtitle ( "") + theme( legend.position = "right")  )/
(right_out$pie.ratio + ggtitle ( "") + theme( legend.position = "right") ) 
#/(left.in$pie.ratio + ggtitle ( "") + theme( legend.position = "right") )  
```

### Expression in.frame 3'

```{r fig=TRUE,fig.width=14, fig.height=10, echo=FALSE, include=TRUE}
(right_in$exp.fusion.plot + ggtitle ( "") + theme( legend.position = "right")  ) 
#|(left.in$exp.fusion.plot + ggtitle ( "") + theme( legend.position = "none") )  
```




```{r}
# more fusion figs for paper 

annt = fusion [ , c("uid", "RNAseq.id", "annt1", "annt2")]
annt = annt[! ( annt$annt1 %in% c('',"0")  | annt$annt2 %in% c('',"0") ), ]

table ( annt$annt1)
table ( annt$annt2)


# now we are going slowly build the dataframe 
anntthis = unique ( c ( annt$annt1,  annt$annt2) )

anntdf = data.frame ( stringsAsFactors = F )

for ( an in anntthis ){
   balloon = c()
   for (an2 in anntthis){
       balloon[an2] = nrow ( annt[annt$annt2 == an & annt$annt1 == an2 ,  ] )
   }
   
   balloon = t ( data.frame ( balloon, stringsAsFactors = F )) 
   row.names ( balloon) = an 
   anntdf = rbind ( anntdf, balloon)
}



# here a trick to know what directio is what 
nrow ( annt[annt$annt1 == "promoters" & annt$annt2 == "exons",  ] )
nrow ( annt[annt$annt2 == "promoters" & annt$annt1 == "exons",  ] )


  
ant = melt ( as.matrix ( anntdf ) )    
ant$X1 = factor ( ant$X1, rev ( c (  "5UTRs","promoters" , "exons" , "3UTRs" )  ) )
ant$X2 = factor ( ant$X2, rev (  "5UTRs","promoters" , "exons" , "3UTRs" ) )


 fusion[ grepl ( "ASNS", fusion$fusion), ]

 ant$X2 = gsub ( "exon", "coding", ant$X2)
  ant$X1 = gsub ( "exon", "coding", ant$X1)
  
  ant$X1 = gsub ( "s$", "", ant$X1)
  ant$X2 = gsub ( "s$", "", ant$X2)
  
  ant$X1 = factor ( ant$X1, levels = c( "promoter" , "5UTR", "coding", "3UTR"  ))
  ant$X2 = factor ( ant$X2, levels = rev ( c( "promoter" , "5UTR", "coding", "3UTR"  )) )
    
balloonplot_type=   ggplot(ant, aes(x =X1, y = X2))  +geom_point( size=0,shape=21, colour="white", fill="#80B1D3", alpha=0 )+
  geom_point( data=ant[ant$value!=0, ], aes(size=value, fill=value),shape=21, colour="white", 
              #fill="#80B1D3", 
              alpha=1) + theme_minimal() +
  geom_tile(color = "black", fill = "transparent") + geom_text(data=ant[ant$value!=0, ], aes(label=value),hjust=.50, vjust=.5, size=10) +
  scale_size_continuous(range = c(8,40)) + theme ( legend.position = "none") + scale_x_discrete(position='top') + xlab("3' Gene") + ylab("5' Gene") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black")) +
   theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 30 ),
        axis.text.x = element_text(angle = 0, size= 30 ),
        axis.title.x = element_text(size=30),
        
        axis.title.y     = element_text(size=30)
  ) +  scale_fill_gradientn(colours=c("#a3c9e3", "#80B1D3","#80B1D3",  "#428dc2",  "#428dc2",  "#428dc2",  "#428dc2", "#428dc2", "#0f72b8") , limits = c(10, 706))
  


tally = ant 
colnames ( tally ) = c("right","left","total")
tally_mis = tally[ tally$total <= 30 , ]
tally = tally[ tally$total > 20, ]
tally = rbind ( tally, data.frame ( "right" = "other", "left"="other", "total" = sum ( tally_mis$total ) ) )   
tally = tally[ tally$total > 0 , ]
tally$type = paste0 ( tally$left, "_", tally$right )  




tally = tally[ , c("type","total")]
colnames ( tally ) = c("Var1","Freq")
tally = tally[ order ( tally$Var1), ]


tab.fusion = make.pie( tally
                , name.first="fusion"
                , cc =rev ( getPalette(8))
                , leg.pos = "right"
                , title = "tally"
                )



```

### Annotation of location of fusion  

```{r fig=TRUE,fig.width=7, fig.height=7, echo=FALSE, include=TRUE}
balloonplot_type
```

### pie chart
```{r fig=TRUE,fig.width=7, fig.height=7, echo=FALSE, include=TRUE}
tab.fusion
```



```{r}
# now lets study the fusions a bit more get more of the biology !

# 5.11.2022 remove this line and use fusion_mod instead which has since fixed the type to is.diag

#fuse2 = merge ( fuse1, rna.key[ , c("sex","sample.type" , "treatment","RNAseq.id", "sample.classification")] , by="RNAseq.id")

fuse2 = temp.all
fuse2$patient.id = as.character ( sapply ( fuse2$RNAseq.id, function (x) rna.key[rna.key$RNAseq.id == x, ]$patient.id))
# correct diagnostics 
fuse2$type = ifelse ( fuse2$type == "is.diag" & fuse2$frameness != "in.frame", "is.novel", fuse2$type )
fuse2$type = ifelse ( fuse2$type == "is.drug" & fuse2$frameness != "in.frame", "is.novel", fuse2$type )

fuse2 = merge ( fuse2, rna.key[ , c("sex","sample.type" , "treatment","RNAseq.id", "sample.classification")] , by="RNAseq.id")
fuse2$sample.type = gsub ( ".*_", "", fuse2$sample.type  )
# get total by treatment and disease then plot 
fuse.by.treatment = fuse2[fuse2$treatment != "Unknown", ]  %>% dplyr::group_by(treatment, RNAseq.id) %>%
  dplyr::summarise(
    Total = n()  
    
  ) %>%  data.frame ( stringsAsFactors = F )

ptreat = wilcox.test(fuse.by.treatment[fuse.by.treatment$treatment=="Naive", ]$Total, fuse.by.treatment[fuse.by.treatment$treatment=="Treated", ]$Total )

fuse.by.treatment = ggplot( fuse.by.treatment  , aes(y=Total, x=treatment))+
  geom_quasirandom(aes(fill = treatment ), size = 3, shape = 21, alpha = .5) +
  theme_bw() +
  xlab("") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") + xlab ( "") +  ylab ( "") +
   scale_fill_manual(values = unique (brewer.pal(7, "Dark2") ))  #+ ylim(0,max ( fuse.by.treatment$Total))  #+ facet_wrap(cancer.subtype~., ncol=3)

fuse.by.treatment = my_add_pval(fuse.by.treatment, pairs = list(c(1, 2)), test='wilcox.test', textsize = 10)


#####  tabulate treatment

fusion_type_ccc = setNames(brewer.pal(4, "Set3") , c("is.diag","is.drug","is.novel", "is.obsv"))
fuse2$is.cancer = ifelse ( fuse2$right %in% genelist$gene , "yes", "no" )
#fuse2$treatment = factor ( fuse2$treatment, c("Treated", "Naive", "Unknown")) 


treat_type_stack=  fuse2[ fuse2$treatment != "Unknown", ]  %>% dplyr::group_by(treatment, type) %>%
  dplyr::summarise(
    Total = n()  
  ) %>% 
    ggplot( .  , aes(x=factor ( treatment, levels = c("Treated", "Naive") ) , y=Total, fill=type   )) +
  geom_bar(position="fill", stat="identity")+
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=15), 
        legend.text      =element_text(size=20)
        , plot.title = element_text(size = 12, face = "bold")
  ) + 
  scale_fill_manual( values= fusion_type_ccc ) + coord_flip()
  


tally = data.frame ( table ( fuse2[fuse2$treatment != "Unknown", c("treatment","is.cancer")]  )  )  
tally$treatment = factor ( tally$treatment, c("Treated", "Naive"))

is_cancer_ccc = c(yes='#a678d6', no='#cfd5e3')
treat_type_cancer  = ggplot( tally  , aes(x=treatment , y=Freq, fill=is.cancer   )) +
  geom_bar(position="fill", stat="identity")+
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=15), 
        legend.text      =element_text(size=20)
        , plot.title = element_text(size = 12, face = "bold")
  ) + coord_flip() + scale_fill_manual(values= is_cancer_ccc  )




 
 
 
## tablulate classifications 

fuse2$sample.classification = gsub ( "NonPrimary" , "Metastasis", fuse2$sample.classification )
fuse2$sample.classification = factor ( fuse2$sample.classification, c(  "Metastasis" , "Primary" , "Unknown"  ))




class_type_stack=  fuse2[fuse2$sample.classification != "Unknown", ]  %>% dplyr::group_by(sample.classification, type) %>%
  dplyr::summarise(
    Total = n()  
  ) %>% 
     ggplot( .  , aes(x=sample.classification , y=Total, fill=type   )) +
  geom_bar(position="fill", stat="identity") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="bottom",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=15), 
        legend.text      =element_text(size=20)
        , plot.title = element_text(size = 12, face = "bold")
  ) + 
  scale_fill_manual( values= fusion_type_ccc ) + coord_flip() 
  



tally = data.frame ( table ( fuse2[, c("sample.classification","is.cancer")]  )  )  
tally=tally[tally$sample.classification != "Unknown", ]
tally$sample.classification = factor ( tally$sample.classification, c('Metastasis',  "Primary"))


class_type_cancer  = ggplot( tally  , aes(x=sample.classification , y=Freq, fill=is.cancer   )) +
  geom_bar(position="fill", stat="identity")+
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="bottom",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=15), 
        legend.text      =element_text(size=20)
        , plot.title = element_text(size = 12, face = "bold")
  ) + coord_flip() + scale_fill_manual(values= is_cancer_ccc  )


(class_type_stack + class_type_cancer) /
(treat_type_stack + treat_type_cancer)



fuse.by.sample.type = fuse2[! ( fuse2$sample.type %in% c ("Chloroma"  ,"Unknown" ) ), ]  %>% dplyr::group_by(sample.type, RNAseq.id, cancer.subtype) %>%
  dplyr::summarise(
    Total = n()  
    
  ) %>%  data.frame ( stringsAsFactors = F )



 
### now plot by classification 



fuse.by.class = fuse2[fuse2$sample.classification != "Unknown", ]  %>% dplyr::group_by(sample.classification, RNAseq.id) %>%
  dplyr::summarise(
    Total = n()  
    
  ) %>%  data.frame ( stringsAsFactors = F )

pclass = wilcox.test( fuse.by.class[fuse.by.class$sample.classification=="Primary", ]$Total ,fuse.by.class[ fuse.by.class$sample.classification=="NonPrimary", ]$Total  )

fuse.by.class$sample.classification = gsub ( "NonPrimary" , "Metastasis", fuse.by.class$sample.classification )
fuse.by.class$sample.classification = factor ( fuse.by.class$sample.classification, c(  "Primary", "Metastasis"    ))

fuse.by.class_plot = ggplot( fuse.by.class  , aes(y=Total, x=sample.classification))+
  geom_quasirandom(aes(fill = sample.classification ), size = 3, shape = 21, alpha = .5) +
  theme_bw() +
  xlab("") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") + xlab ( "") +  ylab ( "") +
   scale_fill_manual(values = unique (brewer.pal(7, "Accent") ))  #+ ylim(0,max ( fuse.by.treatment$Total))  #+ facet_wrap(cancer.subtype~., ncol=3)

fuse.by.class_plot = my_add_pval(fuse.by.class_plot, pairs = list(c(1, 2)), test='wilcox.test', textsize=10)


fuse.by.treatment + fuse.by.class_plot
####


kruskal.test(Total ~ sample.type, data = fuse.by.sample.type)


wilcox.test(fuse.by.sample.type[fuse.by.sample.type$sample.type =="Biopsy", ]$Total
                     , fuse.by.sample.type[fuse.by.sample.type$sample.type =="Metastasis", ]$Total  )

wilcox.test(fuse.by.sample.type[fuse.by.sample.type$sample.type =="Biopsy", ]$Total
                     , fuse.by.sample.type[fuse.by.sample.type$sample.type =="Leukemia", ]$Total  )

wilcox.test(fuse.by.sample.type[fuse.by.sample.type$sample.type =="Resection", ]$Total
                     , fuse.by.sample.type[fuse.by.sample.type$sample.type =="Metastasis", ]$Total  )

wilcox.test(fuse.by.sample.type[fuse.by.sample.type$sample.type %in%  c ( "Resection", "Biopsy" ), ]$Total
                     , fuse.by.sample.type[fuse.by.sample.type$sample.type =="Metastasis", ]$Total  )

wilcox.test(fuse.by.sample.type[fuse.by.sample.type$sample.type %in%  c ( "Resection", "Biopsy" ), ]$Total
                     , fuse.by.sample.type[fuse.by.sample.type$sample.type =="Leukemia", ]$Total  )



fuse.by.sample.type = ggplot(fuse.by.sample.type , aes(y=Total, x=sample.type))+
  geom_quasirandom(aes(fill = sample.type ), size = 3, shape = 21, alpha = .5) +
  theme_bw() +
  xlab("") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") + xlab ( "") +  ylab ( "") +
   scale_fill_manual(values = unique (brewer.pal(7, "Dark2") )) + ylim(0,max ( fuse.by.sample.type$Total))


# lets do wilcox here  wilcox.test(nondup[nondup$group=="fusion", ]$median, nondup[nondup$group=="no.fusion", ]$median)

fusion.frame =  reshape2::melt ( fuse1[ , c( "RNAseq.id","fusion" , "frameness")], id.vars = c("RNAseq.id","fusion" ))

colnames ( fusion.frame) = c("RNAseq.id","fusion" , "extra", "frameness")
#fusion.frame$frameness = ifelse ( grepl("CDS|UTR|un|exo|inter|^\\.$",fusion.frame$frameness ), "unknown", fusion.frame$frameness )
# safer way to do this.  
fusion.frame$frameness = ifelse ( fusion.frame$frameness %in% c("in.frame","out.of.frame"), fusion.frame$frameness, "unknown")

# not goint to add zero 
#zero = setdiff ( rna.key$RNAseq.id, as.character ( unique ( fusion.frame$RNAseq.id) ) )
#for ( z in zero ){
 # fusion.frame = rbind ( fusion.frame, data.frame ( RNAseq.id=z,fusion="zero",extra="frameness", frameness="zero") )     
#}

fusion.frame = plyr::count(fusion.frame, c("RNAseq.id","frameness"))
fusion.frame= merge ( fusion.frame, rna.key[ , c("RNAseq.id", "cancer.subtype")], by = "RNAseq.id")

#fusion.frame$RNAseq.id = factor (fusion.frame$RNAseq.id, levels = fusion.total$RNAseq.id )
#fusion.frame$cancer.subtype = factor ( fusion.frame$cancer.subtype, levels = cancer.subtype.order)


fuse.by.frame = fusion.frame  %>% dplyr::group_by(frameness, RNAseq.id, cancer.subtype) %>%
  dplyr::summarise(
    Total = sum(freq)  
    
  ) %>% ggplot(. , aes(y=Total, x=frameness))+
  geom_quasirandom(aes(fill = frameness ), size = 3, shape = 21, alpha = .5) +
  theme_bw() +
  xlab("") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") +
   scale_fill_manual(values = unique (brewer.pal(7, "Dark2") )) #+ facet_wrap(cancer.subtype~., ncol=3, scales = "free"   )

#fuse.by.sample.type + fuse.by.treatment + lolli_med
#fuse.by.sample.type + fuse.by.treatment + fuse.by.frame

```

### Additional Fusion Tabulation plots

```{r fig=TRUE,fig.width=12, fig.height=10, echo=FALSE, include=TRUE}
 fuse.by.class_plot + fuse.by.treatment
#fuse.by.sample.type + fuse.by.treatment + plot_layout(widths =c ( .4,.2 ) ) #+ fuse.by.frame + plot_layout(widths =c ( .4,.2,.4 ) )
```


### Tabulation {.tabset}


```{r}

tab.fusion.frame = fusion.frame  %>% dplyr::group_by(frameness, cancer.subtype) %>%
    dplyr::summarise(
        Total = sum ( freq) , 

    ) %>% data.frame() 





```




#### by frame 
```{r fig=TRUE,fig.width=6, fig.height=8, echo=FALSE, include=TRUE , results='asis'}
tab.fusion.frame %>%  dplyr::group_by(frameness)  %>%    dplyr::summarise(
        Total = sum ( Total) , 

    ) %>% arrange ( desc ( Total ) ) %>% kable( .  , format = "html" , row.names = F, caption="" , table.attr = "style='width:40%;'" ) %>% kable_classic(full_width = T, position = "center", ) 
```

```{r}
# 5.4.2022 adding exon counts 

exon = readRDS( '/data/SUMMARY/EXON_COUNT/exon_counts_correlations2.rds') 
exon_cor = exon$corr
exon_tpm = exon$data
rm ( exon )
gc()

# subset for cancer genes only 
exon_tpm = exon_tpm[ exon_tpm$gene_name %in% as.character ( genelist$gene ),  c ( names ( exon_tpm)[1:12], trueseq.key$RNAseq.id, "median") ]
length ( unique ( exon_tpm$transcript_id  ) )




exon_cor2 = melt ( as.matrix ( exon_cor ) )
colnames ( exon_cor2) <- c('txn','tube','value')
exon_cor2 = exon_cor2[exon_cor2$tube %in% trueseq.key$RNAseq.id, ]


tp53 = exon_tpm[ exon_tpm$gene_name == "TP53", ]
tp53 = exon_cor2[ exon_cor2$txn %in% unique ( tp53$transcript_id  ), ]
tp53 = tp53 [ order ( tp53$value), ]
tp53= tp53 [ !duplicated ( tp53$tube), ]
#tp53 = tp53[tp53$tube %in% trueseq.key$RNAseq.id,]

# test 
exon_tpm[ exon_tpm$transcript_id == "ENST00000269305.8",  c( "AA3", "AJ10", "Z12", "K2", "median")]

colSums(   exon_tpm[ exon_tpm$transcript_id == "ENST00000269305.8",  c( "AA3", "AJ10", "Z12", "K2") ]
 )


#### plot top against median 
t="ENST00000269305.8" # tp53
#t="ENST00000244741.9"
samples = c( "AA3", "AJ10", "Z12", "K2")
#samples = c( "BK3")

base = melt ( exon_tpm[ exon_tpm$transcript_id == t,  c ( "median", "exon_number", "gene_name", samples ) ] 
              , id = c("exon_number", "gene_name" ) 
              )

base$g = ifelse ( base$variable == "median", 'base', 'study' ) 

base = base %>% group_by(variable) %>% mutate ( 
  tpm_total = sum ( value)
  ) %>% data.frame()

base$tpm_total = paste0 ( "(", as.character ( base$variable) , ")", round ( base$tpm_total, 1) )

gp = ggplot(base, aes(x=exon_number , y=value, group= variable )) +
  geom_line(aes(color=g))+
  geom_point(aes(color=g, alpha=g), size=10) +
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=30),
        axis.title.y     = element_text(size=30), 
        legend.text      =element_text(size=12)
        , plot.title = element_text(size = 30, face = "bold")
  ) + scale_color_manual(values = c(base="#367699", study="grey")) +
  scale_alpha_manual( values = c(base=.9, study=.5) ) + ggtitle ( base$gene_name[1] ) +
   geom_text_repel(data = base[base$exon_number == max ( unique ( base$exon_number) ), ],  
             aes(label =  tpm_total, x = max ( unique ( base$exon_number) ), y = value, hjust = 1),  
             color="red", size=5, 
             box.padding = 0.5, max.overlaps = Inf
             )
             

### end test plot 





exon_tab = data.frame ()
total = length (  unique ( exon_tpm$gene_name )  )
for ( gene in unique ( exon_tpm$gene_name )  ){
  
geneE = exon_tpm[ exon_tpm$gene_name == gene , ]
geneE = exon_cor2[ exon_cor2$txn %in% unique ( geneE$transcript_id  ), ]
geneE = geneE [ order ( geneE$value), ]
geneE = geneE[geneE$value < .3, ]
geneE= geneE [ !duplicated ( geneE$tube), ]

if ( nrow ( geneE ) > 0 ){
  geneE$gene = gene 
  exon_tab = rbind (exon_tab, geneE )
}

print ( total )
total = total - 1 

}

exon_tab = exon_tab[order ( exon_tab$value), ]
exon_tab2 = exon_tab[complete.cases(exon_tab), ]

for ( g in tsg.onco$gene ){

g =   exon_tab2[exon_tab2$gene == g, ]

if ( nrow ( g ) > 0 ){  

  for ( t in as.character ( unique ( g$txn ) ) ) {  

    #t = 'ENST00000244741.9' # ENST00000227507.2
tsample = as.character ( exon_tab[!is.na ( exon_tab$txn) &   exon_tab$txn == t , ]$tube )
base = exon_tpm[ exon_tpm$transcript_id == t,  c ( "median", "exon_number", "gene_name", tsample) ]




base = melt ( base, id = c("exon_number", "gene_name" ) ) 
base$g = ifelse ( base$variable == "median", 'base', 'study' ) 

gp = ggplot(base, aes(x=factor ( exon_number) , y=value, group= variable )) +
  geom_line(aes(color=g))+
  geom_point(aes(color=g, alpha=g), size=10) +
   theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=30),
        axis.title.y     = element_text(size=30), 
        legend.text      =element_text(size=12)
        , plot.title = element_text(size = 30, face = "bold")
  ) + scale_color_manual(values = c(base="#367699", study="grey")) +
  scale_alpha_manual( values = c(base=.9, study=.5) ) + ggtitle ( paste (  t, base$gene_name[1] )  )

plot ( gp )

}

}
  
}






```

```{r}
# find exons that are different BUT overall tpm are the same? 

singleExon = readRDS( '/data/SUMMARY/EXON_COUNT/tpm_normalzied_by_exons.rds')
```


```{r}

### the next few chuncks is just focus on outliers 


setdiff ( trueseq.key$RNAseq.id, unique ( out$RNAseq.id)) 

total_subcancer_truseq = data.frame ( table ( trueseq.key$cancer.subtype ) )
colnames(total_subcancer_truseq) = c("cancer.subtype", "Freq")

out = out [ out$RNAseq.id %in% trueseq.key$RNAseq.id, ]
gtex.out = out [ , c("gene", "RNAseq.id" , "tpm" , "pid", "gtex.xcell")]
gtex.out = gtex.out[gtex.out$gtex.xcell != "NONE", ]
gtex.out$gtex.xcell = gsub ( "OUTLIER", "GTEX", gtex.out$gtex.xcell )
out = out[out$both != "NONE", ]



# gnerates the stack plot for outliers annotated with cosmic 

canlefright = unique ( fusion[ fusion$canon.left == 1 |  fusion$canon.right == 1,  ]$right ) 

# dont subset yet ignore this. 
#this.out =out[out$gene %in% unique ( genelist$gene    ), 
 #                  c ( "gene", "RNAseq.id",  "both", "gtex.median", "median.txn2", "tree.xcell.median", "tpm", "tree.xcell", "tree.TxnSpear2")]

this.out =out[, 
                   c ( "gene", "RNAseq.id",  "both", "gtex.median", "median.txn2", "tree.xcell.median", "tpm", "tree.xcell", "tree.TxnSpear2")]




mintpm = 1.5 # this is for sanity sake. 
normal_check = 1  


# remove low tpm if its outlier high! 
# fix this with both

this.up  = this.out[ ! ( ( this.out$tree.xcell == "OUTLIER_HIGH"|this.out$tree.TxnSpear2 == "OUTLIER_HIGH"  ) & this.out$tpm < mintpm), ]

this.up$gtex.median = as.numeric ( this.up$gtex.median)
this.up$diff_xcell =   this.up$tree.xcell.median-  this.up$gtex.median
this.up$diff_txns = this.up$median.txn2  - this.up$gtex.median  

# we need to use GTEx xcell however keep in mind that means the means might be wayyy off but I think this is ok since I'm trying 
# so to be safe Will set to AND or OR the reason is because are combining - this will probably miss important things 
# not scratch that will use | OR and basically the idea is that since were depending on 2 algo we need to trust that its correct. 
# 5.13.2022 # see notion notes: https://www.notion.so/tree101/2022-March-433d990a84e64dd1be5d5c80d411147f#49ad526cf39649f3a140dd3db4275228

this.up = this.up[ 
                  (this.up$diff_xcell > -normal_check & this.up$tree.xcell == "OUTLIER_HIGH" ) |
                  (this.up$diff_txns > -normal_check & this.up$tree.TxnSpear2 == "OUTLIER_HIGH" ) 
               , ]
  
# 
min ( this.up$tpm )
# this sanity does not work because sometimes the descrpteancy between the two algo are two large. 
min (  this.up[ this.up$tree.xcell == "OUTLIER_HIGH", ]$diff_xcell    )
min (  this.up[ this.up$tree.TxnSpear2 == "OUTLIER_HIGH", ]$diff_txns    )

setdiff ( rna.key$RNAseq.id, unique ( this.up$RNAseq.id)) 
setdiff ( rna.key$RNAseq.id, unique ( out$RNAseq.id)) 

#this.up$RNAseq.id = factor ( as.character ( this.up$RNAseq.id) , levels= fusion.total$RNAseq.id  )

# oncogene 

role = as.character ( sapply(this.up$gene , function (x) {
  
   if ( any ( x %in% oncogene)){
     return ( "oncogene")
   } else if ( any ( x %in% tsg) ){
     return ( "TSG")
   } else if  ( any ( x %in% tsg.onco$gene  ) ){
     return ( "MULTI")
   } else {
        return ( "NO.ROLE" )
   }
   

   
}

)
)


this.up$role = as.character ( role )
#this.up = merge ( this.up, oncogene[ , c("Gene.Symbol", "role")], by.x="gene", by.y= "Gene.Symbol", all.x=T, all.y=F )
rna.up = this.up 
```

```{r}
# let get rna.down as well 

this.down = this.out [ this.out$both == "OUTLIER_LOW" , ]  
# if its low we want it to be lower for something that is meaningful 

this.down = this.down[ (this.down$median.txn2 > mintpm & this.down$tree.TxnSpear2 == "OUTLIER_LOW") | 
                         ( this.down$tree.xcell.median > mintpm & this.down$tree.xcell == "OUTLIER_LOW" ), ]

min ( this.down[ (this.down$tree.TxnSpear2 == "OUTLIER_LOW"), ]$median.txn2  )
min ( this.down[ (this.down$tree.xcell == "OUTLIER_LOW"), ]$tree.xcell.median  )

this.down$gtex.median = as.numeric ( this.down$gtex.median)

this.down$diff_xcell =   this.down$tree.xcell.median-  this.down$gtex.median
this.down$diff_txns = this.down$median.txn2  - this.down$gtex.median  

# big broad filter first, the idea here is we don't care about under expression if the median for the cohort is too low. 
# this.down = this.down[this.down$temp > mintpm, ] 
# 5.13.2022 this is correct, we do want the disease median to be higher, however 1 for mintpm is too stringent. 
# Ok for low I decided to just do a standard cutoff with the median. 
# basically median has to be greater than 1.5 
# **so I decided to do a simple cutoff that is the median needs to be > 1.5 done.** 
# that way the drop in expression will have some meaning to it.**
# see notion notes: https://www.notion.so/tree101/2022-March-433d990a84e64dd1be5d5c80d411147f#49ad526cf39649f3a140dd3db4275228
normal_check_low = 1.5

this.down = this.down[ 
                     (this.down$tree.xcell.median > normal_check_low & this.down$tree.xcell == "OUTLIER_LOW" )|
                       (this.down$median.txn2 > normal_check_low & this.down$tree.TxnSpear2 == "OUTLIER_LOW" ) 
                   , ]



setdiff ( rna.key$RNAseq.id, unique ( this.down$RNAseq.id)) 

#this.down$RNAseq.id = factor ( as.character ( this.down$RNAseq.id) , levels= fusion.total$RNAseq.id  )

# oncogene 
role = as.character ( sapply(this.down$gene , function (x) {
  
  if ( any ( x %in% oncogene)){
    return ( "oncogene")
  } else if ( any ( x %in% tsg) ){
    return ( "TSG")
  } else if  ( any ( x %in% tsg.onco$gene  ) ){
    return ( "MULTI")
  } else {
    return ( "NO.ROLE" )
  }
  
  
  
}

)
)


this.down$role = as.character ( role )
#this.down = merge ( this.down, oncogene[ , c("Gene.Symbol", "role")], by.x="gene", by.y= "Gene.Symbol", all.x=T, all.y=F )
rna.down = this.down 


```


```{r}


rna.up = rna.up [ rna.up$both == "OUTLIER_HIGH", ]
rna.down = rna.down [ rna.down$both == "OUTLIER_LOW", ]


rna.up.all = rna.up  
rna.down.all = rna.down


# actually it should not matter since failedChem was already remove prior

up.all_qc = rna.up[!rna.up$RNAseq.id %in% failedChem, ]
down.all_qc = rna.down[!rna.down$RNAseq.id %in% failedChem, ] 

# sanity check here. 
 setdiff ( rna.up.all$RNAseq.id, up.all_qc$RNAseq.id)

rna.up_cosmic = up.all_qc [ up.all_qc$gene %in% unique ( genelist$gene    ), ]
rna.down_cosmic = down.all_qc [ down.all_qc$gene %in% unique ( genelist$gene    ), ]


```


```{r}

# because this involves outlier we need to now remove
fuseout = temp.all[ ! temp.all$RNAseq.id %in% failedChem, ]
fuseout$right = gsub ( ".*-", '',fuseout$fusion)
fuseout = merge ( fuseout, up.all_qc[ , c("RNAseq.id","both","gene")], by.x=c ("right","RNAseq.id"), c( "gene","RNAseq.id" ) )
fuseout = fuseout[  fuseout$frameness %in% c("in.frame" ),  ]
# remove strange read throughs 


fuseout2 = fuseout %>% 
  group_by(fusion, type )  %>%
  dplyr::summarise  (
    Freq = n()
  )  %>% arrange ( desc ( Freq, right) ) %>% data.frame ( stringsAsFactors = F)


fuseout2$fusion = factor ( fuseout2$fusion , levels = rev ( unique ( fuseout2$fusion) ) )
ggplot(fuseout2, aes(x = fusion, y=Freq,  fill = type)) + 
  scale_fill_manual(values =  cc5  ) +
  geom_bar(stat = 'identity', position = 'stack') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") +
  #scale_y_continuous(breaks= max(fuseout2$Freq)) + 
  coord_flip() 


lolli_right = ggplot(fuseout2, aes(x=fusion, y=Freq)) +
  geom_segment( aes(x=fusion, xend=fusion, y=0, yend=Freq), color="black") +
  geom_point( color=cc5[fuseout2$type], size=7, alpha=0.75) +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  ) + coord_flip() + xlab("") + ylab("Total") +
  theme(legend.position="bottom", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) + ggtitle ( "With Outlier Analysis")






## rethinking outlier fusions by plotting median and the tpm instead 

fuse_out2 = temp.all[ ! temp.all$RNAseq.id %in% failedChem, ]
fuse_out2$right = gsub ( ".*-", '',fuse_out2$fusion)
fuse_out2 = merge ( fuse_out2, up.all_qc, by.x=c ("right","RNAseq.id"), c( "gene","RNAseq.id" ) )
fuse_out2 = fuse_out2[  fuse_out2$frameness %in% c("in.frame" ),  ]

diff = c()
for ( r in 1:nrow ( fuse_out2)){
  r = fuse_out2[r, ]
  d1 = r$tpm - r$median.txn2
  d2 = r$tpm - r$tree.xcell.median
  if ( d2 > d1 ){ d1 = d2 }
  diff = c( diff, d1 )
}
fuse_out2$delta = diff 

fuse_out2 = fuse_out2[ order ( fuse_out2$fusion, fuse_out2$delta ), ]
fuse_out2$fusion = factor ( fuse_out2$fusion, levels = fuse_out2$fusion)

fuse_out2$c = ''; 
fuse_out2$c[1:17] = "1"
fuse_out2$c[18:35] = "2"
fuse_out2$c[36:53] = "3"
fuse_out2$c[54:71] = "4"




fuse_outlier = ggplot( fuse_out2  , aes(x=reorder_within (  RNAseq.id, delta ,  fusion) , y=delta, fill=type   )) +
  geom_bar(stat = 'identity') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("log2 difference") + 
  xlab("") + 
  theme(legend.position="bottom",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=30),
        axis.title.y     = element_text(size=30), 
        legend.text      =element_text(size=12)
        , plot.title = element_text(size = 30, face = "bold")
  ) + 
  scale_x_reordered() +
  scale_fill_manual( values= cc5 ) + 
  coord_flip() + ggtitle ( "Fusion 3' Outliers") +
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  facet_wrap(. ~ type, scales = "free" , ncol=3 ) +
  theme(
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) +
  geom_text(
    aes(label=fusion) 
    #position = position_dodge(width=0.8),
    #,nudge_y = 1
    ,hjust = 1
    ,size=4
  ) +  theme(strip.text.x = element_text(size = 20) )





  #facet_wrap(fusion ~., scales = "free" , ncol=10 )
  #facet_wrap(. ~ type, scales = "free" , ncol=3 )
   #facet_wrap( ~fusion , ncol=1,  scales = "free"  )
  #facet_grid( fusion~. ,  scales = "free", space="free" ,  switch ="both"  ) + theme(strip.text.y = element_text(angle = 0))
  #facet_grid(fusion ~., scales = "free" , space = "free", switch="x"  ) 
#facet_grid(fusion ~ type, scales = "free" , space = "free" )

```


### Fusion right gene with outlier 


```{r fig=TRUE,fig.width=10, fig.height=10, echo=FALSE, include=TRUE , results='asis'}
lolli_right
```

### Is diagnostic or Well known 
```{r fig=TRUE,fig.width=8, fig.height=8, echo=FALSE, include=TRUE , results='asis'}
well.known.diag
```

### Is druggable 
```{r fig=TRUE,fig.width=6, fig.height=8, echo=FALSE, include=TRUE , results='asis'}
# get table from above 
 kable( is.drug , format = "html" , row.names = F, caption="", table.attr = "style='width:40%;'" ) %>% kable_classic(full_width = T, position = "center", )
```



```{r}

# make oncoprint for fusion! 
temp.all$right =  gsub ( ".*-", '',temp.all$fusion)
oncofuse = fusion[ fusion$canon.right == 1, ]$right

setdiff (  oncofuse, temp.all$right)
oncofuse = temp.all[temp.all$right %in% oncofuse, ]
oncofuse$left =  gsub ( "-.*", '',oncofuse$fusion)


oncofuse$wgs.evidence = ifelse (oncofuse$SV_RESULT %in% c( "-", "NO_WGS"), "", "WGS.evidence" )
oncofuse = oncofuse[!oncofuse$RNAseq.id %in% failedChem,   ]


# now lets reorder column and row by highest freq
right_rnk = oncofuse  %>% 
  group_by( right )  %>%
  dplyr::summarise  (
    Freq = n()
  )  %>% arrange ( desc ( Freq, right) ) %>% data.frame ( stringsAsFactors = F) %>% .$right 

left_rnk = oncofuse  %>% 
  group_by( left )  %>%
  dplyr::summarise  (
    Freq = n()
  )  %>% arrange ( desc ( Freq, left) ) %>% data.frame ( stringsAsFactors = F) %>% .$left 



# add in outlier 
rna.both2 = rbind ( up.all_qc, down.all_qc )
oncofuse2 = merge ( oncofuse, rna.both2 [ , c("gene","both", "RNAseq.id")], by.x=c("RNAseq.id", "right" ), by.y=c("RNAseq.id","gene"), all.x=T, all.y=F  )
colnames ( oncofuse2 ) [ colnames ( oncofuse2 ) == "both"] = "OUTLIER"
oncofuse2$OUTLIER = ifelse ( is.na ( oncofuse2$OUTLIER ), "", oncofuse2$OUTLIER)
oncofuse2 =  reshape2::melt ( oncofuse2[ , c( "left","right" , "type", "RNAseq.id", "wgs.evidence", "OUTLIER")], id.vars = c("RNAseq.id", "left","right" ))
setdiff ( unique ( oncofuse2$right), unique ( oncofuse$right)  )

z.up = oncofuse2 %>%
  group_by(left,right) %>%
  summarise(both=unique ( paste0(unique ( value),collapse = ';'))) %>%
  pivot_wider(names_from = right,values_from=both) %>% data.frame(stringsAsFactors = F)

# sanity check 
setdiff ( unique ( oncofuse$right)  ,  colnames ( z.up ) ) 


z.up[is.na(z.up)] <- ""

z.up[] <- lapply(z.up, function(x) as.character(gsub(";$", "", x)))
z.up[] <- lapply(z.up, function(x) as.character(gsub(";;", ";", x)))

# sanity check. 
setdiff ( unique ( oncofuse$right)  ,  colnames ( z.up ) ) 
unique ( as.character ( as.matrix ( z.up ) ))



#z.up <-  reshape2::dcast(biomark , gene ~ RNAseq.id, value.var=c ( "both") )
z.up.name <- z.up$left # grab the gene first 
z.up$left <- NULL
z.up <- as.matrix (z.up)
row.names (z.up) <- z.up.name
z.up[ is.na(z.up) ] <- ""

z.up = z.up [ ]
z.up = z.up[left_rnk, right_rnk]

set.seed(3)
col = sample ( getPalette( 50 ))
names ( col ) = c(1:50)

unique ( as.character ( as.matrix ( z.up ) ))


af = list(
    background = function(x, y, w, h) 
        grid.rect(x, y, w*1, h*1, gp = gpar(fill = "white", col = "black")),
  
    "is.obsv" = function(x, y, w, h) 
        grid.rect(x, y, w*0.90, h*0.90, gp = gpar(fill = "#80B1D3", col = NA)),
    "is.diag" = function(x, y, w, h) 
        grid.rect(x, y, w*0.90, h*0.90, gp = gpar(fill = "#8DD3C7", col = NA)),
    "is.drug" = function(x, y, w, h) 
        grid.rect(x, y, w*0.90, h*0.90, gp = gpar(fill = "#FB8072", col = NA)),
    "is.novel" = function(x, y, w, h) 
        grid.rect(x, y, w*0.90, h*0.90, gp = gpar(fill = "#e8d558", col = NA)),
  
        
    # dots
    WGS.evidence = function(x, y, w, h) 
        grid.points(x, y, pch = 16, size = unit(2.5, "mm") ),
    # crossed lines
    OUTLIER_HIGH = function(x, y, w, h) 
        grid.segments(x - w*0.5, y - h*0.5, x + w*0.5, y + h*0.5, gp = gpar(lwd = 2))
    ,OUTLIER_LOW = function(x, y, w, h) {
        grid.segments(x - w*0.4, y - h*0.4, x + w*0.4, y + h*0.4, gp = gpar(lwd = 2))
        grid.segments(x + w*0.4, y - h*0.4, x - w*0.4, y + h*0.4, gp = gpar(lwd = 2))
    }
)
alter_fun=af


# top annotations 
## we want a stack plot with type 
# input a table with columns matching the oncoprint, in this case genes but each row is an attribute you wanto sum up. 
# the method is to group by the column that is also the column for the oncoprint + the attribute you want to tally and they expand 
# to a wide table.  

topannt = oncofuse[ , c("right","type")]  %>% group_by(right,type) %>% 
  summarise(Freq=n()) %>%
  pivot_wider(names_from = right,values_from=Freq) %>%  mutate_each(funs(replace(., is.na(.), 0))) %>% data.frame(stringsAsFactors = F) 
row.names ( topannt) = topannt[,1]
topannt = topannt[,colnames(z.up)]


# notice how I used cc5 which is the color mathing my other plots. 
top_annt= HeatmapAnnotation(
                total = anno_barplot(t ( topannt) 
                                     #, width = unit(3, "cm")
                                     , gp = gpar(fill = c(cc5[row.names(topannt)]))
                                    , axis_param = list (gp=gpar(fontsize=18), at=c(5,10)  ), bar_width = 1.0 ) 
                # little trick to mask out the annotation. 
                , annotation_name_gp= gpar(fontsize = 0, col="white" ) )


# needs to kep the order of z.up
# fusion.total.disease
# not sure what this is for 
if ( 1> 2 ){
rightannt = fusion.total.disease[ fusion.total.disease$left %in% row.names(z.up), c("left","cancer.subtype")]  %>% group_by(left,cancer.subtype) %>% 
  summarise(Freq=n()) %>%
  pivot_wider(names_from = left,values_from=Freq) %>%  mutate_each(funs(replace(., is.na(.), 0))) %>% data.frame(stringsAsFactors = F) 
row.names ( rightannt) = rightannt[,1]

rightannt = rightannt[,row.names(z.up)]
rightannt$cancer.subtype  = NULL 
right_annt= rowAnnotation(
                total = anno_barplot( t ( rightannt)  
                                     , width = unit(2.2, "cm")
                                     , gp = gpar(fill = c(subcolor[row.names(rightannt)]))
                                    , axis_param = list (gp=gpar(fontsize=18), at=c(5,10)  )
                                    , bar_width = 1.0) 
                # little trick to mask out the annotation. 
                , annotation_name_gp= gpar(fontsize = 0, col="white" ) )
}


# redo left row ranking 





onfuse= oncoPrint(  z.up,
              alter_fun = af
          
            ,show_heatmap_legend = F
            , column_order = colnames ( z.up) 
            , row_order = left_rnk
            , top_annotation = top_annt 
            , show_column_names = T
            ,row_names_side = "left", column_names_side = "top"
            ,show_pct = F 
            #,left_annotation = leftannt 
           # , right_annotation =right_annt
            ,column_title = "3' Gene", column_title_gp = gpar(fontsize = 35)
            , row_title = "5' Gene", row_title_gp = gpar(fontsize = 35)
            
)
 
 # add additional legend 

l3 = c(  col["1"], col["2"], col["4"], col[">5"] )
names ( l3) = c("1","2","4",">5")


l_type = Legend(labels = names ( cc5 )
                #, title = "Type"
                , legend_gp = gpar(fill = cc5,fontsize = 12 ),
             title_position = "leftcenter-rot"
             , labels_gp = gpar(fontsize = 12)
             #, title_gp = gpar(col = "black", fontsize = 15)
             ,grid_height = unit(8, "mm"), grid_width = unit(8, "mm"))

l_annt = Legend(labels = c("OUTLIER_HIGH", "OUTLIER_LOW", "wgs.evidence"), type = "points", pch = c(26, 28, 20),  labels_gp = gpar(fontsize = 12)
                ,grid_height = unit(8, "mm") )

pd = packLegend(l_type, l_annt , row_gap = unit(1, "cm") , direction="vertical")

draw(onfuse,annotation_legend_list = pd)






```


```{r}

# make oncoprint for fusions that are not cannical but just inframe! 
temp.all$right =  gsub ( ".*-", '',temp.all$fusion)
oncoTOP = table (temp.all[ temp.all$frameness == "in.frame", ]$right  )
oncoTOP = sort ( oncoTOP, decreasing = T)


oncoTOP = temp.all[temp.all$right %in% names ( head ( oncoTOP, 12 ) ) , ]
oncoTOP$left =  gsub ( "-.*", '',oncoTOP$fusion)


oncoTOP$wgs.evidence = ifelse (oncoTOP$SV_RESULT %in% c( "-", "NO_WGS"), "", "WGS.evidence" )
oncoTOP = oncoTOP[!oncoTOP$RNAseq.id %in% failedChem,   ]


# now lets reorder column and row by highest freq
right_rnk = oncoTOP  %>% 
  group_by( right )  %>%
  dplyr::summarise  (
    Freq = n()
  )  %>% arrange ( desc ( Freq, right) ) %>% data.frame ( stringsAsFactors = F) %>% .$right 

left_rnk = oncoTOP  %>% 
  group_by( left )  %>%
  dplyr::summarise  (
    Freq = n()
  )  %>% arrange ( desc ( Freq, left) ) %>% data.frame ( stringsAsFactors = F) %>% .$left 



# add in outlier 
rna.both2 = rbind ( up.all_qc, down.all_qc )
oncoTOP2 = merge ( oncoTOP, rna.both2 [ , c("gene","both", "RNAseq.id")], by.x=c("RNAseq.id", "right" ), by.y=c("RNAseq.id","gene"), all.x=T, all.y=F  )
colnames ( oncoTOP2 ) [ colnames ( oncoTOP2 ) == "both"] = "OUTLIER"
oncoTOP2$OUTLIER = ifelse ( is.na ( oncoTOP2$OUTLIER ), "", oncoTOP2$OUTLIER)
oncoTOP2 =  reshape2::melt ( oncoTOP2[ , c( "left","right" , "type", "RNAseq.id", "wgs.evidence", "OUTLIER")], id.vars = c("RNAseq.id", "left","right" ))
setdiff ( unique ( oncoTOP2$right), unique ( oncoTOP$right)  )

z.up = oncoTOP2 %>%
  group_by(left,right) %>%
  summarise(both=unique ( paste0(unique ( value),collapse = ';'))) %>%
  pivot_wider(names_from = right,values_from=both) %>% data.frame(stringsAsFactors = F)

# sanity check 
setdiff ( unique ( oncoTOP$right)  ,  colnames ( z.up ) ) 


z.up[is.na(z.up)] <- ""

z.up[] <- lapply(z.up, function(x) as.character(gsub(";$", "", x)))
z.up[] <- lapply(z.up, function(x) as.character(gsub(";;", ";", x)))

# sanity check. 
setdiff ( unique ( oncoTOP$right)  ,  colnames ( z.up ) ) 
unique ( as.character ( as.matrix ( z.up ) ))



#z.up <-  reshape2::dcast(biomark , gene ~ RNAseq.id, value.var=c ( "both") )
z.up.name <- z.up$left # grab the gene first 
z.up$left <- NULL
z.up <- as.matrix (z.up)
row.names (z.up) <- z.up.name
z.up[ is.na(z.up) ] <- ""

z.up = z.up [ ]
z.up = z.up[left_rnk, right_rnk]

set.seed(3)
col = sample ( getPalette( 50 ))
names ( col ) = c(1:50)

unique ( as.character ( as.matrix ( z.up ) ))


af = list(
  background = function(x, y, w, h) 
    grid.rect(x, y, w*1, h*1, gp = gpar(fill = "white", col = "black")),
  
  "is.obsv" = function(x, y, w, h) 
    grid.rect(x, y, w*0.90, h*0.90, gp = gpar(fill = "#80B1D3", col = NA)),
  "is.diag" = function(x, y, w, h) 
    grid.rect(x, y, w*0.90, h*0.90, gp = gpar(fill = "#8DD3C7", col = NA)),
  "is.drug" = function(x, y, w, h) 
    grid.rect(x, y, w*0.90, h*0.90, gp = gpar(fill = "#FB8072", col = NA)),
  "is.novel" = function(x, y, w, h) 
    grid.rect(x, y, w*0.90, h*0.90, gp = gpar(fill = "#e8d558", col = NA)),
  
  
  # dots
  WGS.evidence = function(x, y, w, h) 
    grid.points(x, y, pch = 16, size = unit(2.5, "mm") ),
  # crossed lines
  OUTLIER_HIGH = function(x, y, w, h) 
    grid.segments(x - w*0.5, y - h*0.5, x + w*0.5, y + h*0.5, gp = gpar(lwd = 2))
  ,OUTLIER_LOW = function(x, y, w, h) {
    grid.segments(x - w*0.4, y - h*0.4, x + w*0.4, y + h*0.4, gp = gpar(lwd = 2))
    grid.segments(x + w*0.4, y - h*0.4, x - w*0.4, y + h*0.4, gp = gpar(lwd = 2))
  }
)
alter_fun=af


# top annotations 
## we want a stack plot with type 
# input a table with columns matching the oncoprint, in this case genes but each row is an attribute you wanto sum up. 
# the method is to group by the column that is also the column for the oncoprint + the attribute you want to tally and they expand 
# to a wide table.  

topannt = oncoTOP[ , c("right","type")]  %>% group_by(right,type) %>% 
  summarise(Freq=n()) %>%
  pivot_wider(names_from = right,values_from=Freq) %>%  mutate_each(funs(replace(., is.na(.), 0))) %>% data.frame(stringsAsFactors = F) 
row.names ( topannt) = topannt[,1]
topannt = topannt[,colnames(z.up)]


# notice how I used cc5 which is the color mathing my other plots. 
top_annt= HeatmapAnnotation(
  total = anno_barplot(t ( topannt) 
                       #, width = unit(3, "cm")
                       , gp = gpar(fill = c(cc5[row.names(topannt)]))
                       , axis_param = list (gp=gpar(fontsize=10), at=c(20, 50)  ), bar_width = 1.0 ) 
  # little trick to mask out the annotation. 
  , annotation_name_gp= gpar(fontsize = 0, col="white" ) )


# needs to kep the order of z.up
# fusion.total.disease
# not sure what this is for 



# redo left row ranking 


w1 = seq ( 1, nrow ( z.up), by=3  )



ha1 = rowAnnotation(genes = anno_mark(at = w1 , labels = row.names(z.up)[w1]  ),   annotation_name_gp= gpar(fontsize = 20)  )


onfuse.highFreq= oncoPrint(  z.up,
                    alter_fun = af
                    
                    ,show_heatmap_legend = F
                    , column_order = colnames ( z.up) 
                    , row_order = left_rnk
                    , top_annotation = top_annt 
                    , show_column_names = T
                    ,row_names_side = "left", column_names_side = "top"
                    ,show_pct = F 
                    #,left_annotation = leftannt 
                    # , right_annotation =right_annt
                    ,column_title = "3' Gene", column_title_gp = gpar(fontsize = 15)
                    , row_title = "5' Gene", row_title_gp = gpar(fontsize = 15)
                    ,  right_annotation = ha1
                    , show_row_names = F 
                    
)

# add additional legend 

l3 = c(  col["1"], col["2"], col["4"], col[">5"] )
names ( l3) = c("1","2","4",">5")


l_type = Legend(labels = names ( cc5 )
                #, title = "Type"
                , legend_gp = gpar(fill = cc5,fontsize = 12 ),
                title_position = "leftcenter-rot"
                , labels_gp = gpar(fontsize = 12)
                #, title_gp = gpar(col = "black", fontsize = 15)
                ,grid_height = unit(8, "mm"), grid_width = unit(8, "mm"))

l_annt = Legend(labels = c("OUTLIER_HIGH", "OUTLIER_LOW", "wgs.evidence"), type = "points", pch = c(26, 28, 20),  labels_gp = gpar(fontsize = 12)
                ,grid_height = unit(8, "mm") )

pd = packLegend(l_type, l_annt , row_gap = unit(1, "cm") , direction="vertical")

draw(onfuse.highFreq,annotation_legend_list = pd)






```






### OUTLIER_FUSION

```{r fig=TRUE,fig.width=8, fig.height=8, echo=FALSE, include=TRUE }
DT::datatable(fuseout,
              caption = '', rownames = T, filter= list ( position="top", clear = FALSE )  , extensions = 'Buttons'
              , options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print')
                               , autoWidth = T, scrollX=T, className = 'dt-left', pageLength =50
              )
              
)



```

### OUTLIER_FUSION in.frame only

```{r fig=TRUE,fig.width=8, fig.height=8, echo=FALSE, include=TRUE }

of = fuseout[fuseout$frameness == "in.frame", ]
of = of[ , -c(1,7,8,10)]

```





```{r}
# tabulation 

combolist=genelist$gene
this.up_tab = up.all_qc [ up.all_qc$gene %in% unique ( c ( drug.up$gene, drug.down$gene, combolist) ),  ]
this.down_tab = down.all_qc [ down.all_qc$gene %in% unique ( c ( drug.up$gene, drug.down$gene, combolist) ),  ]


this.up_tab = merge ( this.up_tab, rna.key[ , c("RNAseq.id", "cancer.subtype")], by="RNAseq.id")

this.up_tab2 =  this.up_tab[this.up_tab$role %in%  c (  "oncogene") | this.up_tab$gene %in% drug.up$gene, ]   %>% dplyr::group_by (cancer.subtype, gene) %>%
 #this.up_tab2 =  this.up_tab %>% dplyr::group_by (cancer.subtype, gene) %>%
 dplyr::summarise(
    total = n ( ) # 
  ) %>% data.frame()

this.up_tab2 = this.up_tab2[ order ( this.up_tab2$total, decreasing = T), ]

up.out = data.frame()

for ( ut in unique ( this.up_tab2$cancer.subtype)){
  ut = this.up_tab2[this.up_tab2$cancer.subtype == ut, ]
  ut = ut [ ut$total > 2, ]
  up.out = rbind ( head ( ut, 8),up.out )
}




up.out$total =  as.numeric( up.out$total) 
up.out = merge ( up.out, total_subcancer_truseq, by="cancer.subtype")
up.out$percent = round ( up.out$total / up.out$Freq, 2)

up_oncogene = ggplot( up.out  , aes(x=reorder_within ( gene, percent , cancer.subtype) , y=percent, fill=cancer.subtype   )) +
  geom_bar(stat = 'identity', position = 'stack') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 7.5 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=15), 
        legend.text      =element_text(size=12)
        , plot.title = element_text(size = 12, face = "bold")
  ) + 
  scale_x_reordered() +
  scale_fill_manual( values= subcolor ) + coord_flip() + ggtitle ( "Recurrent UP Oncogenes") +
  # scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) + # plot in intergers
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  facet_grid(cancer.subtype~., scales = "free" , space = "free" ) #+  theme (strip.text.y = element_text(size = 6, colour = "black") )




```


```{r}
## down TSG 

this.down_tab = down.all_qc [ down.all_qc$gene %in% unique ( c ( drug.down$gene, drug.down$gene, combolist) ),  ]


this.down_tab = merge ( this.down_tab, rna.key[ , c("RNAseq.id", "cancer.subtype")], by="RNAseq.id")

this.down_tab2 =  this.down_tab[this.down_tab$role %in%  c (  "TSG"), ]   %>% dplyr::group_by (cancer.subtype, gene) %>%
  #this.down_tab2 =  this.down_tab %>% dplyr::group_by (cancer.subtype, gene) %>%
  dplyr::summarise(
    total = n ( ) # 
  ) %>% data.frame()

this.down_tab2 = this.down_tab2[ order ( this.down_tab2$total, decreasing = T), ]

down.out = data.frame()

for ( ut in unique ( this.down_tab2$cancer.subtype)){
  ut = this.down_tab2[this.down_tab2$cancer.subtype == ut, ]
  ut = ut [ ut$total > 2, ]
  down.out = rbind ( head ( ut, 8),down.out )
}




down.out$total =  as.numeric( down.out$total) 
down.out = merge ( down.out, total_subcancer_truseq, by="cancer.subtype")
down.out$percent = round ( down.out$total / down.out$Freq, 2)

down_tsg = ggplot( down.out  , aes(x=reorder_within ( gene, percent , cancer.subtype) , y=percent, fill=cancer.subtype   )) +
  geom_bar(stat = 'identity', position = 'stack') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 7.5 ),
        axis.text.x = element_text(angle = 0, size= 15 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12)
        , plot.title = element_text(size = 12, face = "bold")
  ) + 
  scale_x_reordered() +
  scale_fill_manual( values= subcolor ) + coord_flip() + ggtitle ( "Recurrent DOWN TSGs") +
  # scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) + # plot in intergers
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  facet_grid(cancer.subtype~., scales = "free" , space = "free" ) 

```

## OUTLIER oncogene/tsg 
```{r}
up_oncogene + down_tsg
```

```{r}
# draw grid 

up_onc_grid = dcast( up.out 
              , gene  ~ cancer.subtype, value.var= "percent"  )

up_onc_grid[is.na(up_onc_grid)] = 0 


order_f = data.frame ( table ( up.out$gene ) )
order_f = order_f[ order(order_f$Freq, decreasing = F), ]

up.out$gene = factor ( up.out$gene, levels= as.character ( order_f$Var1))

out_up_grid = ggplot(up.out,
                         aes(x =  cancer.subtype, 
                             y =  gene ,
                             colour = cancer.subtype ,
                             size = percent )) +
  geom_point() +   scale_size_continuous(range = c(.2,5.2)) +
  labs(x = NULL, y = NULL) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")
        , legend.position = "none") +
  scale_colour_manual( values= subcolor ) + 
  geom_hline(yintercept = seq(1.5, (length(unique(up.out$gene))-0.5), 1) , colour="grey" , alpha=.5) + 
  geom_vline(xintercept = seq(1.5, (length(unique(up.out$cancer.subtype))-0.5), 1), colour="grey" , alpha=.5 )






down_onc_grid = dcast( down.out 
                     , gene  ~ cancer.subtype, value.var= "percent"  )

down_onc_grid[is.na(down_onc_grid)] = 0 


order_f = data.frame ( table ( down.out$gene ) )
order_f = order_f[ order(order_f$Freq, decreasing = F), ]

down.out$gene = factor ( down.out$gene, levels= as.character ( order_f$Var1))

out_down_grid = ggplot(down.out,
                     aes(x =  cancer.subtype, 
                         y =  gene ,
                         colour = cancer.subtype ,
                         size = percent )) +
  geom_point() +   scale_size_continuous(range = c(.2,7.2)) +
  labs(x = NULL, y = NULL) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")
        , legend.position = "none") +
  scale_colour_manual( values= subcolor ) + 
  geom_hline(yintercept = seq(1.5, (length(unique(down.out$gene))-0.5), 1) , colour="grey" , alpha=.5) + 
  geom_vline(xintercept = seq(1.5, (length(unique(down.out$cancer.subtype))-0.5), 1), colour="grey" , alpha=.5 )



```


```{r}
# cell surface 
# may be we need another set of outliers, one that is > than gtex pan? 
gtex_out = out[ out$gtex.pan == "OUTLIER_HIGH", ]
cell_surface = readRDS(  "/ehome/resource/markers/cell_surface.rds" )



this.up_surface_tab = up.all_qc [ up.all_qc$gene %in% unique ( c ( cell_surface$cell_surface$gene ) ),  ]



this.up_surface_tab = merge ( this.up_surface_tab, rna.key[ , c("RNAseq.id", "cancer.subtype")], by="RNAseq.id")

this.up_surface_tab2 =  this.up_surface_tab  %>% dplyr::group_by (cancer.subtype, gene) %>%
  #this.up_surface_tab2 =  this.up_surface_tab %>% dplyr::group_by (cancer.subtype, gene) %>%
  dplyr::summarise(
    total = n ( ) # 
  ) %>% data.frame()

this.up_surface_tab2 = this.up_surface_tab2[ order ( this.up_surface_tab2$total, decreasing = T), ]

up.out = data.frame()

for ( ut in unique ( this.up_surface_tab2$cancer.subtype)){
  ut = this.up_surface_tab2[this.up_surface_tab2$cancer.subtype == ut, ]
  ut = ut [ ut$total > 3, ]
  up.out = rbind ( head ( ut, 8),up.out )
}




up.out$total =  as.numeric( up.out$total) 
up.out = merge ( up.out, total_subcancer, by="cancer.subtype")
up.out$percent = round ( up.out$total / up.out$Freq, 2)

cell_mem_disease = ggplot( up.out  , aes(x=reorder_within ( gene, percent , cancer.subtype) , y=percent, fill=cancer.subtype   )) +
  geom_bar(stat = 'identity', position = 'stack') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 7.5 ),
        axis.text.x = element_text(angle = 0, size= 15 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12)
        , plot.title = element_text(size = 12, face = "bold")
  ) + 
  scale_x_reordered() +
  scale_fill_manual( values= subcolor ) + coord_flip() + ggtitle ( "Recurrent UP Disease (Cell-membrane)") +
  # scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) + # plot in intergers
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  facet_grid(cancer.subtype~., scales = "free" , space = "free" ) 



```


```{r}
# cell surface 
# may be we need another set of outliers, one that is > than gtex pan? 
gtex_out = out[ out$gtex.pan == "OUTLIER_HIGH", ]




this.up_surface_tab = gtex_out [ gtex_out$gene %in% unique ( c ( cell_surface$cell_surface$gene ) ),  ]



this.up_surface_tab = merge ( this.up_surface_tab, rna.key[ , c("RNAseq.id", "cancer.subtype")], by="RNAseq.id")

this.up_surface_tab2 =  this.up_surface_tab  %>% dplyr::group_by (cancer.subtype, gene) %>%
  #this.up_surface_tab2 =  this.up_surface_tab %>% dplyr::group_by (cancer.subtype, gene) %>%
  dplyr::summarise(
    total = n ( ) # 
  ) %>% data.frame()

this.up_surface_tab2 = this.up_surface_tab2[ order ( this.up_surface_tab2$total, decreasing = T), ]

up.out = data.frame()

for ( ut in unique ( this.up_surface_tab2$cancer.subtype)){
  ut = this.up_surface_tab2[this.up_surface_tab2$cancer.subtype == ut, ]
  ut = ut [ ut$total > 3, ]
  up.out = rbind ( head ( ut, 8),up.out )
}




up.out$total =  as.numeric( up.out$total) 
up.out$total =  as.numeric( up.out$total) 
up.out = merge ( up.out, total_subcancer, by="cancer.subtype")
up.out$percent = round ( up.out$total / up.out$Freq, 2)


cell_mem_gtex = ggplot( up.out  , aes(x=reorder_within ( gene, percent , cancer.subtype) , y=percent, fill=cancer.subtype   )) +
  geom_bar(stat = 'identity', position = 'stack') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 7.5 ),
        axis.text.x = element_text(angle = 0, size= 15 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12)
        , plot.title = element_text(size = 12, face = "bold")
  ) + 
  scale_x_reordered() +
  scale_fill_manual( values= subcolor ) + coord_flip() + ggtitle ( "Recurrent UP GTEx (Cell-membrane)") +
  # scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) + # plot in intergers
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  facet_grid(cancer.subtype~., scales = "free" , space = "free" ) 

```


### OUTLIER cell membrane 
```{r}

cell_mem_disease + cell_mem_gtex 

```


```{r}

kinase = read.xlsx( "/ehome/resource/markers/Kincat_Hsap.08.02.xlsx" ) 



this.up_kinase_tab =  merge (  up.all_qc  , kinase [ , 1:5], by.x="gene", by.y="Name" ) 



this.up_kinase_tab = merge ( this.up_kinase_tab, rna.key[ , c("RNAseq.id", "cancer.subtype")], by="RNAseq.id")

this.up_kinase_tab2 =  this.up_kinase_tab  %>% dplyr::group_by (cancer.subtype, gene) %>%
  #this.up_kinase_tab2 =  this.up_kinase_tab %>% dplyr::group_by (cancer.subtype, gene) %>%
  dplyr::summarise(
    total = n ( ) # 
  ) %>% data.frame()

this.up_kinase_tab2 = this.up_kinase_tab2[ order ( this.up_kinase_tab2$total, decreasing = T), ]





# 2/21/2022 changed to only cancer genes 
# also instea we should stack it 
this.up_kinase_tab2 = this.up_kinase_tab2[this.up_kinase_tab2$gene %in% genelist$gene, ]


this.up_kinase_tab2 = this.up_kinase_tab2  %>% dplyr::group_by ( gene) %>%
  dplyr::mutate(
    order = sum (total ) 
  ) %>% arrange(desc(order)) %>% select ( -order) %>% 
  data.frame()

stackout_kinase = this.up_kinase_tab2[this.up_kinase_tab2$total >=2, ]
stackout_kinase$gene = factor ( stackout_kinase$gene, levels = rev ( unique ( stackout_kinase$gene)) )

out_kinase_stack = ggplot(data=stackout_kinase, aes(x=gene, y=total,  fill=cancer.subtype  )) +
geom_bar(stat="identity",  width = 1, position = "stack", color="grey") +  # color=""
theme_bw() +
theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
      panel.background = element_blank(), axis.line = element_line(colour = "black")) +
ylab("Top Recurrent Kinase Outliers") + 
xlab("") + 
 scale_y_continuous( expand = c(0, .30), breaks = c(5,10,15,20,25,30))   + 
  coord_flip() + scale_fill_manual( values = subcolor) +
  theme(legend.position="bottom",  legend.key = element_blank(),
      
      axis.text.y = element_text(size= 20),
      axis.text.x = element_text(angle = 0, size= 15 ),
      axis.title.x = element_text(size=15),
      axis.title.y     = element_text(size=20), 
      legend.text      =element_text(size=12)
      , plot.title = element_text(size = 12, face = "bold")
) 

 
 
 



up.out = data.frame()

for ( ut in unique ( this.up_kinase_tab2$cancer.subtype)){
  ut = this.up_kinase_tab2[this.up_kinase_tab2$cancer.subtype == ut, ]
  ut = ut [ ut$total > 2, ]
  up.out = rbind ( head ( ut, 8),up.out )
}




up.out$total =  as.numeric( up.out$total) 
up.out = merge ( up.out, total_subcancer, by="cancer.subtype")
up.out$percent = round ( up.out$total / up.out$Freq, 2)

top_kinase = ggplot( up.out  , aes(x=reorder_within ( gene, percent , cancer.subtype) , y=percent, fill=cancer.subtype   )) +
geom_bar(stat = 'identity', position = 'stack') +
theme_bw() +
theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
      panel.background = element_blank(), axis.line = element_line(colour = "black")) +
ylab("") + 
xlab("") + 
theme(legend.position="none",  legend.key = element_blank(),
      
      axis.text.y = element_text(size= 7.5 ),
      axis.text.x = element_text(angle = 0, size= 15 ),
      axis.title.x = element_text(size=12),
      axis.title.y     = element_text(size=20), 
      legend.text      =element_text(size=12)
      , plot.title = element_text(size = 12, face = "bold")
) + 
scale_x_reordered() +
scale_fill_manual( values= subcolor ) + coord_flip() + ggtitle ( "Recurrent UP Kinase Gene") +
# scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) + # plot in intergers
scale_y_continuous(expand = expansion(mult = c(0, .1))) +
facet_grid(cancer.subtype~., scales = "free" , space = "free" )


## get kinase family 





```


### OUTLIER Kinase


```{r}
top_kinase +top_kinase_family
```



```{r}

# objective study outliers by disease, however the no.role column does not 
# distinguish between up or down 
table ( up.all_qc[  up.all_qc$gene %in%  combolist, ] $role)
table ( up.all_qc $role)   

up.all_qc_annt = up.all_qc
up.all_qc_annt$role = NULL 
up.all_qc_annt$cancer = ifelse ( up.all_qc_annt$gene %in% combolist, "is.cancer", "not.cancer" )
up.all_qc_annt = merge ( up.all_qc_annt, genelist [, c("gene", "cosmic.annot")], by="gene", all.x=T, all.y=F)
up.all_qc_annt$cosmic.annot = ifelse ( is.na(up.all_qc_annt$cosmic.annot), "non.cancer", up.all_qc_annt$cosmic.annot )
table ( up.all_qc_annt[ up.all_qc_annt$cancer == "not.cancer", ]$cosmic.annot )
table ( up.all_qc_annt[ up.all_qc_annt$cancer != "not.cancer", ]$cosmic.annot )
up.all_qc_annt$cosmic.annot = ifelse ( grepl ( ",", up.all_qc_annt$cosmic.annot), "MULTI", up.all_qc_annt$cosmic.annot)


up.all_qc_annt = merge ( up.all_qc_annt, rna.key[ , c ( "RNAseq.id", "cancer.subtype") ], by="RNAseq.id")

# FIND top for each disease. 
# most be at least 35% 
up.out_tab  = data.frame()
up.out_tab_all = data.frame()
up.out_tab_nofilter = data.frame()
for ( ut2 in unique ( rna.key$cancer.subtype )     ){

  ut = up.all_qc_annt[up.all_qc_annt$cancer.subtype == ut2, ]
  fq = data.frame( table ( ut$gene))
  fq = fq [ order ( -fq$Freq), ]
  # remove "RP" 
  fq = fq [ !grepl ( "^RP", fq$Var1), ]
  
  ut_total = length ( unique ( ut$RNAseq.id) )  
  ut_trueseq = length ( trueseq.key[ trueseq.key$cancer.subtype == ut2, ]$RNAseq.id )
  
  if (ut_total !=  ut_trueseq ){
    print ( ut2 )
    print ( 'something wrong with total' )
    stop 
    
  }
  
  fq$per = fq$Freq /  ut_total
  colnames ( fq )[1] = "gene"
  fq = merge ( fq, ut[ , c ( "gene", "cancer", "cosmic.annot", "cancer.subtype") ]  , by="gene" )
  fq = fq [!duplicated (fq$gene), ]
  fq = fq [ order ( -fq$Freq), ]
  
  up.out_tab =  rbind ( up.out_tab,  rbind ( head ( fq[ fq$per > .35 & fq$cancer != "not.cancer", ], 10), 
                       head ( fq[ fq$per > .35 & fq$cancer == "not.cancer", ], 10 ) 
                       )
  )
  
 up.out_tab_all =  rbind ( up.out_tab_all,  rbind ( fq[ fq$per > .35 & fq$cancer != "not.cancer", ] , 
                            fq[ fq$per > .35 & fq$cancer == "not.cancer", ] 
                       )
  )
 
 
  up.out_tab_nofilter =  rbind ( up.out_tab_nofilter,  rbind ( fq ) )
  
  
  
 
 
}



cancer_tab = up.out_tab_all[ up.out_tab_all$cancer == "is.cancer", ]
cancer_tab = cancer_tab [ ! grepl ( "\\.|-", cancer_tab$gene), ]

cancer_tab$is.kinase = ifelse ( cancer_tab$gene %in% kinase$Name, "*", "")

outR_cancer =  cancer_tab %>% group_by(cancer.subtype) %>% top_n(wt = per, n = 6)  %>%
   slice(tail(row_number(), 10)) %>% 
  ggplot( . , aes(x=reorder_within ( gene, per , cancer.subtype) , y=per, fill=cosmic.annot   )) +
   geom_bar(stat = 'identity', position = 'stack') +
   geom_text(aes(label=is.kinase), vjust=.8, hjust=- .8  , size=10) +
    theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black")) +
    ylab("") + 
    xlab("") + 
    theme(legend.position="bottom",  legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=12),
          axis.title.y     = element_text(size=20), 
          legend.text      =element_text(size=12)
          , plot.title = element_text(size = 30, face = "bold")
    ) + 
    scale_x_reordered() +
  scale_fill_manual( values= getPalette( 10 ) ) + 
  coord_flip() + ggtitle ( "Most Recurrent Cosmic cancer UP outliers") +
 # scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) + # plot in intergers
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  facet_wrap(cancer.subtype~.
             #, scales = "free" 
             , scales = "free_y" 
            , ncol=5
             )

nocancer_tab = up.out_tab_all[ up.out_tab_all$cancer == "not.cancer", ]
nocancer_tab = nocancer_tab [ ! grepl ( "\\.|-", nocancer_tab$gene), ]

outR_nocancer = nocancer_tab %>% group_by(cancer.subtype) %>% top_n(wt = per, n = 6)  %>%
   slice(tail(row_number(), 10)) %>% 
ggplot( . , aes(x=reorder_within ( gene, per , cancer.subtype) , y=per, fill=cosmic.annot   )) +
   geom_bar(stat = 'identity', position = 'stack') +
    theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black")) +
    ylab("") + 
    xlab("") + 
    theme(legend.position="none",  legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=12),
          axis.title.y     = element_text(size=20), 
          legend.text      =element_text(size=12)
          , plot.title = element_text(size = 30, face = "bold")
    ) + 
    scale_x_reordered() +
  scale_fill_manual( values="#afb6ba") + 
  coord_flip() + ggtitle ( "Most Recurrent Non-Cosmic cancer UP outliers") +
 # scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) + # plot in intergers
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  facet_wrap(cancer.subtype~.
             #, scales = "free" 
             , scales = "free_y" 
            , ncol=5
             )


# which one genes shows up the most 
nocancer_tab_f = nocancer_tab %>% group_by(  gene ) %>% 
  summarise ( 
    total = n()
    ) %>% arrange(desc(total)) %>% data.frame ()

# which genes are present in more than 6 or more cancer subtypes? 
nocancer_tab_f2 = nocancer_tab_f[nocancer_tab_f$total >=6, ]

nocancer_tab_f2  = merge ( nocancer_tab_f2 , nocancer_tab[ , c("gene", "cancer.subtype", "per")], by="gene"  )
nocancer_tab_f2$gene = as.character ( nocancer_tab_f2$gene)

nocancer_tab_f2$per = round ( nocancer_tab_f2$per, 1 )
nocancer_tab_f2 = nocancer_tab_f2 [ order ( -nocancer_tab_f2$total), ]

nocancer_tab_f2$cancer.subtype = factor ( nocancer_tab_f2$cancer.subtype, levels= rev ( names ( subcolor))  )
nocancer_tab_f2$gene = factor ( nocancer_tab_f2$gene, levels=rev ( unique ( nocancer_tab_f2$gene )  ) )



nocancer_tab_f2_plot = ggplot(data=nocancer_tab_f2, aes(x=gene, y=1,  fill=cancer.subtype, label=per  )) +
  geom_bar(stat="identity",  width = 1, position = "stack", color="grey") +  # color=""
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"),
          axis.text.y = element_text(size= 20 ),
      axis.text.x =  element_text(size= 20 ),
      axis.title.x = element_text(size=20),
      axis.title.y     = element_text(size=25), 
      legend.text      =element_text(size=25),
      legend.title = element_text(size=25),
      plot.title = element_text(size = 25, face = "bold")
        
        ) +
  ylab("") + 
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  coord_flip() + scale_fill_manual(values=subcolor) + 
  geom_text(size = 10, position = position_stack(vjust = 0.5) ) + ggtitle ( "Non-Cosmic outliers most prevelant across dieases") +
  scale_x_discrete(expand = c(0, 0)) +
   scale_y_discrete(expand = c(0, 0))


### cancer genes 

cancer_tab_f = cancer_tab %>% group_by(  gene ) %>% 
  summarise ( 
    total = n()
  ) %>% arrange(desc(total)) %>% data.frame ()

# which genes are present in more than 6 or more cancer subtypes? 
cancer_tab_f2 = cancer_tab_f[cancer_tab_f$total >=3, ]

cancer_tab_f2  = merge ( cancer_tab_f2 , cancer_tab[ , c("gene", "cancer.subtype", "per")], by="gene"  )
cancer_tab_f2$gene = as.character ( cancer_tab_f2$gene)

cancer_tab_f2$per = round ( cancer_tab_f2$per, 1 )
cancer_tab_f2 = cancer_tab_f2 [ order ( -cancer_tab_f2$total), ]

cancer_tab_f2$cancer.subtype = factor ( cancer_tab_f2$cancer.subtype, levels= rev ( names ( subcolor))  )
cancer_tab_f2$gene = factor ( cancer_tab_f2$gene, levels=rev ( unique ( cancer_tab_f2$gene )  ) )



cancer_tab_f2_plot = ggplot(data=cancer_tab_f2, aes(x=gene, y=1,  fill=cancer.subtype, label=per  )) +
  geom_bar(stat="identity",  width = 1, position = "stack", color="grey") +  # color=""
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"),
        axis.text.y = element_text(size= 20 ),
        axis.text.x =  element_text(size= 20 ),
        axis.title.x = element_text(size=20),
        axis.title.y     = element_text(size=25), 
        legend.text      =element_text(size=25),
        legend.title = element_text(size=25),
        plot.title = element_text(size = 25, face = "bold"  )
        ,  legend.position="none"  
        
  ) +
  ylab("") + 
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  coord_flip() + scale_fill_manual(values=subcolor)  +
  geom_text(size = 10, position = position_stack(vjust = 0.5) ) + ggtitle ( "Cosmic outliers most prevelant across dieases") +
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0))






```




```{r}
combolist=genelist$gene
this.up = up.all_qc [ up.all_qc$gene %in% unique ( c ( drug.up$gene, drug.down$gene, combolist) ),  ]
this.down = down.all_qc [ down.all_qc$gene %in% unique ( c ( drug.up$gene, drug.down$gene, combolist) ),  ]


this.up_unique =  merge ( this.up, rna.key[ ,c("RNAseq.id", "cancer.subtype")], by="RNAseq.id")
this.up_unique =  plyr::count(this.up_unique[!duplicated ( paste( this.up_unique$gene, this.up_unique$cancer.subtype  )), ], c("cancer.subtype","role"))
colnames( this.up_unique ) = c("cancer.subtype", "role","total.up")
this.up_unique = this.up_unique[this.up_unique$role %in% c( "oncogene" ), ]

this.up = plyr::count(this.up, c("RNAseq.id","role"))
colnames ( this.up) = c("RNAseq.id", "role","total.up")

# add zeros 
zero = setdiff (  as.character ( unique ( trueseq.key$RNAseq.id) ), as.character ( this.up$RNAseq.id )  )
for ( z in zero ){
  this.up = rbind ( this.up, data.frame ( RNAseq.id=z, role="NO.ROLE", total.up=0) )     
}




this.up = merge ( this.up, trueseq.key[ , c("RNAseq.id", "cancer.subtype")], by = "RNAseq.id")
outorder = c()

this.up = this.up [ order ( this.up$cancer.subtype, this.up$total.up), ]
outorder =  this.up %>% dplyr::group_by(cancer.subtype, RNAseq.id) %>%
  dplyr::summarise(
    total_order = sum ( total.up  ) # count occurence but leave out pizzly
  ) %>% data.frame()

outorder = outorder [ order ( outorder$cancer.subtype, outorder$total_order), ]


this.up$RNAseq.id = factor ( this.up$RNAseq.id, levels= outorder$RNAseq.id )
this.up$cancer.subtype = factor ( this.up$cancer.subtype, levels = cancer.subtype.order)



out.up.media = this.up[this.up$total.up != 0, ] %>% dplyr::group_by (RNAseq.id) %>%
  dplyr::summarise(
    total = sum ( total.up  ) # count occurence but leave out pizzly
    ,cancer.subtype = unique ( cancer.subtype)
  ) %>% data.frame()



out.up.media = out.up.media %>% dplyr::group_by (cancer.subtype) %>%
  dplyr::summarise(
    med = median ( total  ) # count occurence but leave out pizzly
  ) %>% data.frame()



this.up$role = factor (this.up$role, levels =  c(  "oncogene", "TSG", "MULTI",  "NO.ROLE") )



out.total.bar.up = merge ( this.up, out.up.media, by="cancer.subtype")
out.total.bar.up = ggplot(data=out.total.bar.up, aes(x=RNAseq.id, y=total.up,  fill=role  )) +
  geom_bar(stat="identity",  width = 1, position = "stack") +  # color=""
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("Total UP outliers") + 
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))  +
  facet_grid(~cancer.subtype, scales = "free" , space = "free" ) +
  scale_fill_manual(values = c (  "#fc7e08", "#77bbd1", "#8cd468" , "#e1e4e6")   ) +
 # geom_hline(data = out.up.media, aes(yintercept =med), color="#6d706d", size=1) 
 geom_hline( aes(yintercept =med), color="#2e2c2d", size=1) +
   scale_y_continuous( expand = expansion(mult = c(.007, .1)))

 
```


```{r}


this.down_unique =  merge ( this.down, rna.key[ ,c("RNAseq.id", "cancer.subtype")], by="RNAseq.id")
this.down_unique =  plyr::count(this.down_unique[!duplicated ( paste( this.down_unique$gene, this.down_unique$cancer.subtype  )), ], c("cancer.subtype","role"))
colnames( this.down_unique ) = c("cancer.subtype", "role","total.down")
this.down_unique = this.down_unique[this.down_unique$role %in% c(  "TSG"), ]



this.down = plyr::count(this.down, c("RNAseq.id","role"))
colnames ( this.down) = c("RNAseq.id", "role","total.down")

# add zeros 
zero = setdiff (  as.character ( unique ( trueseq.key$RNAseq.id) ), as.character ( this.down$RNAseq.id )  )
for ( z in zero ){
  this.down = rbind ( this.down, data.frame ( RNAseq.id=z, role="NO.ROLE", total.down=0) )     
}




this.down = merge ( this.down, trueseq.key[ , c("RNAseq.id", "cancer.subtype")], by = "RNAseq.id")


outorder_down =  this.down %>% dplyr::group_by(cancer.subtype, RNAseq.id) %>%
  dplyr::summarise(
    total_order = sum ( total.down ) # count occurence but leave out pizzly
  ) %>% data.frame()

outorder_down = outorder_down [ order ( outorder_down$cancer.subtype, outorder_down$total_order), ]


this.down$RNAseq.id = factor ( this.down$RNAseq.id, levels= outorder_down$RNAseq.id )





#this.down$RNAseq.id = factor ( this.down$RNAseq.id, levels= fusion.total$RNAseq.id  )
this.down$cancer.subtype = factor ( this.down$cancer.subtype, levels = cancer.subtype.order)



out.down.media = this.down[this.down$total.down != 0, ] %>% dplyr::group_by (RNAseq.id) %>%
  dplyr::summarise(
    total = sum ( total.down  ) # count occurence but leave out pizzly
    ,cancer.subtype = unique ( cancer.subtype)
  ) %>% data.frame()



out.down.media = out.down.media %>% dplyr::group_by (cancer.subtype) %>%
  dplyr::summarise(
    med = median ( total  ) # count occurence but leave out pizzly
  ) %>% data.frame()



this.down$role = factor (this.down$role, levels =  c(  "oncogene", "TSG", "MULTI",  "NO.ROLE") )


out.total.bar.down = ggplot(data=this.down, aes(x=RNAseq.id, y=total.down,  fill=role  )) +
  geom_bar(stat="identity",  width = 1, position = "stack") +  # color=""
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("Total DOWN outliers") + 
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))  +
  facet_grid(~cancer.subtype, scales = "free" , space = "free" ) +
  scale_fill_manual(values = c (  "#fc7e08", "#77bbd1", "#8cd468" , "#e1e4e6")   ) +
  geom_hline(data = out.down.media, aes(yintercept =med), color="#6d706d", size=1)



# fusion.total.bar / out.total.bar.up / out.total.bar.down
```






```{r}
# merge 
this.down2 = this.down
this.down2$total.down = -1 * this.down2$total.down 
colnames ( this.down2)[3] = "total.up"

this.up$type = "up"
this.down2$type = "down"

outlier.both = rbind ( this.down2, this.up )
colnames ( outlier.both)[2] = "alteration"
outlier.both$role.in.cosmic =  as.character( outlier.both$alteration )

## change course here, 2.18.2022
## keeping everything that means up oncogene as well as TSG and vice versa
## the commented were the older version
outlier.both = outlier.both [
 # ( outlier.both$type == "down" & (outlier.both$role.in.cosmic %in% c("TSG")) ) | 
  #( outlier.both$type == "up" & (outlier.both$role.in.cosmic %in% c("oncogene")) ) ,
outlier.both$role.in.cosmic %in% c("oncogene", "TSG"), 
  ]


unique ( outlier.both[outlier.both$type == "up",  ]$role.in.cosmic )
unique ( outlier.both[outlier.both$type == "down",  ]$role.in.cosmic )

outlier.both= outlier.both[ order ( outlier.both$RNAseq.id), ]




# now we need to clean up the genes a bit to only include druggable and/or combolist 

out.up_down.media = outlier.both %>% dplyr::group_by (cancer.subtype, type ) %>%
  dplyr::summarise(
    med = median ( total.up  ) # count occurence but leave out pizzly
  ) %>% data.frame()

outcolor = c( oncogene = "#fc7e08", TSG = "#77bbd1" )

out.both.bar = ggplot(data=outlier.both, aes(x=RNAseq.id, y=total.up , fill=alteration  )) +
  geom_bar(stat="identity",  width = 1, position ="stack") +  # color=""
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("patient sample") + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm"))  +
  facet_grid(~cancer.subtype, scales = "free" , space = "free" ) + 
 # scale_fill_manual(values = c ( "#fc7e08", "#77bbd1", "#8cd468" , "#e1e4e6", "white")    ) +
  scale_fill_manual(values = outcolor    ) +
  geom_hline(data = out.up_down.media, aes(yintercept =med), color="#6d706d", size=1)



# tabulate all up and down, however need to associate up and down and then rbind 


tab.outlier = outlier.both
tab.outlier$total.up = ifelse ( tab.outlier$total.up < 0, tab.outlier$total.up * -1, tab.outlier$total.up )


plot_updown <- function ( tab.outlier = tab.outlier, dir = "up", col = 3 , cp = c("oncogene" ="#bf3434", "TSG" = "#c27070")){

m1= tab.outlier[ tab.outlier$type == dir , ]   %>%
  group_by(cancer.subtype, alteration, type ) %>%
  summarise( 
          countT= sum(total.up),
          median.type = median ( total.up )
         ) %>%  group_by(cancer.subtype)  %>% 
        mutate ( 
          per = sum ( countT)  
          ) %>% group_by(cancer.subtype, alteration ) %>%  mutate (
            per = round(100*countT/per,2)
          ) %>% data.frame( stringsAsFactors = F ) 

# need extra step in order to get the true median and not the median by type, oncogene and tsg
m2 = tab.outlier   %>%
  group_by(cancer.subtype)   %>% 
 summarise(
 median = median ( total.up )
   )%>% data.frame( stringsAsFactors = F ) 

m1 = merge ( m1, m2, by="cancer.subtype")


m1 =  m1   %>%
  group_by(cancer.subtype ) %>%
  mutate( 
          bothsum = sum(countT),
         )%>% data.frame( stringsAsFactors = F ) 


### tabulate unique 
colnames ( this.down_unique )[3] = "total"
colnames ( this.up_unique )[3] = "total"
m1_unique = rbind ( this.down_unique, this.up_unique )
m1_unique = m1_unique   %>%
  group_by(cancer.subtype)   %>% 
 summarise(
 totalu = sum ( total )
   )%>% data.frame( stringsAsFactors = F ) 

m1 = merge ( m1, m1_unique, by="cancer.subtype")



m1$label = paste0( m1$cancer.subtype, "(",m1$bothsum, ";", m1$totalu,  ")" )
# order facet by ocogene ratio 
m1_ratio = dcast( m1 , cancer.subtype ~ alteration, value.var= "countT"  ) 
m1_ratio$to = m1_ratio$TSG / ( m1_ratio$oncogene + m1_ratio$TSG )
m1 = merge ( m1, m1_ratio[ , c("cancer.subtype", "to")], by="cancer.subtype" )

m1 = m1[order ( m1$to ), ]

m1$label = factor ( as.character ( m1$label), levels = unique ( m1$label) )

#
pie.out =  ggplot(m1, aes(x="", y=per, fill=alteration )) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0)     +
  xlab("") + ylab("") +
   scale_fill_manual(values= outcolor ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), 
       
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) + theme_void() +  facet_grid(. ~ label   )

m1$label = factor ( as.character ( m1$label) , levels = sort ( unique ( as.character ( m1$label) ) )    )


pie.out2 =  ggplot(m1, aes(x="", y=per, fill=alteration )) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0)     +
  xlab("") + ylab("") +
   scale_fill_manual(values= outcolor ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), 
       
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
  ) + theme_void() +
   theme(legend.position="bottom",  legend.key = element_blank(),
          
          axis.text.y = element_text(size= 20 ),
          #axis.text.x = element_blank(),
          axis.title.x = element_text(size=20),
          
          axis.title.y     = element_text(size=20), 
          legend.text      =element_text(size=20), 
         legend.key.size = unit(2, 'cm'),
         plot.margin=grid::unit(c(0,0,0,0), "mm")
    )+  facet_wrap  (label~., ncol = col  ) + 
  theme(strip.text.x = element_text(size = 15, colour = "black", angle = 0)) +
  scale_fill_manual(values = cp )






#(out.both.bar + theme(legend.position = "none") ) +  ( pie.out ) + plot_layout(widths =c ( .8,.2 ) )



m3 =  m1   %>%
  group_by(alteration  ) %>%
  summarise( 
          bothsum = sum(countT),
         )

total_pie = ggplot(m3, aes(x="", y=bothsum, fill=alteration )) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0)     +
  xlab("") + ylab("") +
   scale_fill_manual(values= outcolor ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), 
       
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
  )  + theme_void() + ggtitle ( paste ( "total", sum ( m3$bothsum)) )

return ( list (
  pie.out = pie.out
  , pie.out2 = pie.out2
  , tab = m3 
  , total_pie = total_pie
  
)
         
         )


}


up_plots = plot_updown( tab.outlier = tab.outlier, dir = "up", col = 3, cp = c("oncogene" ="#bf3434", "TSG" = "#eb9898")   )
down_plots = plot_updown( tab.outlier = tab.outlier, dir = "down", col = 3 , cp = c("oncogene" ="#00a1f2", "TSG" = "#aed3e6") )

```

## OUTLIER UP (oncogene), DOWN ( TSG)

```{r}
up_plots$pie.out2 + down_plots$pie.out2

```

```{r}


# 2.21.2022 change to only study cancer genes 
this.up_family_tab =  merge (  up.all_qc[up.all_qc$gene %in% genelist$gene , ] , familyclass [ , c("family","Symbol")], by.x="gene", by.y="Symbol" ) 



this.up_family_tab = merge ( this.up_family_tab, rna.key[ , c("RNAseq.id", "cancer.subtype")], by="RNAseq.id")

this.up_family_tab2 =  this.up_family_tab  %>% dplyr::group_by (cancer.subtype, family) %>%
  #this.up_family_tab2 =  this.up_family_tab %>% dplyr::group_by (cancer.subtype, gene) %>%
  dplyr::summarise(
    total = n ( ) # 
  ) %>% data.frame()

this.up_family_tab2 = this.up_family_tab2[ order ( this.up_family_tab2$total, decreasing = T), ]

up.out = data.frame()

for ( ut in unique ( this.up_family_tab2$cancer.subtype)){
  ut = this.up_family_tab2[this.up_family_tab2$cancer.subtype == ut, ]
  ut = ut [ ut$total > 5, ]
  up.out = rbind ( head ( ut, 10),up.out )
}


# decided to remove any gene that does not make the 3 cutoff 
#up.out = this.up_family_tab2[this.up_family_tab2$family %in% as.character ( up.out$family), ]


up.out$total =  as.numeric( up.out$total) 
up.out = merge ( up.out, total_subcancer, by="cancer.subtype")
up.out$percent = round ( up.out$total / up.out$Freq, 2)
#up.out = up.out[up.out$percent > .20, ]

ggplot( up.out  , aes(x=family , y=total, fill=cancer.subtype   )) +
  geom_bar(stat = 'identity', position = 'stack') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 7.5 ),
        axis.text.x = element_text(angle = 90, size= 15 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12)
        , plot.title = element_text(size = 12, face = "bold")
  ) + 

  scale_fill_manual( values= subcolor )  + ggtitle ( "Recurrent UP family Gene") 


up_f = dcast( up.out
       , family  ~ cancer.subtype, value.var= "percent"  )

up_f[is.na(up_f)] = 0 


order_f = data.frame ( table ( up.out$family ) )
order_f = order_f[ order(order_f$Freq, decreasing = F), ]

up.out$family = factor ( up.out$family, levels= as.character ( order_f$Var1))


out_family_grid = ggplot(up.out,
       aes(x =  cancer.subtype, 
           y =  family ,
           colour = cancer.subtype ,
           size = percent )) +
  geom_point() +   scale_size_continuous(range = c(3.5,6.5)) +
  labs(x = NULL, y = NULL) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")
        , legend.position = "none") +
  scale_colour_manual( values= subcolor ) + 
  geom_hline(yintercept = seq(1.5, (length(unique(up.out$family))-0.5), 1) , colour="grey" , alpha=.5) + 
  geom_vline(xintercept = seq(1.5, (length(unique(up.out$cancer.subtype))-0.5), 1), colour="grey" , alpha=.5 ) +
  theme(
      # element_blank()
      #axis.text.y = element_text(size= 25 ),
      axis.text.x =  element_text(size= 12 , angle=30, hjust = 1 ),
      #axis.title.x = element_text(size=25),
      #axis.title.y     = element_text(size=25), 
      #legend.text      =element_text(size=25),
      #legend.title = element_text(size=25),
      #plot.title = element_text(size = 40, face = "bold")
) 

 


```


```{r}

# druggable outliers 
temp = drug.up[ grepl ( "Dexamethasone", drug.up$drugs ),   ]
temp$drugs = gsub ( "Dexamethasone", "", temp$drugs)
temp$drugs = gsub ( "^,|,$", "", temp$drugs)
tempdrugs = drug.up

drug.up$drugs = gsub ( "Dexamethasone", "", drug.up$drugs)
drug.up$drugs = gsub ( "^,|,$", "", drug.up$drugs)



unique ( drug.up$alteration )
this.up_drug_tab =  merge (  up.all_qc , drug.up, by.x="gene", by.y="gene" ) 



this.up_drug_tab = merge ( this.up_drug_tab, rna.key[ , c("RNAseq.id", "cancer.subtype")], by="RNAseq.id")


# remove duplicates but we need to be careful to remove in this order, L1,L2, L3 



this.up_drug_tab$dup = paste ( this.up_drug_tab$RNAseq.id, this.up_drug_tab$gene )
this.up_drug_tab = this.up_drug_tab[ !duplicated ( this.up_drug_tab$dup), ]



this.up_drug_tab2 =  this.up_drug_tab  %>% dplyr::group_by (cancer.subtype, gene) %>%
  #this.up_drug_tab2 =  this.up_drug_tab %>% dplyr::group_by (cancer.subtype, gene) %>%
  dplyr::summarise(
    total = n ( ) # 
  ) %>% data.frame()

this.up_drug_tab2 = this.up_drug_tab2[ order ( this.up_drug_tab2$total, decreasing = T), ]

up.out = data.frame()

for ( ut in unique ( this.up_drug_tab2$cancer.subtype)){
  ut = this.up_drug_tab2[this.up_drug_tab2$cancer.subtype == ut, ]
  ut = ut [ ut$total > 2, ]
  up.out = rbind ( head ( ut, 8),up.out )
}


## remove line below because the small circles is too messy! 
#up.out = this.up_drug_tab2[ this.up_drug_tab2$gene %in% as.character ( up.out$gene), ]

# we should study the drugs thought to make sure they make sense. 

study_drug = drug.up [ drug.up$gene %in% as.character ( unique ( up.out$gene)),  ]
View ( study_drug )



up.out$total =  as.numeric( up.out$total) 
up.out = merge ( up.out, total_subcancer_truseq, by="cancer.subtype")
up.out$percent = round ( up.out$total / up.out$Freq, 2)

top_drug = ggplot( up.out  , aes(x=reorder_within ( gene, percent , cancer.subtype) , y=percent, fill=cancer.subtype   )) +
  geom_bar(stat = 'identity', position = 'stack') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 7.5 ),
        axis.text.x = element_text(angle = 0, size= 15 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12)
        , plot.title = element_text(size = 12, face = "bold")
  ) + 
  scale_x_reordered() +
  scale_fill_manual( values= subcolor ) + coord_flip() + ggtitle ( "Recurrent UP drug Gene") +
  # scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) + # plot in intergers
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  facet_grid(cancer.subtype~., scales = "free" , space = "free" )



order_f = data.frame ( table ( up.out$gene ) )
order_f = order_f[ order(order_f$Freq, decreasing = F), ]

up.out$gene = factor ( up.out$gene, levels= as.character ( order_f$Var1))

out_drug_grid = ggplot(up.out,
                         aes(x =  cancer.subtype, 
                             y =  gene ,
                             colour = cancer.subtype ,
                             size = percent )) +
  geom_point() +   scale_size_continuous(range = c(1,6)  ) +
  labs(x = NULL, y = NULL) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")
        , legend.position = "none") +
  scale_colour_manual( values= subcolor ) + 
  geom_hline(yintercept = seq(1.5, (length(unique(up.out$gene))-0.5), 1) , colour="grey" , alpha=.5) + 
  geom_vline(xintercept = seq(1.5, (length(unique(up.out$cancer.subtype))-0.5), 1), colour="grey" , alpha=.5 )


top_druggable_up_outliers = unique ( up.out$gene )
top_druggable_up_outliers = rna.up[ rna.up$both == "OUTLIER_HIGH" & rna.up$gene %in% as.character ( top_druggable_up_outliers), ]

```

### top Drugs 

```{r}
out_drug_grid
```

### drug tabulations 
```{r}

# tabulate druggable outliers 


drug.color = setNames( c("#fc7e08", "#77bbd1"), c("OUTLIER_HIGH", "OUTLIER_LOW") )
drug.out = this.up_drug_tab 
drug.out = drug.out[ drug.out$level != "", ]



tally = as.data.frame ( 
  table ( drug.out$level), stringAsFactor=F
)
tab.drug.level = make.pie( tally
                           , name.first="levels"
                           #, cc = drug.color
                           , leg.pos = "right"
                           , title = "Druggable Levels"
)





```

```{r}


this.up_drugName_tab = this.up_drug_tab %>% 
  mutate(drug = strsplit(as.character(drugs), ",")) %>% 
  unnest(drug) %>% data.frame()


#this.up_drugName_tab = this.up_drugName_tab[this.up_drugName_tab$level != "L3", ]

this.up_drugName_tab$drug = gsub ( " ", "", this.up_drugName_tab$drug )

drugtop = data.frame ( table ( this.up_drugName_tab$drug))
drugtop = drugtop [ order ( drugtop$Freq, decreasing = T ), ]


this.up_drugName_tab$drugs = NULL 

this.up_drugName_tab2 =  this.up_drugName_tab  %>% dplyr::group_by (cancer.subtype, drug) %>%
  #this.up_drugName_tab2 =  this.up_drugName_tab %>% dplyr::group_by (cancer.subtype, gene) %>%
  dplyr::summarise(
    total = n ( ) # 
  ) %>% data.frame()

this.up_drugName_tab2 = this.up_drugName_tab2[ order ( this.up_drugName_tab2$total, decreasing = T), ]





up.drugName = data.frame()

for ( ut in unique ( this.up_drugName_tab2$cancer.subtype )){
  ut = this.up_drugName_tab2[this.up_drugName_tab2$cancer.subtype == ut, ]
  ut = ut [ ut$total > 2, ]
  up.drugName = rbind ( head ( ut, 8),up.drugName )
}


#up.drugName = this.up_drugName_tab2[this.up_drugName_tab2$drug %in% as.character ( unique ( up.drugName$drug) ), ]

up.drugName$total =  as.numeric( up.drugName$total) 
up.drugName = merge ( up.drugName, total_subcancer_truseq, by="cancer.subtype")
up.drugName$percent = round ( up.drugName$total / up.drugName$Freq, 2)





#up.drugName = up.drugName[ !grepl ( "VEGFReceptorTyrosine|PD173074|CDKInhibitorSNS|METTyrosineKinaseI", up.drugName$drug), ]



up.drugName$level = sapply ( up.drugName$drug, function(x)
{
  l = unique ( this.up_drugName_tab [ this.up_drugName_tab$drug == x, ]$level )
  l = sort ( l )
  return ( l[1])
}
  
  )
# sub, 2-23-2022
up.drugName$drug = gsub ( "^TyrosineKinaseInhibitor", "TKI ", up.drugName$drug )
up.drugName$drug = gsub ( "^METTyrosineKinaseInhibitor", "c-MET kinase_i ", up.drugName$drug )
up.drugName$drug = gsub ( "^CDKInhibitor", "CDK_i ", up.drugName$drug )



plot_drug_pie <- function ( up.drugName = up.drugName , l = "L1", ft = 10 , c=NULL ){

m1= up.drugName[up.drugName$level %in% l , ]  %>%
  group_by(cancer.subtype, drug ) %>%
  mutate( 
    countT= sum(total),
    median.type = median ( total )
  ) %>%  group_by(drug)  %>% 
  mutate ( 
    per = sum ( total)  
  ) %>% group_by(cancer.subtype, drug ) %>%  mutate (
    per = round(100*countT/per,2)
  ) %>% data.frame( stringsAsFactors = F ) 

if ( is.na ( c)){
lm = length ( unique (m1$drug))
}
gpd = ggplot(m1, aes(x="", y=per, fill=cancer.subtype )) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0)     +
  xlab("") + ylab("") +
  #scale_fill_manual(values= outcolor ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(), 
        
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
  ) + theme_void() +
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 100 ),
        #axis.text.x = element_blank(),
        axis.title.x = element_text(size=40),
        
        axis.title.y     = element_text(size=40), 
        legend.text      =element_text(size=40), 
        legend.key.size = unit(2, 'cm'),
        plot.margin=grid::unit(c(0,0,0,0), "mm")
  )+  facet_wrap  ( . ~ drug , ncol = c  ) + scale_fill_manual(values=subcolor) +
  theme(strip.text.x = element_text(size = ft, colour = "black", angle = 0))

return ( gpd )

}


l1_d = plot_drug_pie(up.drugName = up.drugName , l = "L1", ft=15, c=2)
l2_d = plot_drug_pie(up.drugName = up.drugName , l = "L2", ft=40, c=1)
l3_d = plot_drug_pie(up.drugName = up.drugName , l = "L3", ft=15 , c=2)




top_drug = ggplot( up.drugName  , aes(x=reorder_within ( drug, percent , cancer.subtype) , y=percent, fill=cancer.subtype   )) +
  geom_bar(stat = 'identity', position = 'stack') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="none",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 7.5 ),
        axis.text.x = element_text(angle = 0, size= 15 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12)
        , plot.title = element_text(size = 12, face = "bold")
  ) + 
  scale_x_reordered() +
  scale_fill_manual( values= subcolor ) + coord_flip() + ggtitle ( "Recurrent UP drug Gene") +
  # scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) + # plot in intergers
  scale_y_continuous(expand = expansion(mult = c(0, .1))) +
  facet_grid(cancer.subtype~., scales = "free" , space = "free" )


#up.drugName$drug  = str_trunc(as.character ( up.drugName$drug), 25 ) 

order_f = data.frame ( table ( up.drugName$drug ) )
order_f = order_f[ order(order_f$Freq, decreasing = F), ]

up.drugName$drug = factor ( up.drugName$drug, levels= as.character ( order_f$Var1) )


 
drug_level.color = c( "#990000", "#8DB8E0", "#EBCF46", "#c7c9c9") 
names ( drug_level.color ) = c( "L1","L2", "L3", "no.drug" )


out_drug_grid2 = ggplot(up.drugName,
                       aes(x =  cancer.subtype, 
                           y =  drug ,
                           colour = cancer.subtype ,
                           size = percent )) +
  geom_point() +      scale_size_continuous(range = c(1,6)  )+
  labs(x = NULL, y = NULL) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")
        , legend.position = "none", 
        axis.text.y = element_text(size= 15 ), 
        axis.text.x = element_text(size= 15 )
        
        ) +
  scale_colour_manual( values= subcolor ) + 
  geom_hline(yintercept = seq(1.5, (length(unique(up.drugName$drug))-0.5), 1) , colour="grey" , alpha=.5) + 
  geom_vline(xintercept = seq(1.5, (length(unique(up.drugName$cancer.subtype))-0.5), 1), colour="grey" , alpha=.5 ) +
 facet_grid(level ~., scales = "free" , space = "free", switch = "y"  ) + scale_y_discrete(position = "right") # +theme(strip.text.y = element_blank())

# change color 


g <- ggplot_gtable(ggplot_build(out_drug_grid2))
stripr <- which(grepl('strip-l', g$layout$name))
fills <- c(drug_level.color[1:3] )
k <- 1
for (i in stripr) {
  j <- which(grepl('rect', g$grobs[[i]]$grobs[[1]]$childrenOrder))
  g$grobs[[i]]$grobs[[1]]$children[[j]]$gp$fill <- fills[k]
  k <- k+1
}

out_drug_grid2 = g
grid::grid.draw(out_drug_grid2 )






```

### Drug Grid 
```{r}
out_drug_grid / out_drug_grid2
```


### Drug level outlier
```{r}



library ( ggwaffle )


drug.color = c( "#990000", "#8DB8E0", "#EBCF46", "#c7c9c9") 
names ( drug.color ) = c( "L1","L2", "L3", "no.drug" )



#drug.color = setNames( c("#fc7e08", "#77bbd1"), c("OUTLIER_HIGH", "OUTLIER_LOW") )
drug.out = this.up_drug_tab 
drug.out = drug.out[ drug.out$level != "", ]


#################

missing = setdiff ( trueseq.key$RNAseq.id, unique ( drug.out$RNAseq.id))


drug.out_waffle = drug.out [ , c("cancer.subtype", "RNAseq.id", "level", "gene", "drugs" )]

# select out unique sample --> gene --> drug each of these triplets needs to have only one unique Levels  
# the reason for this is because we need make sure not to count the levels for a single target and drug twice. Here we should take the top only so we do this
# by ordering the levels. 
drug.out_waffle = drug.out_waffle[ order(drug.out_waffle$RNAseq.id, drug.out_waffle$gene, drug.out_waffle$drugs,  drug.out_waffle$level ), ]

# test 
drug.out_waffle [ duplicated ( paste ( drug.out_waffle$RNAseq.id, drug.out_waffle$gene, drug.out_waffle$drugs ) ),  ]
# not that L1 comes out first 
drug.out_waffle [ drug.out_waffle$RNAseq.id == "V6" & drug.out_waffle$gene == "KIT", ]
drug.out_waffle [ drug.out_waffle$RNAseq.id == "S8" & drug.out_waffle$gene == "KIT", ]

# remove here 
drug.out_waffle = drug.out_waffle [ ! duplicated ( paste ( drug.out_waffle$RNAseq.id, drug.out_waffle$gene, drug.out_waffle$drugs ) ),  ]

# test 
drug.out_waffle [ drug.out_waffle$RNAseq.id == "V6" & drug.out_waffle$gene == "KIT", ]
drug.out_waffle [ drug.out_waffle$RNAseq.id == "S8" & drug.out_waffle$gene == "KIT", ]

# count how many unique drug -> levels per sample. 
drug.out_waffle2 = drug.out_waffle  %>% dplyr::group_by( RNAseq.id, level) %>%
  dplyr::summarise(
     total = n() , 
     cancer.subtype = unique ( cancer.subtype )
  ) %>% data.frame ( stringsAsFactors = F )

# insert zero, that is samples that do not have any targetable drugs at all, like nada zero for outliers. 
for ( z in missing ){
  zc = trueseq.key[ trueseq.key$RNAseq.id == z, ]$cancer.subtype
  drug.out_waffle2 = rbind ( drug.out_waffle2, data.frame ( RNAseq.id=z, level="NONE", total=0, cancer.subtype = zc ) ) 
}



# ordering this is tricky because we need to consider the sum of the levels so you can't just order by total since total for a sample can have poentially 3 total one for each level. 
rorder = drug.out_waffle2 %>% dplyr::group_by( RNAseq.id ) %>%
  dplyr::summarise(
    total = sum(total) , 
  ) %>% data.frame ( stringsAsFactors = F )

rorder = rorder[ order ( rorder$total), ]


drug.out_waffle2 = drug.out_waffle2[ order ( drug.out_waffle2$RNAseq.id, drug.out_waffle2$total), ]
drug.out_waffle2$cancer.subtype = factor ( drug.out_waffle2$cancer.subtype, levels = cancer.subtype.order )
drug.out_waffle2$level = factor( drug.out_waffle2$level, levels = c("NONE","L1","L2","L3"))

drug.out_waffle2$RNAseq.id = factor ( drug.out_waffle2$RNAseq.id, levels = unique ( rorder$RNAseq.id) )



ggplot(drug.out_waffle2, aes(fill=level,  x=RNAseq.id , y=total )) + 
  geom_bar(stat="identity",  width = 1, position = "stack") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        , axis.ticks.y=element_blank()
        , axis.text.y = element_text(size= 18 )
  ) +
  #theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
  scale_y_continuous( expand = c(0, 0)) +
  scale_fill_manual( values= drug.color )+
  facet_grid(~cancer.subtype, scales = "free", space = "free"   ) 


####











tally = data.frame ( table ( drug.out_waffle2$level ) )
colnames ( tally ) = c(  "Drug Level", "Freq" )

level_data = setNames(tally$Freq, tally$`Drug Level`)
drug.color["NONE"] = "grey" 

waffle::waffle(level_data, rows=12 , size=1, 
       colors= c( "grey", "#990000", "#8DB8E0", "#EBCF46" ), 
       title="Drug Levels", 
       xlab="",  legend_pos = "bottom") +
    theme(legend.position="bottom",  legend.key = element_blank(),
          
          legend.text      =element_text(size=30), 
         legend.key.size = unit(1.2, 'cm'),
         plot.margin=grid::unit(c(2,2,2,2), "mm")
    )


# pie chart 
drug_unq = drug.out_waffle2
# tabulate unique patient only but rank by L1, L2, L3 
drug_unq = merge ( drug_unq, trueseq.key[ , c("patient.id", "RNAseq.id")], by="RNAseq.id")
drug_unq = drug_unq[order ( drug_unq$level), ]
drug_unq = drug_unq[!duplicated(drug_unq$patient.id), ]

# sanity check 
setdiff ( drug_unq$patient.id, trueseq.key$patient.id  )


drugpie = data.frame ( table ( drug_unq$level )  )
colnames ( drugpie ) = c("Level", "Freq")


drugpie$pct_sample = round ( drugpie$Freq /  nrow ( trueseq.key[!duplicated ( trueseq.key$patient.id), ]), 3)


kable( drugpie , format = "html" , row.names = F, caption="", table.attr = "style='width:40%;'" ) %>% kable_classic(full_width = F, position = "center", )  %>%
  column_spec(1, bold = T, border_right = T) %>%
  row_spec(2, background = drug.color["L1"])  %>%
  row_spec(3, background = drug.color["L2"])  %>%
 row_spec(4, background = drug.color["L3"])  


### create a ggplot table drug pie 
drugpie_t1 <- ggtexttable(drugpie, rows = NULL, theme = ttheme("blank")) %>%
  tab_add_hline(at.row = c( 1:2 ) , row.side = "top", linewidth = 2) %>% table_cell_bg (
    row=2, fill="#d7d7d9", color = "#d7d7d9",  column = 1:ncol(drugpie)  
  ) %>% table_cell_bg (
    row=3, fill=drug.color["L1"], color = drug.color["L1"],  column = 1:ncol(drugpie)  
  )%>% table_cell_bg (
    row=4, fill=drug.color["L2"], color = drug.color["L2"],  column = 1:ncol(drugpie)  
  )%>% table_cell_bg (
    row=5, fill=drug.color["L3"], color = drug.color["L3"],  column = 1:ncol(drugpie)  
  )




# by sample 

drugpie2 = data.frame ( table ( drug.out_waffle2$level )  )
colnames ( drugpie2 ) = c("Level", "Freq")


drugpie2 = drugpie2[drugpie2$Level != "NONE", ]
# get total patient instead 

drug_per_patient = make.pie( drugpie2[, 1:2]
                , name.first="Level"
                , cc = drug.color
                , leg.pos = "right"
                , title = "Drug Levels"
                , pietext = 30
                )


```

```{r}
# study pdl 

immuno.panel = c( "PDCD1LG2","CD276","CTLA4")
imm = out [ out$gene %in% immuno.panel & out$both == "OUTLIER_HIGH", ]







```


```{r}

lolli.up = out.up_down.media[out.up_down.media$type == "up", ]

lolli.up= lolli.up %>%  arrange ( desc ( med) ) %>%  dplyr::mutate(cancer.subtype = factor ( cancer.subtype, levels=rev ( cancer.subtype) )) %>% data.frame()


lolli.up= ggplot(lolli.up, aes(x=cancer.subtype, y=med)) +
    geom_segment( aes(x=cancer.subtype, xend=cancer.subtype, y=0, yend=med), color="black") +
    geom_point( color=subcolor [ rev ( levels ( lolli.up$cancer.subtype)) ], size=10, alpha=0.75) +
    theme_light() +
    theme(
        panel.grid.major.y = element_blank(),
        panel.border = element_blank(),
        axis.ticks.y = element_blank()
    ) + coord_flip() + xlab("") + ylab("median") +
   theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) 



lolli.down = out.up_down.media[out.up_down.media$type == "down", ]
lolli.down$med = lolli.down$med * -1 
lolli.down= lolli.down %>%  arrange ( desc ( med) ) %>%  dplyr::mutate(cancer.subtype = factor ( cancer.subtype, levels=rev ( cancer.subtype) )) %>% data.frame()


lolli.down= ggplot(lolli.down, aes(x=cancer.subtype, y=med)) +
    geom_segment( aes(x=cancer.subtype, xend=cancer.subtype, y=0, yend=med), color="black") +
    geom_point( color=subcolor [ rev ( levels ( lolli.down$cancer.subtype)) ], size=10, alpha=0.75) +
    theme_light() +
    theme(
        panel.grid.major.y = element_blank(),
        panel.border = element_blank(),
        axis.ticks.y = element_blank()
    ) + coord_flip() + xlab("") + ylab("median") +
   theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) 



# testing out different orienations 
( lolli.up + ggtitle ("Oncogene Up" ) ) | ( lolli.down + ggtitle ("TSG Down" )  ) | fusion.plot$lolli_med + ggtitle ( "Fusions (in.frame )")


( lolli.up + ggtitle ("Oncogene Up" ) )  | fusion.plot$lolli_med + ggtitle ( "Fusions (in.frame )")


```

## Outlier Analysis {.tabset}

### Stack plots 
```{r fig=TRUE,fig.width=20, fig.height=8, echo=FALSE, include=TRUE }
(out.both.bar + theme(legend.position = "none")  ) / 
  ( lolli.up + ggtitle ("Median Activating Oncogene" ) + lolli.down + ggtitle ("Median Disrupting TSG" ) + pie.out2 )

```

```{r}
# are there any trends in treated vs. untreated?  
# biopsy vs met ? 

outlier.both2 = outlier.both 
outlier.both2$total.up = ifelse( outlier.both2$total.up < 0, outlier.both2$total.up *-1, outlier.both2$total.up) 
outlier.both2 = merge ( outlier.both2, trueseq.key[, c("RNAseq.id", "treatment", "sample.type")], by="RNAseq.id")
outlier.both2$sample.type = gsub ( ".*_", "", outlier.both2$sample.type )

out_treat = outlier.both2  %>% dplyr::group_by(type, RNAseq.id, treatment) %>%
  dplyr::summarise(
    Total = sum(total.up)  
    
  ) %>% ggplot(. , aes(y=Total, x=treatment))+
  geom_quasirandom(aes(fill = treatment ), size = 3, shape = 21, alpha = .5) +
  theme_bw() +
  xlab("") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") +
  scale_fill_manual(values = unique (brewer.pal(7, "Dark2") )) + facet_grid(type~.  )



out_type = outlier.both2  %>% dplyr::group_by(type, RNAseq.id, sample.type) %>%
  dplyr::summarise(
    Total = sum(total.up)  
    
  ) %>% ggplot(. , aes(y=Total, x=sample.type))+
  geom_quasirandom(aes(fill = sample.type ), size = 3, shape = 21, alpha = .5) +
  theme_bw() +
  xlab("") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") +
  scale_fill_manual(values = unique (brewer.pal(7, "Dark2") )) + facet_grid(type~.  )




```

### Outlier Trends
```{r fig=TRUE,fig.width=20, fig.height=10, echo=FALSE, include=TRUE }
out_type + ( out_treat + ylab("")  )+ plot_layout(widths =c ( .7,.3 ) )


```

```{r}
#### now lets incorporate CNA 
rna.down$type = "down"
rna.up$type = "up"
rna.both = rbind ( rna.down, rna.up )

# we need to determine what is activating and what is not. 

rna.both.ac.dis = rna.both  [
  ( rna.both$type == "down" & (rna.both$role %in% c("TSG")) ) | 
  ( rna.both$type == "up" & (rna.both$role %in% c("oncogene")) ) ,
]



colnames ( cna ) = c( "wgs.id","gene","cna","logratio")
# lets break this up 
cna = cna %>% 
  mutate(role = strsplit(as.character(cna), ";")) %>% 
  unnest(role) %>% data.frame()
# don't know what to do with break 
cna = cna[ cna$role != "CNA_BREAK", ]
cna$type =  stringr::str_match(cna$role , "CNA_(.*)\\d+")[, 2]
wgs.id = gsub( "_.*?G", "", cna$wgs.id)
intersect ( wgs.id, trueseq.key$WGS.id)
cna$wgs.id = wgs.id 

zero.wgs = setdiff(trueseq.key$WGS.id,wgs.id  )
#zero.wgs = zero.wgs[zero.wgs!="None"] # some are not completed and some just has None! 
zero.wgs = trueseq.key[ trueseq.key$WGS.id %in% zero.wgs, ]$RNAseq.id


############# # no need to subset because cna will merge into them 


############################## plot cna now 
# by subsetting with oncogene or tsg instead off all$gene this ensures that only activating (gains ) are assocaited 

cna.up = cna[ cna$type %in% c( "GAIN" , "GAIN1") & cna$gene %in%  unique ( oncogene) , c("wgs.id","gene", "type"), ]
cna.down = cna[ cna$type == "LOSS" & cna$gene %in% unique (  tsg)   , c("wgs.id","gene", "type"), ]

cna.up = merge ( cna.up, trueseq.key[ , c("RNAseq.id","cancer.subtype","WGS.id")], by.x="wgs.id", by.y="WGS.id")
cna.down = merge ( cna.down, trueseq.key[ , c("RNAseq.id","cancer.subtype","WGS.id")], by.x="wgs.id", by.y="WGS.id")



# major change here is to only merge when its true. 
cna.up = merge ( cna.up, rna.both.ac.dis[ rna.both.ac.dis$type == "up", ], by=c ( "RNAseq.id", "gene" ),  all.x=F, all.y=F )
cna.down = merge ( cna.down, rna.both.ac.dis[ rna.both.ac.dis$type == "down", ], by=c ( "RNAseq.id", "gene" ),  all.x=F, all.y=F )
# after merging I don't need of the columns, so to simplify I'm going to cut back here. 

cna.up = cna.up [ , c("RNAseq.id", "gene", "type.x", "cancer.subtype", "both", "role")]
cna.down = cna.down [ , c("RNAseq.id", "gene", "type.x", "cancer.subtype", "both", "role")]

cna.both = rbind ( cna.up, cna.down) 
cna.both$type.x = gsub ( "GAIN1", "GAIN", cna.both$type.x)

cna.both.tab = cna.both  %>%
  group_by(cancer.subtype, type.x ) %>%
  summarise( 
          countT= n()
         ) %>%  group_by(cancer.subtype)  %>% 
        mutate ( 
          per = sum ( countT)  
          ,median = median ( countT )
          ) %>% group_by(cancer.subtype, type.x ) %>%  mutate (
            per = round(100*countT/per,2)
          ) %>% data.frame( stringsAsFactors = F ) 

cna.both.tab$CNA_RNA = cna.both.tab$type.x
cna.color = c(GAIN="#cce876", LOSS="#c7778f" )
cna.both.tab$label = paste0 ( cna.both.tab$cancer.subtype, "(", cna.both.tab$median, ")" ) 


cna.pie =   ggplot(cna.both.tab, aes(x="", y=per, fill=CNA_RNA )) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0)     +
  xlab("") + ylab("") +
 scale_fill_manual(values= cna.color ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
       legend.position = "bottom"
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
  ) +
  theme( plot.margin=grid::unit(c(0,0,0,0), "mm")) + theme_void() +  facet_wrap  (label~., ncol = 2  )
  


cna.both.tab.m1 =  cna.both.tab   %>%
  group_by(cancer.subtype ) %>%
  mutate( 
    bothsum = sum(countT),
  )

cna.both.tab.m1$label = paste0( cna.both.tab.m1$cancer.subtype, "(",cna.both.tab.m1$bothsum, ")" )
#temp = fusion.frame.per[!duplicated ( m1$cancer.subtype), ]

ggplot(cna.both.tab.m1, aes(x="", y=per, fill=CNA_RNA )) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0)     +
  xlab("") + ylab("") +
  scale_fill_manual(values= cna.color ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        legend.position = "bottom"
        #, axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
  ) +
  theme( plot.margin=grid::unit(c(0,0,0,0), "mm")) + theme_void() +  facet_wrap  (label~., ncol = 2  )


(out.both.bar + theme(legend.position = "none")  ) / 
 ( (  (lolli.up + ggtitle ("Median Activating Oncogene" ) ) | ( lolli.down + ggtitle ("Median Disrupting TSG" )) 
   |  ( pie.out2 + ggtitle ( "RNA outlier") +  theme(legend.position = "bottom") )  
   |  ( cna.pie + ggtitle ( "RNA + CNA") +  theme(legend.position = "bottom") )  
   
   ) + plot_layout(widths =c ( .2,.2, .3, .25 ) ) ) 



```



```{r}

# next calculate CNA and gene expression. 


cna.rna = merge ( cna, trueseq.key, by.x="wgs.id", by.y="WGS.id")

cna.tab = cna.rna[ cna.rna$gene %in%  unique ( c ( genelist$gene  ) ) &
                   cna.rna$type %in% c("GAIN", "GAIN1")
                   , ] %>% dplyr::group_by( gene ) %>%
  dplyr::summarise(
    freq = n()  
  ) %>% arrange ( desc ( freq) ) %>% data.frame ( stringsAsFactors = F)


 
mincna_s = 2 
statsdo = cna.tab[ cna.tab$freq >= mincna_s, ]$gene 

# going to use  Mann-Whitney U Test 
statswu = data.frame ( stringsAsFactors = F  )
# to be unbias we don't select for median > 0 
cna.tab.gain.plot  = data.frame(stringsAsFactors = F)


for ( test.gene in statsdo){
  
 
  test.sample = cna.rna[cna.rna$gene == test.gene & 
                          cna.rna$type %in% c("GAIN", "GAIN1")
                        , c("RNAseq.id", "gene")]
  # see above, again we need to double check if there is at least a sample > 3 
  
  if ( length ( unique ( test.sample$RNAseq.id) ) < mincna_s ){
    next 
  }
  
   if ( ! any ( tpm$gene %in% test.gene )){
      next 
   }
  
  
  
  test.tpm = melt ( tpm[tpm$gene==test.gene, test.sample$RNAseq.id] )
  test.tpm$group = "GAIN"
  type = trueseq.key[ trueseq.key$RNAseq.id %in% unique ( as.character ( test.tpm$variable) ),  ]$cancer.type
  
  # this is important to make sure we are comparing solid to solid and liquid to liquid 
  
  if ( any(type %in% c("CNS","Solid"))){
    type =   unique ( c("CNS","Solid", type) ) 
  }
  
  
  test.tpm = merge ( test.tpm, trueseq.key[ , c("RNAseq.id", "cancer.subtype") ], by.x="variable",  by.y="RNAseq.id" )
  
  nofusesample = setdiff(trueseq.key$RNAseq.id, test.sample$RNAseq.id) 
  # the first line removes any nofusesample that does not have the same type 
  nofusesample = trueseq.key[trueseq.key$cancer.type %in% unique ( type) & trueseq.key$RNAseq.id %in% nofusesample , ]$RNAseq.id
  
  noCNA.tpm = melt ( tpm[tpm$gene==test.gene, nofusesample  ] )
  noCNA.tpm$group = "NO.GAIN"
  noCNA.tpm$cancer.subtype = "none"
  
  testexpCNA = rbind ( noCNA.tpm, test.tpm )
  testexpCNA$value = log2 ( testexpCNA$value + 1 )
  testexpCNA[1] = "sample"
  testexpCNA$CNA = test.gene
  # Mann-Whitney U
  
  x1 = testexpCNA[testexpCNA$group == "NO.GAIN", ]$value
  x2 = testexpCNA[testexpCNA$group == "GAIN", ]$value
  wu = wilcox.test(x2,x1)
  diff = median ( x2 ) - median ( x1 )
 
  

  statswu = rbind ( statswu, data.frame ( p=wu$p.value, plog2 = -log2(wu$p.value), median = diff, gene = test.gene
                                          , freq = length ( unique ( test.sample$RNAseq.id) )   )
                    
  )
  
    testexpCNA$zscale = as.numeric ( scale ( testexpCNA$value ) )
  testexpCNA$percentile = as.numeric ( perc.rank(testexpCNA$value) )
  
  
  cna.tab.gain.plot = rbind ( cna.tab.gain.plot, testexpCNA )
}


statswu$fdr = p.adjust(statswu$p, n=nrow ( statswu) )

# plotting the top 10 
cna.tab.gain.plot = merge ( cna.tab.gain.plot , statswu, by.x="CNA", by.y="gene")
cna.tab.gain.plot = cna.tab.gain.plot[order ( -cna.tab.gain.plot$median ), ]

topdo = as.character ( unique ( cna.tab.gain.plot[cna.tab.gain.plot$p < .05, ]$CNA) [1:12] )
todo2 = as.character ( unique ( cna.tab.gain.plot[ cna.tab.gain.plot$CNA %in% topdo & cna.tab.gain.plot$p < .05, ]$CNA) )  # this will be used to label the scatter 



cna.tab.gain.plot$annt = paste ( "p:" , round(cna.tab.gain.plot$p, 2),    "fdr:", round(cna.tab.gain.plot$fdr, 2))
cna.tab.gain.plot$CNA = factor ( cna.tab.gain.plot$CNA , levels = unique ( cna.tab.gain.plot$CNA  ))

manual_over_up =  unique ( cna.tab.gain.plot[cna.tab.gain.plot$p < .05, ]$CNA) [1:6]

logit_df2 = cna.tab.gain.plot  %>% dplyr::group_by(CNA, group )  %>%  dplyr::summarise(
  median = median ( value )
) %>% data.frame( stringsAsFactors = F)


kt_gain = kruskal.test(logit_df2$median~logit_df2$group)


kt_gain =round ( kt_gain$p.value, 3 ) 
if ( kt_gain < .001){
  kt_gain = "p< .001"
}else {
  kt_gain = paste ( "p=", kt_gain)
}



cna.tab.gain.plot2 = ggplot(cna.tab.gain.plot[cna.tab.gain.plot$CNA %in% manual_over_up , ], aes(y=value, x=group))+ #geom_boxplot(outlier.shape = NA)+ 
  geom_violin(alpha = 0.5) +
  geom_point(aes(fill = cancer.subtype), size = 4, shape = 21, position = position_jitterdodge(dodge.width = 0.25), alpha = 0.75 )+
  theme_bw() +
  ylab("log2 ( tpm +1 ) ") +
  xlab("") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 12 ),
        axis.text.x = element_text(angle = 0, size= 12 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") + xlab ( "") +  ylab ( "") +
  scale_fill_manual(values = subcolor2 ) + 
  geom_text( aes(x = 1.5, y = 0, label = annt), vjust = .20)  +
  facet_wrap(~ CNA, scales = "free" ,  ncol = 3 ) + ggtitle ( "CNA GAIN  With pv < .05")


# plot scatter 
exp.CNA = cna.tab.gain.plot
exp.CNA$diffabs = abs ( exp.CNA$median)
exp.CNA2 = exp.CNA[ order ( exp.CNA$diffabs), ]
exp.CNA2$CNA = factor ( exp.CNA2$CNA, levels = unique ( exp.CNA2$CNA))
#exp.CNA2$diffabs = factor ( exp.CNA2$diffabs, levels = unique ( exp.CNA2$diffabs))
exp.CNA2$direction = ifelse ( exp.CNA2$median > 0, "higher", "lower")
exp.CNA.gain = exp.CNA2 

exp.CNA.gain$dir2 = ifelse ( exp.CNA.gain$p < .05, paste0 ( exp.CNA.gain$direction, "_", "Sig"),  paste0 ( exp.CNA.gain$direction, "_", "noSig") )


expCNA.scat.plot.gain = ggplot(exp.CNA.gain[!duplicated ( exp.CNA.gain$CNA), ], aes(y=plog2, x=median))+ #geom_boxplot(outlier.shape = NA)+ 
  geom_point(aes ( fill = dir2), size = 5, shape = 21, position = position_jitterdodge(dodge.width = 0.25, jitter.height =.2), alpha = 0.75 )+
  theme_bw() +
  ylab(" -log2 ( pv)") +
  xlab(" log2FC ") +
  theme(legend.position="bottom", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 12 ),
        axis.text.x = element_text(angle = 0, size= 12 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  )  + geom_hline(yintercept= 4.3, color = "grey", linetype = "dashed", size = 3.0 ) +  scale_fill_manual(values = piecc ) +
  #annotate(geom="text", x=3, y=4.8,  label="pv < .05", size=10 ) +
  #annotate(geom="text", x=3, y=20,  label=kt_gain, size=10 ) +
  geom_label_repel(
    data = exp.CNA.gain[!duplicated ( exp.CNA.gain$CNA) & exp.CNA.gain$CNA %in% todo2, ],
    aes(
      label = CNA
    )
    , 
    fontface="bold", 
    color="black",
    size = 5, 
    point.padding = unit(.25, "lines"),
    #  box.padding = unit(.25, "lines"),
    # nudge_x = .3,
    show.legend = F,
    #segment.color = cc5
    #segment.alpha = 0
    #segment.linetype = 2, 
    #segment.curvature = 40, 
    #arrow = arrow(length = unit(0.015, "npc"))
  ) + ggtitle ( "CNA GAIN and Expression ")


# do pie chart here and chisquare.

tally = exp.CNA.gain[!duplicated ( exp.CNA.gain$CNA), ]

tally$piegroup = ifelse ( tally$p < .05, paste0(tally$direction,"_Sig"), paste0(tally$direction,"_noSig") )



  tally = data.frame ( table ( tally$piegroup))
  #piecc = c( higher_Sig = "#cf4051", higher_noSig = "#e0c126", lower_Sig = "#83bf43", lower_noSig= "#6ec0db" )
  
  
  addthis = setdiff ( names ( piecc), tally$Var1 )
  tally2 = tally 
  for ( a in addthis){
    tally2 = rbind ( tally2, data.frame(Var1=a, Freq=0))
  }
  
  
  tally2$state = gsub ( ".*_", "", tally2$Var1)
  tally2$group = gsub ( "_.*", "", tally2$Var1)
  tally2$Var1 = NULL 
  tally2 =  pivot_wider(data = tally2, names_from = group,values_from=Freq) %>% data.frame(stringsAsFactors = F)
  row.names ( tally2) = tally2$state
  tally2$state = NULL 
  chi = chisq.test(tally2, simulate.p.value=T)
  if ( is.na ( chi )){
    chi = chisq.test(tally2[,1], simulate.p.value=T)
  }
  
  if ( is.na ( chi )){
    chi = chisq.test(tally2[,2], simulate.p.value=T)
  }
  
  
  
  
  pie.gain = make.pie( tally
                       , name.first= paste ( "chisq=", round ( chi$p.value, 4 ))
                       #,name.first= paste ( "")
                       , cc = piecc
                       , leg.pos = "right"
                       , title = ""
  )
  






```

```{r}
cna.rna = merge ( cna, trueseq.key, by.x="wgs.id", by.y="WGS.id")

cna.LOSS = cna.rna[ cna.rna$gene %in%  unique ( genelist$gene ) &
                   !cna.rna$type %in% c("GAIN", "GAIN1")
                   , ] %>% dplyr::group_by( gene ) %>%
  dplyr::summarise(
    freq = n()  
  ) %>% arrange ( desc ( freq) ) %>% data.frame ( stringsAsFactors = F)



mincna_s = 2 
statsdo = cna.LOSS[ cna.LOSS$freq >= mincna_s, ]$gene 



# going to use  Mann-Whitney U Test 
statswu = data.frame ( stringsAsFactors = F  )
# to be unbias we don't select for median > 0 
cna.tab.loss.plot  = data.frame(stringsAsFactors = F)


for ( test.gene in statsdo){
  
  
  test.sample = cna.rna[cna.rna$gene == test.gene & 
                          cna.rna$type %in% c("LOSS", "LOSS1")
                        , c("RNAseq.id", "gene")]
  # see above, aloss we need to double check if there is at least a sample > 3 
  
  if ( length ( unique ( test.sample$RNAseq.id) ) < mincna_s ){
    next 
  }
  
  if ( ! any ( tpm$gene %in% test.gene )){
    next 
  }
  
  
  
  test.tpm = melt ( tpm[tpm$gene==test.gene, test.sample$RNAseq.id] )
  test.tpm$group = "LOSS"
  type = trueseq.key[ trueseq.key$RNAseq.id %in% unique ( as.character ( test.tpm$variable) ),  ]$cancer.type
  
  # this is important to make sure we are comparing solid to solid and liquid to liquid 
  
  if ( any(type %in% c("CNS","Solid"))){
    type =   unique ( c("CNS","Solid", type) ) 
  }
  
  
  test.tpm = merge ( test.tpm, trueseq.key[ , c("RNAseq.id", "cancer.subtype") ], by.x="variable",  by.y="RNAseq.id" )
  
  nofusesample = setdiff(trueseq.key$RNAseq.id, test.sample$RNAseq.id) 
  # the first line removes any nofusesample that does not have the same type 
  nofusesample = trueseq.key[trueseq.key$cancer.type %in% unique ( type) & trueseq.key$RNAseq.id %in% nofusesample , ]$RNAseq.id
  
  noCNA.tpm = melt ( tpm[tpm$gene==test.gene, nofusesample  ] )
  noCNA.tpm$group = "NO.LOSS"
  noCNA.tpm$cancer.subtype = "none"
  
  testexpCNA = rbind ( noCNA.tpm, test.tpm )
  testexpCNA$value = log2 ( testexpCNA$value + 1 )
  testexpCNA[1] = "sample"
  testexpCNA$CNA = test.gene
  # Mann-Whitney U
  
  x1 = testexpCNA[testexpCNA$group == "NO.LOSS", ]$value
  x2 = testexpCNA[testexpCNA$group == "LOSS", ]$value
  wu = wilcox.test(x2,x1)
  diff = median ( x2 ) - median ( x1 )
  
  
  
  statswu = rbind ( statswu, data.frame ( p=wu$p.value, plog2 = -log2(wu$p.value), median = diff, gene = test.gene
                                          , freq = length ( unique ( test.sample$RNAseq.id) )   )
                    
  )
  
  testexpCNA$zscale = as.numeric ( scale ( testexpCNA$value ) )
  testexpCNA$percentile = as.numeric ( perc.rank(testexpCNA$value) )
  
  
  cna.tab.loss.plot = rbind ( cna.tab.loss.plot, testexpCNA )
}


statswu$fdr = p.adjust(statswu$p, n=nrow ( statswu) )

# plotting the top 10 
cna.tab.loss.plot = merge ( cna.tab.loss.plot , statswu, by.x="CNA", by.y="gene")
cna.tab.loss.plot = cna.tab.loss.plot[order ( cna.tab.loss.plot$median), ]

topdo = as.character ( unique ( cna.tab.loss.plot[cna.tab.loss.plot$p < .05, ]$CNA) [1:12] )
todo2 = as.character ( unique ( cna.tab.loss.plot[ cna.tab.loss.plot$CNA %in% topdo & cna.tab.loss.plot$p < .05, ]$CNA) )  # this will be used to label the scatter 



cna.tab.loss.plot$annt = paste ( "p:" , round(cna.tab.loss.plot$p, 2),    "fdr:", round(cna.tab.loss.plot$fdr, 2))
cna.tab.loss.plot$CNA = factor ( cna.tab.loss.plot$CNA , levels = unique ( cna.tab.loss.plot$CNA  ))

manual_over_up =  unique ( cna.tab.loss.plot[cna.tab.loss.plot$p < .05, ]$CNA) [1:6]

logit_df2 = cna.tab.loss.plot  %>% dplyr::group_by(CNA, group )  %>%  dplyr::summarise(
  median = median ( value )
) %>% data.frame( stringsAsFactors = F)


kt_loss = kruskal.test(logit_df2$median~logit_df2$group)


kt_loss =round ( kt_loss$p.value, 3 ) 
if ( kt_loss < .001){
  kt_loss = "p< .001"
}else {
  kt_loss = paste ( "p=", kt_loss)
}



cna.tab.loss.plot2 = ggplot(cna.tab.loss.plot[cna.tab.loss.plot$CNA %in% manual_over_up , ], aes(y=value, x=group))+ #geom_boxplot(outlier.shape = NA)+ 
  geom_violin(alpha = 0.5) +
  geom_point(aes(fill = cancer.subtype), size = 4, shape = 21, position = position_jitterdodge(dodge.width = 0.25), alpha = 0.75 )+
  theme_bw() +
  ylab("log2 ( tpm +1 ) ") +
  xlab("") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 12 ),
        axis.text.x = element_text(angle = 0, size= 12 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") + xlab ( "") +  ylab ( "") +
  scale_fill_manual(values = subcolor2 ) + 
  geom_text( aes(x = 1.5, y = 0, label = annt), vjust = .20)  +
  facet_wrap(~ CNA, scales = "free" ,  ncol = 3 ) + ggtitle ( "CNA LOSS and expression")


# plot scatter 
exp.CNA = cna.tab.loss.plot
exp.CNA$diffabs = abs ( exp.CNA$median)
exp.CNA2 = exp.CNA[ order ( exp.CNA$diffabs), ]
exp.CNA2$CNA = factor ( exp.CNA2$CNA, levels = unique ( exp.CNA2$CNA))
#exp.CNA2$diffabs = factor ( exp.CNA2$diffabs, levels = unique ( exp.CNA2$diffabs))
exp.CNA2$direction = ifelse ( exp.CNA2$median > 0, "higher", "lower")
exp.CNA.loss = exp.CNA2 

exp.CNA.loss$dir2 = ifelse ( exp.CNA.loss$p < .05, paste0 ( exp.CNA.loss$direction, "_", "Sig"),  paste0 ( exp.CNA.loss$direction, "_", "noSig") )

expCNA.scat.plot.loss = ggplot(exp.CNA.loss[!duplicated ( exp.CNA.loss$CNA), ], aes(y=plog2, x=median))+ #geom_boxplot(outlier.shape = NA)+ 
  geom_point(aes ( fill = dir2), size = 5, shape = 21, position = position_jitterdodge(dodge.width = 0.25, jitter.height =.2), alpha = 0.75 )+
  theme_bw() +
  ylab(" -log2 ( pv)") +
  xlab(" log2FC ") +
  theme(legend.position="bottom", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 12 ),
        axis.text.x = element_text(angle = 0, size= 12 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  )  + geom_hline(yintercept= 4.3, color = "grey", linetype = "dashed", size = 3.0 ) +  scale_fill_manual(values = piecc ) +
  #annotate(geom="text", x=3, y=4.8,  label="pv < .05", size=10 ) +
  #annotate(geom="text", x=3, y=20,  label=kt_loss, size=10 ) +
  geom_label_repel(
    data = exp.CNA.loss[!duplicated ( exp.CNA.loss$CNA) & exp.CNA.loss$CNA %in% todo2, ],
    aes(
      label = CNA
    )
    , 
    fontface="bold", 
    color="black",
    size = 5, 
    point.padding = unit(.25, "lines"),
    #  box.padding = unit(.25, "lines"),
    # nudge_x = .3,
    show.legend = F,
    #segment.color = cc5
    #segment.alpha = 0
    #segment.linetype = 2, 
    #segment.curvature = 40, 
    #arrow = arrow(length = unit(0.015, "npc"))
  ) + ggtitle ( "Recurrent CNA LOSS >= 2 unique pid ")


# do pie chart here and chisquare.

tally = exp.CNA.loss[!duplicated ( exp.CNA.loss$CNA), ]

tally$piegroup = ifelse ( tally$p < .05, paste0(tally$direction,"_Sig"), paste0(tally$direction,"_noSig") )



tally = data.frame ( table ( tally$piegroup))
#piecc = c( higher_Sig = "#cf4051", higher_noSig = "#e0c126", lower_Sig = "#83bf43", lower_noSig= "#6ec0db" )


addthis = setdiff ( names ( piecc), tally$Var1 )
tally2 = tally 
for ( a in addthis){
  tally2 = rbind ( tally2, data.frame(Var1=a, Freq=0))
}


tally2$state = gsub ( ".*_", "", tally2$Var1)
tally2$group = gsub ( "_.*", "", tally2$Var1)
tally2$Var1 = NULL 
tally2 =  pivot_wider(data = tally2, names_from = group,values_from=Freq) %>% data.frame(stringsAsFactors = F)
row.names ( tally2) = tally2$state
tally2$state = NULL 
chi = chisq.test(tally2, simulate.p.value=T)
if ( is.na ( chi )){
  chi = chisq.test(tally2[,1], simulate.p.value=T)
}

if ( is.na ( chi )){
  chi = chisq.test(tally2[,2], simulate.p.value=T)
}




pie.loss = make.pie( tally
                     , name.first= paste ( "chisq=", round ( chi$p.value, 4 ))
                     #,name.first= paste ( "")
                     , cc = piecc
                     , leg.pos = "right"
                     , title = ""
)



```


### Copy Number with Expression Trends
```{r fig=TRUE,fig.width=20, fig.height=10, echo=FALSE, include=TRUE }
#out_type + ( out_treat + ylab("")  )+ plot_layout(widths =c ( .7,.3 ) )
( expCNA.scat.plot.gain + theme ( legend.position = "none") + xlab("") )/
expCNA.scat.plot.loss



```

### Pie Chart

```{r fig=TRUE,fig.width=20, fig.height=10, echo=FALSE, include=TRUE }
#out_type + ( out_treat + ylab("")  )+ plot_layout(widths =c ( .7,.3 ) )
pie.gain /
pie.loss



```

### TOP Copy Number Expression

```{r fig=TRUE,fig.width=20, fig.height=10, echo=FALSE, include=TRUE }

 ( cna.tab.gain.plot2 + ggtitle ( "GAIN: top most significant genes" ) ) /
cna.tab.loss.plot2 + ggtitle ( "LOSS: top most significant genes" )

```



```{r}
prep=0
if ( prep == 1 ){

  snv3 =  read.table ( 
    paste0( "/data/SUMMARY/WGS/SPCG-UCPG.mutect.post.filtered.annotated.txt.gz" ) 
    ,header=TRUE,sep="\t",stringsAsFactors = FALSE,
    na.strings=".", quote = "", fill = TRUE)
  

  


snv3_id = gsub( "_.*?G", "", snv3$sample )
snv3$WGS.id = snv3_id

# check to see how complete this is. 
intersect ( snv3_id, trueseq.key$WGS.id)
setdiff ( trueseq.key$WGS.id,snv3_id )
length (  setdiff ( trueseq.key$WGS.id,unique ( snv3_id ) ) )

# remove na 
dim ( snv3 )
snv3 =snv3 [ !is.na(snv3$RNA_REF_COUNT),    ]
dim ( snv3 )
length (  setdiff ( trueseq.key$WGS.id,snv3_id ) )

length (  setdiff ( trueseq.key$WGS.id,snv3$WGS.id ) )
# clean up 
snv3 = snv3[snv3$GENE != "", ]


### calculate raf/daf
view_snv = c( "sample",   
              "VEP_Worst_Gene", "RNA_PVALUE"  ,     "RNA_DEPTH"   ,  "RNA_REF_COUNT"  , "RNA_ALT_COUNT" , "RNA_ALT_AF",  "WGS_REF_COUNT", "WGS_ALT_COUNT", "TUMOR.AF", "TUMOR.AD" , "balance", "allbalance","is.sig")

snv3$rafdaf =  snv3$RNA_ALT_AF / snv3$TUMOR.AF 

snv3$rafdaf_annt = sapply ( snv3$rafdaf, function(x) {
    n=""
    if ( x <= .25 ){n="EXTREME_LOW"} else if ( x > .25 & x <= 1 )
    {n="LOW"} else if ( x > 1 & x <= 2 )
    {n="HIGH"} else if ( x>2 ){ n="EXTREME_HIGH"}
})


snv3$is.sig = ifelse ( snv3$RNA_PVALUE < .05, "sig", "no.sig")



snv4 = snv3 

snv4$is.cancer = as.character ( sapply ( snv4$VEP_Worst_Gene, function (x){
  n="no"
  if ( x %in% genelist$gene ){
    n="yes"
  }
  return ( n )
}
  
  ) )



# add wgs ref and alt 
snv4$WGS_REF_COUNT = round ( snv4$TUMOR.AD - ( snv4$TUMOR.AF * snv4$TUMOR.AD  ) )
snv4$WGS_ALT_COUNT = round ( snv4$TUMOR.AF * snv4$TUMOR.AD   )


snv4$balance = sapply ( 1:nrow(snv4), function(x) {
    
    n="no.sig"
    r = snv4[x:x,]
    if (r$is.sig == "sig"){
      x= r$rafdaf
    if ( x <= .25 ){n="EXTREME_LOW"} else if ( x > .25 & x <= 1 )
    {n="LOW"} else if ( x > 1 & x <= 2 )
    {n="HIGH"} else if ( x>2 ){ n="EXTREME_HIGH"}
    }
    return ( n )
})




snv4$logp = -log(snv4$RNA_PVALUE)
# above will produce Inf because p value is 0 so we will mode it a bit with the max
mp = max ( snv4$logp [ snv4$logp != Inf]) + 0.5
# sanity check 
max ( snv4$logp  )
tail ( sort ( snv4$logp))

snv4$logp = ifelse ( snv4$logp == Inf, mp, snv4$logp ) 
tail ( sort ( snv4$logp))
head ( sort ( snv4$logp))


snv4$neg_log_ratio = -log2 ( snv4$rafdaf )

saveRDS(  snv4, config$SNV_2021_rds, compress = F )
 saveRDS(  gSNV, 'sSNV.rds', compress = F ) 
}else {
  snv4 = readRDS( config$SNV_2021_rds)
}

dim ( snv4 )
snv4 = merge ( trueseq.key[ , c("RNAseq.id", "WGS.id", "cancer.subtype")], snv4 , by.x= "WGS.id", by.y= "WGS.id")

length (  setdiff ( trueseq.key$WGS.id,snv4$WGS.id ) )




snv4$uid = make.unique(paste ( snv4$sample, snv4$VEP_Worst_Gene, snv4$pos))
```

```{r}

# 4/21/2022 
# fix eliminate RNA depth < 6
cosmic = snv4 
cosmic$logratio = cosmic$neg_log_ratio
cosmic$RNA_DEPTH = cosmic$RNA_ALT_COUNT + cosmic$RNA_REF_COUNT
dim ( cosmic )

cosmic = cosmic[cosmic$RNA_DEPTH >=5, ] # make sure there is enough depth here
dim ( cosmic )
quantile ( cosmic$RNA_DEPTH)
#cosmic$logratio = ifelse ( cosmic$logratio== -Inf, max( ))


cosmic = cosmic [ order ( cosmic$logp, abs ( cosmic$logratio), decreasing = T ), ]
cosmic$balance = factor ( cosmic$balance, levels= c( "EXTREME_HIGH" ,"HIGH" ,"LOW" ,"EXTREME_LOW"  , "no.sig"  ) )





### What are the classification for up regulators vs down? 
view_snv = c( "sample",   "VEP_Worst_Gene", "RNA_PVALUE"  ,     "RNA_DEPTH"   ,  "RNA_REF_COUNT"  , "RNA_ALT_COUNT" , "RNA_ALT_AF",  "WGS_REF_COUNT", "WGS_ALT_COUNT", "TUMOR.AF", "TUMOR.AD" , "balance","is.sig","cancer.subtype")
cosmic$balance
cosmic$balance_direction = ifelse ( cosmic$rafdaf <=1, "down", "up") 
cosmic$balance_simple = gsub ( "EXTREME_", "", cosmic$balance)
cosmic$VEP_Worst_Impact_simple = gsub ( "HIGH|MODERATE","High/Moderate", cosmic$VEP_Worst_Impact)
cosmic$VEP_Worst_Impact_simple = gsub ( "LOW|MODIFIER","LOW/MODIFIER", cosmic$VEP_Worst_Impact_simple)

# what is 
balance_tab = data.frame ( table ( cosmic [  ,  "balance" ]   ))
colnames ( balance_tab) = c("state", "Freq")

#### ALL with worse impact ell

rafdaf_color = c( EXTREME_HIGH="#d65175", HIGH="#f5d9cb", LOW="#b8cce3", EXTREME_LOW="#46a6eb", no.sig="grey" )

lthis = cosmic [ ! duplicated ( paste ( cosmic$WGS.id, cosmic$VEP_Worst_Gene ) ) & cosmic$is.sig == "sig"  & cosmic$VEP_Worst_Impact %in% c("HIGH", "MODERATE")  , ] 
lthis = lthis [ order (   abs (lthis$neg_log_ratio) , lthis$logp, decreasing = T ), ]

#View ( lthis[ lthis$is.cancer == "yes" & grepl ( "LOW", lthis$balance), c(view_snv,"rafdaf","neg_log_ratio","logp") ])
#View ( lthis[ lthis$is.cancer == "yes" & grepl ( "HIGH", lthis$balance), c(view_snv,"rafdaf","neg_log_ratio","logp") ])



### lets study the RAF/DAF a bit here. 


library(networkD3)
##### https://stackoverflow.com/questions/63255589/create-a-sankey-diagram-with-tidy-data-in-r

# get the columns we need 
# going to calculate frequency for each combinations 
# the way this works is that each column will be a seperate node
# the frequency table will give you the total combination for each combination 
set.seed(123)
ar_ratio= cosmic [ , c("balance_simple", "VEP_Worst_Impact_simple",   "is.cancer" , "VEP_Worst_Gene",   "uid" )]

# below has nothing to do with plot. I'm just trying to associate the gene with a cosmic annotation 
ar_ratio = merge ( ar_ratio, genelist[ , c("cosmic.annot", "gene")], by.x="VEP_Worst_Gene", by.y="gene", all.x=T, all.y=F )
ar_ratio$cosmic.annot = gsub ( ",FUSION" , "", ar_ratio$cosmic.annot)
ar_ratio$cosmic.annot = gsub ( "^ONCOGENE$|^TSG$" , "ONCOGENE,TSG", ar_ratio$cosmic.annot)
ar_ratio[ is.na ( ar_ratio$cosmic.annot), ]$cosmic.annot  <- "not.cancer"

# have started here is easier. 
#ar_ratio= cosmic [ , c("balance_simple", "VEP_Worst_Impact_simple",   "is.cancer" ,   "uid" )]
ar_ratio <- data.frame ( table (  ar_ratio[ , colnames ( ar_ratio) != c("uid")]  ))
ar_ratio = ar_ratio[ar_ratio$Freq != 0, ]
ar_ratio$VEP_Worst_Gene = NULL 
# rename to make things clearer 
colnames ( ar_ratio) = c( "allelic balance (RAF/DAF)", "Predicted Impact", "Cancer Gene", "annt", "Freq")
# rename values, 
ar_ratio$`allelic balance (RAF/DAF)` = gsub ( "no.sig", "concordant", ar_ratio$`allelic balance (RAF/DAF)` )

# create a linkage table from left to right, that is left is source and right column is target

links <-
  ar_ratio %>% 
  as_tibble() %>%
  mutate(row = row_number()) %>% 
  pivot_longer(cols = c(-row, -Freq),
               names_to = 'column', values_to = 'source') %>% 
  mutate(column = match(column, names(ar_ratio))) %>% 
  mutate(source = paste0(source, '__', column)) %>% 
  group_by(row) %>% 
  mutate(target = lead(source, order_by = column)) %>% 
  drop_na(target, source) %>% 
  group_by(source, target) %>% 
  summarise(value = sum(Freq), .groups = 'drop')

# create a table for nodes 
nodes <- data.frame(name = unique(c(links$source, links$target)))
links$source <- match(links$source, nodes$name) - 1
links$target <- match(links$target, nodes$name) - 1
nodes$name <- sub('__[0-9]+$', '', nodes$name)
## each node should have a color 


nodes$colors = getPalette( nrow ( nodes) )


# set manual colors 
nodes[nodes$name == "concordant", ]$colors = as.character ( rafdaf_color[ "no.sig"])
nodes[nodes$name == "HIGH", ]$colors = as.character ( rafdaf_color[ "EXTREME_HIGH"])
nodes[nodes$name == "LOW", ]$colors =as.character (  rafdaf_color[ "EXTREME_LOW"])
nodes[nodes$name == "High/Moderate", ]$colors ="#e3a944"
nodes[nodes$name == "LOW/MODIFIER", ]$colors ="#708c5b"
nodes[nodes$name == "no", ]$colors ="#c288b4"
nodes[nodes$name == "yes", ]$colors ="#7685a3"
nodes[nodes$name == "not.cancer", ]$colors ="#ad8c51"
nodes[nodes$name == "N", ]$colors ="#7d7b78"
nodes[nodes$name == "FUSION", ]$colors ="#708c5b"
nodes[nodes$name == "ONCOGENE,TSG", ]$colors ="#68a0a8"

# add color to JS which is what is under this plot function 

colors <- paste(nodes$colors, collapse = '", "')
colorJS <- paste('d3.scaleOrdinal(["', colors, '"])')

sankeyPlot = sankeyNetwork(Links = links, Nodes = nodes, Source = "source", 
              Target = "target", Value = "value", NodeID = "name", colourScale = colorJS )

 
sankeyPlot$x$options$fontSize = 20

# I'm going to manaully change the plot's labels, adding in total for each node and percentage 
sankeyPlot2 = sankeyPlot 


# get total to add to label 
# each column is a new level 
# we want to get total at each leve. 
df = data.frame ()

for (n  in colnames ( ar_ratio)[-ncol(ar_ratio)] )
{
n = ar_ratio %>% dplyr::group_by_at(n) %>%
  dplyr::summarise(
    total = sum ( Freq )  
  ) %>%  
  dplyr::mutate( 
    Percentage= paste0 ( round ( total / sum ( total ), 2) * 100, "%")
    ) %>% data.frame()
colnames ( n) = c("name","total","per")
df = rbind ( df, n )
}


nodes2 = merge ( nodes, df, by="name")
nodes2$name2 = paste0 ( nodes2$name, " (total ",nodes2$total,"| ", nodes2$per, " )"  )

# reorder nodes so that it matches name on the nodes
#sankeyPlot$x$nodes$name=  "Concordant ( p > .05 ) 5000"
# the plot will get the name from node but right now is not  in the same order. 

nodes2 = nodes2[match(   sankeyPlot2$x$nodes$name, nodes2$name  )      ,] 
sankeyPlot2$x$nodes$name = nodes2$name2
sankeyPlot2

### now drill down on the high/moderate 


# find top genes for each category 
nt=15
# just plot in the above plot object 
plot_top <- function ( ddd, title = "", legendp="none" , ax=30, ay=30, tt=20){
top_all = ddd
top_all = top_all [ ! duplicated ( paste ( top_all$WGS.id, top_all$VEP_Worst_Gene ) ) , ]

top_all_10 = data.frame ( table ( top_all$VEP_Worst_Gene ) )
top_all_10 = top_all_10[top_all_10$Freq > 1, ]


top_all_10 = top_all_10[ order ( top_all_10$Freq, decreasing = T), ]
top_all = top_all[top_all$VEP_Worst_Gene %in% as.character(top_all_10$Var1)[1:nt], ]
top_all$direction = ifelse ( top_all$rafdaf <=1, "down","up")
# count by gene and disease 
top_all = top_all[top_all$direction == "up", ]
top_all = top_all %>% dplyr::group_by(VEP_Worst_Gene, cancer.subtype, direction ) %>%
  dplyr::summarise(
    total = n (   ) 
  ) %>% data.frame()

top_all$total = ifelse ( top_all$direction == "up", top_all$total, top_all$total * -1 )

# reorder by total 
ro = top_all %>% dplyr::group_by(VEP_Worst_Gene ) %>%
  dplyr::summarise(
    total = sum ( abs ( total)   ) 
  ) %>% arrange(total) %>% data.frame()

top_all$VEP_Worst_Gene = factor (top_all$VEP_Worst_Gene, levels = as.character ( ro$VEP_Worst_Gene) )


freq_cancer = ggplot() + 
  geom_bar(data = top_all, aes(x=VEP_Worst_Gene, y=total, fill=cancer.subtype),stat = "identity") +
  #geom_bar(data = top_all, aes(x=VEP_Worst_Gene, y=total, fill=cancer.subtype),stat = "identity") +
  scale_fill_manual(values=subcolor) + coord_flip() + ylab ( " RAF/DAF  ") + xlab ('') +
  theme_minimal()  +
  theme(legend.position=legendp,  legend.key = element_blank(),
        axis.text.x = element_text(size= ax ), 
        axis.text.y = element_text(size= ay ),
        #axis.text.x = element_blank(),
        axis.title.x = element_text(size=ax),
        plot.title = element_text(size=tt), 
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12)
  ) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  geom_hline(yintercept=0, color="grey", size=1.5) + ggtitle ( title )

return ( list ( plot = freq_cancer, df = top_all )  )

}


# none duplicted, significant high to moderate prediction only 
# tempc= cosmic [ ! duplicated ( paste ( cosmic$WGS.id, cosmic$VEP_Worst_Gene ) ), ]
tempc = cosmic[ cosmic$VEP_Worst_Gene != "HLA-A", ]
# the function will remove duplicates so don't do it here. 

snv_gene_cancer = plot_top ( ddd=tempc[ tempc$is.sig == "sig"  
                    & tempc$VEP_Worst_Impact %in% c("HIGH", "MODERATE")
                    & tempc$is.cancer == "yes"
                    , ] , title = "Cosmic Cancer Genes"  )

snv_gene_non_cancer = plot_top ( 
                    tempc[ tempc$is.sig == "sig"  
                    & tempc$VEP_Worst_Impact %in% c("HIGH", "MODERATE")
                    & tempc$is.cancer == "no"
                    , ] , title = "Non-Cancer Genes",legendp="none"  )



snv_gene_cancer_df = snv_gene_cancer$df
snv_gene_non_cancer_df = snv_gene_non_cancer$df 

snv_gene_cancer =snv_gene_cancer$plot
snv_gene_non_cancer =snv_gene_non_cancer$plot


snv_gene_cancer$labels$y=""



( snv_gene_cancer + xlab("")  )/ snv_gene_non_cancer + ylab ( "Frequency")

######################################

up_high_onco =  tempc[ tempc$balance_simple == "HIGH"  
                    & tempc$VEP_Worst_Impact %in% c("HIGH", "MODERATE")
                    & tempc$is.cancer == "yes"
                    , ] 
  
up_high_onco = up_high_onco[ order ( up_high_onco$rafdaf, decreasing = T), ]
## we should label some of the 5% 




lthis = up_high_onco$uid[1:15]
  
dafraf_plot_all = ggplot(cosmic, aes(x=TUMOR.AF, y= RNA_ALT_AF, shape = is.cancer)) + 
    geom_point( aes ( color=balance , alpha=is.sig, size=logp ), position=position_jitter(h=0.01, w=0.01)  )  + 
  scale_color_manual(values = rafdaf_color  , labels = c(EXTREME_HIGH="> 2", HIGH="[1-2]", LOW="[0.25-1]", EXTREME_LOW="< 0.25", no.sig="not significant")      ) +
     theme_minimal()   + xlab("DAF") + ylab ( "RAF") +
  theme(legend.position="right",  legend.key = element_blank(),
          
          axis.text.y = element_text(size= 15 ),
          #axis.text.x = element_blank(),
          axis.title.x = element_text(size=20),
          
          axis.title.y     = element_text(size=20), 
          legend.text      =element_text(size=12)
    ) + scale_alpha_manual(values=c(  EXTREME_HIGH=.6, HIGH=.6, LOW=.6, EXTREME_LOW=.6, no.sig=.18  )) +
  guides(
        shape = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="cancer gene")
        , size = FALSE
        #, fill = guide_legend(title.position="top", title.hjust = 0.5)
        , color = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="RAF/DAF")
        , alpha = FALSE
    ) +
  geom_text_repel(
        data = . %>% 
            mutate(label = ifelse(
                uid %in% lthis,
                     VEP_Worst_Gene, "")), 
        aes(
            label = label, size=10
        ), 
        box.padding = 0.01, size = 4, colour = "black", segment.size  = 0.1, seed = 123
    ) + ggtitle ( "All SNVs ")



dafraf_plot_all_log = ggplot(cosmic, aes(x=log2 ( RNA_REF_COUNT + RNA_ALT_COUNT ), y=  rafdaf  , shape = is.cancer)) + 
    geom_point( aes ( color=balance , alpha=is.sig, size=logp ), position=position_jitter(h=0.01, w=0.01)  )  + 
  scale_color_manual(values = rafdaf_color  , labels = c(EXTREME_HIGH="> 2", HIGH="[1-2]", LOW="[0.25-1]", EXTREME_LOW="< 0.25", no.sig="not significant")      ) +
     theme_minimal()   + xlab("log2 ( RNA Depth)") + ylab ( "log2 ( RAF/DAF)") +
  theme(legend.position="none",  legend.key = element_blank(),
          
          axis.text.y = element_text(size= 15 ),
          #axis.text.x = element_blank(),
          axis.title.x = element_text(size=20),
          
          axis.title.y     = element_text(size=20), 
          legend.text      =element_text(size=12)
    ) + scale_alpha_manual(values=c(  EXTREME_HIGH=.6, HIGH=.6, LOW=.6, EXTREME_LOW=.6, no.sig=.18  )) +
  guides(
        shape = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="cancer gene")
        , size = FALSE
        #, fill = guide_legend(title.position="top", title.hjust = 0.5)
        , color = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="RAF/DAF")
        , alpha = FALSE
    ) +
  geom_text_repel(
        data = . %>% 
            mutate(label = ifelse(
                uid %in% lthis,
                     VEP_Worst_Gene, "")), 
        aes(
            label = label, size=10
        ), 
        box.padding = 0.01, size = 4, colour = "black", segment.size  = 0.1, seed = 123
    ) + ggtitle ( "All SNVs ")


# lets get all the genes median 
outmedian = out %>% group_by(gene) %>% 
  summarise ( 
    mediantpm = median ( tpm)
    ) %>% data.frame()
outmedian$ztpm = scale ( outmedian$mediantpm)

cosmicz = merge ( cosmic, outmedian, by.x="GENE", by.y="gene")



dafraf_plot_all_zscale = ggplot(cosmicz, aes(x=ztpm , y=  rafdaf  , shape = is.cancer)) + 
    geom_point( aes ( color=balance , alpha=is.sig, size=logp ), position=position_jitter(h=0.01, w=0.01)  )  + 
  scale_color_manual(values = rafdaf_color  , labels = c(EXTREME_HIGH="> 2", HIGH="[1-2]", LOW="[0.25-1]", EXTREME_LOW="< 0.25", no.sig="not significant")      ) +
     theme_minimal()   + xlab("zscale TPM of Gene") + ylab ( "log2 ( RAF/DAF)") +
  theme(legend.position="none",  legend.key = element_blank(),
          
          axis.text.y = element_text(size= 15 ),
          #axis.text.x = element_blank(),
          axis.title.x = element_text(size=20),
          
          axis.title.y     = element_text(size=20), 
          legend.text      =element_text(size=12)
    ) + scale_alpha_manual(values=c(  EXTREME_HIGH=.6, HIGH=.6, LOW=.6, EXTREME_LOW=.6, no.sig=.18  )) +
  guides(
        shape = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="cancer gene")
        , size = FALSE
        #, fill = guide_legend(title.position="top", title.hjust = 0.5)
        , color = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="RAF/DAF")
        , alpha = FALSE
    ) +
  geom_text_repel(
        data = . %>% 
            mutate(label = ifelse(
                uid %in% lthis,
                     VEP_Worst_Gene, "")), 
        aes(
            label = label, size=10
        ), 
        box.padding = 0.01, size = 4, colour = "black", segment.size  = 0.1, seed = 123
    ) + ggtitle ( "All SNVs ")







# simplify and most likely will be used in paper: 4/21/2022
# remove cancer annotation shape 
# remove varying color 
# remove labels

rafdaf_color_simple = c ( HIGH=  "#d65175", LOW="#46a6eb", no.sig = "grey")


total_high = nrow ( cosmic[ cosmic$balance_simple == "HIGH", ]) 
total_nosig = nrow ( cosmic[ cosmic$balance_simple == "no.sig", ]) 
total_low =  nrow ( cosmic[ cosmic$balance_simple == "LOW", ]) 

daf_vs_raf = ggplot(cosmic, aes(x=TUMOR.AF, y= RNA_ALT_AF)) + 
  geom_point( aes ( color=balance_simple , alpha=balance), position=position_jitter(h=0.01, w=0.01)  )  + 
  scale_color_manual(values = rafdaf_color_simple  
                     , labels = c(HIGH=paste0 ( "RAF > DAF (" , total_high, ")") , 
                                LOW=paste0 ( "RAF < DAF (", total_low, ")" ) , 
                                no.sig=paste0 ( "not significant (", total_nosig, ")")      ) ) +
  scale_alpha_manual ( values =  c(EXTREME_HIGH=".7", HIGH=".5", LOW=".5", EXTREME_LOW=".7", no.sig=".3")  ) +
  theme_minimal()   + xlab("DAF") + ylab ( "RAF") +
  theme(legend.position="right",  legend.key = element_blank(),
        axis.text.x = element_text(size= 20 ), 
        axis.text.y = element_text(size= 20 ),
        axis.title.x = element_text(size=20),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12)
  ) + scale_alpha_manual(values=c(  EXTREME_HIGH=.6, HIGH=.6, LOW=.6, EXTREME_LOW=.6, no.sig=.18  )) +
  guides(
    shape = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="cancer gene")
    , size = FALSE
    , color = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="RAF/DAF")
    , alpha = FALSE
  ) + ggtitle ( "All SNVs ") + geom_abline(intercept =0 , slope = 1) +
  theme(panel.grid.major = element_blank()
          , panel.grid.minor = element_blank()
          ,panel.background = element_blank()
          , axis.line = element_line(colour = "black") ) +
  scale_y_continuous(expand = expansion(mult = c(0, .01)))

# add porportion

af_port = data.frame (  table ( cosmic [ , c ( "balance_simple", "cancer.subtype") ]  ), stringsAsFactors = FALSE )  
af_port$balance_simple = factor (  af_port$balance_simple , levels = c( "HIGH", "no.sig", "LOW") )

# lets try to rank this by highest porportion of high 
rankhigh = af_port %>% group_by( cancer.subtype )  %>% 
  mutate ( 
    total = sum ( Freq )
    ) %>% group_by( cancer.subtype, balance_simple ) %>%
  mutate (
    percent = round ( Freq / total, 3 )  
  ) %>% 
  subset(., balance_simple == "HIGH") %>% 
  arrange(.,desc(percent)) 

af_port$cancer.subtype = factor (  af_port$cancer.subtype 
                                   , levels = as.character ( rankhigh$cancer.subtype )
                                   )


af_port_plot <- ggplot(af_port, aes(fill=balance_simple, y=Freq, x=cancer.subtype) )+ 
    geom_bar(position="fill", stat="identity", width=1, color='#e8eded') +
  # add text but position needs to be set to the same, which is filled
    geom_text(aes(label = Freq),
            position = position_fill(vjust = 0.5)
  ) +  
  scale_y_continuous(expand = expansion(mult = c(0, 0))) + # remove gaps 
   theme(legend.position="right",  legend.key = element_blank(),
        axis.text.x = element_text(size= 20 , angle=30, hjust = 1  ), 
        axis.text.y = element_text(size= 20 ),
        axis.title.x = element_text(size=20),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12), 
        panel.grid.major = element_blank()
          , panel.grid.minor = element_blank()
          ,panel.background = element_blank()
          , axis.line = element_line(colour = "white") # this will give the illusion to fill in any remaining gaps
          , panel.border=element_blank()
        
  ) + ylab("percentage") + xlab ( "") +
  scale_fill_manual(values = rafdaf_color_simple  , labels = c(HIGH="RAF > DAF", LOW="RAF < DAF", no.sig="not significant")      ) +
   guides(
    fill = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="RAF/DAF balance")
    ) 
    
  
# remove none - significance and plot 

nosig_port = af_port[ af_port$balance_simple != "no.sig", ] 

rankhigh2 = nosig_port %>% group_by( cancer.subtype )  %>% 
  mutate ( 
    total = sum ( Freq )
    ) %>% group_by( cancer.subtype, balance_simple ) %>%
  mutate (
    percent = round ( Freq / total, 3 )  
  ) %>% 
  subset(., balance_simple == "HIGH") %>% 
  arrange(.,desc(percent)) 

nosig_port$cancer.subtype = factor (  nosig_port$cancer.subtype 
                                   , levels = rev ( as.character ( rankhigh2$cancer.subtype ) )
                                   )




af_port_plot_nosig <- ggplot( nosig_port , aes(fill=balance_simple, y=Freq, x=cancer.subtype) )+ 
    geom_bar(position="fill", stat="identity", width=1, color='#e8eded') +
  # add text but position needs to be set to the same, which is filled
    geom_text(aes(label = Freq),
            position = position_fill(vjust = 0.5)
  ) +  
  scale_y_continuous(expand = expansion(mult = c(0, 0))) + # remove gaps 
   theme(legend.position="right",  legend.key = element_blank(),
        axis.text.x = element_text(size= 20   ), 
        axis.text.y = element_text(size= 20 ),
        axis.title.x = element_text(size=20),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12), 
        panel.grid.major = element_blank()
          , panel.grid.minor = element_blank()
          ,panel.background = element_blank()
          , axis.line = element_line(colour = "white") # this will give the illusion to fill in any remaining gaps
          , panel.border=element_blank()
        
  ) + ylab("percentage") + xlab ( "") +
  scale_fill_manual(values = rafdaf_color_simple  , labels = c(HIGH="RAF > DAF", LOW="RAF < DAF", no.sig="not significant")      ) +
   guides(
    fill = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="RAF/DAF balance")
    ) + coord_flip() +geom_hline(yintercept=.5, color="#868786", size=1.5) 


### add proportion but in the context of type

table ( cosmic$VEP_Worst_Consequence)

ap2 = cosmic 
ap2$VEP_Worst_Consequence = gsub ( "synonymous_variant", "Synonymous", ap2$VEP_Worst_Consequence)
ap2 = ap2[ap2$VEP_Worst_Consequence %in% c ( 'missense_variant', 'frameshift_variant', 'inframe_insertion', 'Synonymous' ),  ]


ap2$synonymous = ifelse ( ap2$VEP_Worst_Consequence == "Synonymous", "Synonymous", "NonSynonymous"  )


ap2 = data.frame (  table ( ap2 [ , c ( "synonymous", "balance_simple") ]  ), stringsAsFactors = FALSE )  
ap2$balance_simple = factor (  ap2$balance_simple , levels = c( "HIGH", "no.sig", "LOW") )




nonsym_sym = ggplot(ap2 , aes(fill=balance_simple, y=Freq, x=synonymous ) )+ 
    geom_bar(position="fill", stat="identity", width=1, color='#e8eded') +
  # add text but position needs to be set to the same, which is filled
    geom_text(aes(label = Freq),
            position = position_fill(vjust = 0.5)
  ) +  
  scale_y_continuous(expand = expansion(mult = c(0, 0))) + # remove gaps 
   theme(legend.position="right",  legend.key = element_blank(),
        axis.text.x = element_text(size= 20 , angle=30, hjust = 1  ), 
        axis.text.y = element_text(size= 20 ),
        axis.title.x = element_text(size=20),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12), 
        panel.grid.major = element_blank()
          , panel.grid.minor = element_blank()
          ,panel.background = element_blank()
          , axis.line = element_line(colour = "white") # this will give the illusion to fill in any remaining gaps
          , panel.border=element_blank()
        
  ) + ylab("percentage") + xlab ( "") +
  scale_fill_manual(values = rafdaf_color_simple  , labels = c(HIGH="RAF > DAF", LOW="RAF < DAF", no.sig="not significant")      ) +
   guides(
    fill = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="RAF/DAF balance")
    ) 


#### what are the AF percentage between comsic genes and non-cancer genes. 

rafdaf_cosmic_plot = data.frame (  table ( cosmic [ , c ( "balance_simple", "is.cancer") ]  ), stringsAsFactors = FALSE )   %>% 
   subset(., balance_simple != "no.sig") %>% 
   group_by( is.cancer  ) %>% 
   mutate(  total = sum ( Freq), 
            percent = round (  Freq / total, 2 )
            , is.cancer = factor ( is.cancer, levels= c("yes", "no"))
            ) %>% 
   ggplot( . , aes(fill=balance_simple, y=percent, x=is.cancer) )+ 
   geom_bar(position="fill", stat="identity", width=1, color='#e8eded') +
   # add text but position needs to be set to the same, which is filled
   geom_text(aes(label = Freq),
             position = position_fill(vjust = 0.5)
             , size = 20
   ) +  
   scale_y_continuous(expand = expansion(mult = c(0, 0))) + # remove gaps 
   theme(legend.position="none",  legend.key = element_blank(),
         axis.text.x = element_text(size= 20 , vjust=0, angle=0  ), 
         axis.text.y = element_text(size= 20 ),
         axis.title.x = element_text(size=20),
         axis.title.y     = element_text(size=20), 
         legend.text      =element_text(size=12), 
         panel.grid.major = element_blank()
         , panel.grid.minor = element_blank()
         ,panel.background = element_blank()
         , axis.line = element_line(colour = "white") # this will give the illusion to fill in any remaining gaps
         , panel.border=element_blank()
         
   ) + ylab("percentage") + xlab ( "") +
   scale_fill_manual(values = rafdaf_color_simple  , labels = c(HIGH="RAF > DAF", LOW="RAF < DAF", no.sig="not significant")      ) +
   guides(
     fill = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="RAF/DAF balance")
   ) +geom_hline(yintercept=.5, color="#868786", size=1.5) +  xlab("Cosmic Cancer Gene") 
 




# what are the percent AF between oncogene and TSG? 

ctemp = merge ( cosmic,  genelist[  , c("gene", "cosmic.annot" )], by.x='VEP_Worst_Gene', by.y='gene')
ctemp = ctemp[ctemp$RNA_PVALUE < .05,   ]
ctemp = ctemp [ ctemp$is.cancer == "yes", ]
ctemp$cosmic.annot = ifelse ( grepl ( ",", ctemp$cosmic.annot ), "MULTI", ctemp$cosmic.annot )
 
 
rafdaf_oncTSG_plot =  data.frame (  table ( ctemp [ , c ( "balance_simple", "cosmic.annot") ]  ), stringsAsFactors = FALSE )   %>% 
    subset(., balance_simple != "no.sig") %>% 
    group_by( cosmic.annot  ) %>% 
    mutate(  total = sum ( Freq), 
             percent = round (  Freq / total, 2 )  
    )  %>% 
    ggplot( . , aes(fill=balance_simple, y=percent, x=cosmic.annot) )+ 
    geom_bar(position="fill", stat="identity", width=1, color='#e8eded') +
    # add text but position needs to be set to the same, which is filled
    geom_text(aes(label = Freq),
              position = position_fill(vjust = 0.5)
              , size = 16 
    ) +  
    scale_y_continuous(expand = expansion(mult = c(0, 0))) + # remove gaps 
    theme(legend.position="none",  legend.key = element_blank(),
          axis.text.x = element_text(size= 20, hjust = 1, angle=20   ), 
          axis.text.y = element_text(size= 20 ),
          axis.title.x = element_text(size=20),
          axis.title.y     = element_text(size=20), 
          legend.text      =element_text(size=12), 
          panel.grid.major = element_blank()
          , panel.grid.minor = element_blank()
          ,panel.background = element_blank()
          , axis.line = element_line(colour = "white") # this will give the illusion to fill in any remaining gaps
          , panel.border=element_blank()
          
    ) + ylab("percentage") + xlab ( "") +
    scale_fill_manual(values = rafdaf_color_simple  , labels = c(HIGH="RAF > DAF", LOW="RAF < DAF", no.sig="not significant")      ) +
    guides(
       fill = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="RAF/DAF balance")
    ) + 
   #coord_flip() +
   geom_hline(yintercept=.5, color="#868786", size=1.5) 
 
 
rafdaf_cosmic_plot + rafdaf_oncTSG_plot

# save this for later for stack plot of cosmic annotations for each type of aa vep consequences. 
byCancer_Gene_df = ctemp



# ? what is the ratio of AF for each vep consquences? 
## only vep consequences with > 50 are plotted. 

pretty_var_names = c( synonymous_variant = "synonymous", non_coding_transcript_exon_variant = "noncoding_exon_variant" )

ap2 = cosmic [ cosmic$RNA_PVALUE < .05, ]

for ( f in names ( pretty_var_names) ){
  ap2$VEP_Worst_Consequence = gsub ( f, as.character ( pretty_var_names[f] ) , ap2$VEP_Worst_Consequence)
}

#ap2$VEP_Worst_Consequence = gsub ( "synonymous_variant", "synonymous", ap2$VEP_Worst_Consequence)
#ap2$VEP_Worst_Consequence = gsub ( "non_coding_transcript_exon_variant", "noncoding_exon_variant", ap2$VEP_Worst_Consequence)



ftemp = data.frame ( table ( ap2$VEP_Worst_Consequence) )
aa_conq = as.character ( ftemp[ftemp$Freq > 50, ]$Var1 )

ap2 = ap2[ap2$VEP_Worst_Consequence %in%  aa_conq , ]
ap2$synonymous = ap2$VEP_Worst_Consequence
ap2 = data.frame (  table ( ap2 [ , c ( "synonymous", "balance_simple") ]  ), stringsAsFactors = FALSE )  
ap2$balance_simple = factor (  ap2$balance_simple , levels = c( "HIGH", "no.sig", "LOW") )



rankhigh2 = ap2 %>% group_by( synonymous  )  %>% 
  mutate ( 
    total = sum ( Freq )
    ) %>% group_by( synonymous , balance_simple ) %>%
  mutate (
    percent = round ( Freq / total, 3 )  
  ) %>% 
  subset(., balance_simple == "HIGH") %>% 
  arrange(.,desc(percent)) 

ap2$synonymous = factor (  ap2$synonymous 
                                   , levels = rev ( as.character ( rankhigh2$synonymous ) )
                                   )



portion_consequence_stack = ggplot(ap2 , aes(fill=balance_simple, y=Freq, x=synonymous ) )+ 
    geom_bar(position="fill", stat="identity", width=1, color='#e8eded') +
  # add text but position needs to be set to the same, which is filled
    geom_text(aes(label = Freq),
            position = position_fill(vjust = .7)
  ) +  
  scale_y_continuous(expand = expansion(mult = c(0, 0))) + # remove gaps 
   theme(legend.position="right",  legend.key = element_blank(),
        axis.text.x = element_text(size= 20 , angle=00, hjust = 0  ), 
        axis.text.y = element_text(size= 20 ),
        axis.title.x = element_text(size=20),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12), 
        panel.grid.major = element_blank()
          , panel.grid.minor = element_blank()
          ,panel.background = element_blank()
          , axis.line = element_line(colour = "white") # this will give the illusion to fill in any remaining gaps
          , panel.border=element_blank()
        
  ) + ylab("percentage") + xlab ( "") +
  scale_fill_manual(values = rafdaf_color_simple  , labels = c(HIGH="RAF > DAF", LOW="RAF < DAF", no.sig="not significant")      ) +
   guides(
    fill = guide_legend(title.position="top", title.vjust = .5, override.aes = list(size=10) , title ="RAF/DAF balance")
    ) + coord_flip()



### AF stack but comparing LOW vs HIGH  

set.seed(12)
rcolor2 = sample ( getPalette(20 ) , 8 ) 
plotByrafdafType <- ggplot(ap2 , aes(fill=synonymous, y=Freq, x=  balance_simple) )+ 
    geom_bar(position="fill", stat="identity", width=1, color='#e8eded') +
  # add text but position needs to be set to the same, which is filled
    geom_text(aes(label = Freq),
            position = position_fill(vjust = .7)
  ) +  
  scale_y_continuous(expand = expansion(mult = c(0, 0))) + # remove gaps 
   theme(legend.position="right",  legend.key = element_blank(),
        axis.text.x = element_text(size= 20 , angle=00, hjust = 0  ), 
        axis.text.y = element_text(size= 20 ),
        axis.title.x = element_text(size=20),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12), 
        panel.grid.major = element_blank()
          , panel.grid.minor = element_blank()
          ,panel.background = element_blank()
          , axis.line = element_line(colour = "white") # this will give the illusion to fill in any remaining gaps
          , panel.border=element_blank()
        
  ) + ylab("percentage") + xlab ( "") +
  scale_fill_manual(values = rcolor2 , labels = c(HIGH="RAF > DAF", LOW="RAF < DAF", no.sig="not significant")      ) +
   guides(
    fill = guide_legend(title.position="top", title.vjust = .5, override.aes = list(size=10) , title ="RAF/DAF balance")
    ) 


## vep_consquences bar version

compareTypes <- ap2 %>% group_by( synonymous  )  %>% 
  subset( balance_simple != "no.sig") %>%
  mutate ( 
    total = sum ( Freq )
    ) %>% group_by( synonymous , balance_simple ) %>%
  
  mutate (
    percent = round ( Freq / total, 3 )  
  )  %>%  
  arrange(factor(synonymous, levels = unique(synonymous[balance_simple == "HIGH"
                                            ][order(dense_rank(-percent[balance_simple == "HIGH"]))]
  )  )  ) %>% mutate ( balance_simple = factor ( balance_simple , levels = c("LOW", "HIGH")) 
                       ) %>% data.frame() 



compareTypes$synonymous = factor ( compareTypes$synonymous, levels= rev ( unique ( as.character(compareTypes$synonymous))) )
#%>% mutate(synonymous = factor(synonymous, levels =  unique (  synonymous)  )) %>% 

var_outcome= ggplot(compareTypes, aes(synonymous , percent)) +   
  geom_bar(aes(fill = balance_simple), position = "dodge", stat="identity") +  
  scale_y_continuous(expand = expansion(mult = c(0, 0))) + # remove gaps 
   theme(legend.position="none",  legend.key = element_blank(),
        axis.text.x = element_text(size= 20 , angle=30, hjust = 1  ), 
        axis.text.y = element_text(size= 20 ),
        axis.title.x = element_text(size=20),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12), 
        panel.grid.major = element_blank()
          , panel.grid.minor = element_blank()
          ,panel.background = element_blank()
          , axis.line = element_line(colour = "grey") # this will give the illusion to fill in any remaining gaps
          , panel.border=element_blank()
        
  ) + ylab("percentage") + xlab ( "") +
  scale_fill_manual(values = rafdaf_color_simple  , labels = c(HIGH="RAF > DAF", LOW="RAF < DAF", no.sig="not significant")      ) +
   guides(
    fill = guide_legend(title.position="top", title.vjust = .5, override.aes = list(size=10) , title ="RAF/DAF balance")
    ) 

## for each var_outcome group what is the top cancer gene and top non-cancer gene? 


temp = compareTypes[ !duplicated ( compareTypes$synonymous), c("synonymous", "total")]
colnames ( temp) = c("Var1", "Freq")


temp$per = temp$Freq / sum ( temp$Freq)

temp$Var1 = gsub ( "_variant", "", temp$Var1)
temp$Var1 = gsub ( "_prime", "'", temp$Var1)




vep_consq_pie = ggplot(data = temp,
       aes(x = "1", y = per, fill = Var1)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = Freq  ),
            position = position_stack(vjust = 0.5),
            color = "grey20", size = 12, fontface = "bold"  ) +
  coord_polar(theta = "y" ) +
  scale_fill_manual(values= brewer.pal(8, "Dark2") )  +
  theme_void() +
  theme(legend.position = "none",
        legend.direction = "vertical")+  

  theme ( legend.text      =element_text(size=12),
          legend.title = element_text(size=12) ) +
  geom_label(aes(x = 1.55, label = Var1), 
             position = position_stack(vjust = 0.75) 
             ,label.padding = unit(.25, "lines")
             , size = 6.5
             , label.size = NA
             )


# need to change back the names 


compareTypes$VEP_Worst_Consequence = gsub (  "synonymous", "synonymous_variant", compareTypes$synonymous)
compareTypes$VEP_Worst_Consequence = gsub ( "noncoding_exon_variant", "non_coding_transcript_exon_variant" , compareTypes$VEP_Worst_Consequence)


tab_gene = data.frame ()

for ( v in unique ( as.character ( compareTypes$VEP_Worst_Consequence ) )){
  v2 = cosmic[ cosmic$VEP_Worst_Consequence %in% v, ]
  v2_cancer = data.frame ( table ( v2[v2$is.cancer == "yes" & v2$RNA_PVALUE < .05, ]$VEP_Worst_Gene) )
  v2_cancer = v2_cancer[order ( v2_cancer$Freq, decreasing = T), ]
  v2_nocancer = data.frame ( table ( v2[v2$is.cancer == "no" & v2$RNA_PVALUE < .05, ]$VEP_Worst_Gene) )
  v2_nocancer = v2_nocancer[order ( v2_nocancer$Freq, decreasing = T), ]
  
  v2_cancer$type = v 
  v2_cancer$is.cancer = "yes"
  
  v2_nocancer$type = v 
  v2_nocancer$is.cancer = "no"
  
  tab_gene = rbind (tab_gene, head ( v2_cancer, 1 ) )
  tab_gene = rbind (tab_gene, head ( v2_nocancer, 1 ) )
}



kable( tab_gene[ tab_gene$Freq > 2, ]  , format = "html" , row.names = F, caption = "Most frequent gene in each category > 2" ) %>% kable_classic(full_width = F, position = "center")

## plot by genes. 

# stop here - next should remove some aa consequence categories. 

highFreqGene <- cosmic[cosmic$VEP_Worst_Gene %in% as.character ( tab_gene[ tab_gene$Freq > 2, ]$Var1 ), ]

highFreqGenePlot <-highFreqGene %>% group_by( VEP_Worst_Gene, VEP_Worst_Consequence, balance_simple ) %>% 
  summarise ( total = n() ) %>% 
  ggplot(.) + 
    geom_bar( 
             mapping = aes(x=VEP_Worst_Gene , 
                           y = total, fill = as.factor(balance_simple)  ), 
             stat="identity", 
             position='stack', 
             width = 1, color = "black"
             
             ) + scale_fill_manual(values = rafdaf_color_simple ) +
   scale_y_continuous(expand = expansion(mult = c(0, 0))) + # remove gaps 
   theme(legend.position="bottom",  legend.key = element_blank(),
        axis.text.x = element_text(size= 15 , angle=30, hjust = 1  ), 
        axis.text.y = element_text(size= 20 ),
        axis.title.x = element_text(size=20),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12), 
        panel.grid.major = element_blank()
          , panel.grid.minor = element_blank()
          ,panel.background = element_blank()
          , axis.line = element_line(colour = "grey") # this will give the illusion to fill in any remaining gaps
          , panel.border=element_blank()
        
  ) + # scale_x_reordered() +
  ggforce::facet_row(vars(VEP_Worst_Consequence), scales = 'free', space = 'free') + xlab ("")



topCancerGenes = data.frame ( table ( cosmic[ cosmic$RNA_PVALUE< .05 & cosmic$is.cancer == "yes", ]$VEP_Worst_Gene)) 
topCancerGenes= topCancerGenes[ order ( -topCancerGenes$Freq ), ]




topCancerGenes <- cosmic[cosmic$VEP_Worst_Gene %in% as.character ( head ( topCancerGenes$Var1, 6) ), ]

topCancerGenesPlot <-topCancerGenes %>% group_by( VEP_Worst_Gene, VEP_Worst_Consequence, balance_simple ) %>% 
  summarise ( total = n() ) %>% 
  ggplot(.) + 
    geom_bar( 
             mapping = aes(x=VEP_Worst_Gene , 
                           y = total, fill = as.factor(balance_simple)  ), 
             stat="identity", 
             position='stack', 
             width = 1, color = "black"
             
             ) + scale_fill_manual(values = rafdaf_color_simple ) +
   scale_y_continuous(expand = expansion(mult = c(0, 0))) + # remove gaps 
   theme(legend.position="bottom",  legend.key = element_blank(),
        axis.text.x = element_text(size= 15 , angle=30, hjust = 1  ), 
        axis.text.y = element_text(size= 20 ),
        axis.title.x = element_text(size=20),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12), 
        panel.grid.major = element_blank()
          , panel.grid.minor = element_blank()
          ,panel.background = element_blank()
          , axis.line = element_line(colour = "grey") # this will give the illusion to fill in any remaining gaps
          , panel.border=element_blank()
        
  ) + # scale_x_reordered() +
  ggforce::facet_row(vars(VEP_Worst_Consequence), scales = 'free', space = 'free') + xlab ("")




topNonCancerGenes = data.frame ( table ( cosmic[ cosmic$RNA_PVALUE< .05 & cosmic$is.cancer == "no", ]$VEP_Worst_Gene)) 
topNonCancerGenes= topNonCancerGenes[ order ( -topNonCancerGenes$Freq ), ]

topNonCancerGenes <- cosmic[cosmic$VEP_Worst_Gene %in% as.character ( head ( topNonCancerGenes$Var1, 6) ), ]

topNonCancerGenesPlot <-topNonCancerGenes %>% group_by( VEP_Worst_Gene, VEP_Worst_Consequence, balance_simple ) %>% 
  summarise ( total = n() ) %>% 
  ggplot(.) + 
    geom_bar( 
             mapping = aes(x=VEP_Worst_Gene , 
                           y = total, fill = as.factor(balance_simple)  ), 
             stat="identity", 
             position='stack', 
             width = 1, color = "black"
             
             ) + scale_fill_manual(values = rafdaf_color_simple ) +
   scale_y_continuous(expand = expansion(mult = c(0, 0))) + # remove gaps 
   theme(legend.position="bottom",  legend.key = element_blank(),
        axis.text.x = element_text(size= 15 , angle=30, hjust = 1  ), 
        axis.text.y = element_text(size= 20 ),
        axis.title.x = element_text(size=20),
        axis.title.y     = element_text(size=20), 
        legend.text      =element_text(size=12), 
        panel.grid.major = element_blank()
          , panel.grid.minor = element_blank()
          ,panel.background = element_blank()
          , axis.line = element_line(colour = "grey") # this will give the illusion to fill in any remaining gaps
          , panel.border=element_blank()
        
  ) + # scale_x_reordered() +
  ggforce::facet_row(vars(VEP_Worst_Consequence), scales = 'free', space = 'free') + xlab ("")







# how does it look like with cosmo 


 
quantile ( byCancer_Gene_df$RNA_PVALUE)

# convert all to "other" if not TSG or oncgogen
byCancer_Gene_df$cosmic.annot = ifelse ( byCancer_Gene_df$cosmic.annot %in% c("TSG", "ONCOGENE"), byCancer_Gene_df$cosmic.annot, 'OTHER')

temp = byCancer_Gene_df %>% 
   group_by(VEP_Worst_Consequence ) %>% 
   mutate(n = n())  %>%
   group_by(  VEP_Worst_Consequence, cosmic.annot  ) %>% 
   summarise ( 
      percent = n() /n,
      total = n 
      ) %>% filter ( !duplicated ( VEP_Worst_Consequence, cosmic.annot) ) %>% data.frame()
# sanity check total should equal previous cancer gene yes catgories 
sum ( temp[!duplicated ( temp$VEP_Worst_Consequence), ]  $total)
# now only use vep that was used previously 

temp = temp [ temp$VEP_Worst_Consequence %in% compareTypes$VEP_Worst_Consequence, ]
sum ( temp[!duplicated ( temp$VEP_Worst_Consequence), ]  $total)





#byCancer_Gene_df = temp 
byCancer_Gene_df$label = paste0 ( round ( byCancer_Gene_df$percent * 100 , 1 ), "%" ) 



for ( f in names ( pretty_var_names) ){
   byCancer_Gene_df$VEP_Worst_Consequence = gsub ( f, as.character ( pretty_var_names[f] ) , byCancer_Gene_df$VEP_Worst_Consequence)
}

byCancer_Gene_df$VEP_Worst_Consequence = factor ( byCancer_Gene_df$VEP_Worst_Consequence, levels = levels  ( var_outcome$data$synonymous)   )   

ccc_tsgOnc = c( ONCOGENE='#bf4f3b', OTHER='#ced6db', TSG = '#619dc7')

   
vep_aa_cosmic_plot =    ggplot(byCancer_Gene_df, aes(fill=cosmic.annot, y=percent,  x=VEP_Worst_Consequence) )+ 
   geom_bar(position="stack", stat="identity", width=1, color='#e8eded') +
   # add text but position needs to be set to the same, which is filled
   geom_text(aes(label = label),
             position = position_stack(vjust = 0.5)
             , size = 8.5
   ) +  
   scale_y_continuous(expand = expansion(mult = c(0, 0))) + # remove gaps 
   theme(legend.position="bottom",  legend.key = element_blank(),
         axis.text.x = element_text(size= 20 , hjust = 1, angle=30  ), 
         axis.text.y = element_text(size= 20 ),
         axis.title.x = element_text(size=20),
         axis.title.y     = element_text(size=20), 
         legend.text      =element_text(size=12), 
         panel.grid.major = element_blank()
         , panel.grid.minor = element_blank()
         ,panel.background = element_blank()
         , axis.line = element_line(colour = "white") # this will give the illusion to fill in any remaining gaps
         , panel.border=element_blank()
         
   ) + ylab("percentage") + xlab ( "") + scale_fill_manual(values = ccc_tsgOnc )


### do some diseases have varying degrees of allelic imbalance? 
#### these are not useful. 
ddd = cosmic [ ! duplicated ( paste ( cosmic$WGS.id, cosmic$VEP_Worst_Gene ) ) 
                    #& cosmic$is.sig == "sig"  
                    & cosmic$VEP_Worst_Impact %in% c("HIGH", "MODERATE")
                    #& cosmic$is.cancer == "yes"
                    , ]
ddd$balance = as.character ( ddd$balance)
dddf = data.frame ( table ( ddd$cancer.subtype, ddd$balance))

colnames ( dddf ) = c(  "cancer.subtype","balance","Freq"  )
dddf$balance = factor ( dddf$balance, levels= c("EXTREME_HIGH", "HIGH", "LOW", "EXTREME_LOW", "no.sig")   )

dddfi = dddf[ dddf$balance != "no.sig", ]

dddf$ratio = ifelse ( dddf$balance == "no.sig", "concordant", "discordant")

by_disease = ggplot(dddf, aes(fill=ratio, y=Freq, x=cancer.subtype)) + 
    geom_bar(position="fill", stat="identity") +
    scale_fill_manual(values = c(concordant="grey", discordant = "black" )        ) +
    theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black")) +
    ylab("") + 
    xlab("") + coord_flip()  +
    theme(legend.position="bottom", legend.justification = "left"
          ,axis.text.y = element_text(size=20),
          axis.text.x = element_text(size=20),
          axis.title.x = element_text(size=22),
          axis.title.y     = element_text(size=22), 
          legend.text      =element_text(20), 
          panel.border = element_blank()
    )+ 
  scale_y_continuous( expand = c(.03, 0.033), breaks=c(0, .5, 1) ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) + theme ( legend.text      =element_text(size=35),
            legend.title = element_text(size=25) ) +
  guides( fill = guide_legend(override.aes = list(size=20) , title =""  ) ) 




by_disease_ins = ggplot(dddfi[ dddfi$balance != "no.sig", ], aes(fill=balance, y=Freq, x=cancer.subtype)) + 
    geom_bar(position="fill", stat="identity") +
    scale_fill_manual(values = rafdaf_color  , labels = c(EXTREME_HIGH="> 2", HIGH="[1-2]", LOW="[0.25-1]", EXTREME_LOW="< 0.25", no.sig="not significant")      ) +
    theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black")) +
    ylab("") + 
    xlab("") + coord_flip()  + theme ( legend.position = "bottom", legend.justification = "left") +
    theme(
          #axis.text.x = element_blank(),
          #axis.text.y = element_blank(),
          axis.title.x = element_blank(),
          axis.title.y     = element_blank(), 
          panel.border = element_blank(), legend.position = "right"
          ,axis.text.y = element_text(size=20),
          axis.text.x = element_text(size=20)
          
    )+  geom_hline(yintercept=.5, color="#868786", size=1.5) +
  scale_y_continuous( expand = c(.03, 0.033), breaks=c(0, .5, 1) ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) + theme ( legend.text      =element_text(size=15),
            legend.title = element_text(size=25) ) +
  guides( fill = guide_legend(override.aes = list(size=15) , title =""  ) ) 


by_disease + by_disease_ins + plot_layout(widths  =c ( .6,.4 ) )

###################### ************************* ###########################
## how many are tsg or oncogenes? 
down = cosmic [ grepl ( "LOW", cosmic$balance)   , ]
# how many are TSG, how many oncogene and how many are not even listed? 
down_tsg = merge ( down, genelist, by.x = "VEP_Worst_Gene", by.y= "gene" )
df_down = data.frame (  table ( down_tsg$cosmic.annot) )
df_down = df_down[ df_down$Var1 %in% c("ONCOGENE",'TSG'), ]

up = cosmic [ grepl ( "HIGH", cosmic$balance)   , ]
# how many are TSG, how many oncogene and how many are not even listed? 
up_tsg = merge ( up, genelist, by.x = "VEP_Worst_Gene", by.y= "gene" )
df_up = data.frame (  table ( up_tsg$cosmic.annot) )
df_up = df_up[ df_up$Var1 %in% c("ONCOGENE",'TSG'), ]

#################### ************************** ##########################

make.pie( df_down
                , name.first="type"
                #, cc = c("#83bf43","#cf4051", "#6ec0db", "grey")
                , leg.pos = "right"
                , title = "Lower allelic imbalance"
                )

make.pie( df_up
                , name.first="type"
                #, cc = c("#83bf43","#cf4051", "#6ec0db", "grey")
                , leg.pos = "right"
                , title = "Higher allelic imbalance"
                )


### by disease 



#########################################################################
if ( 1 > 2 ){
topsomatic = data.frame ( table ( gSNV[ !duplicated ( paste ( gSNV$sample, gSNV$VEP_Worst_Gene  ) )
                                        & gSNV$VEP_Worst_Gene %in% genelist$gene 
                                        & gSNV$VEP_Worst_Impact %in% c("HIGH", "MODERATE")
                                        , ]$VEP_Worst_Gene))
topsomatic = topsomatic[ order ( topsomatic$Freq, decreasing = T), ]

topgene_snv = cosmic [ cosmic$VEP_Worst_Gene %in% as.character ( topsomatic$Var1)[1:10],   ]
plot_top(topgene_snv)
}
#######################################






#View ( snv4 [ snv4$VEP_Worst_Gene %in% as.character ( dfreq$Var1[1:3] )  & snv4$is.sig == "sig", view_snv ]    )
View ( head ( cosmic [ , view_snv]))




### longitunal studies. 
rna_long = multi_g1$data
# remove anything that does not have "RW"
rna_long = rna_long[ rna_long$label == "R/W", ]
# remove anything without dups 
keep = as.character ( rna_long$patient.id [ duplicated ( rna_long$patient.id) ] )
rna_long = rna_long[ rna_long$patient.id %in% keep, ]
# do snv 

high_mod = list()
match_mod = list()

for ( sm in unique ( rna_long$patient.id)){

#sm = "293"
smdf = rna_long[ rna_long$patient.id == sm , ]
smdg = cosmic[ cosmic$WGS.id %in% smdf$WGS.id, ]
smdg$uid = paste ( smdg$chrom, smdg$pos, smdg$ref, smdg$VEP_Worst_Gene)

if ( length ( unique (smdg$WGS.id )) <= 1  ) {
  next
}


smdg = merge ( smdf[ , c("WGS.id", "temporal.order")],smdg, by="WGS.id" )
smdg$x = paste0 ( smdg$WGS.id, "_T=",  smdg$temporal.order )
smdg$x =gsub ( "^.*?_", "",  smdg$x)
gsub ( "^.*?_", "",  smdg$x)

# we cannot plot all 

smdg = smdg[ order ( smdg$temporal.order), ]
smdg$x =factor ( smdg$x, levels= unique ( smdg$x) )

inboth = smdg$uid [ duplicated ( smdg$uid) ]




g1 = ggplot( smdg[  smdg$VEP_Worst_Impact %in% c("HIGH", "MODERATE") , ] , aes(x = x, y = rafdaf
                          , color = balance
                          , shape = is.cancer
                        , group = uid
                   )
        ) + 
geom_jitter(size=8, position=position_jitter(w = 0.07, h = 0), alpha=.5) + 
  #geom_line() + 
  geom_path(arrow = arrow(length = unit(0.55, "cm") )  ) +
  ylab( paste0 (  "RAF/DAF" )  )+ xlab("temporal.order") +
 theme_bw() +
theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
      panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
      
      axis.text.y = element_text(size= 12 ),
      axis.text.x = element_text(angle = 0, size= 12 ),
      axis.title.x = element_text(size=18),
      
      axis.title.y     = element_text(size=18), 
      legend.text      =element_text(size=12)
) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          axis.line = element_line(colour = "black")) +
  xlab ('')  + 
  scale_color_manual(values = rafdaf_color  , labels = c(EXTREME_HIGH="> 2", HIGH="[1-2]", LOW="[0.25-1]", EXTREME_LOW="< 0.25", no.sig="not significant")      ) +
  ggtitle ( paste ( "All High/Moderate variant pid=", sm)  )

high_mod[[sm]] = g1 


g2 = ggplot( smdg[  smdg$uid %in% inboth & smdg$VEP_Worst_Impact %in% c("HIGH", "MODERATE")  , ] , aes(x = x, y = rafdaf
                          , color = balance
                          , shape = is.cancer
                        , group = uid
                   )
        ) + 
geom_jitter(size=8, position=position_jitter(w = 0.01, h = 0), alpha=.5) + 
  #geom_line() + 
  geom_path(arrow = arrow(length = unit(0.55, "cm") )  ) +
  ylab( paste0 (  "RAF/DAF" )  )+ xlab("temporal.order") +
 theme_bw() +
theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
      panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
      
      axis.text.y = element_text(size= 12 ),
      axis.text.x = element_text(angle = 0, size= 12 ),
      axis.title.x = element_text(size=18),
      
      axis.title.y     = element_text(size=18), 
      legend.text      =element_text(size=12)
) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          axis.line = element_line(colour = "black")) +
  xlab ('')  + 
  scale_color_manual(values = rafdaf_color  , labels = c(EXTREME_HIGH="> 2", HIGH="[1-2]", LOW="[0.25-1]", EXTREME_LOW="< 0.25", no.sig="not significant")      ) +
  ggtitle ( paste ( "Matched High/Moderate All variant pid=", sm )  )

match_mod[[sm]] = g2 

}


subset_hm <- high_mod[seq(from=1, to=length(high_mod), by=4)] #

for ( n in subset_hm){
#do.call("grid.arrange", c(match_mod[1:4], ncol=2))
}
```


```{r}
## make pie 


pfont = 16 

pie_distr = ar_ratio 
pie_distr$`Predicted Impact` = toupper(pie_distr$`Predicted Impact`)
head ( pie_distr )


pfont = 16 

level1 = pie_distr %>% group_by( `allelic balance (RAF/DAF)`  )  %>%
  summarise( 
    total = sum ( Freq)
    ) %>% data.frame( stringsAsFactors = F  )
  
colnames ( level1 ) = c( "Var", "Freq")  

raf1 = make.pie( level1
          , name.first="balance"
          , cc = c ( "concordant" = "#8D8D8D", "HIGH"= "#DA6383", "LOW"="#59AFED")
          , leg.pos = "bottom"
          , title = ""
          , pietext = pfont
) +theme ( legend.text      =element_text(size=pfont),
           legend.title = element_text(size=19) ) + 
  guides( fill = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=20) , title =""  ) ) +
  scale_fill_manual(values =  c( "concordant" = "#8D8D8D", "HIGH"= "#DA6383", "LOW"="#59AFED"), labels=c("concordant", "RAF>DAF", "RAF < DAF") ) 





### high level 2 

                      
level2_high_HIGH = pie_distr[pie_distr$`allelic balance (RAF/DAF)` == "HIGH" ,  ]  %>% group_by( `Predicted Impact`  )  %>%
  summarise( 
    total = sum ( Freq)
    ) %>% data.frame( stringsAsFactors = F  )


colnames ( level2_high_HIGH  ) = c( "Var1", "Freq") 


l2_raf_HIGH2 =  make.pie2(
          ftype.freq = level2_high_HIGH 
         ,name.first = "Impact" # label for legend 
         , title = "" 
         , cc = c( "HIGH/MODERATE" = '#E6B257', 'LOW/MODIFIER' ='#7F976C') 
         , leg.pos = "bottom"
         ,pietext= 10
         ,legtext = 20 
         , dir = -1 
 ) +   guides( fill = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=20) , title ="Impact" , nrow = 1  ) )  



 
 
 
 


# high level 3 

level3_high_HIGH = pie_distr[pie_distr$`Predicted Impact` == "HIGH/MODERATE" & pie_distr$`allelic balance (RAF/DAF)` == "HIGH",  ]  %>% group_by( `Cancer Gene`  )  %>%
  summarise( 
    total = sum ( Freq)
    ) %>% data.frame( stringsAsFactors = F  )

colnames ( level3_high_HIGH  ) = c( "Var1", "Freq") 
# 2.23.22
# change yes/no to is cancer not.cancer 
level3_high_HIGH$Var1 = ifelse ( level3_high_HIGH$Var1 == "yes", "is.cancer", "not.cancer")
level3_high_HIGH = level3_high_HIGH[ order ( level3_high_HIGH$Freq), ]
level3_high_HIGH$Var1 = factor ( level3_high_HIGH$Var1, levels = c("is.cancer", "not.cancer"))

l2_raf_HIGH3 = 
 make.pie2(
          ftype.freq = level3_high_HIGH 
         ,name.first = "Impact" # label for legend 
         , title = "" 
         , cc= c( "not.cancer" = '#B59863', 'is.cancer' ='#77AAB1') 
         , leg.pos = "bottom"
         ,pietext= 10
         ,legtext = 20 
         , dir = -1 
 ) + guides( fill = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=20) , title ="Cosmic cancer", nrow = 1  ) )  








################ 

level2_low_LOW = pie_distr[pie_distr$`allelic balance (RAF/DAF)` == "LOW" ,  ]  %>% group_by( `Predicted Impact`  )  %>%
  summarise( 
    total = sum ( Freq)
  ) %>% data.frame( stringsAsFactors = F  )


colnames ( level2_low_LOW  ) = c( "Var1", "Freq") 


l2_raf_LOW2 =  make.pie2(
  ftype.freq = level2_low_LOW 
  ,name.first = "Impact" # label for legend 
  , title = "" 
  , cc =  c( "HIGH/MODERATE" = '#E6B257', 'LOW/MODIFIER' ='#7F976C') 
  , leg.pos = "bottom"
  ,pietext= 10
  ,legtext = 20 
  , dir = -1 
) +   guides( fill = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=20) , title ="Impact" , nrow = 1  ) )  









# low level 3 

level3_low_LOW = pie_distr[pie_distr$`Predicted Impact` == "HIGH/MODERATE" & pie_distr$`allelic balance (RAF/DAF)` == "LOW",  ]  %>% group_by( `Cancer Gene`  )  %>%
  summarise( 
    total = sum ( Freq)
  ) %>% data.frame( stringsAsFactors = F  )

colnames ( level3_low_LOW  ) = c( "Var1", "Freq") 
# 2.23.22
# change yes/no to is cancer not.cancer 
level3_low_LOW$Var1 = ifelse ( level3_low_LOW$Var1 == "yes", "is.cancer", "not.cancer")
level3_low_LOW = level3_low_LOW[ order ( level3_low_LOW$Freq), ]
level3_low_LOW$Var1 = factor ( level3_low_LOW$Var1, levels = c("is.cancer", "not.cancer"))

l2_raf_LOW3 = 
  make.pie2(
    ftype.freq = level3_low_LOW 
    ,name.first = "Impact" # label for legend 
    , title = "" 
    , cc= c( "not.cancer" = '#B59863', 'is.cancer' ='#77AAB1') 
    , leg.pos = "bottom"
    ,pietext= 10
    ,legtext = 20 
    , dir = -1 
  ) + guides( fill = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=20) , title ="Cosmic cancer", nrow = 1  ) )  







 (raf1 ) + (l2_raf_HIGH2 / l2_raf_LOW2 )  + (   l2_raf_HIGH3 / l2_raf_LOW3    ) 


 raf1  + l2_raf_HIGH2    +  l2_raf_HIGH3    

```

```{r}

plot_out = "./pdfOUT/"
dir.create(plot_out)

  pdf(file = paste0(plot_out,  "longi_RAFdaf.pdf" ),   # The directory you want to save the file in
    width = 8, # The width of the plot in inches
    height = 6) # The height of the plot in inches
  
for ( n in unique ( as.character(  rna_long$patient.id )   ) ){
  
  print ( 
    high_mod[[n]] | match_mod[[n]]
  )
  
}

dev.off()



## plot just the delta and combine all earlier temporal orders 
delta = data.frame(stringsAsFactors = F)
for ( sm in unique ( rna_long$patient.id)){

#sm = "293"
smdf = rna_long[ rna_long$patient.id == sm , ]
smdg = cosmic[ cosmic$WGS.id %in% smdf$WGS.id, ]
smdg$uid = paste ( smdg$chrom, smdg$pos, smdg$ref, smdg$VEP_Worst_Gene)

if ( length ( unique (smdg$WGS.id )) <= 1  ) {
  next
}


smdg = merge ( smdf[ , c("WGS.id", "temporal.order")],smdg, by="WGS.id" )
smdg$x = paste0 ( smdg$WGS.id, "_T=",  smdg$temporal.order )
smdg$x =gsub ( "^.*?_", "",  smdg$x)


# check how many temporal order, the condition is that it has to be more than 2 
if ( length ( unique ( smdg$temporal.order ) ) <= 1) {
  next; 
}

# combine all earlier
tm = sort (unique ( smdg$temporal.order ) )
last = tail  ( tm, 1)
early = tm [ tm!=last]
earlygrep = paste ( early, collapse="|")

smdg$temporal.order = gsub ( earlygrep, "T0",smdg$temporal.order )
smdg$temporal.order = gsub ( last, "T1",smdg$temporal.order )
meta = smdg[ !duplicated ( smdg$uid), c("VEP_Worst_Gene", "VEP_Worst_Impact", "is.cancer","uid")] 
# now we should take the median of the raf 
smdg =  smdg  %>% dplyr::group_by ( temporal.order , uid ) %>%
  #this.left_fusion_tab2 =  this.left_fusion_tab %>% dplyr::group_by (cancer.subtype, gene) %>%
  dplyr::summarise(
    median = median ( rafdaf ) # 
  ) %>% data.frame()


inboth = smdg$uid [ duplicated ( smdg$uid) ]
smdg = smdg[smdg$uid %in% inboth, ]

if ( nrow ( smdg) == 0 ){
  next;
}

smdg =smdg[ order ( smdg$uid), ]


smdg = smdg %>% 
  group_by(uid) %>%
  #summarise(value = median[temporal.order == "T1"] / median[temporal.order == "T0"] - 1 ) %>% data.frame()
  summarise(value = median[temporal.order == "T1"] - median[temporal.order == "T0"]  ) %>% data.frame()
smdg = merge ( smdg, meta, by="uid")

smdg$sample = sm 


delta = rbind ( delta, smdg)





}



delta$VEP_Worst_Impact = gsub ( "MODERATE", "HIGH",delta$VEP_Worst_Impact  )
delta$VEP_Worst_Impact = gsub ( "LOW", "MODIFIER",delta$VEP_Worst_Impact  )


delta2 = merge ( delta, trueseq.key [ !duplicated ( trueseq.key$patient.id), c("cancer.subtype", "patient.id")], by.x="sample", by.y="patient.id")
delta2 = delta2[order ( delta2$cancer.subtype), ]
delta2$sample = factor ( delta2$sample, levels = unique ( delta2$sample) )





x1= seq ( .5, length ( unique ( delta2$sample) )  , by= 1) 

x2 = x1 + .85

minannt = min ( delta2$value) - .5 

anntcolor = as.character ( sapply ( unique ( delta2$sample) , function ( x) {
  return ( subcolor [ unique ( delta2[delta2$sample == x, ]$cancer.subtype )  ] )
}
) )


delta2$g = delta2$VEP_Worst_Impact

deltaplot_worsthigh = ggplot( delta2[ delta2$VEP_Worst_Impact == "HIGH", ] , aes(x = sample, y = value
                     , color = VEP_Worst_Impact
                     , shape = is.cancer
                     #, group = uid
)
) + 
  geom_jitter(size=4, position=position_jitter(w = 0.1, h = 0), alpha=.5) + 
  ylab( paste0 (  "allellic balance percent shift  T2-T1 " )  )+ xlab("sample") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
  theme(legend.position="bottom", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 12 ),
        axis.text.x = element_text(angle = 0, size= 12 ),
        axis.title.x = element_text(size=18),
        
        axis.title.y     = element_text(size=18), 
        legend.text      =element_text(size=12)
  ) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
            axis.line = element_line(colour = "black"))   +
  scale_color_manual(values = c(HIGH="#e3a944", MODIFIER="#708c5b")  , 
                     labels = c( HIGH="HIGH/MODERATE", MODIFIER="LOW/MODIFIER")      ) +
  guides(
    shape =  ( guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="is.cancer" , alpha = FALSE) ),
    #, size = FALSE
    #, fill = guide_legend(title.position="top", title.hjust = 0.5)
    color = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="VEP IMPACT")
    , alpha = FALSE
  ) +
  geom_hline(yintercept=0, linetype="dashed", 
             color = "grey", size=2) +
  annotate(
    xmin = x1, 
    xmax = x2,
    ymin = rep(-Inf, length ( unique ( delta2$sample))),
    ymax = rep(minannt, length ( unique ( delta2$sample))),
    geom = "rect",
    fill = anntcolor
  ) 


deltaplot_worstlow = ggplot( delta2[ delta2$VEP_Worst_Impact == "MODIFIER", ] , aes(x = sample, y = value
                     , color = VEP_Worst_Impact
                     , shape = is.cancer
                     #, group = uid
)
) + 
  geom_jitter(size=4, position=position_jitter(w = 0.1, h = 0), alpha=.5) + 
  ylab( paste0 (  "allellic balance percent shift  T2-T1 " )  )+ xlab("sample") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
  theme(legend.position="bottom", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 12 ),
        axis.text.x = element_text(angle = 0, size= 12 ),
        axis.title.x = element_text(size=18),
        
        axis.title.y     = element_text(size=18), 
        legend.text      =element_text(size=12)
  ) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
            axis.line = element_line(colour = "black"))   +
  scale_color_manual(values = c(HIGH="#e3a944", MODIFIER="#708c5b")  , 
                     labels = c( HIGH="HIGH/MODERATE", MODIFIER="LOW/MODIFIER")      ) +
  geom_hline(yintercept=0, linetype="dashed", 
             color = "grey", size=2) +
  guides(
    shape =  ( guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="is.cancer" , alpha = FALSE) ),
    #, size = FALSE
    #, fill = guide_legend(title.position="top", title.hjust = 0.5)
    color = guide_legend(title.position="top", title.hjust = 0.5, override.aes = list(size=10) , title ="VEP IMPACT")
    , alpha = FALSE
  )



deltaplot_worsthigh / deltaplot_worstlow

##########################  
snvread = readRDS( config$SNV.hugo )
snv2 = snvread$snvtotal


snv2.id = gsub( "_.*?G", "", snv2$wgs.id)
intersect ( snv2.id, trueseq.key$WGS.id)
snv2$wgs.id = snv2.id 
snv2 = merge ( snv2, trueseq.key, by.x="wgs.id", by.y= "WGS.id")
# are these dups
snv2 = snv2[ order ( snv2$level), ]
snv2$dup = paste ( snv2$wgs.id, snv2$gene)

snv2.all = snv2
snv2 = snv2[!duplicated ( snv2$dup), ]
snv2$dup = NULL 


snv2.drug = snv2 


snv = snv2 


snv2 = data.frame ( table ( snv2$RNAseq.id) )
snv2 =snv2[snv2$Freq >0, ]
colnames ( snv2) = c("RNAseq.id", "total")
snv2$cat = "SNV"


# big fix for drug here.  


exact.drug = 1




if ( exact.drug == 1){
  
  total.drug = data.frame ( stringsAsFactors = F )
  out.up.drug = c()
  pos = 1 
  
  for ( ge in rna.up$gene ){
    b2 = drug.up[drug.up$gene  == ge & drug.up$clinical_sig == "positive", ]
    b = sort ( unique ( b2$level) )
    if ( is.na(b[1])){
      b[1]=""
    }else {
      tube = as.character ( rna.up[pos:pos, ]$RNAseq.id )
      b2$RNAseq.id = tube[1]
      total.drug = rbind ( total.drug, b2 )
    }
    out.up.drug = c( out.up.drug, b[1])
    pos = pos + 1 
    
  }
  
  rna.up$level = out.up.drug
  
  #### down expression drug 
  out.down.drug = c()
  pos = 1 
  for ( ge in rna.down$gene ){
    b2 = drug.down[drug.down$gene  == ge & drug.down$clinical_sig == "positive" , ]
    b = sort ( unique ( b2$level) )
    if ( is.na(b[1])){
      b[1]=""
    }else {
      tube = as.character ( rna.down[pos:pos, ]$RNAseq.id )
      b2$RNAseq.id = tube[1]
      total.drug = rbind ( total.drug, b2 )
    }
    out.down.drug = c( out.down.drug, b[1])
    pos = pos + 1 
  }
  
  rna.down$level = out.down.drug
  
  #### CNA 
  drug.up.CNA = drug[drug$type == "CNA" & drug$alteration %in% c( "AMPLIFICATION" , "Amplification" ) & drug$clinical_sig == "positive", ]
  drug.down.CNA = drug[drug$type == "CNA" & drug$alteration %in% c( "LOSS"  ) & drug$clinical_sig == "positive", ]
  
  cna.up.drug = cna[ cna$type == "GAIN" & cna$gene %in% unique ( drug.up.CNA$gene), ]
  cna.down.drug = cna[ cna$type == "LOSS" %in% unique ( drug.down.CNA$gene) , ]
  
  
  ## do gain 
  levels.cna.drug.up = c()
  pos = 1 
  
  for ( ge in cna.up.drug$gene ){
    b2 = drug.up.CNA[drug.up.CNA$gene  == ge, ]
    b = sort ( unique ( b2$level) )
    if ( is.na(b[1])){
      b[1]=""
    }else {
      tube = as.character ( cna.up.drug[pos:pos, ]$wgs.id )
      # tube=gsub( "_.*?G", "", tube)
      tube=trueseq.key[trueseq.key$WGS.id == tube[1],   ]$RNAseq.id
      b2$RNAseq.id = tube[1]
      total.drug = rbind ( total.drug, b2 )
    }
    levels.cna.drug.up = c( levels.cna.drug.up, b[1])
    pos = pos + 1 
  }
  
  cna.up.drug$level = levels.cna.drug.up
  
  
  ## do loss 
  levels.cna.drug.down = c()
  pos =1 
  for ( ge in cna.down.drug$gene ){
    b2 = drug.down.CNA[drug.down.CNA$gene  == ge, ]
    b = sort ( unique ( b2$level) )
    if ( is.na(b[1])){
      b[1]=""
    }else {
      tube = as.character ( cna.down.drug[pos:pos, ]$wgs.id )
      # tube=gsub( "_.*?G", "", tube)
      tube=trueseq.key[trueseq.key$WGS.id == tube[1],   ]$RNAseq.id
      b2$RNAseq.id = tube[1]
      
      total.drug = rbind ( total.drug, b2 )
    }
    levels.cna.drug.down = c( levels.cna.drug.down, b[1])
    pos = pos + 1 
  }
  
  cna.down.drug$level = levels.cna.drug.down
  
  
}
```

```{r}

## create oncoprint for druggable outliers! 

snv3 = snv[ , c("RNAseq.id", "gene", "alteration")  ]
colnames ( snv3 )= c("RNAseq.id", "gene", "snv")
snv3$match = paste ( snv3$RNAseq.id, snv3$gene)

wgs.gene = unique ( snv3$gene)

# rna.up, is there SV evidence? 
sv.alt = read.table("/data/SUMMARY/WGS/SV.alterations.txt", header=F,sep="\t",stringsAsFactors = FALSE,na.strings=".",  quote = "", fill = TRUE)
colnames ( sv.alt ) = c( "wgs.id","gene","sv","type","value")

wgs.id = gsub( "_.*?G", "", sv.alt$wgs.id)
sv.alt$wgs.id = wgs.id 
intersect ( unique ( sv.alt$wgs.id), unique( trueseq.key$WGS.id))
sv.alt = merge ( sv.alt, trueseq.key[ , c("RNAseq.id", "WGS.id")] , by.x='wgs.id', by.y="WGS.id" )
sv.alt$match = paste ( sv.alt$RNAseq.id, sv.alt$gene)

wgs.gene = unique ( c( wgs.gene, sv.alt$gene) )

rna.up$uid = seq ( 1, nrow(rna.up), 1)
rna.up$match = paste ( rna.up$RNAseq.id, rna.up$gene)

rna.upWG = merge ( rna.up, sv.alt [ , c( "RNAseq.id", "match", "sv", "wgs.id" ) ], by = "match", all.x=T, all.y=F  )
rna.upWG$RNAseq.id.y = NULL 
colnames ( rna.upWG )[ grepl ( "RNAseq.id",colnames ( rna.upWG ) )] = "RNAseq.id"
rna.upWG[is.na(rna.upWG)] = ''
rna.upWG = rna.upWG[!duplicated ( rna.upWG$uid), ]
colnames ( rna.upWG )[ grepl ( "type",colnames ( rna.upWG ) )] = "out.type"

cna.rna$match = paste ( cna.rna$RNAseq.id, cna.rna$gene)
rna.upWG = merge ( rna.upWG, cna.rna[ cna.rna$type %in% c("GAIN", "GAIN1"), c("match", "type")], by = "match", all.x=T, all.y=F  )
colnames ( rna.upWG )[ colnames ( rna.upWG ) == "type"] = "cna.type"
rna.upWG[is.na(rna.upWG)] = ''

# merge SNV now 
rna.upWG = merge ( rna.upWG, snv3[ , c("match", "snv")], by = "match", all.x=T, all.y=F  )
rna.upWG[is.na(rna.upWG)] = ''
rna.upWG = rna.upWG[!duplicated ( rna.upWG$uid), ]

wgs.gene = unique ( c( wgs.gene, cna.rna$gene) )


########### do down now 

rna.down$uid = seq ( 1, nrow(rna.down), 1)
rna.down$match = paste ( rna.down$RNAseq.id, rna.down$gene)

rna.downWG = merge ( rna.down, sv.alt [ , c( "RNAseq.id", "match", "sv", "wgs.id" ) ], by = "match", all.x=T, all.y=F  )
rna.downWG$RNAseq.id.y = NULL 
colnames ( rna.downWG )[ grepl ( "RNAseq.id",colnames ( rna.downWG ) )] = "RNAseq.id"
rna.downWG[is.na(rna.downWG)] = ''
rna.downWG = rna.downWG[!duplicated ( rna.downWG$uid), ]
colnames ( rna.downWG )[ grepl ( "type",colnames ( rna.downWG ) )] = "out.type"


rna.downWG = merge ( rna.downWG, cna.rna[ cna.rna$type %in% c("LOSS"), c("match", "type")], by = "match", all.x=T, all.y=F  )
colnames ( rna.downWG )[ colnames ( rna.downWG ) == "type"] = "cna.type"
rna.downWG[is.na(rna.downWG)] = ''

# merge SNV now 
rna.downWG = merge ( rna.downWG, snv3[ , c("match", "snv")], by = "match", all.x=T, all.y=F  )
rna.downWG[is.na(rna.downWG)] = ''
rna.downWG = rna.downWG[!duplicated ( rna.downWG$uid), ]


out.up.down.wgs = rbind ( rna.upWG, rna.downWG)
out.up.down.wgs$match = NULL 
# don't care about what type of evidence so long as there is. 
out.up.down.wgs$wgs.evidence = ifelse ( (out.up.down.wgs$sv != "" | out.up.down.wgs$snv != "" | out.up.down.wgs$cna.type != "" ), 
                                        'WGS.evidence', ''
  
  )

out.up.down.wgs$drug = ifelse ( out.up.down.wgs$level != "" , 
                                        'druggable', ''
                                        
)


out.up.down.wgs$value = paste (out.up.down.wgs$both, out.up.down.wgs$wgs.evidence, sep=";" )
out.up.down.wgs$value = gsub ( ";+", ";", out.up.down.wgs$value )
out.up.down.wgs$value = gsub ( ";$", "", out.up.down.wgs$value )

out.up.down.wgs = out.up.down.wgs[!out.up.down.wgs$RNAseq.id %in% failedChem,   ]

setdiff( out.up.down.wgs$gene,  wgs.gene)

# now lets reorder column and row by highest freq

left_rnk = out.up.down.wgs[out.up.down.wgs$level != "", ]  %>% 
  group_by( gene )  %>%
  dplyr::summarise  (
    Freq = n()
  )  %>% arrange ( desc ( Freq, left) ) %>% data.frame ( stringsAsFactors = F) %>% .$gene 



use.this.gene = left_rnk[1:50]
use.this.gene = use.this.gene[!is.na(use.this.gene)]
#use.this.gene = c("VEGFA","STK11", "SMARCB1")
outm  =   out.up.down.wgs[ out.up.down.wgs$gene %in% use.this.gene, ]

outm = out.up.down.wgs[ out.up.down.wgs$gene %in% as.character (  out_drug_grid$data$gene  ), ]              
outm =outm[outm$value == "OUTLIER_HIGH", ]


z.up = outm %>%
  group_by(RNAseq.id,gene) %>%
  summarise(both=unique ( paste0(unique ( value),collapse = ';'))) %>%
  pivot_wider(names_from = RNAseq.id,values_from=both) %>% data.frame(stringsAsFactors = F)

z.up[is.na(z.up)] <- ""

z.up[] <- lapply(z.up, function(x) as.character(gsub(";$", "", x)))
z.up[] <- lapply(z.up, function(x) as.character(gsub(";;", ";", x)))

z.up.name <- z.up$gene # grab the gene first 
z.up$gene <- NULL
z.up <- as.matrix (z.up)
row.names (z.up) <- z.up.name
z.up[ is.na(z.up) ] <- ""

z.up = z.up [ ]
#col_rnk = intersect (unique ( as.character ( fusion.total$RNAseq.id ) ), colnames ( z.up ) ) 
#z.up = z.up[use.this.gene, unique (col_rnk  )]

set.seed(3)
col = sample ( getPalette( 50 ))
names ( col ) = c(1:50)

unique ( as.character ( as.matrix ( z.up ) ))


af = list(
  background = function(x, y, w, h) 
    grid.rect(x, y, w*1, h*1, gp = gpar(fill = "#f2f2f2", col = NA)),
  
  "is.obsv" = function(x, y, w, h) 
    grid.rect(x, y, w*0.90, h*0.90, gp = gpar(fill = "#80B1D3", col = NA)),
  "is.diag" = function(x, y, w, h) 
    grid.rect(x, y, w*0.90, h*0.90, gp = gpar(fill = "#537cbd", col = NA)),
  "OUTLIER_HIGH" = function(x, y, w, h) 
    grid.rect(x, y, w*0.90, h*0.90, gp = gpar(fill = "#537cbd", col = NA)),
  "OUTLIER_LOW" = function(x, y, w, h) 
    grid.rect(x, y, w*0.90, h*0.90, gp = gpar(fill = "#5B734B", col = NA)),
  
  
  # dots
  WGS.evidence = function(x, y, w, h) 
    grid.points(x, y, pch = 16, size = unit(2.5, "mm") ),
  # crossed lines
  druggable = function(x, y, w, h) 
    grid.segments(x - w*0.5, y - h*0.5, x + w*0.5, y + h*0.5, gp = gpar(lwd = 2))
  ,HOLD = function(x, y, w, h) {
    grid.segments(x - w*0.4, y - h*0.4, x + w*0.4, y + h*0.4, gp = gpar(lwd = 2))
    grid.segments(x + w*0.4, y - h*0.4, x - w*0.4, y + h*0.4, gp = gpar(lwd = 2))
  }
)
alter_fun=af


# top annotations 
## we want a stack plot with type 
# input a table with columns matching the oncoprint, in this case genes but each row is an attribute you wanto sum up. 
# the method is to group by the column that is also the column for the oncoprint + the attribute you want to tally and they expand 
# to a wide table.  

onco.key = data.frame ( RNAseq.id = colnames ( z.up), stringsAsFactors = F )
onco.key$cancer.subtype = as.character ( sapply ( onco.key$RNAseq.id , function(x) {
  trueseq.key[trueseq.key$RNAseq.id == x, ]$cancer.subtype
} )
  )


onco.key$to = as.character ( sapply ( onco.key$RNAseq.id , function(x) {
   df = data.frame ( table ( z.up [, x] ) )
   return ( as.numeric ( df[df$Var1 != "", ]$Freq) )
} )
  )




onco.key = onco.key[ order ( match(onco.key$cancer.subtype,c(  unique ( out_drug_grid$data$cancer.subtype ) ) ), onco.key$to ), ]



z.up = z.up [ , onco.key$RNAseq.id]




# rename columb groups 
rename_cs = unique ( onco.key$cancer.subtype )
#rename_cs[rename_cs=="OTHER_HEME"] = "OH"

############ drug annoation
druggable = as.character ( sapply ( row.names ( z.up ) , function(x) {
  x1 = unique (  out.up.down.wgs[out.up.down.wgs$gene == x, ]$level )
  if ( any ( grepl ( "L", x1) )  ){
    x1 = x1[x1!=""]
    x1 = sort ( x1 )
    return ( x1 )
  } else {
    return ("no.drug")
  }
})
)

bcc = c( "#990000", "#8DB8E0", "#EBCF46", "#c7c9c9") 
names ( bcc ) = c( "L1","L2", "L3", "no.drug" )
drugannt = rowAnnotation(
  Drug = druggable
  
  ,col = list ( 
    Drug= bcc
  ) , annotation_name_gp= gpar(fontsize = 0, col="white" )
, show_legend =  F )


# get top barplot 
df = data.frame ()
for ( x in 1:ncol(z.up)){
  x1 = z.up[ , x]
  x1 = names ( x1 [ grepl ( "OUT", x1)] )
  df = rbind ( 
    df, 
    out.up.down.wgs[out.up.down.wgs$gene %in% x1 & out.up.down.wgs$RNAseq.id == colnames ( z.up )[x], c("RNAseq.id","level")] 
  )
  
}

#df = df[df$level != "", ]

df2  =  df %>% group_by(RNAseq.id, level) %>% 
  summarise(Freq=n()) %>%
  pivot_wider(names_from = RNAseq.id,values_from=Freq) %>%  mutate_each(funs(replace(., is.na(.), 0))) %>% data.frame(stringsAsFactors = F)

row.names ( df2) = df2[,1]
df2 = df2[,colnames(z.up)]
df2 = df2[-4, ]



topannt2 = HeatmapAnnotation(
  total = anno_barplot(t ( df2) 
                       #, width = unit(3, "cm")
                       , gp = gpar(fill = bcc)
                       , axis_param = list (gp=gpar(fontsize=10), at=c(2,4)  ), bar_width = 1.0 ) 
  # little trick to mask out the annotation. 
  , annotation_name_gp= gpar(fontsize = 0, col="white" ) , show_legend =  F )



# resort row order? 
rorder = c ( sapply ( row.names ( z.up ) , function(x) {
  x1 = unique (  out.up.down.wgs[out.up.down.wgs$gene == x, ]$level )
  if ( any ( grepl ( "L", x1) )  ){
    x1 = x1[x1!=""]
    x1 = sort ( x1 )
    return ( x1 )
  } else {
    return ("no.drug")
  }
})
)

rorder = sort ( rorder)

oncoOUT = oncoPrint(  z.up,
                    alter_fun = af
                    
                    ,show_heatmap_legend = F
            , column_split =  factor ( onco.key$cancer.subtype, levels=unique ( onco.key$cancer.subtype ))
                    , column_order =onco.key$RNAseq.id
            , column_title_gp = gpar( 
              fill = unique(subcolor[ onco.key$cancer.subtype]  ), 
              fontsize = c( rep ( 10, 2), 8,rep ( 10, 10)    ))
            , column_title =rename_cs 
            , left_annotation = drugannt
            ,show_pct = F # pct on the left 
        
            , top_annotation = topannt2 
            , row_order = names (rorder )
            
                    
) 
           

oncoOUT_drugOutlier = oncoOUT 



wg.legend = Legend(labels = c( "wgs.evidence"), type = "points", pch = c(20),  labels_gp = gpar(fontsize = 12)
                ,grid_height = unit(8, "mm") )

pd2 = packLegend(wg.legend , row_gap = unit(1, "cm") , direction="horizontal")

draw(oncoOUT,annotation_legend_list = pd2)



```

### Oncoprint for Druggable Outliers 

```{r fig=TRUE,fig.width=22, fig.height=18.5, echo=FALSE, include=TRUE , results='asis'}

draw(oncoOUT_drugOutlier)
```


```{r}

# get top outlier by disease 

outtemp = merge ( trueseq.key [ , c ( "RNAseq.id", "cancer.subtype") ],  out.up.down.wgs [ , c("both", "gene", "RNAseq.id")] , by="RNAseq.id")
outtemp2 = outtemp %>% 
  group_by(gene, cancer.subtype )  %>%
  dplyr::summarise  (
    Freq = n()
)  %>% arrange ( desc ( Freq) ) %>% data.frame ( stringsAsFactors = F)

topout = data.frame ( stringsAsFactors = F )

for ( c in  unique ( outtemp2$cancer.subtype) ){
  top = head ( outtemp2[outtemp2$cancer.subtype == c , ]$gene, 5 ) 
  top = outtemp[ outtemp$gene %in% top & outtemp$cancer.subtype == c, ] %>% 
   group_by(gene, both )  %>%
  dplyr::summarise  (
    Freq = n(),
    cancer.subtype = c
)  %>% arrange ( desc ( Freq) ) %>% data.frame ( stringsAsFactors = F)
  
  topout = rbind ( topout, top )
  
}

outcolor = c("OUTLIER_HIGH"= "#FB8072",  "OUTLIER_LOW"  = "#5B734B" )
topout = topout[ order ( topout$cancer.subtype, topout$Freq), ]  

#topout$gene = factor ( topout$gene, levels = unique ( topout$gene))
topout$cancer.subtype = factor ( topout$cancer.subtype, levels=cancer.subtype.order )

topout <- topout %>% 
  # 1. Remove grouping
  ungroup() %>% 
  arrange(cancer.subtype, Freq) %>%
  # 3. Add order column of row numbers
  mutate(order = row_number())





p2 = ggplot(topout, aes(x =  order, y=Freq,  fill = both) ) + 
 scale_fill_manual(values =  outcolor  ) +
  geom_bar(stat = 'identity', position = 'stack') +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"),
        legend.position = "bottom"
        ) +
  ylab("") + 
  xlab("") +
  #scale_y_continuous(breaks= max(fuseout2$Freq)) + 
  coord_flip() + facet_wrap(cancer.subtype~., scales = "free", ncol=2 ) +
  scale_x_continuous(
    breaks = topout$order,
    labels = topout$gene) 



if ( 1 > 2){
g <- ggplot_gtable(ggplot_build(p2))

strips <- which(grepl('strip', g$layout$name))

pal <- subcolor[ cancer.subtype.order]

for (i in seq_along(strips)) {
  k <- which(grepl('text', g$grobs[[strips[i]]]$grobs[[1]]$childrenOrder))
  l <- which(grepl('text', g$grobs[[strips[i]]]$grobs[[1]]$childrenOrder))
  g$grobs[[strips[i]]]$grobs[[1]]$children[[k]]$gp$fill <- pal[i]
  g$grobs[[strips[i]]]$grobs[[1]]$children[[l]]$children[[1]]$gp$col <- 'black'
}

plot ( g )
}

```


### Most Recurrent Outliers by disease

```{r fig=TRUE,fig.width=12, fig.height=15.5, echo=FALSE, include=TRUE , results='asis'}

p2
```


```{r}

if ( 1 > 2){

t1_hold = cna.up.drug 
t2_hold = cna.down.drug
t3_hold = best.out.up.drug 
t4_hold = best.out.down.drug
}
# rna find level by sample 

best.out.up.drug=rna.up[rna.up$level != "",  ]
best.out.down.drug=rna.down[rna.down$level != "",  ]

### WGS 

## WGS continue 
# we should not need for trueseq key here since CNA is not dependent on CNA 
 
cna.up.drug = merge ( cna.up.drug, rna.key[ , c("RNAseq.id","cancer.subtype","WGS.id")], by.x="wgs.id", by.y="WGS.id")
cna.down.drug = merge ( cna.down.drug, rna.key[ , c("RNAseq.id","cancer.subtype","WGS.id")], by.x="wgs.id", by.y="WGS.id")

if ( nrow ( cna.up.drug ) > 0 ){
  cna.up.drug$match = paste ( cna.up.drug$wgs.id, cna.up.drug$gene)
  cna.up.drug = cna.up.drug[!duplicated ( cna.up.drug$match), ]
  cna.up.drug$match=NULL 
}

if ( nrow ( cna.down.drug ) > 0 ){
cna.down.drug$match = paste ( cna.down.drug$wgs.id, cna.down.drug$gene)
  cna.down.drug = cna.down.drug[!duplicated ( cna.down.drug$match), ]
  cna.down.drug$match=NULL 
  }


# 
best.cna.up.drug=cna.up.drug
best.cna.down.drug=cna.down.drug

# match it up 

out.up.drug = best.out.up.drug
out.down.drug = best.out.down.drug

out.up.drug$match = paste ( out.up.drug$gene, out.up.drug$RNAseq.id)
out.down.drug$match = paste ( out.down.drug$gene, out.down.drug$RNAseq.id)

cna.up.drug$match = paste ( cna.up.drug$gene, cna.up.drug$RNAseq.id)
cna.down.drug$match = paste ( cna.down.drug$gene, cna.down.drug$RNAseq.id)

# match both 
int.drug.up = intersect (out.up.drug$match,  cna.up.drug$match )
int.drug.down = intersect (out.down.drug$match,  cna.down.drug$match )

int.drug.up = out.up.drug[out.up.drug$match %in% int.drug.up, ]
int.drug.down = out.down.drug[out.down.drug$match %in% int.drug.down, ]



# remove from rna and wgs 
out.up.drug = out.up.drug[!out.up.drug$match %in% int.drug.up$match, ]
cna.up.drug = cna.up.drug[!cna.up.drug$match %in% int.drug.up$match, ]

out.down.drug = out.down.drug[!out.down.drug$match %in% int.drug.down$match, ]
cna.down.drug = cna.down.drug[!cna.down.drug$match %in% int.drug.down$match, ]

# tables 

out.up.drug.freq = data.frame ( table ( out.up.drug$RNAseq.id) )
colnames ( out.up.drug.freq) = c("RNAseq.id", "total")
out.up.drug.freq$cat = "RNA-UP"

out.down.drug.freq = data.frame ( table ( out.down.drug$RNAseq.id) )
colnames ( out.down.drug.freq) = c("RNAseq.id", "total")
out.down.drug.freq$cat = "RNA-DOWN"


cna.up.drug.freq = data.frame ( table ( cna.up.drug$RNAseq.id) )
colnames ( cna.up.drug.freq) = c("RNAseq.id", "total")
cna.up.drug.freq$cat = "WGS-GAIN"

cna.down.drug.freq = data.frame ( table ( cna.down.drug$RNAseq.id) )
if ( nrow ( cna.down.drug.freq) > 0 ){
  colnames ( cna.down.drug.freq) = c("RNAseq.id", "total")
  cna.down.drug.freq$cat = "WGS-LOSS"
}


## both 
int.drug.up.tab = data.frame  (table ( int.drug.up$RNAseq.id))
colnames ( int.drug.up.tab) = c("RNAseq.id", "total")
int.drug.up.tab$cat = "RNA-DNA-UP"

if ( nrow ( int.drug.down ) > 0 ){
  int.drug.down.tab = data.frame  (table ( int.drug.down$RNAseq.id))
  colnames ( int.drug.down.tab) = c("RNAseq.id", "total")
  int.drug.down.tab$cat = "RNA-DNA-DOWN"
}



drug.final = rbind (out.up.drug.freq, out.down.drug.freq )
drug.final = rbind ( drug.final,cna.up.drug.freq )
drug.final = rbind ( drug.final,cna.down.drug.freq )
drug.final = rbind ( drug.final,int.drug.up.tab )
if ( nrow ( int.drug.down ) > 0 ){
  drug.final = rbind ( drug.final,int.drug.down.tab )
}

drug.final = rbind ( drug.final,drug.fuse )
drug.final = rbind ( drug.final, snv2)

drug.final = drug.final[drug.final$total != 0, ]
zero = setdiff(trueseq.key$RNAseq.id, unique ( drug.final$RNAseq.id) )
for ( z in zero ){
  drug.final = rbind ( drug.final, data.frame ( RNAseq.id=z,total=1,cat="no-action") )     
}


drug.final = merge ( drug.final, rna.key[ , c("RNAseq.id", "cancer.subtype")], by = "RNAseq.id")
drug.final$RNAseq.id = factor ( drug.final$RNAseq.id, levels= fusion.total$RNAseq.id  )
drug.final$cancer.subtype = factor ( drug.final$cancer.subtype, levels = cancer.subtype.order)

cc2 = brewer.pal(7, "Set3")
cc2 = c( cc2, "#d742f5","#999494")

drug.final$action = factor ( drug.final$cat, levels = c(  "RNA-UP", "RNA-DOWN" , "RNA-DNA-UP", 
                                                           "SNV", "WGS-GAIN",
                                                          "RNA-DNA-DOWN" ,"fusion"  ,  "WGS-LOSS"  , "no-action"  
))

names ( cc2 ) = c(  
  "fusion" , "RNA-UP" , "RNA-DOWN", "WGS-GAIN", "WGS-LOSS"  , "RNA-DNA-UP" ,  "RNA-DNA-DOWN"  , "SNV", "no-action"  
)
cc2["RNA-DNA-UP"] = "#80B1D3"
cc2["SNV"] = "#FDB462"


stack_drug = ggplot(drug.final, aes(fill=action,  x=RNAseq.id , y=total)) + 
  geom_bar(position="fill", stat="identity", width = 1) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("Percent Actionable") + 
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        , axis.ticks.y=element_blank()
        , axis.text.y=element_blank()
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
  scale_y_continuous( expand = c(0, 0)) +
  scale_fill_manual(values=cc2 )+
  facet_grid(~cancer.subtype, scales = "free" , space = "free" )


# fun with grid 



pal <- subcolor[ cancer.subtype.order]

g <- ggplot_gtable(ggplot_build(stack_drug))

strips <- which(grepl('strip-', g$layout$name))

for (i in seq_along(strips)) {
  k <- which(grepl('rect', g$grobs[[strips[i]]]$grobs[[1]]$childrenOrder))
  l <- which(grepl('titleGrob', g$grobs[[strips[i]]]$grobs[[1]]$childrenOrder))
  g$grobs[[strips[i]]]$grobs[[1]]$children[[k]]$gp$fill <- pal[i]
  g$grobs[[strips[i]]]$grobs[[1]]$children[[l]]$children[[1]]$gp$col <- pal[i]
}

plot(g)




# create pie chart here. 
pie.drug= drug.final %>%
  group_by(action) %>%
  summarise(count = sum(total) ) %>%
  mutate( prop = count / sum(count) ) %>% data.frame()

pie.drug$action = factor ( pie.drug$action, levels = c(  
  levels(drug.final$action )    
))



# this part is necessary to place the label rotationally correctly 

mpp <- pie.drug %>% 
  # group_by(type )  %>%
  mutate(
    cs = rev(cumsum(rev(prop))),
    text_yp = prop/2 + lead(cs, 1),
    text_yp = if_else(is.na(text_yp), prop/2, text_yp)
  )  %>%  data.frame()




pie.drug = ggplot(mpp, aes(x="", y=prop, fill=action )) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0)   + 
  theme_void()   +
  geom_label_repel(
    aes(
      label = count, y = text_yp
    ), show_guide  = F
    
  ) + xlab("") + ylab("") +
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_rect(fill="lightblue")
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
  theme(
    strip.background = element_blank(),
    strip.text = element_blank()
  ) + theme_void()  +
  scale_fill_manual(values = cc2)











```





```{r}
# find "best of" 

#best.out.up.drug = merge ( best.out.up.drug , drug.up , by.x="gene", by.y="gene1" )
#best.out.down.drug = merge ( best.out.down.drug , drug.down , by.x="gene", by.y="gene1" )
#best.fusion
#best.cna.up.drug = merge ( best.cna.up.drug , drug.up , by.x="gene", by.y="gene1" )
#best.cna.down.drug  =   merge ( best.cna.down.drug , drug.down , by.x="gene", by.y="gene1" )



best.out.up.drug$module = "OUTLIER.up"
best.out.down.drug$module =  "OUTLIER.down"
best.cna.up.drug$module = "CNA.gain"
if ( nrow (best.cna.down.drug) ){
  best.cna.down.drug$module = "CNA.loss"
}

best.drug = rbind ( best.out.up.drug , best.out.down.drug ) 

best.drug = rbind ( best.drug [ , c("gene","RNAseq.id", "module", "level")]
                    , best.cna.up.drug [ , c("gene","RNAseq.id", "module", "level" )]
)



best.fusion$module = "FUSION"
colnames ( best.fusion )[1] = "gene"

# fixing AE1 will come back some other time 
#temp = best.fusion[1:1, ]
#temp$RNAseq.id = "AE1"

#best.drug = rbind ( best.drug
 #                   , rbind ( best.fusion [ , c("gene","RNAseq.id", "module",  "level") ], temp[ , c("gene","RNAseq.id", "module",  "level") ] ) 
#)

best.drug = rbind ( best.drug, best.fusion [ , c("gene","RNAseq.id", "module",  "level") ]  )

## add snv 

snv2.drug$module = "SNV"
snv2.drug$gene = snv2.drug$gene


best.drug = rbind ( best.drug
                    , snv2.drug [ , unique ( names ( best.drug) )]
)

## calculate zero! 

zero = setdiff(rna.key$RNAseq.id, unique ( best.drug$RNAseq.id) )
for ( z in zero ){
  best.drug = rbind ( best.drug, data.frame ( gene = "NONE", RNAseq.id=z, module="NONE",level="no.drug") )     
}


# plot pie-chart 

best.drug.cat = as.character ( best.drug$level )


#missing = length ( setdiff ( fusion.type$RNAseq.id, best.drug$RNAseq.id) )
#best.drug.cat = c( best.drug.cat, rep ( "no.drug", missing ))


drug.cat.freq = data.frame ( table ( best.drug.cat), stringsAsFactors = F)
# continue pie chart below because I want to match the color 


zero = as.character ( unique ( drug.final$RNAseq.id) )
bestof.final = data.frame ( stringsAsFactors = F)

for ( z in zero ){
  zid = z
  z = best.drug [ best.drug$RNAseq.id == z, ]
  bestof = "no.drug"
  if ( nrow (z) == 0 ) {
    bestof = "no.drug" 
  } else {
    b = sort ( unique ( z$level) )
    bestof = b[1]
  }
  # what module? 
  best.module = "no.module"
  if ( nrow ( z) > 0 ){
    best.module = paste ( unique ( z[z$level == bestof, ]$module ) , collapse = "," )
    
  }
  
  bestof.final = rbind ( bestof.final , data.frame ( RNAseq.id = zid, best.drug.cat = bestof, module= best.module, stringsAsFactors = F ) )
  
}

bestof.final = merge ( bestof.final, rna.key[ , c("RNAseq.id","cancer.subtype")])
setdiff ( rna.key$RNAseq.id, unique ( bestof.final$RNAseq.id))

bestof.final$RNAseq.id = factor ( bestof.final$RNAseq.id, levels= levels ( drug.final$RNAseq.id)  )
bestof.final$cancer.subtype = factor ( bestof.final$cancer.subtype, levels = cancer.subtype.order)


unique ( bestof.final$best.drug.cat ) 

bestof.final$best.drug.cat = factor ( bestof.final$best.drug.cat , levels= c( "L1","L2", "L3", "no.drug"))
setdiff(bestof.final$RNAseq.id,   drug.final$RNAseq.id )
setdiff(drug.final$RNAseq.id  , bestof.final$RNAseq.id )


bestof.plot = ggplot(bestof.final, aes(fill=best.drug.cat,  x=RNAseq.id , y=1)) + 
  geom_bar(position="fill", stat="identity", width = 1) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        , axis.ticks.y=element_blank()
        , axis.text.y=element_blank()
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
  scale_y_continuous( expand = c(0, 0)) +
  scale_fill_manual(values=bcc )+
  facet_grid(~cancer.subtype, scales = "free" , space = "free" ) 


# continue pie chart 
#drug.cat.freq$drug.cat.freq = factor ( drug.cat.freq$best.drug.cat , levels= c( "L1","L2", "L3", "no.drug"))





## now create bestof unique class pie chart the above is mainly for any type while this pie below is unique, the best 
bestofunique.freq = data.frame ( table ( as.character ( bestof.final$best.drug.cat) ), stringsAsFactors = F)
bestofunique.freq = bestofunique.freq[ order ( match(bestofunique.freq$Var1,levels( bestof.final$best.drug.cat ) ) ), ]



bestofunique.freq$Var1 = factor ( bestofunique.freq$Var1, levels = levels( bestof.final$best.drug.cat ))



mpp = bestofunique.freq %>%
  mutate(countT= sum(Freq)) %>%
  mutate(per=round(100*Freq/countT,2))%>% data.frame()




# this part is necessary to place the label rotationally correctly 

mpp <- mpp %>% 
  # group_by(type )  %>%
  mutate(
    cs = rev(cumsum(rev(per))),
    text_yp = per/2+ lead(cs, 1),
    text_yp = if_else(is.na(text_yp), per/2, text_yp)
  )  %>%  data.frame()


mpp$Freq = as.character(mpp$Freq )
mpp$Freq = factor ( mpp$Freq, levels = c( mpp$Freq ))

mpp$best.in.class = as.character(bestofunique.freq$Var1 )
mpp$best.in.class = factor ( mpp$best.in.class , levels = levels( bestof.final$best.drug.cat ) )




pie.bestofunique = ggplot(mpp, aes(x="", y=per, fill=best.in.class )) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0)   + 
  theme_void()   +
  geom_label_repel(
    aes(
      label = Freq, y = text_yp
    ), show_guide  = F,
    color="black",
    size = 5,
    point.padding = unit(12.25, "lines"),
    #  box.padding = unit(.25, "lines"),
    nudge_x = .5,
    show.legend = F,
    #segment.color = cc5
    segment.alpha = 0
    
  ) + xlab("") + ylab("") +
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_rect(fill="lightblue")
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
  theme(
    strip.background = element_blank(),
    strip.text = element_blank()
  ) + theme_void()  +
  scale_fill_manual(values = bcc)








```


## Drug Summary {.tabset}

### Best off 

```{r fig=TRUE,fig.width=12, fig.height=15.5, echo=FALSE, include=TRUE , results='asis'}


  (stack_drug + theme(legend.position = "none" )  + ( pie.drug  )  + plot_layout(widths =c ( 1,.2 ) ) )  /
  ( bestof.plot + theme ( legend.position = "bottom",  strip.background = element_blank(),
                          strip.text.x = element_blank() )      + ( pie.bestofunique + theme ( legend.position = "none")  )  + plot_layout(widths =c ( 1,.2 )   ) ) + 
  plot_layout(heights =c ( 1,.2) )



```

```{r}
# redo best of drug plot 


out.up.drug.level =  out.up.drug [ , c("RNAseq.id","level")]
out.up.drug.level$cat = "RNA-UP"

out.down.drug.level = out.down.drug [ , c("RNAseq.id","level")]
out.down.drug.level$cat = "RNA-DOWN"


cna.up.drug.level = cna.up.drug[ , c("RNAseq.id","level")]
cna.up.drug.level$cat = "WGS-GAIN"


if ( nrow ( cna.down.drug)   == 0 ){
  cna.down.drug.level =  cna.down.drug[ , 1:3]
  colnames ( cna.down.drug.level) = c("RNAseq.id", "level", "WGS-LOSS")
  
}else{
  cna.down.drug.level = cna.down.drug[ , c("RNAseq.id","level")]
  cna.down.drug.level$cat = "WGS-LOSS"
}


snv2.drug.level = snv2.drug [ , c("RNAseq.id","level")]
snv2.drug.level$cat = "SNV"



## both 
int.drug.up.tab = int.drug.up[ , c("RNAseq.id","level")]
int.drug.up.tab$cat = "RNA-DNA-UP"

best.fusion.level = best.fusion[ , c("RNAseq.id","level")]
best.fusion.level = best.fusion.level[ order ( best.fusion.level$level), ]
best.fusion.level = best.fusion.level[!duplicated ( best.fusion.level$RNAseq.id), ]
best.fusion.level$cat = "fusion"

drug.levels = rbind (out.up.drug.level, out.down.drug.level )
drug.levels = rbind ( drug.levels,cna.up.drug.level )
drug.levels = rbind ( drug.levels,cna.down.drug.level )
drug.levels = rbind ( drug.levels,int.drug.up.tab )
drug.levels = rbind ( drug.levels,best.fusion.level )
drug.levels = rbind ( drug.levels,snv2.drug.level )
drug.levels= merge ( drug.levels, rna.key[, c("RNAseq.id", "cancer.subtype")], by="RNAseq.id")

drug.levels.comb = drug.levels

drug.levels = drug.levels %>% dplyr::group_by ( cat, level) %>%
  dplyr::summarise(
    total = n(   ) 
    #,cancer.subtype = unique ( cancer.subtype)
  ) %>% data.frame()

drug.levels = rbind ( drug.levels, c("no-action", "L3", 0 ))


drug.levels$cat = factor ( drug.levels$cat, levels = c("fusion", "RNA-UP" , 'RNA-DOWN',  "RNA-DNA-UP", "WGS-GAIN",'SNV', 'no-action'    ))

drug.levels$total = as.numeric ( drug.levels$total )
stack.level.plot = ggplot(drug.levels, aes(fill=level,  x=cat , y=total)) + 
  geom_bar(position="stack", stat="identity") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("Clinical Level") + 
  xlab("") + #ylim(1,71) + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        , axis.ticks.y=element_blank()
        , axis.text.y=element_blank()
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
  #scale_y_continuous( expand = c(0, 0)) +
  scale_fill_manual(values=bcc )

# sum by cancer.subtype 


comb.drug.final = drug.final 
comb.drug.final$cat = gsub ( "RNA-UP", "RNA", comb.drug.final$cat)
comb.drug.final$cat = gsub ( "RNA-DOWN", "RNA", comb.drug.final$cat)
comb.drug.final$cat = gsub ( "RNA-DNA-UP", "CNA", comb.drug.final$cat)
comb.drug.final$cat = gsub ( "RNA-DNA-DOWN", "CNA", comb.drug.final$cat)
comb.drug.final$cat = gsub ( "WGS-GAIN", "CNA", comb.drug.final$cat)


drug.final2 = drug.final %>% dplyr::group_by (cancer.subtype, cat) %>%
  dplyr::summarise(
    total = sum ( total  ) 
    #,cancer.subtype = unique ( cancer.subtype)
  ) %>% data.frame()


drug.final2$cancer.subtype = factor ( drug.final2$cancer.subtype, levels = cancer.subtype.order)
drug.final2$cat = factor ( drug.final2$cat, levels = c("fusion", "RNA-UP" , 'RNA-DOWN',  "RNA-DNA-UP", "WGS-GAIN",'SNV', 'no-action'    ))



totaldisease = data.frame ( table ( rna.key$cancer.subtype ) )
drug.final3 =  merge ( drug.final2, totaldisease, by.x="cancer.subtype", by.y="Var1")
drug.final3$percent = round ( (drug.final3$total / drug.final3$Freq  ), 2) 

drug.grid = ggplot(drug.final3, aes(x=cat, y=cancer.subtype, fill = percent, label = total)) +
  geom_tile( col = 'grey') +
  geom_text(col = "black") +
  theme_minimal() + xlab("Modality") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  scale_fill_gradient2(low = "white", mid = "yellow", high = "red")


stack.level.plot / drug.grid



#### redo but combine 

comb.drug.final = drug.final 
comb.drug.final$cat = gsub ( "RNA-UP", "RNA", comb.drug.final$cat)
comb.drug.final$cat = gsub ( "RNA-DOWN", "RNA", comb.drug.final$cat)
comb.drug.final$cat = gsub ( "RNA-DNA-UP", "CNA", comb.drug.final$cat)
comb.drug.final$cat = gsub ( "RNA-DNA-DOWN", "CNA", comb.drug.final$cat)
comb.drug.final$cat = gsub ( "WGS-GAIN", "CNA", comb.drug.final$cat)


drug.final2 = comb.drug.final %>% dplyr::group_by (cancer.subtype, cat) %>%
  dplyr::summarise(
    total = sum ( total  ) 
    #,cancer.subtype = unique ( cancer.subtype)
  ) %>% data.frame()


drug.final2$cancer.subtype = factor ( drug.final2$cancer.subtype, levels = cancer.subtype.order)
drug.final2$cat = factor ( drug.final2$cat, levels = c("fusion", "RNA" , 'CNA','SNV', 'no-action'    ))



drug.grid = ggplot(drug.final2, aes(x=cat, y=cancer.subtype, fill = total, label = total)) +
  geom_tile( col = 'grey') +
  geom_text(col = "black") +
  theme_minimal() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  scale_fill_gradient2(low = "white", mid = "yellow", high = "red")






drug.levels.comb$cat = gsub ( "RNA-UP", "RNA", drug.levels.comb$cat)
drug.levels.comb$cat = gsub ( "RNA-DOWN", "RNA", drug.levels.comb$cat)
drug.levels.comb$cat = gsub ( "RNA-DNA-UP", "CNA", drug.levels.comb$cat)
drug.levels.comb$cat = gsub ( "RNA-DNA-DOWN", "CNA", drug.levels.comb$cat)
drug.levels.comb$cat = gsub ( "WGS-GAIN", "CNA", drug.levels.comb$cat)

drug.levels = drug.levels.comb %>% dplyr::group_by ( cat, level) %>%
  dplyr::summarise(
    total = n(   ) 
    #,cancer.subtype = unique ( cancer.subtype)
  ) %>% data.frame()

drug.levels = rbind ( drug.levels, c("no-action", "L3", 0 ))


drug.levels$cat = factor ( drug.levels$cat, levels = c("fusion", "RNA" , 'CNA','SNV', 'no-action'    ) )

drug.levels$total = as.numeric ( drug.levels$total )
stack.level.plot = ggplot(drug.levels, aes(fill=level,  x=cat , y=total)) + 
  geom_bar(position="stack", stat="identity") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("Clinical Level") + 
  xlab("") + #ylim(1,71) + 
  # scale_color_manual(values= rep("black", 13) ) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
        , axis.ticks.y=element_blank()
        #, axis.text.y=element_blank()
  ) +
  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
  #scale_y_continuous( expand = c(0, 0)) +
  scale_fill_manual(values=bcc )





stack.level.plot / drug.grid

```

### Best by Disease
* some modalities are combined 

```{r fig=TRUE,fig.width=12, fig.height=10, echo=FALSE, include=TRUE , results='asis'}
stack.level.plot / drug.grid

```

### Drug database output

* Notes lots of repeat here since drugs are not collapse

```{r fig=TRUE, echo=FALSE, include=TRUE }


#best.drug2 = merge (  trueseq.key[ , c("RNAseq.id", "patient.id","cancer.subtype")], best.drug, by="RNAseq.id")
#colnames ( best.drug2) [4:5] = paste ( colnames ( best.drug2) [4:5], "___________")
snvdrug.table = snvread$snvtotal 
snvdrug.table$wgs.id = gsub( "_.*?G", "", snvdrug.table$wgs.id)


#intersect ( snvdrug.table$wgs.id, trueseq.key$WGS.id)

snvdrug.table = merge ( snvdrug.table, trueseq.key, by.x="wgs.id", by.y= "WGS.id")


sharen = intersect ( names ( best.fusion), names ( snvdrug.table))
sharen = intersect( sharen, names ( total.drug ))
best1 = rbind (best.fusion[ ,c ("RNAseq.id", sharen ) ], snvdrug.table[ ,c ("RNAseq.id",sharen ) ]  )
best1 = rbind ( best1, total.drug[, c ("RNAseq.id", sharen ) ])
best1 = best1[best1$clinical_sig == "positive", ]

nameorder = c( "RNAseq.id"
               ,"gene"
               ,"type"
               ,"alteration"
               ,"drugs"
               ,"level"
               ,"source"
               ,"clinical_sig"
               ,"chromosome"
               ,"start"
               ,"stop"
               ,"reference_bases"
               ,"variant_bases"
               ,"citation_id"
               ,"disease"
               ,"drug_interaction_type"
               ,"representative_transcript"
               ,"reference_build"
               ,"evidence_civic_url"
               , "evidence_level"
               ,"variant"
               
)

DT::datatable(best1[, nameorder],
              caption = 'Druggable alterations', rownames = T, filter= list ( position="top", clear = FALSE )  
              , extensions = 'Buttons'
              , options = list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print')
                               , autoWidth = T, scrollX=T, className = 'dt-left'
              )
              
)

```

```{r}




bestof.final.freq = plyr::count(bestof.final, c("best.drug.cat","cancer.subtype"))
bestof.final.freq$cancer.subtype = factor ( bestof.final.freq$cancer.subtype , levels= cancer.subtype.order   )
# tabulate by category 

tab.plot.drugcat = ggplot(bestof.final.freq, aes(area = freq, fill = best.drug.cat, label = cancer.subtype, subgroup= best.drug.cat ) ) +
  geom_treemap(position = "identity" ) + 
  geom_treemap_text(colour = "black", place = "bottomright", min.size=0, grow=F, reflow = T, size= 12, angle=0) +
  geom_treemap_subgroup_text(place = "centre", reflow = F, alpha = 1, colour =
                               "white", fontface = "italic", min.size = 0,  angle=0, size= 20 ) +
  geom_treemap_subgroup_border(colour = "black", size = 3) + scale_fill_manual(values = bcc ) 


tab.plot.drugcat.cancer_facet = ggplot(bestof.final.freq, aes(area = freq, fill = best.drug.cat, label = "", subgroup= best.drug.cat ) ) +
  geom_treemap(position = "identity" ) + 
  geom_treemap_text(colour = "black", place = "bottomright", min.size=0, grow=F, reflow = T, size= 12, angle=90) +
  geom_treemap_subgroup_text(place = "centre", reflow = F, alpha = 0, colour =
                               "white", fontface = "italic", min.size = 0,  angle=0, size= 20 ) +
  geom_treemap_subgroup_border(colour = "black", size = .5) + scale_fill_manual(values = bcc )  + 
  theme(legend.position="bottom") +  facet_grid(cancer.subtype ~., scales = "free" , space = "free" ) + theme( legend.position = "none")


tab.plot.drugcat.cancer_sub = ggplot(bestof.final.freq, aes(area = freq, fill =  cancer.subtype, label = best.drug.cat , subgroup=cancer.subtype  ) ) +
  geom_treemap(position = "identity" ) + 
  geom_treemap_text(colour = "black", place = "bottomright", min.size=0, grow=F, reflow = T, size= 12, angle=0) +
  geom_treemap_subgroup_text(place = "centre", reflow = F, alpha = 1, colour =
                               "black", fontface = "italic", min.size = 0,  angle=0, size= 20 ) +
  geom_treemap_subgroup_border(colour = "black", size = 1) + scale_fill_manual(values = subcolor[ cancer.subtype.order ]  )  + 
  theme(legend.position="right") 


( bestof.plot + theme ( legend.position = "none") ) / tab.plot.drugcat  / tab.plot.drugcat.cancer_sub +
  plot_layout(heights =c ( .1,1,1) )






```



### Drug plots
* First plot describes the bestoff by sample.  These are in the same order as above.
+ best is ranked by FDA, Clinical, then pre-clinical
+ this is not a total tabulation, only the single best is shown, even if there are ties. 
* The other two plots are treemaps by either drug category or cancer subtype.  

```{r fig=TRUE,fig.width=22, fig.height=17, echo=FALSE, include=TRUE , results='asis'}
( bestof.plot + theme ( legend.position = "none") ) / tab.plot.drugcat  / tab.plot.drugcat.cancer_sub +
  plot_layout(heights =c ( .1,1,1) )

```



```{r}
#genetab = data.frame ( table ( best1$gene ) ) 

ubest1 = best1
#ubest1$un = paste ( ubest1$RNAseq.id, ubest1$gene, ubest1$alteration, ubest1$clinical_sig )
ubest1$un = paste ( ubest1$RNAseq.id, ubest1$gene)
ubest1 =ubest1[!duplicated( ubest1$un), ]
genetab =  plyr::count(ubest1, c(" type","gene")) 

# find total 
gt = data.frame ( table ( ubest1$gene ) ) 
gt = gt[order ( -gt$Freq), ]
colnames ( gt ) = c("gene","total")

genetab = merge ( genetab, head ( gt, 20 ) , by="gene")
genetab = genetab[ order ( -genetab$total), ]


genetab$gene = factor ( genetab$gene, levels= rev ( unique ( genetab$gene)) )

tabc = c( "CNA"="#FB8072", "expression"="#dbd59a","VARIANT"="#80B1D3")
genetab$type = factor ( genetab$type, levels=c("expression","CNA","VARIANT", "fusion"))

tabc["fusion"]="#b9de85"

drug.tab.gene.type  = ggplot(genetab, aes(fill=type,  x=gene , y=freq )) + 
  geom_bar(position="stack", stat="identity") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + coord_flip() +
  scale_fill_manual(values=tabc) 


# scale_fill_manual(values=cc2 )


ubestcomb = best1[ !best1$drug_interaction_type %in% c("Combination", "Sequential" ), ]


ubestcomb = ubestcomb %>% 
  mutate(drugs = strsplit(as.character(drugs), ",")) %>% 
  unnest(drugs) %>%  data.frame(stringsAsFactors = F)
ubestcomb$drugs = gsub ( "^ ", "", ubestcomb$drugs)

ubestcomb =  merge ( ubestcomb, trueseq.key[ , c ( "RNAseq.id","cancer.subtype") ], by="RNAseq.id")

ubestcomb$dup = paste ( ubestcomb$RNAseq.id, ubestcomb$drugs)
ubestcomb = ubestcomb[!duplicated ( ubestcomb$dup), ]

ubestcomb = ubestcomb[!ubestcomb$drugs == "Chemotherapy", ]


genetab =  plyr::count(ubestcomb, c("cancer.subtype","drugs")) 
# find total 
gt = data.frame ( table ( ubestcomb$drugs ) ) 
gt = gt[order ( -gt$Freq), ]
colnames ( gt ) = c("drugs","total")

genetab = merge ( genetab, head ( gt, 20 ) , by="drugs")
genetab = genetab[ order ( -genetab$total), ]

genetab$drugs = gsub ( ".*GSK3377794", "GSK3377794", genetab$drugs)
genetab$drugs = factor ( genetab$drugs, levels= rev ( unique ( genetab$drugs)) )
genetab$cancer.subtype = factor ( genetab$cancer.subtype, levels = cancer.subtype.order  )

genetab$label = as.character ( sapply (as.character ( genetab$drugs), function (x) { 
  
  toString( unique  ( drug[grepl  ( x, drug$drugs), ]$type) )
} )  )



drug.tab.drug.type  = ggplot(genetab, aes(fill=cancer.subtype,  x=drugs , y=freq )) + 
  geom_bar(position="stack", stat="identity") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  scale_fill_manual(values = subcolor[ cancer.subtype.order ]  ) + coord_flip() 
 # geom_text(data=genetab[!duplicated ( genetab$drugs), ]  , aes(fill=cancer.subtype,  x=drugs ,label=label),hjust=-1)



```


## TOP druggable genes/drugs


```{r fig=TRUE,fig.width=16, fig.height=12, echo=FALSE, include=TRUE , results='asis'}

drug.tab.gene.type + drug.tab.drug.type


```




```{r}

### functions for multi sample plotting! 

pcolor = c (   brewer.pal(n = 5, name = "Set1") , brewer.pal(n = 8, name = "RdBu") ) 
 plot.long.1gene = function ( gene, disease ){
   
gene.tpm = reshape2::melt ( tpm[tpm$gene == gene, temporal.key[temporal.key$cancer.subtype %in% disease, ]$RNAseq.id] )
colnames ( gene.tpm ) = c("RNAseq.id", "tpm")
gene.tpm = merge ( gene.tpm, temporal.key[ , c(1,2,5,25)] ,by ="RNAseq.id" )
gene.tpm = gene.tpm[ order ( gene.tpm$cancer.subtype, gene.tpm$patient.id,  gene.tpm$temporal.order) , ]
gene.tpm$tpm = log2 ( gene.tpm$tpm + 1 )
gene.tpm$temporal.order = as.character (gene.tpm$temporal.order)
# plot by samples 


# plot each sample but ratain all temporal orders 

m2 = ggplot( gene.tpm , aes(x = temporal.order, y = tpm, color = cancer.subtype, group = patient.id)) + 
    geom_point(size=6 , alpha = 0.9    ) + geom_line() + ylab( paste0 ( gene, ": log2(tpm + 1 )" ) )+ xlab("temporal.order: RNA-seq") +
     theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
    theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=18),
          
          axis.title.y     = element_text(size=18), 
          legend.text      =element_text(size=12)
    ) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
              axis.line = element_line(colour = "black"))+ scale_color_manual(values= subcolor) +
  facet_wrap(   patient.id ~. , scale="free") 

return ( m2 )

}
# plot all temporal but mean all 


 plot.long.Mgene = function ( gene, disease, pid=NA ){
   

if ( is.na(pid)) {gene.tpm = reshape2::melt ( tpm[tpm$gene %in%  gene, c( "gene", temporal.key[temporal.key$cancer.subtype %in% disease, ]$RNAseq.id) ] )}else
{
  gene.tpm = reshape2::melt ( tpm[tpm$gene %in%  gene, c( "gene", temporal.key[temporal.key$cancer.subtype %in% disease & temporal.key$patient.id == pid, ]$RNAseq.id) ] )
}

colnames ( gene.tpm ) = c("gene", "RNAseq.id", "tpm")
gene.tpm = merge ( gene.tpm, temporal.key[ , c(1,2,5,34)] ,by ="RNAseq.id" )
gene.tpm = gene.tpm[ order ( gene.tpm$cancer.subtype, gene.tpm$patient.id,  gene.tpm$temporal.order) , ]
gene.tpm$tpm = log2 ( gene.tpm$tpm + 1 )
gene.tpm$temporal.order = as.character (gene.tpm$temporal.order)

m1 = gene.tpm  %>% 
dplyr::group_by(patient.id, temporal.order , gene ) %>% 
dplyr::mutate(
              tpm = mean(tpm)
              ) %>% ggplot( . , aes(x = temporal.order, y = tpm, color = patient.id, group = patient.id)) + 
geom_point(size=6 , alpha = 0.75    ) + 
  #geom_line() + 
  geom_path(arrow = arrow(length = unit(0.55, "cm") )  ) +
  ylab( paste0 (  "log2(tpm + 1 )" )  )+ xlab("temporal.order: RNA-seq") +
 theme_bw() +
theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
      panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
      
      axis.text.y = element_text(size= 12 ),
      axis.text.x = element_text(angle = 0, size= 12 ),
      axis.title.x = element_text(size=18),
      
      axis.title.y     = element_text(size=18), 
      legend.text      =element_text(size=12)
) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          axis.line = element_line(colour = "black")) + 
  facet_wrap(   gene ~., scales="free" )   + 
  scale_color_manual(values= pcolor  ) 
  #scale_color_manual(values= getPalette(34))



coef = gene.tpm  %>% 
  dplyr::group_by(patient.id, temporal.order, gene ) %>% 
  dplyr::mutate(
                tpm = mean(tpm)
                ) %>% dplyr::group_by(patient.id, gene ) %>%
  dplyr::mutate(
                coef = summary ( lm ( tpm   ~  temporal.order ))$coef[2],
                )  %>% data.frame(stringsAsFactors = F)



    
return ( list ( plot = m1, df = coef ) )    
   
 }
 
 
 
 plot.long.Mgene_merged = function ( gene, disease ){
  
  gene.tpm = reshape2::melt ( tpm[tpm$gene %in%  gene, c( "gene", temporal.key[temporal.key$cancer.subtype %in% disease, ]$RNAseq.id) ] )
  colnames ( gene.tpm ) = c("gene", "RNAseq.id", "tpm")
  gene.tpm = merge ( gene.tpm, temporal.key[ , c(1,2,5,34)] ,by ="RNAseq.id" )
  gene.tpm = gene.tpm[ order ( gene.tpm$cancer.subtype, gene.tpm$patient.id,  gene.tpm$temporal.order) , ]
  gene.tpm$tpm = log2 ( gene.tpm$tpm + 1 )
  gene.tpm$temporal.order = as.character (gene.tpm$temporal.order)
  
  gene.tpm2 = data.frame ()
  for ( p in unique ( gene.tpm$patient.id)){
    p = gene.tpm[ gene.tpm$patient.id == p, ]
    mp = min ( p$temporal.order)
    p$temporal.order = ifelse(p$temporal.order > mp, 2, 1)
    gene.tpm2 = rbind ( gene.tpm2, p)
  }
  
  gene.tpm2$temporal.order = as.character ( gene.tpm2$temporal.order)
  
  m1 = gene.tpm2  %>% 
    dplyr::group_by(patient.id, temporal.order , gene ) %>% 
    dplyr::mutate(
      tpm = median(tpm)
    ) %>% ggplot( . , aes(x = temporal.order, y = tpm, color = patient.id, group = patient.id)) + 
    geom_point(size=6 , alpha = 0.75    ) + geom_line() + ylab( paste0 (  ": log2(tpm + 1 )" )  )+ xlab("temporal.order: RNA-seq") +
    theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
    theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=18),
          
          axis.title.y     = element_text(size=18), 
          legend.text      =element_text(size=12)
    ) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
              axis.line = element_line(colour = "black")) + 
    facet_wrap(   gene ~., scales="free" )   + scale_color_manual(values= pcolor  ) 
  
  
  dfr =  gene.tpm2  %>% 
    dplyr::group_by( patient.id, temporal.order , gene ) %>% 
    dplyr::summarise(
      tpm = median(tpm)
    )  %>% data.frame( stringsAsFactors = F  )
  
 dfr$gene = paste (  dfr$gene, dfr$patient.id  )
 dfr= dfr[ , c("gene", "temporal.order", "tpm")] %>%
  group_by(gene,temporal.order)  %>%
  pivot_wider(names_from = temporal.order, values_from=tpm) %>% data.frame(stringsAsFactors = F)
  dfr = dfr[ order ( dfr$gene), ]
   
  row.names ( dfr ) = dfr$gene
  dfr$gene = NULL 
  colnames ( dfr ) = c("T1", "T2")
  return ( list ( plot = m1, df = dfr ) )    
  
 }
 
 
 plot.long.Mgene.singlepid = function ( gene, disease, pid=NA ){
  
  
  if ( is.na(pid)) {gene.tpm = reshape2::melt ( tpm[tpm$gene %in%  gene, c( "gene", temporal.key[temporal.key$cancer.subtype %in% disease, ]$RNAseq.id) ] )}else
  {
    gene.tpm = reshape2::melt ( tpm[tpm$gene %in%  gene, c( "gene", temporal.key[temporal.key$cancer.subtype %in% disease & temporal.key$patient.id == pid, ]$RNAseq.id) ] )
  }
  
  colnames ( gene.tpm ) = c("gene", "RNAseq.id", "tpm")
  gene.tpm = merge ( gene.tpm, temporal.key[ , c(1,2,5,34)] ,by ="RNAseq.id" )
  gene.tpm = gene.tpm[ order ( gene.tpm$cancer.subtype, gene.tpm$patient.id,  gene.tpm$temporal.order) , ]
  gene.tpm$tpm = log2 ( gene.tpm$tpm + 1 )
  gene.tpm$temporal.order = as.character (gene.tpm$temporal.order)
  
  m1 = gene.tpm  %>% 
    dplyr::group_by(patient.id, temporal.order , gene ) %>% 
    dplyr::mutate(
      tpm = mean(tpm)
    ) %>% ggplot( . , aes(x = temporal.order, y = tpm, color = cancer.subtype, group = patient.id)) + 
    geom_point(size=6 , alpha = 0.75    ) + geom_line() + ylab( paste0 (  "log2(tpm + 1 )" )  )+ xlab("temporal.order: RNA-seq") +
    theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
    theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=18),
          
          axis.title.y     = element_text(size=18), 
          legend.text      =element_text(size=12)
    ) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
              axis.line = element_line(colour = "black")) + 
    facet_wrap(   gene ~., scales="free" , ncol=1)   + scale_color_manual(values= subcolor  ) 
  
  
  
  
  coef = gene.tpm  %>% 
    dplyr::group_by(patient.id, temporal.order, gene ) %>% 
    dplyr::mutate(
      tpm = mean(tpm)
    ) %>% dplyr::group_by(patient.id, gene ) %>%
    dplyr::mutate(
      coef = summary ( lm ( tpm   ~  temporal.order ))$coef[2],
    )  %>% data.frame(stringsAsFactors = F)
  
  
  
  
  return ( list ( plot = m1, df = coef ) )    
  
 }
 
 
  plot.long.Sgene.singlepid = function ( gene, disease, pid=NA, sz=4 ){
  
  
  if ( is.na(pid)) {gene.tpm = reshape2::melt ( tpm[tpm$gene ==  gene, c( "gene", temporal.key[temporal.key$cancer.subtype %in% disease, ]$RNAseq.id) ] )}else
  {
    gene.tpm = reshape2::melt ( tpm[tpm$gene %in%  gene, c( "gene", temporal.key[temporal.key$cancer.subtype %in% disease & temporal.key$patient.id == pid, ]$RNAseq.id) ] )
  }
  
  colnames ( gene.tpm ) = c("gene", "RNAseq.id", "tpm")
  gene.tpm = merge ( gene.tpm, temporal.key[ , c(1,2,5,34)] ,by ="RNAseq.id" )
  gene.tpm = gene.tpm[ order ( gene.tpm$cancer.subtype, gene.tpm$patient.id,  gene.tpm$temporal.order) , ]
  gene.tpm$tpm = log2 ( gene.tpm$tpm + 1 )
  gene.tpm$temporal.order = as.character (gene.tpm$temporal.order)
  
  gene.tpm$gene = paste ( pid, gene )
  
  m1 = gene.tpm  %>% 
    dplyr::group_by(patient.id, temporal.order , gene ) %>% 
    dplyr::mutate(
      tpm = mean(tpm)
    ) %>% ggplot( . , aes(x = temporal.order, y = tpm, color = cancer.subtype, group = patient.id)) + 
    geom_point(size=sz , alpha = 0.75    ) + geom_line() + ylab("")+ xlab("" ) +
    theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
    theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=18),
          
          axis.title.y     = element_text(size=18), 
          legend.text      =element_text(size=12)
    ) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
              axis.line = element_line(colour = "black")) + 
    facet_wrap(   gene ~., scales="free" , ncol=1)   + scale_color_manual(values= subcolor  )  +
    theme(strip.background =element_rect(fill=subcolor[disease ])) 
  
  
  
  
  coef = gene.tpm  %>% 
    dplyr::group_by(patient.id, temporal.order, gene ) %>% 
    dplyr::mutate(
      tpm = mean(tpm)
    ) %>% dplyr::group_by(patient.id, gene ) %>%
    dplyr::mutate(
      coef = summary ( lm ( tpm   ~  temporal.order ))$coef[2],
    )  %>% data.frame(stringsAsFactors = F)
  
  
  
  
  return ( list ( plot = m1, df = coef ) )    
  
  }
  
  

```

```{r}

fusion_mod = temp.all
fusion_mod$patient.id = as.character ( sapply ( fusion_mod$RNAseq.id, function (x) rna.key[rna.key$RNAseq.id == x, ]$patient.id))
# correct diagnostics 
fusion_mod$type = ifelse ( fusion_mod$type == "is.diag" & fusion_mod$frameness != "in.frame", "is.novel", fusion_mod$type )
fusion_mod$type = ifelse ( fusion_mod$type == "is.drug" & fusion_mod$frameness != "in.frame", "is.novel", fusion_mod$type )



# updated 5.11.2022, new longtudinal analysis
compute_longi <- function (dff , annt = NA ) {
  
if ( !is.na ( annt) ){
  dff$fusion = dff[[annt]]
}  
  
  
  


longi.key_work = rna.key[ rna.key$patient.id %in% key.long$patient.id, 
                            c ( "patient.id" ,"temporal.order" ,"cancer.subtype", "sample.classification" ,"treatment"  ,  "WGS.id" , "RNAseq.id" ) ]

# order by disease, then by temporal order 
longi.key_work$temporal.order = as.numeric(longi.key_work$temporal.order)
longi.key_work = longi.key_work[ order ( longi.key_work$cancer.subtype, longi.key_work$patient.id,  longi.key_work$temporal.order), ]
head ( longi.key_work )

# lets get samples that only have temporal orders 
temporal = c()
for ( tu in unique ( longi.key_work$patient.id)){
  t = unique ( longi.key_work[longi.key_work$patient.id == tu, ]$temporal.order )
  if ( length ( t ) > 1  ){
    temporal = c( temporal, tu  )
  }
}

# only include samples that have temporal ordering

temporal.key = longi.key_work[ longi.key_work$patient.id %in% temporal, ]


## get total fusion for each sample 
tfusion_long = dff %>% dplyr::group_by(RNAseq.id) %>%
  dplyr::summarise(
    total_fusion = n()  
    
  ) %>%  data.frame ( stringsAsFactors = F )




fusion_total2 = merge ( tfusion_long, temporal.key[ , c("patient.id", "cancer.subtype", "temporal.order", "RNAseq.id")] ,by ="RNAseq.id", all.x =F, all.y=T )
fusion_total2 = fusion_total2[ order ( fusion_total2$cancer.subtype, fusion_total2$patient.id,  fusion_total2$temporal.order) , ]
fusion_total2[is.na(fusion_total2)] <- 0 
fusion_total2$temporal.order = as.character (fusion_total2$temporal.order)


## here be dragons 
fusion_total2 = fusion_total2[fusion_total2$patient.id != "40", ]




# change all first sample to 1 
fusion_total2$temporal.order = as.numeric ( fusion_total2$temporal.order )

for ( p in unique ( fusion_total2$patient.id)){
  m = min ( fusion_total2[ fusion_total2$patient.id == p, ]$temporal.order) 
  fusion_total2[ fusion_total2$patient.id == p & fusion_total2$temporal.order == m , ]$temporal.order  = 1 
}


# remove all middle samples. 
fusion_total2$temporal.order = ifelse ( fusion_total2$temporal.order == 1, 1, 2 )
fusion_total2 = fusion_total2 [ order ( fusion_total2$patient.id, fusion_total2$temporal.order, -fusion_total2$total_fusion), ]
fusion_total2 = fusion_total2[ ! duplicated ( paste ( fusion_total2$patient.id, fusion_total2$temporal.order) ), ]

# median center all total 

long_total_fusion = fusion_total2  %>% 
  dplyr::group_by(patient.id, temporal.order ) %>% 
  dplyr::mutate(
    total_fusion = median(total_fusion)
  ) %>% 
  ggplot( . , aes(x = temporal.order, y = total_fusion, color = cancer.subtype, group = patient.id)) + 
  geom_point(size=6 , alpha = 0.75    ) + geom_line() + ylab(" log2(cpm)")+ xlab("temporal.order: RNA-seq") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 12 ),
        axis.text.x = element_text(angle = 0, size= 12 ),
        axis.title.x = element_text(size=18),
        
        axis.title.y     = element_text(size=18), 
        legend.text      =element_text(size=12)
  ) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
            axis.line = element_line(colour = "black"))+ scale_color_manual(values= subcolor) +
  facet_wrap(   cancer.subtype ~. , scale="free") +   scale_x_continuous(  breaks=c(1, 2) ) +
  scale_y_continuous(limits = c(0, NA) )


### create dataframe for heatmaps

long_out = data.frame ()
long_info = data.frame()

for ( tu in unique ( fusion_total2$patient.id)){
  t = fusion_total2 [ fusion_total2$patient.id == tu, ]
  tmin = min ( t$temporal.order ) 
  tmax = max  ( t$temporal.order)
  
  tmin = fusion_total2 [ fusion_total2$patient.id == tu & fusion_total2$temporal.order == tmin, ]
  tmax = fusion_total2 [ fusion_total2$patient.id == tu & fusion_total2$temporal.order == tmax, ]
  
  tprimary = dff[dff$RNAseq.id == tmin$RNAseq.id[1], ] 
  tsecondary = dff[dff$RNAseq.id == tmax$RNAseq.id[1], ] 
  
  primary_only = setdiff ( tprimary$fusion, tsecondary$fusion)
  secondary_only =  setdiff (  tsecondary$fusion, tprimary$fusion )
  both = intersect(  tsecondary$fusion, tprimary$fusion  )
  
  #long_info[[tu]] = c("primary_only"= toString( primary_only), "secondary_only"=toString(secondary_only), "both"=toString(both) )
  
    for ( p in primary_only ){
    long_info= rbind ( long_info, data.frame ( pid=tu, type="primary_only", info = p ) )
    }
  
  for ( p in secondary_only ){
    long_info= rbind ( long_info, data.frame ( pid=tu, type="secondary_only", info = p ) )
  }

  for ( p in both ){
    long_info= rbind ( long_info, data.frame ( pid=tu, type="both",info =  p ) )
  }

  
  
  long_out = rbind ( long_out, data.frame ( pid = tu
                                            , cancer.subtype = t$cancer.subtype[1]
                                            , primary_only = length ( primary_only )
                                            , secondary_only = length ( secondary_only )
                                            , shared= length ( both ) 
  ) )
  
  
  
}


# rehash details into a dataframe
long_details = data.frame()
for ( n in names ( long_info)){
  
  info_temp = long_info[[n]]
  
  data.frame ( 
    
    )
  
}

# sanity check, need to 2x intersect
# this summation shoud equal aboe 
p = 10
sum ( as.numeric ( long_out[long_out$pid %in% c(p), c(3:4) ] ) , as.numeric ( long_out[long_out$pid %in% c(p), c(5) ] * 2  ) )-
sum ( fusion_total2 [fusion_total2$patient.id == p, ]$total_fusion )


long_out = data.frame ( t ( long_out ))
# get disease and pid orders, sine these will be removed later 
diseaseorder = as.character ( long_out[2, ] ) 
pidorder = as.character( long_out[ 1, ])
long_out = long_out[ - c( 1,2), ]
# make into numeric 
long_out[] <- lapply(long_out, function(x) {as.numeric(as.character(x)) } ) 

colnames ( long_out) = paste0( diseaseorder, "_", pidorder )


cm2 <- apply(long_out, 2,  function(x) (x - mean(x)) / sd(x)) 
cm2 <- data.frame ( cm2 )

cm2 <-  cm2 %>% select_if(~all(!is.na(.))) 


ht_long_fusion <- ComplexHeatmap::Heatmap(  data.matrix( cm2 ) 
                          ,  name = ' '
                          , row_order = c( "primary_only", "shared", "secondary_only")
                          #, column_order = colnames ( long_out)
                          #, show_row_dend = T
                          #, show_column_names = F
                          #, show_row_names = F
                          #, name= heat.title
                          #,column_split = km2
                          #,top_annotation = topannt
                          #, col = pals::brewer.rdylbu(25) 
                          , col =  colorRampPalette(rev(brewer.pal(n = 7, name ="RdYlBu")))(25)
                          #, cluster_columns = sort ( dend2$dend , decreasing = T)
                          #, column_split = 2
                          #, cluster_rows  = dend_gene$dend 
                          #, row_split = 7
                          #,row_title = NULL
                          #,column_title = NULL
                          ,show_heatmap_legend = T
                          , row_names_gp = grid::gpar(fontsize = 28) 
                          ,heatmap_legend_param = list(at = c(-2, 0, 2), labels = c("low", "medium", "high"),  labels_gp = gpar(col = c("#4575B4", "#E3E31F", "#D73027"), fontsize = 14)   
                                                       , legend_height = unit(6, "cm") , legend_width = unit(2, "cm")
                                                       ) 
                          
)



return ( list ( ht=ht_long_fusion, long_info_details = long_info, singleg =long_total_fusion, temporal.key=temporal.key , modkey =fusion_total2  ))


}

fuse_longout = compute_longi(  fusion_mod  )
outUP_longout = compute_longi( up.all_qc[ up.all_qc$both == "OUTLIER_HIGH" & up.all_qc$gene %in%  genelist$gene, ] , "gene" )

temp = data.frame ( table ( outUP_longout$long_info_details [ outUP_longout$long_info_details$type == "secondary_only", ]$info ) )
temp = temp [ order( -temp$Freq), ]


## Examples of how to draw a gene with modified longi key outputed from compute_longi 
###

# 
tkey = outUP_longout$modkey
gene = "NONO"
gene.tpm = reshape2::melt ( tpm[tpm$gene == gene, tkey$RNAseq.id] )
colnames ( gene.tpm ) = c("RNAseq.id", "tpm")

gene.tpm = merge ( gene.tpm, tkey[ , c("patient.id", "cancer.subtype", "temporal.order", "RNAseq.id")] ,by ="RNAseq.id" )
gene.tpm = gene.tpm[ order ( gene.tpm$cancer.subtype, gene.tpm$patient.id,  gene.tpm$temporal.order) , ]
gene.tpm$tpm = log2 ( gene.tpm$tpm + 1 )
gene.tpm$temporal.order = as.character (gene.tpm$temporal.order)
# plot by samples 


gene.tpm  %>% 
  dplyr::group_by(patient.id, temporal.order ) %>% 
  dplyr::mutate(
    tpm = mean(tpm)
  ) %>% dplyr::group_by(patient.id) %>%
  #  dplyr::mutate(
  #               coef = summary ( lm ( tpm   ~  temporal.order ))$coef[2],
  #              ) %>% 
  ggplot( . , aes(x = temporal.order, y = tpm, color = cancer.subtype, group = patient.id)) + 
  geom_point(size=6 , alpha = 0.75    ) + geom_line() + ylab(" log2(cpm)")+ xlab("temporal.order: RNA-seq") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 12 ),
        axis.text.x = element_text(angle = 0, size= 12 ),
        axis.title.x = element_text(size=18),
        
        axis.title.y     = element_text(size=18), 
        legend.text      =element_text(size=12)
  ) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
            axis.line = element_line(colour = "black"))+ scale_color_manual(values= subcolor) +
  facet_wrap(   cancer.subtype ~. ) 

#############


## study longitudinal for fusion 
psb_fusions = merge ( fuse_longout$long_info_details , fusion_mod, by.x = c("pid","info"), by.y=c ( "patient.id", "fusion") )
psb_fusions = psb_fusions[psb_fusions$RNAseq.id %in% fuse_longout$modkey$RNAseq.id, ]
psb_fusions$is.cancer = ifelse ( psb_fusions$right %in% genelist$gene , "yes", "no" )
psb_results1 = as.data.frame.matrix ( as.matrix ( table ( psb_fusions[ , c("type.x", "type.y")]) ) )
# shared needs to be divided by 2 so that it does not count twice 
# dont divide by two - just because its shared does not mean it needs to be counted only once. 
psb_results1[which ( row.names ( psb_results1) == "both") , ] = psb_results1[which ( row.names ( psb_results1) == "both") , ] 
psb_results1$total = as.numeric ( rowSums(psb_results1) )

row.names ( psb_results1)[which ( row.names(psb_results1) == "both" ) ]  = "shared"

kable( psb_results1[c("primary_only", "shared", "secondary")  , ] , format = "html" , row.names = T, caption="", table.attr = "style='width:40%;'" ) %>% 
  kable_classic(full_width = F, position = "center", )  %>%
    column_spec(1, bold = T, border_right = T) 

# plot total by type  
temps = psb_fusions
temps$type.x = gsub ( "both", "shared", temps$type.x )
temps$type.x  = factor ( temps$type.x , levels = c( "primary_only", "shared", "secondary_only" ))
long_cmp_fusion <- temps %>% group_by(type.x, pid) %>% summarise ( TotalbyType = n()
                                                                         ) %>%
ggplot( .  , aes(y=TotalbyType, x=type.x))+
  geom_quasirandom(aes(fill = type.x ), size = 6, shape = 21, alpha = .5) +
  theme_bw() +
  xlab("") +
  theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=22),
        
        axis.title.y     = element_text(size=22), 
        legend.text      =element_text(size=12)
  ) + stat_summary(fun.y='mean', geom = "crossbar", width = .5, color="black") + xlab ( "") +  ylab ( "") +
  scale_fill_manual(values = unique (brewer.pal(7, "Dark2") ))  #+ ylim(0,max ( fuse.by.treatment$Total))  #+ facet_wrap(cancer.subtype~., ncol=3)

## added post=2 to correct for pvalue
long_cmp_fusion = my_add_pval(long_cmp_fusion, pairs = list(c(1, 2)  ), test='wilcox.test', post=1, textsize = 10)
long_cmp_fusion = my_add_pval(long_cmp_fusion, pairs = list(c(2, 3)  ), test='wilcox.test', post=1, textsize = 10) 


temps %>% group_by(pid, type.x) %>% summarise ( Total = n()) %>% 
kruskal.test(Total ~ type.x, data = .)





tally = data.frame ( table ( psb_fusions[ , c("type.x", "is.cancer")]) ) 
tally$type.x = gsub ( "both", "shared", tally$type.x)
tally$type.x  = factor ( tally$type.x , levels = c(  "secondary_only" , "shared", "primary_only"))

long_stack_is_cancer <- ggplot( tally  , aes(x=type.x  , y=Freq, fill=is.cancer   )) +
  geom_bar(position="fill", stat="identity")+
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  ylab("") + 
  xlab("") + 
  theme(legend.position="bottom",  legend.key = element_blank(),
        
        axis.text.y = element_text(size= 20 ),
        axis.text.x = element_text(angle = 0, size= 20 ),
        axis.title.x = element_text(size=12),
        axis.title.y     = element_text(size=15), 
        legend.text      =element_text(size=20)
        , plot.title = element_text(size = 12, face = "bold")
  ) + coord_flip() + scale_fill_manual(values= is_cancer_ccc  )


# worth mentioning 
psb_fusions$left = gsub ( "-.*$", '', psb_fusions$info)

## shared 
long_fusion_ofInt = psb_fusions [ (psb_fusions$type.x == "both" & psb_fusions$is.cancer == "yes" | psb_fusions$type.y == "is.diag") & psb_fusions$left != psb_fusions$right , ]
long_fusion_ofInt=long_fusion_ofInt[ order ( long_fusion_ofInt$type.y), ]
kable ( long_fusion_ofInt[!duplicated ( paste(long_fusion_ofInt$pid, long_fusion_ofInt$info )) , c("pid", "info","cancer.subtype", "type.y")], format="html", row.names=F, table.attr = "style='width:40%;'" ) %>% 
  kable_classic(full_width = F, position = "center", )  %>%
    column_spec(1, bold = T, border_right = T) 

## secondary 

long_fusion_ofInt = psb_fusions [ (psb_fusions$type.x == "secondary_only" & psb_fusions$is.cancer == "yes" & psb_fusions$frameness == "in.frame" | psb_fusions$type.y == "is.obsv") & psb_fusions$left != psb_fusions$right | grepl ( "NTRK3", psb_fusions$right), ]
long_fusion_ofInt=long_fusion_ofInt[ order ( long_fusion_ofInt$type.y), ]
kable ( long_fusion_ofInt[!duplicated ( paste(long_fusion_ofInt$pid, long_fusion_ofInt$info )) , c("pid", "info","cancer.subtype", "type.y")], format="html", row.names=F, table.attr = "style='width:40%;'" ) %>%
  kable_classic(full_width = F, position = "center", )  %>%
    column_spec(1, bold = T, border_right = T) 





prime_fusion_only = fuse_longout$long_info_details [fuse_longout$long_info_details$type == "primary_only", ]
prime_fusion_only = merge ( prime_fusion_only, fusion_mod, by.x = c("pid","info"), by.y=c ( "patient.id", "fusion") )

secondary_fusion_only = fuse_longout$long_info_details [fuse_longout$long_info_details$type == "secondary_only", ]
secondary_fusion_only = merge ( secondary_fusion_only, fusion_mod, by.x = c("pid","info"), by.y=c ( "patient.id", "fusion") )

both_fusion_only = fuse_longout$long_info_details [fuse_longout$long_info_details$type == "both", ]
both_fusion_only = merge ( both_fusion_only, fusion_mod, by.x = c("pid","info"), by.y=c ( "patient.id", "fusion") )


```


```{r}
# older logititudinal 
# now we get to longitudanal stuff! 
longi.key = trueseq.key[duplicated(trueseq.key$patient.id), ]$patient.id 
longi.key = trueseq.key[trueseq.key$patient.id %in% longi.key, ]
# order by disease, then by temporal order 
longi.key$temporal.order = as.numeric(longi.key$temporal.order)
longi.key = longi.key[ order ( longi.key$cancer.subtype, longi.key$patient.id,  longi.key$temporal.order), ]
head ( longi.key[ , c ( 1:10, 34) ] )

# lets get samples that only have temporal orders 
temporal = c()
for ( tu in unique ( longi.key$patient.id)){
  t = unique ( longi.key[longi.key$patient.id == tu, ]$temporal.order )
  if ( length ( t ) > 1  ){
    temporal = c( temporal, tu  )
  }
}

# only include samples that have temporal ordering
temporal.key = longi.key[ longi.key$patient.id %in% temporal, ]

# example of JAK2 
gene = "JAK2"
gene.tpm = reshape2::melt ( tpm[tpm$gene == gene, temporal.key$RNAseq.id] )
colnames ( gene.tpm ) = c("RNAseq.id", "tpm")

gene.tpm = merge ( gene.tpm, temporal.key[ , c(1,2,5,34)] ,by ="RNAseq.id" )
gene.tpm = gene.tpm[ order ( gene.tpm$cancer.subtype, gene.tpm$patient.id,  gene.tpm$temporal.order) , ]
gene.tpm$tpm = log2 ( gene.tpm$tpm + 1 )
gene.tpm$temporal.order = as.character (gene.tpm$temporal.order)
# plot by samples 


# plot each sample but ratain all temporal orders 

m2 = ggplot( gene.tpm , aes(x = temporal.order, y = tpm, color = cancer.subtype, group = patient.id)) + 
    geom_point(size=6 , alpha = 0.75    ) + 
  geom_line() + ylab(" log2(cpm)")+ xlab("temporal.order: RNA-seq") +
     theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
    theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=18),
          
          axis.title.y     = element_text(size=18), 
          legend.text      =element_text(size=12)
    ) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
              axis.line = element_line(colour = "black"))+ scale_color_manual(values= subcolor) +
  facet_wrap(   cancer.subtype ~. , scales = "free"  ) 

# plot all temporal but mean all 



    gene.tpm  %>% 
    dplyr::group_by(patient.id, temporal.order ) %>% 
    dplyr::mutate(
                  tpm = mean(tpm)
                  ) %>% dplyr::group_by(patient.id) %>%
  #  dplyr::mutate(
   #               coef = summary ( lm ( tpm   ~  temporal.order ))$coef[2],
    #              ) %>% 
      ggplot( . , aes(x = temporal.order, y = tpm, color = cancer.subtype, group = patient.id)) + 
    geom_point(size=6 , alpha = 0.75    ) + geom_line() + ylab(" log2(cpm)")+ xlab("temporal.order: RNA-seq") +
     theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
    theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=18),
          
          axis.title.y     = element_text(size=18), 
          legend.text      =element_text(size=12)
    ) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
              axis.line = element_line(colour = "black"))+ scale_color_manual(values= subcolor) +
  facet_wrap(   cancer.subtype ~. ) 


# LETS FIGURE OUT WHICH GENES HAVE THE HIGHEST SLOPE? 
tab.long = data.frame(stringsAsFactors = F)
for ( gene in  c( oncogene,tsg ) ){

gene.tpm = reshape2::melt ( tpm[tpm$gene == gene, temporal.key$RNAseq.id] )
if ( nrow ( gene.tpm) == 0 ){
  next
}
colnames ( gene.tpm ) = c("RNAseq.id", "tpm")
gene.tpm = merge ( gene.tpm, temporal.key[ , c(1,2,5,34)] ,by ="RNAseq.id" )
gene.tpm = gene.tpm[ order ( gene.tpm$cancer.subtype, gene.tpm$patient.id,  gene.tpm$temporal.order) , ]
gene.tpm$tpm = log2 ( gene.tpm$tpm + 1 )
gene.tpm$RNAseq.id = as.character ( gene.tpm$RNAseq.id)
gene.tpm = gene.tpm[ !is.na(gene.tpm$temporal.order), ]
#gene.tpm$temporal.order = as.character (gene.tpm$temporal.order)


df = gene.tpm  %>% 
  dplyr::group_by(patient.id, temporal.order ) %>% 
  dplyr::mutate(
                tpm = mean(tpm)
                ) %>% dplyr::group_by(patient.id) %>%
  dplyr::mutate(
                coef = median ( as.numeric ( summary ( lm ( tpm  ~  temporal.order ))$coef[-1, 1] ) ) ,
                ) %>% dplyr::group_by(cancer.subtype) %>% distinct ( patient.id , .keep_all = TRUE ) %>% dplyr::group_by(cancer.subtype) %>%
  dplyr::mutate( 
     median.slope = median ( coef),
     gene = gene, 
     total = n(),
     total.up = length ( coef[ coef > 0 ] ) / total, 
     total.down = length ( coef[ coef < 0 ] ) / total
    ) %>% dplyr::group_by(cancer.subtype) %>% distinct ( cancer.subtype , .keep_all = TRUE ) %>%
  data.frame(stringsAsFactors = F)
  df$RNAseq.id = NULL 
  
  if ( any(oncogene %in% gene )){
    df$type = "oncogene"
  } else if ( any(tsg %in% gene )) {
    df$type = "tsg"
  }
  
tab.long = rbind ( df, tab.long )
  
  

}

tab.long$state = ifelse ( tab.long$coef > 0, "pos", "neg")

os.long = tab.long[tab.long$cancer.subtype == "OS", ]    
os.long = os.long[ order ( os.long$total.up, decreasing = T), ]
os.long$state = ifelse ( os.long$coef > 0, "pos", "neg")

testgene = c( os.long[!duplicated ( os.long$gene), ]$gene )

mtest = plot.long.Mgene( testgene[1:6], "OS")
mtest$plot
#plot.long.1gene ( testgene[1], "OS")



#### redo this but I will median all samples 

tab.long_double = data.frame(stringsAsFactors = F)
for ( gene in  c( oncogene )){
  
  gene.tpm = reshape2::melt ( tpm[tpm$gene == gene, temporal.key$RNAseq.id] )
  if ( nrow ( gene.tpm) == 0 ){
    next
  }
  colnames ( gene.tpm ) = c("RNAseq.id", "tpm")
  gene.tpm = merge ( gene.tpm, temporal.key[ , c("RNAseq.id","patient.id","cancer.subtype","temporal.order" )] ,by ="RNAseq.id" )
  gene.tpm = gene.tpm[!is.na(gene.tpm$temporal.order), ]
  
  gene.tpm = gene.tpm[ order ( gene.tpm$cancer.subtype, gene.tpm$patient.id,  gene.tpm$temporal.order) , ]
  gene.tpm$tpm = log2 ( gene.tpm$tpm + 1 )
  gene.tpm$temporal.order = as.character (gene.tpm$temporal.order)
  gene.tpm$temporal.order = as.numeric ( gene.tpm$temporal.order)
  gene.tpm2 = data.frame ()
  for ( p in unique ( gene.tpm$patient.id)){
    p = gene.tpm[ gene.tpm$patient.id == p, ]
    mp = min ( p$temporal.order)
    p$temporal.order = ifelse(p$temporal.order > mp, 2, 1)
    gene.tpm2 = rbind ( gene.tpm2, p)
  }
  
  df = gene.tpm2  %>% 
    dplyr::group_by(patient.id, temporal.order  ) %>% 
    dplyr::mutate(
      tpm = median(tpm)
    ) %>% dplyr::group_by(patient.id) %>%
    dplyr::mutate(
      coef = summary ( lm ( tpm   ~  temporal.order ))$coef[2],
    ) %>% dplyr::group_by(cancer.subtype) %>% distinct ( patient.id , .keep_all = TRUE ) %>% dplyr::group_by(cancer.subtype) %>%
    dplyr::mutate( 
      median.slope = median ( coef),
      gene = gene, 
      total = n(),
      total.up = length ( coef[ coef > 0 ] ) / total, 
      total.down = length ( coef[ coef < 0 ] ) / total
    ) %>% dplyr::group_by(cancer.subtype) %>% distinct ( cancer.subtype , .keep_all = TRUE ) %>%
    data.frame(stringsAsFactors = F)
  df$RNAseq.id = NULL 
  tab.long_double = rbind ( df, tab.long_double )
  
  
  
}

os.long = tab.long_double[tab.long_double$cancer.subtype == "OS", ]    
os.long = os.long[ order ( os.long$total.up, decreasing = T), ]

testgene = c( os.long[!duplicated ( os.long$gene) & os.long$total.up > .8, ]$gene )

pthis = plot.long.Mgene_merged ( testgene, "OS")
pthis$plot

head ( pthis$df)



# try normalizing by gene only 

sheat2 = data.frame ()
for ( sthis in unique ( rename_cs ) ){
sc = pthis$df [ grepl ( sthis , row.names ( pthis$df )),      ]
sheat2 = rbind ( sheat2, as.data.frame ( scale ( sc ) ) )
}

# try taking the differnce only 
sheat3 = pthis$df 
sheat3$diff = sheat3$T2 - sheat3$T1


hm = 1



col_fun = colorRamp2(c(-2, -.1, .1 ,2), c("#eff5cb", "#eff5cb", "#f5946e", "#f5946e"))


hmint = ComplexHeatmap::Heatmap( sheat3[ , 3, drop=F] 
                                    , name= paste ( " Delta: Last-First " )
                                    #, row_order = type_row_order
                                    
                                    #, column_order = colnames ( pthis$df)
                                    , cluster_columns = F
                                    
                                  , show_row_names = F
                                    #, row_labels = ens_gene
                                    , show_row_dend = F
                                    , show_column_names = F
                                 #   , col= c("#eff5cb", "#f5946e")
                                    , col = col_fun                                 
                                    ,cluster_column_slices = F
                                    , cluster_rows =F
                                    ,row_title = NULL
                                    #, right_annotation = r_annt
                                    #,column_split =  typekey$type1
                                    , column_title = NULL  
                                    #, top_annotation = ha
                                    #,  right_annotation = haside 
                                    
                                    
  )



################  heatmap by slope

os.long2 = tab.long[tab.long$cancer.subtype == "OS" & tab.long$type == "oncogene",  ]    
os.long2 = os.long2[ order ( os.long2$total.up, decreasing = T), ]
os.long2$state = ifelse ( os.long2$coef > 0, "pos", "neg")

testgene = c( os.long2[!duplicated ( os.long2$gene) & os.long2$total.up > .85, ]$gene )

mtest = plot.long.Mgene( testgene, "OS")
mtest$plot

sheat = mtest$df 

sheat$type = "onc"
left.annt = rep("onc", nrow(sheat))

os.long2 = tab.long[tab.long$cancer.subtype == "OS" & tab.long$type == "tsg",  ]    
os.long2 = os.long2[ order ( os.long2$total.up, decreasing = F), ]
os.long2$state = ifelse ( os.long2$coef > 0, "pos", "neg")

testgene = c( os.long2[!duplicated ( os.long2$gene) & os.long2$total.down > .85, ]$gene )

mtest = plot.long.Mgene( testgene, "OS")
mtest$plot
mtest$df$type = "tsg"

left.annt = c ( left.annt, rep("tsg", nrow(mtest$df)) )


sheat = rbind ( sheat, mtest$df)

sheat$dup = paste (  sheat$gene, sheat$patient.id )
sheat = sheat[!duplicated(sheat$dup), ]
sheat = sheat[ order ( sheat$type, sheat$gene, sheat$patient.id), ]

row.names ( sheat) = sheat$dup
pid = unique ( sheat$patient.id)
wl = which (   grepl ( pid , row.names ( sheat)) )
rename = row.names(sheat)
rename = gsub ( " .+", "", rename)               
r_annt = rowAnnotation(genes = anno_mark(at = wl , labels = rename[wl]  ))


left.annt = ifelse ( 
 rename %in% oncogene, "onc", "tsg"
    )


leftannt = rowAnnotation(
  Type =  left.annt
  
  ,col = list ( 
    Type= c(onc="#faee43", tsg= "#8cc5fa" )
  ),  annotation_name_gp= gpar(fontsize = 0, col="white" )
)



os.top.tempo = ComplexHeatmap::Heatmap( sheat[ , "coef", drop=F] 
                                    , name= paste ( " Temporal Slope " )
                                    #, row_order = type_row_order
                                    
                                    #, column_order = colnames ( pthis$df)
                                    , cluster_columns = F
                                    
                                  , show_row_names = F
                                    #, row_labels = ens_gene
                                    , show_row_dend = F
                                    , show_column_names = F
                                 #   , col= c("#eff5cb", "#f5946e")
                                    , col = col_fun                                 
                                    ,cluster_column_slices = F
                                    , cluster_rows =F
                                    ,row_title = NULL
                                    , right_annotation = r_annt
                                    #,column_split =  typekey$type1
                                    , column_title = NULL  
                                    #, top_annotation = ha
                                   ,  left_annotation = leftannt 
                                    
                                    
  )




## now you can plot these 
################################ seperate functions 
gene = "HOXD13"
disease = "OS"

plot.single.temporal <- function ( gene, disease ){

gene.tpm = reshape2::melt ( tpm[tpm$gene == gene, temporal.key$RNAseq.id] )
colnames ( gene.tpm ) = c("RNAseq.id", "tpm")
gene.tpm = merge ( gene.tpm, temporal.key[ , c(1,2,5,34)] ,by ="RNAseq.id" )
gene.tpm = gene.tpm[ gene.tpm$cancer.subtype == disease, ]
gene.tpm = gene.tpm[ order ( gene.tpm$cancer.subtype, gene.tpm$patient.id,  gene.tpm$temporal.order) , ]
gene.tpm$tpm = log2 ( gene.tpm$tpm + 1 )
gene.tpm$temporal.order = as.character (gene.tpm$temporal.order)
# plot by samples 


g2 = ggplot( gene.tpm[gene.tpm$patient.id %in% gene.tpm$patient.id [ duplicated ( gene.tpm$patient.id)] , ]  , aes(x = temporal.order, y = tpm, color = patient.id, group = patient.id)) + 
    geom_point(size=6 , alpha = 0.75    ) + 
  geom_path(arrow = arrow(length = unit(0.55, "cm") )  )+
  ylab(" log2(cpm)")+ xlab("temporal.order: RNA-seq") +
     theme_bw() +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.background = element_blank(), axis.line = element_line(colour = "black"))  +
    theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
          
          axis.text.y = element_text(size= 12 ),
          axis.text.x = element_text(angle = 0, size= 12 ),
          axis.title.x = element_text(size=18),
          
          axis.title.y     = element_text(size=18), 
          legend.text      =element_text(size=12)
    ) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
              axis.line = element_line(colour = "black"))+ scale_color_manual(values= getPalette(34)) +
  #facet_wrap(   cancer.subtype ~. , scales = "free"  ) 
  ggtitle ( paste ( gene, "Disease: ", disease ))

return ( g2 )
}

# find correlations same as above! 

find_cor <- function ( disease) {
os.long2 = tab.long[tab.long$cancer.subtype == disease & tab.long$type == "oncogene",  ]    
os.long2 = os.long2[ order ( os.long2$total.up, decreasing = T), ]
os.long2$state = ifelse ( os.long2$coef > 0, "pos", "neg")

testgene = c( os.long2[!duplicated ( os.long2$gene) & os.long2$total.up > .85, ]$gene )

mtest = plot.long.Mgene( testgene, disease)


sheat.top = mtest$df 

sheat.top$type = "onc"
left.annt = rep("onc", nrow(sheat.top))

os.long2 = tab.long[tab.long$cancer.subtype == disease & tab.long$type == "tsg",  ]    
os.long2 = os.long2[ order ( os.long2$total.up, decreasing = F), ]
os.long2$state = ifelse ( os.long2$coef > 0, "pos", "neg")

testgene = c( os.long2[!duplicated ( os.long2$gene) & os.long2$total.down > .85, ]$gene )

mtest = plot.long.Mgene( testgene, disease)

mtest$df$type = "tsg"

left.annt = c ( left.annt, rep("tsg", nrow(mtest$df)) )


sheat.top = rbind ( sheat.top, mtest$df)

sheat.top$dup = paste (  sheat.top$gene, sheat.top$patient.id )
sheat.top = sheat.top[!duplicated(sheat.top$dup), ]
sheat.top = sheat.top[ order ( sheat.top$type, sheat.top$gene, sheat.top$patient.id), ]

row.names ( sheat.top) = sheat.top$dup
pid = unique ( sheat.top$patient.id)
wl = which (   grepl ( paste0 ( " ",pid,"$") , row.names ( sheat.top)  ) )
rename = row.names(sheat.top)
rename = gsub ( " .+", "", rename)               
r_annt = rowAnnotation(genes = anno_mark(at = wl , labels = rename[wl]  ))


left.annt = ifelse ( 
  rename %in% oncogene, "onc", "tsg"
)


leftannt = rowAnnotation(
  Type =  left.annt
  
  ,col = list ( 
    Type= c(onc="#faee43", tsg= "#8cc5fa" )
  ),  annotation_name_gp= gpar(fontsize = 0, col="white" )
)



top.cor.ht = ComplexHeatmap::Heatmap( sheat.top[ , "coef", drop=F] 
                                        , name= paste ( " Temporal Slope " )
                                        #, row_order = type_row_order
                                        
                                        #, column_order = colnames ( pthis$df)
                                        , cluster_columns = F
                                        
                                        , show_row_names = F
                                        #, row_labels = ens_gene
                                        , show_row_dend = F
                                        , show_column_names = F
                                        #   , col= c("#eff5cb", "#f5946e")
                                        , col = col_fun                                 
                                        ,cluster_column_slices = F
                                        , cluster_rows =F
                                        ,row_title = NULL
                                        , right_annotation = r_annt
                                        #,column_split =  typekey$type1
                                        , column_title = NULL  
                                        #, top_annotation = ha
                                        ,  left_annotation = leftannt 
                                        
                                        
)

return ( list ( sheat = sheat.top, ht=top.cor.ht) )
}


os.cor = find_cor ( "OS")
nrsts.cor = find_cor ( "NRSTS")
wilms.cor = find_cor ( "WILMS")
plot.single.temporal ( gene = "CDH11", "OS")

```


## TOP trending genes {.tabset}

### OS {.tabset}

#### Heatmap 

* these are the genes with the most consistent change in direction from low to high temporal order

```{r fig=TRUE,fig.width=16, fig.height=12, echo=FALSE, include=TRUE , results='asis'}
draw ( os.cor$ht ) 
```

```{r fig=TRUE,fig.width=8, fig.height=8, echo=FALSE, include=TRUE , results='asis'}
 # for ( gene in unique ( sheat$gene)) {
  #  cat ( paste ( "####", gene , "\n\n"))
   # plot ( plot.single.temporal ( gene = gene, "OS") )
    #cat ( "\n\n")
  #}
```

#### Oncogenes
```{r fig=TRUE,fig.width=16, fig.height=12, echo=FALSE, include=TRUE , results='asis'}
mtest = plot.long.Mgene( unique ( os.cor$sheat[os.cor$sheat$type == "onc", ]$gene ) , "OS")
mtest$plot
```

#### TSG
```{r fig=TRUE,fig.width=16, fig.height=12, echo=FALSE, include=TRUE , results='asis'}
mtest = plot.long.Mgene( unique ( os.cor$sheat[os.cor$sheat$type != "onc", ]$gene ) , "OS")
mtest$plot
```

### NRSTS {.tabset}

#### Heatmap 

* these are the genes with the most consistent change in direction from low to high temporal order

```{r fig=TRUE,fig.width=16, fig.height=12, echo=FALSE, include=TRUE , results='asis'}
draw ( nrsts.cor$ht ) 
```

#### Oncogenes
```{r fig=TRUE,fig.width=16, fig.height=12, echo=FALSE, include=TRUE , results='asis'}
mtest = plot.long.Mgene( unique ( nrsts.cor$sheat[nrsts.cor$sheat$type == "onc", ]$gene ) , "NRSTS")
mtest$plot
```

#### TSG
```{r fig=TRUE,fig.width=16, fig.height=12, echo=FALSE, include=TRUE , results='asis'}
mtest = plot.long.Mgene( unique ( nrsts.cor$sheat[nrsts.cor$sheat$type != "onc", ]$gene ) , "NRSTS")
mtest$plot
```


